// tdesign-react/es/color-picker/components/trigger.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import React from 'react';
import classNames from 'classnames';
import { Input } from '../../input/index.js';
import { Color } from '../../_common/js/color-picker/color.js';
import useClassNames from '../hooks/useClassNames.js';
import useControlled from '../../hooks/useControlled.js';
import '../../_chunks/dep-7909fefa.js';
import '../../input/Input.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-8368bb87.js';
import 'tdesign-icons-react';
import '../../_chunks/dep-e430e2ee.js';
import '../../_chunks/dep-1cce5a60.js';
import '../../_chunks/dep-4a64abb1.js';
import '../../_chunks/dep-e54bb0dc.js';
import '../../_chunks/dep-b2626e28.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../hooks/useConfig.js';
import '../../_chunks/dep-13a70a46.js';
import '../../_chunks/dep-dc1ce7ed.js';
import '../../_chunks/dep-4e2181de.js';
import '../../_chunks/dep-c197ce4f.js';
import '../../_chunks/dep-5a2778c1.js';
import '../../_chunks/dep-a2b97df8.js';
import '../../_chunks/dep-7aeeb170.js';
import '../../_chunks/dep-2afd805c.js';
import '../../_chunks/dep-e237f0f2.js';
import '../../_chunks/dep-4ed2dce5.js';
import '../../_chunks/dep-c8db35c0.js';
import '../../_chunks/dep-96aef4a9.js';
import '../../_chunks/dep-557a1b0f.js';
import '../../_chunks/dep-264b0e55.js';
import '../../_chunks/dep-f5693bf6.js';
import '../../_chunks/dep-fdd1d1a6.js';
import '../../_chunks/dep-7ff8b18e.js';
import '../../_chunks/dep-ebf3a272.js';
import '../../_chunks/dep-6a2482c6.js';
import '../../_chunks/dep-a980ac05.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import '../../hooks/useGlobalIcon.js';
import '../../_common/js/utils/helper.js';
import '../../_chunks/dep-59b9253f.js';
import '../../_chunks/dep-d1b4ab24.js';
import '../../input/InputGroup.js';
import '../../locale/LocalReceiver.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../input/defaultProps.js';
import '../../_chunks/dep-f54ff1e5.js';
import '../../_chunks/dep-e1d4dd61.js';
import '../../_chunks/dep-236ef238.js';
import '../../_chunks/dep-c8bfa11e.js';
import '../../_chunks/dep-199b1581.js';
import '../../_util/noop.js';
import '../../input/style/css.js';
import '../../input/type.js';
import '../../_chunks/dep-864ee0c2.js';
import 'tinycolor2';
import '../../_common/js/color-picker/cmyk.js';
import '../../_common/js/color-picker/gradient.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var ColorPickerTrigger = function ColorPickerTrigger(props) {
  var baseClassName = useClassNames();
  var _props$disabled = props.disabled,
      disabled = _props$disabled === void 0 ? false : _props$disabled,
      _props$inputProps = props.inputProps,
      inputProps = _props$inputProps === void 0 ? {
    autoWidth: true
  } : _props$inputProps;

  var _useControlled = useControlled(props, "value", props.onChange),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      value = _useControlled2[0],
      setValue = _useControlled2[1];

  var handleChange = function handleChange(input, ctx) {
    if (Color.isValid(input)) {
      setValue(input, ctx);
    }
  };

  return /* @__PURE__ */React.createElement("div", {
    className: "".concat(baseClassName, "__trigger--default")
  }, /* @__PURE__ */React.createElement(Input, _objectSpread(_objectSpread({}, inputProps), {}, {
    value: value,
    disabled: disabled,
    label: /* @__PURE__ */React.createElement("div", {
      className: classNames("".concat(baseClassName, "__trigger--default__color"), "".concat(baseClassName, "--bg-alpha"))
    }, /* @__PURE__ */React.createElement("span", {
      className: "color-inner",
      style: {
        background: value
      }
    })),
    onBlur: handleChange,
    onChange: function onChange(v) {
      return setValue(v);
    }
  })));
};

var ColorTrigger = /*#__PURE__*/React.memo(ColorPickerTrigger);

export { ColorTrigger as default };
