// tdesign-react/lib/watermark/Watermark.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import React, { useState, useRef, useEffect } from 'react';
import generateBase64Url from '../_common/js/watermark/generateBase64Url.js';
import randomMovingStyle from '../_common/js/watermark/randomMovingStyle.js';
import injectStyle from '../_common/js/utils/injectStyle.js';
import useConfig from '../hooks/useConfig.js';
import useMutationObservable from '../_util/useMutationObserver.js';
import { watermarkDefaultProps } from './defaultProps.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../_chunks/dep-05d10538.js';
import '../_chunks/dep-652c1652.js';
import '../_chunks/dep-b989427d.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var Watermark = function Watermark(_ref) {
  var _ref$alpha = _ref.alpha,
      alpha = _ref$alpha === void 0 ? watermarkDefaultProps.alpha : _ref$alpha,
      _ref$x = _ref.x,
      x = _ref$x === void 0 ? 200 : _ref$x,
      _ref$y = _ref.y,
      y = _ref$y === void 0 ? 210 : _ref$y,
      _ref$width = _ref.width,
      width = _ref$width === void 0 ? 120 : _ref$width,
      _ref$height = _ref.height,
      height = _ref$height === void 0 ? 60 : _ref$height,
      _ref$rotate = _ref.rotate,
      tempRotate = _ref$rotate === void 0 ? watermarkDefaultProps.rotate : _ref$rotate,
      _ref$zIndex = _ref.zIndex,
      zIndex = _ref$zIndex === void 0 ? 10 : _ref$zIndex,
      _ref$lineSpace = _ref.lineSpace,
      lineSpace = _ref$lineSpace === void 0 ? watermarkDefaultProps.lineSpace : _ref$lineSpace,
      _ref$isRepeat = _ref.isRepeat,
      isRepeat = _ref$isRepeat === void 0 ? watermarkDefaultProps.isRepeat : _ref$isRepeat,
      _ref$removable = _ref.removable,
      removable = _ref$removable === void 0 ? watermarkDefaultProps.removable : _ref$removable,
      _ref$movable = _ref.movable,
      movable = _ref$movable === void 0 ? watermarkDefaultProps.movable : _ref$movable,
      _ref$moveInterval = _ref.moveInterval,
      moveInterval = _ref$moveInterval === void 0 ? watermarkDefaultProps.moveInterval : _ref$moveInterval,
      _ref$offset = _ref.offset,
      offset = _ref$offset === void 0 ? [] : _ref$offset,
      content = _ref.content,
      children = _ref.children,
      watermarkContent = _ref.watermarkContent,
      className = _ref.className,
      _ref$style = _ref.style,
      style = _ref$style === void 0 ? {} : _ref$style;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var gapX = x;
  var gapY = y;
  var rotate = tempRotate;

  if (movable) {
    gapX = 0;
    gapY = 0;
    rotate = 0;
  }

  var clsName = "".concat(classPrefix, "-watermark");

  var _useState = useState(""),
      _useState2 = _slicedToArray(_useState, 2),
      base64Url = _useState2[0],
      setBase64Url = _useState2[1];

  var watermarkRef = useRef();
  var offsetLeft = offset[0] || gapX / 2;
  var offsetTop = offset[1] || gapY / 2;
  useEffect(function () {
    generateBase64Url({
      width: width,
      height: height,
      rotate: rotate,
      lineSpace: lineSpace,
      alpha: alpha,
      gapX: gapX,
      gapY: gapY,
      watermarkContent: watermarkContent,
      offsetLeft: offsetLeft,
      offsetTop: offsetTop
    }, function (url) {
      setBase64Url(url);
    });
  }, [width, height, rotate, zIndex, lineSpace, alpha, offsetLeft, offsetTop, gapX, gapY, watermarkContent]);
  useMutationObservable(watermarkRef.current, function (mutations) {
    if (removable) return;
    mutations.forEach(function (mutation) {
      if (mutation.type === "childList") {
        var removeNodes = mutation.removedNodes;
        removeNodes.forEach(function (node) {
          watermarkRef.current.appendChild(node);
        });
      }
    });
  });
  var parent = useRef();
  useEffect(function () {
    parent.current = watermarkRef.current.parentElement;
    var keyframesStyle = randomMovingStyle();
    injectStyle(keyframesStyle);
  }, []);
  useMutationObservable(document.body, function (mutations) {
    if (removable) return;
    mutations.forEach(function (mutation) {
      if (mutation.type === "childList") {
        var removeNodes = mutation.removedNodes;
        removeNodes.forEach(function (node) {
          var element = node;

          if (element === watermarkRef.current) {
            parent.current.appendChild(element);
          }
        });
      }
    });
  });
  return /* @__PURE__ */React.createElement("div", {
    style: {
      position: "relative",
      overflow: "hidden"
    },
    className: clsName,
    ref: watermarkRef
  }, children || content, /* @__PURE__ */React.createElement("div", {
    className: className,
    style: _objectSpread({
      zIndex: zIndex,
      position: "absolute",
      left: 0,
      right: 0,
      top: 0,
      bottom: 0,
      width: "100%",
      height: "100%",
      backgroundSize: "".concat(gapX + width, "px"),
      pointerEvents: "none",
      backgroundRepeat: movable ? "no-repeat" : isRepeat ? "repeat" : "no-repeat",
      backgroundImage: "url('".concat(base64Url, "')"),
      animation: movable ? "watermark infinite ".concat(moveInterval * 4 / 60, "s") : "none"
    }, style)
  }));
};

export { Watermark as default };
