// tdesign-react/esm/table/hooks/usePagination.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import { _ as _toConsumableArray } from '../../_chunks/dep-59b9253f.js';
import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import React, { useState, useCallback, useEffect } from 'react';
import useConfig from '../../hooks/useConfig.js';
import { Pagination } from '../../pagination/index.js';
import '../../_chunks/dep-7909fefa.js';
import '../../_chunks/dep-d1b4ab24.js';
import '../../_chunks/dep-13a70a46.js';
import '../../_chunks/dep-1cce5a60.js';
import '../../_chunks/dep-dc1ce7ed.js';
import '../../_chunks/dep-4e2181de.js';
import '../../_chunks/dep-c197ce4f.js';
import '../../_chunks/dep-e430e2ee.js';
import '../../_chunks/dep-4a64abb1.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-e54bb0dc.js';
import '../../_chunks/dep-b2626e28.js';
import '../../_chunks/dep-5a2778c1.js';
import '../../_chunks/dep-a2b97df8.js';
import '../../_chunks/dep-7aeeb170.js';
import '../../_chunks/dep-2afd805c.js';
import '../../_chunks/dep-e237f0f2.js';
import '../../_chunks/dep-4ed2dce5.js';
import '../../_chunks/dep-c8db35c0.js';
import '../../_chunks/dep-96aef4a9.js';
import '../../_chunks/dep-557a1b0f.js';
import '../../_chunks/dep-264b0e55.js';
import '../../_chunks/dep-f5693bf6.js';
import '../../_chunks/dep-fdd1d1a6.js';
import '../../_chunks/dep-7ff8b18e.js';
import '../../_chunks/dep-ebf3a272.js';
import '../../_chunks/dep-6a2482c6.js';
import '../../_chunks/dep-a980ac05.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import '../../pagination/Pagination.js';
import '../../_chunks/dep-8368bb87.js';
import 'classnames';
import '../../_chunks/dep-27f96743.js';
import '../../_chunks/dep-c8bfa11e.js';
import '../../_chunks/dep-199b1581.js';
import '../../_chunks/dep-1a36642a.js';
import '../../_chunks/dep-614f9a08.js';
import '../../_chunks/dep-aadcae4d.js';
import '../../_chunks/dep-93a10d8c.js';
import '../../_chunks/dep-236ef238.js';
import '../../_chunks/dep-3576b6cd.js';
import '../../_util/noop.js';
import '../../hooks/useControlled.js';
import '../../_chunks/dep-f54ff1e5.js';
import '../../_chunks/dep-e1d4dd61.js';
import '../../select/index.js';
import '../../select/base/Select.js';
import '../../_chunks/dep-ddb0f1c1.js';
import '../../locale/LocalReceiver.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../select/util/helper.js';
import '../../select/base/OptionGroup.js';
import '../../select/defaultProps.js';
import '../../select/base/Option.js';
import '../../_chunks/dep-a5f8b6ec.js';
import '../../_chunks/dep-3c98169f.js';
import '../../_util/useRipple.js';
import '../../_util/useAnimation.js';
import '../../_util/setStyle.js';
import '../../_util/dom.js';
import 'raf';
import '../../_util/easing.js';
import '../../common/FakeArrow.js';
import '../../loading/index.js';
import '../../loading/Loading.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../loading/gradient.js';
import '../../_common/js/loading/circle-adapter.js';
import '../../_common/js/utils/set-style.js';
import '../../_common/js/utils/helper.js';
import '../../loading/defaultProps.js';
import '../../loading/plugin.js';
import '../../loading/style/index.js';
import '../../loading/type.js';
import '../../select-input/index.js';
import '../../select-input/SelectInput.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../../popup/hooks/useTrigger.js';
import 'react-is';
import '../../popup/utils/ref.js';
import '../../_util/composeRefs.js';
import '../../popup/utils/transition.js';
import '../../_util/useMutationObserver.js';
import '../../_chunks/dep-236b0359.js';
import '../../_chunks/dep-6d5c0a90.js';
import '../../_util/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../popup/style/index.js';
import '../../popup/type.js';
import '../../select-input/useSingle.js';
import '../../_chunks/dep-c776db26.js';
import '../../_chunks/dep-afbd242f.js';
import '../../input/index.js';
import '../../input/Input.js';
import 'tdesign-icons-react';
import '../../hooks/useGlobalIcon.js';
import '../../input/InputGroup.js';
import '../../input/defaultProps.js';
import '../../input/style/index.js';
import '../../input/type.js';
import '../../select-input/useMultiple.js';
import '../../tag-input/index.js';
import '../../tag-input/TagInput.js';
import '../../_util/useDragSorter.js';
import '../../tag-input/useTagScroll.js';
import '../../tag-input/useTagList.js';
import 'lodash';
import '../../tag/index.js';
import '../../tag/Tag.js';
import '../../tag/CheckTag.js';
import '../../tag/defaultProps.js';
import '../../tag/style/index.js';
import '../../tag/type.js';
import '../../tag-input/useHover.js';
import '../../tag-input/defaultProps.js';
import '../../tag-input/style/index.js';
import '../../tag-input/type.js';
import '../../select-input/useOverlayInnerStyle.js';
import '../../select-input/defaultProps.js';
import '../../select-input/style/index.js';
import '../../select-input/type.js';
import '../../select/base/PopupContent.js';
import '../../select/style/index.js';
import '../../select/type.js';
import '../../input-number/index.js';
import '../../input-number/InputNumber.js';
import '../../button/index.js';
import '../../button/Button.js';
import '../../button/defaultProps.js';
import '../../button/style/index.js';
import '../../button/type.js';
import '../../input-number/useInputNumber.js';
import '../../_util/useCommonClassName.js';
import '../../_common/js/input-number/number.js';
import '../../_common/js/input-number/large-number.js';
import '../../_common/js/log/log.js';
import '../../_common/js/log/index.js';
import '../../input-number/defaultProps.js';
import '../../input-number/style/index.js';
import '../../input-number/type.js';
import '../../input-adornment/index.js';
import '../../input-adornment/InputAdornment.js';
import '../../input-adornment/style/index.js';
import '../../input-adornment/type.js';
import '../../pagination/hooks/useBoundaryJumper.js';
import '../../pagination/hooks/usePrevNextJumper.js';
import '../../pagination/hooks/usePageNumber.js';
import '../../pagination/hooks/useTotal.js';
import '../../pagination/validators.js';
import '../../pagination/defaultProps.js';
import '../../pagination/style/index.js';
import '../../pagination/type.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function usePagination(props) {
  var pagination = props.pagination,
      data = props.data,
      disableDataPage = props.disableDataPage;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useState = useState([]),
      _useState2 = _slicedToArray(_useState, 2),
      dataSource = _useState2[0],
      setDataSource = _useState2[1];

  var _useState3 = useState(false),
      _useState4 = _slicedToArray(_useState3, 2),
      isPaginateData = _useState4[0],
      setIsPaginateData = _useState4[1];

  var updateDataSourceAndPaginate = useCallback(function () {
    var current = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    var pageSize = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 10;
    var isPaginateData2 = Boolean(!disableDataPage && data.length > pageSize);
    setIsPaginateData(isPaginateData2);

    if (isPaginateData2) {
      var start = (current - 1) * pageSize;
      var end = current * pageSize;
      setDataSource(_toConsumableArray(data.slice(start, end)));
    } else {
      setDataSource(data);
    }
  }, [data, disableDataPage]);
  useEffect(function () {
    if (!pagination || !pagination.current) return;
    updateDataSourceAndPaginate(pagination.current, pagination.pageSize);
  }, [pagination, updateDataSourceAndPaginate]);
  useEffect(function () {
    if (!pagination || !pagination.defaultCurrent) return;
    updateDataSourceAndPaginate(pagination.defaultCurrent, pagination.defaultPageSize);
  }, [updateDataSourceAndPaginate]);

  var renderPagination = function renderPagination() {
    if (!pagination) return null;
    return /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-table__pagination")
    }, /* @__PURE__ */React.createElement(Pagination, _objectSpread(_objectSpread({}, pagination), {}, {
      onChange: function onChange(pageInfo) {
        var _props$pagination, _props$pagination$onC, _props$onPageChange;

        (_props$pagination = props.pagination) === null || _props$pagination === void 0 ? void 0 : (_props$pagination$onC = _props$pagination.onChange) === null || _props$pagination$onC === void 0 ? void 0 : _props$pagination$onC.call(_props$pagination, pageInfo);
        (_props$onPageChange = props.onPageChange) === null || _props$onPageChange === void 0 ? void 0 : _props$onPageChange.call(props, pageInfo, dataSource);

        if (pagination && !pagination.current && pagination.defaultCurrent) {
          updateDataSourceAndPaginate(pageInfo.current, pageInfo.pageSize);
        }
      }
    })));
  };

  return {
    isPaginateData: isPaginateData,
    dataSource: dataSource,
    renderPagination: renderPagination
  };
}

export { usePagination as default };
