// tdesign-react/lib/calendar/CalendarCellComp.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import React from 'react';
import useConfig from '../hooks/useConfig.js';
import usePrefixClass from './hooks/usePrefixClass.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import { blockName } from './_util.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-1e32006c.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import 'dayjs';

var CalendarCellComp = function CalendarCellComp(props) {
  var mode = props.mode,
      cell = props.cell,
      cellAppend = props.cellAppend,
      theme = props.theme,
      _props$isDisabled = props.isDisabled,
      isDisabled = _props$isDisabled === void 0 ? false : _props$isDisabled,
      cellData = props.cellData,
      isCurrent = props.isCurrent,
      isNow = props.isNow,
      fillWithZero = props.fillWithZero,
      createCalendarCell = props.createCalendarCell,
      onCellClick = props.onCellClick,
      onCellDoubleClick = props.onCellDoubleClick,
      onCellRightClick = props.onCellRightClick;

  var _useLocaleReceiver = useLocaleReceiver("calendar"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var monthLabelList = t(local.cellMonth).split(",");

  var getMonthCN = function getMonthCN(month) {
    return monthLabelList[month];
  };

  var _useConfig = useConfig(),
      calendarConfig = _useConfig.calendar;

  var fix0 = function fix0(num) {
    var _ref;

    var fillZero = num < 10 && ((_ref = fillWithZero !== null && fillWithZero !== void 0 ? fillWithZero : calendarConfig.fillWithZero) !== null && _ref !== void 0 ? _ref : true);
    return fillZero ? "0".concat(num) : num;
  };

  var prefixCls = usePrefixClass();
  return /* @__PURE__ */React.createElement("td", {
    className: prefixCls([blockName, "table-body-cell"], isDisabled ? "is-disabled" : null, isCurrent ? "is-checked" : null, isNow ? [blockName, "table-body-cell--now"] : null),
    onClick: onCellClick,
    onDoubleClick: onCellDoubleClick,
    onContextMenu: onCellRightClick
  }, function () {
    if (cell && typeof cell === "function") return cell(mode === "month" ? createCalendarCell(cellData) : cellData);
    if (cell && typeof cell !== "function") return cell;
    var cellCtx;

    if (mode === "year") {
      var mIndex = cellData.date.getMonth();
      cellCtx = theme === "full" ? getMonthCN(mIndex) : t(local.monthSelection, {
        month: "".concat(mIndex + 1)
      });
    } else {
      cellCtx = fix0(cellData.date.getDate());
    }

    return /* @__PURE__ */React.createElement("div", {
      className: prefixCls([blockName, "table-body-cell-display"])
    }, cellCtx);
  }(), function () {
    var cellCtx;

    if (cellAppend && typeof cellAppend === "function") {
      cellCtx = cellAppend(mode === "month" ? createCalendarCell(cellData) : cellData);
    }

    if (cellAppend && typeof cellAppend !== "function") {
      cellCtx = cellAppend;
    }

    return cellAppend && /* @__PURE__ */React.createElement("div", {
      className: prefixCls([blockName, "table-body-cell-content"])
    }, cellCtx);
  }());
};

export { CalendarCellComp as default };
