// tdesign-react/lib/steps/StepItem.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import React, { useContext } from 'react';
import classNames from 'classnames';
import { CloseIcon, CheckIcon } from 'tdesign-icons-react';
import useConfig from '../hooks/useConfig.js';
import useGlobalIcon from '../hooks/useGlobalIcon.js';
import StepsContext from './StepsContext.js';
import { stepItemDefaultProps } from './defaultProps.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';

var StepItem = function StepItem(props) {
  var _classNames;

  var index = props.index,
      icon = props.icon,
      title = props.title,
      content = props.content,
      value = props.value,
      children = props.children,
      style = props.style,
      status = props.status;

  var _useContext = useContext(StepsContext),
      current = _useContext.current,
      theme = _useContext.theme,
      onChange = _useContext.onChange,
      readonly = _useContext.readonly;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix,
      globalStepsConfig = _useConfig.steps;

  var _useGlobalIcon = useGlobalIcon({
    CloseIcon: CloseIcon,
    CheckIcon: CheckIcon
  }),
      CloseIcon$1 = _useGlobalIcon.CloseIcon,
      CheckIcon$1 = _useGlobalIcon.CheckIcon;

  var canClick = status !== "process" && !readonly;

  function renderIcon() {
    if (!icon) return null;
    var iconCls = "".concat(classPrefix, "-steps-item__icon--number");
    if (icon && icon !== true) return /* @__PURE__ */React.createElement("span", {
      className: iconCls
    }, icon);
    if (theme !== "default") return null;
    if (status === "error") return /* @__PURE__ */React.createElement("span", {
      className: iconCls
    }, globalStepsConfig.errorIcon || /* @__PURE__ */React.createElement(CloseIcon$1, null));
    if (status === "finish") return /* @__PURE__ */React.createElement("span", {
      className: iconCls
    }, /* @__PURE__ */React.createElement(CheckIcon$1, null));
    return /* @__PURE__ */React.createElement("span", {
      className: iconCls
    }, Number(index) + 1);
  }

  function onStepClick(e) {
    if (!canClick) return;
    var currentValue = value !== null && value !== void 0 ? value : index;
    onChange(currentValue, current, {
      e: e
    });
  }

  return /* @__PURE__ */React.createElement("div", {
    style: style,
    className: classNames((_classNames = {}, _defineProperty(_classNames, "".concat(classPrefix, "-steps-item"), true), _defineProperty(_classNames, "".concat(classPrefix, "-steps-item--wait"), status === "default"), _defineProperty(_classNames, "".concat(classPrefix, "-steps-item--error"), status === "error"), _defineProperty(_classNames, "".concat(classPrefix, "-steps-item--finish"), status === "finish"), _defineProperty(_classNames, "".concat(classPrefix, "-steps-item--process"), status === "process"), _defineProperty(_classNames, props.className, !!props.className), _classNames))
  }, /* @__PURE__ */React.createElement("div", {
    className: classNames("".concat(classPrefix, "-steps-item__inner"), _defineProperty({}, "".concat(classPrefix, "-steps-item--clickable"), canClick)),
    onClick: onStepClick
  }, /* @__PURE__ */React.createElement("div", {
    className: classNames("".concat(classPrefix, "-steps-item__icon"), _defineProperty({}, "".concat(classPrefix, "-steps-item-").concat(status), status))
  }, renderIcon()), /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-steps-item__content")
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-steps-item__title")
  }, title), /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-steps-item__description")
  }, content), children ? /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-steps-item__extra")
  }, children) : null)));
};

StepItem.displayName = "StepItem";
StepItem.defaultProps = stepItemDefaultProps;

export { StepItem as default };
