// tdesign-react/lib/input-adornment/InputAdornment.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _objectWithoutProperties } from '../_chunks/dep-8368bb87.js';
import React, { forwardRef } from 'react';
import classNames from 'classnames';
import useConfig from '../hooks/useConfig.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';

var _excluded = ["prepend", "append", "children", "className", "style"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var InputAdornment = /*#__PURE__*/forwardRef(function (props, ref) {
  var _classNames;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var prepend = props.prepend,
      append = props.append,
      children = props.children,
      className = props.className,
      style = props.style,
      wrapperProps = _objectWithoutProperties(props, _excluded);

  var renderAddon = function renderAddon(type, classPrefix2, Content) {
    var result;

    if (typeof Content === "function") {
      result = /* @__PURE__ */React.createElement(Content, null);
    } else if (typeof Content !== "undefined") {
      result = Content;
    }

    if (result || typeof result === "number") {
      result = /* @__PURE__ */React.createElement("span", {
        className: "".concat(classPrefix2, "-input-adornment__").concat(type)
      }, result);
    }

    return result;
  };

  var renderChildren = function renderChildren() {
    return React.Children.map(children, function (child) {
      if (!child) return null;

      if ( /*#__PURE__*/React.isValidElement(child)) {
        return /*#__PURE__*/React.cloneElement(child, _objectSpread(_objectSpread(_objectSpread({}, wrapperProps), child.props), {}, {
          onChange: function onChange() {
            var _wrapperProps$onChang, _wrapperProps$onChang2, _child$props, _child$props$onChange, _child$props$onChange2;

            for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
              args[_key] = arguments[_key];
            }

            wrapperProps === null || wrapperProps === void 0 ? void 0 : (_wrapperProps$onChang = wrapperProps.onChange) === null || _wrapperProps$onChang === void 0 ? void 0 : (_wrapperProps$onChang2 = _wrapperProps$onChang.call) === null || _wrapperProps$onChang2 === void 0 ? void 0 : _wrapperProps$onChang2.call.apply(_wrapperProps$onChang2, [_wrapperProps$onChang, null].concat(args));
            (_child$props = child.props) === null || _child$props === void 0 ? void 0 : (_child$props$onChange = _child$props.onChange) === null || _child$props$onChange === void 0 ? void 0 : (_child$props$onChange2 = _child$props$onChange.call) === null || _child$props$onChange2 === void 0 ? void 0 : _child$props$onChange2.call.apply(_child$props$onChange2, [_child$props$onChange, null].concat(args));
          }
        }));
      }

      return child;
    });
  };

  return /* @__PURE__ */React.createElement("div", {
    ref: ref,
    style: style,
    className: classNames("".concat(classPrefix, "-input-adornment"), className, (_classNames = {}, _defineProperty(_classNames, "".concat(classPrefix, "-input-adornment--prepend"), !!prepend), _defineProperty(_classNames, "".concat(classPrefix, "-input-adornment--append"), !!append), _classNames))
  }, renderAddon("prepend", classPrefix, prepend), renderChildren(), renderAddon("append", classPrefix, append));
});
InputAdornment.displayName = "InputAdornment";

export { InputAdornment as default };
