// tdesign-react/cjs/menu/HeadMenu.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var classNames = require('classnames');
var React = require('react');
var isObject = require('../_chunks/dep-aab723b3.js');
var tabs_index = require('../tabs/index.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var menu_hooks_useMenuContext = require('./hooks/useMenuContext.js');
var menu_MenuContext = require('./MenuContext.js');
var menu__util_checkSubMenuActive = require('./_util/checkSubMenuActive.js');
require('../_chunks/dep-2a90f794.js');
require('../tabs/Tabs.js');
require('../_chunks/dep-9c5d4f93.js');
require('../_chunks/dep-2205decf.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../tabs/TabNav.js');
require('tdesign-icons-react');
require('../_chunks/dep-db2b3323.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-52c5813c.js');
require('../_chunks/dep-09805e35.js');
require('../_chunks/dep-1fff9729.js');
require('../_util/noop.js');
require('../tabs/useTabClass.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../tabs/TabNavItem.js');
require('../_util/useRipple.js');
require('../_util/useAnimation.js');
require('../_util/setStyle.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../hooks/useGlobalIcon.js');
require('../tabs/TabBar.js');
require('../_common/js/tabs/base.js');
require('../tabs/TabPanel.js');
require('../tabs/defaultProps.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('./_util/checkSubMenuChildExpanded.js');
require('./_util/type.js');
require('./_util/checkMenuType.js');
require('./_util/checkSubMenuChildrenActive.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var TabPanel = tabs_index.Tabs.TabPanel;

var HeadMenu = function HeadMenu(props) {
  var children = props.children,
      className = props.className,
      _props$theme = props.theme,
      theme = _props$theme === void 0 ? "light" : _props$theme,
      style = props.style,
      logo = props.logo,
      operations = props.operations;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useMenuContext = menu_hooks_useMenuContext["default"](_objectSpread(_objectSpread({}, props), {}, {
    children: children,
    mode: "title"
  })),
      value = _useMenuContext.value;

  var childList = React.useMemo(function () {
    if (value.expandType === "popup") return null;
    var activeMenu = menu__util_checkSubMenuActive["default"](children, value.active);
    if (!activeMenu) return null;
    var child = activeMenu.props.children;
    if (Array.isArray(child)) return child;
    if (isObject.isObject_1(child)) return [child];
    return activeMenu.props.children;
  }, [children, value.expandType, value.active]);
  var currentChildListValues = (childList === null || childList === void 0 ? void 0 : childList.length) > 0 ? childList.map(function (item) {
    return item.props.value;
  }) : [];
  return /* @__PURE__ */React__default["default"].createElement(menu_MenuContext.MenuContext.Provider, {
    value: value
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-head-menu"), "".concat(classPrefix, "-menu--").concat(theme), className),
    style: _objectSpread({}, style)
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-head-menu__inner")
  }, logo && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-menu__logo")
  }, logo), /* @__PURE__ */React__default["default"].createElement("ul", {
    className: "".concat(classPrefix, "-menu")
  }, children), operations && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-menu__operations")
  }, operations)), (childList === null || childList === void 0 ? void 0 : childList.length) > 0 && /* @__PURE__ */React__default["default"].createElement("ul", {
    className: "".concat(classPrefix, "-head-menu__submenu ").concat(classPrefix, "-submenu")
  }, /* @__PURE__ */React__default["default"].createElement(tabs_index.Tabs, {
    value: currentChildListValues.includes(value.active) ? value.active : currentChildListValues[0],
    onChange: value.onChange
  }, childList.map(function (_ref) {
    var props2 = _ref.props;
    return /* @__PURE__ */React__default["default"].createElement(TabPanel, {
      value: props2.value,
      key: props2.value,
      label: props2.children
    });
  })))));
};

exports["default"] = HeadMenu;
