// tdesign-react/cjs/menu/MenuItem.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var _util_useRipple = require('../_util/useRipple.js');
var menu_MenuContext = require('./MenuContext.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../_util/useAnimation.js');
require('../_util/setStyle.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../_util/noop.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var MenuItem = function MenuItem(props) {
  var _classNames;

  var content = props.content,
      _props$children = props.children,
      children = _props$children === void 0 ? content : _props$children,
      disabled = props.disabled,
      href = props.href,
      _props$target = props.target,
      target = _props$target === void 0 ? "_self" : _props$target,
      value = props.value,
      className = props.className,
      style = props.style,
      icon = props.icon,
      onClick = props.onClick;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var menuItemRef = React.useRef();
  _util_useRipple["default"](menuItemRef);

  var _useContext = React.useContext(menu_MenuContext.MenuContext),
      onChange = _useContext.onChange,
      setState = _useContext.setState,
      active = _useContext.active;

  var handleClick = function handleClick(e) {
    e.stopPropagation();
    if (disabled) return;
    onClick && onClick({
      e: e
    });
    onChange(value);
    setState({
      active: value
    });
  };

  return /* @__PURE__ */React__default["default"].createElement("li", {
    ref: menuItemRef,
    className: classNames__default["default"]("".concat(classPrefix, "-menu__item"), className, (_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-is-disabled"), disabled), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-is-active"), value === active), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-menu__item--plain"), !icon), _classNames)),
    style: _objectSpread({}, style),
    onClick: handleClick
  }, icon, href ? /* @__PURE__ */React__default["default"].createElement("a", {
    href: href,
    target: target,
    className: classNames__default["default"]("".concat(classPrefix, "-menu__item-link"))
  }, /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-menu__content")
  }, children)) : /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-menu__content")
  }, children));
};

MenuItem.displayName = "MenuItem";

exports["default"] = MenuItem;
