// tdesign-react/cjs/menu/SubMenu.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var menu_MenuContext = require('./MenuContext.js');
var _util_useRipple = require('../_util/useRipple.js');
var menu__util_getSubMenuChildStyle = require('./_util/getSubMenuChildStyle.js');
var menu__util_checkSubMenuChildrenActive = require('./_util/checkSubMenuChildrenActive.js');
var common_FakeArrow = require('../common/FakeArrow.js');
var menu__util_checkMenuType = require('./_util/checkMenuType.js');
var menu__util_cacularPaddingLeft = require('./_util/cacularPaddingLeft.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../_util/noop.js');
require('../_util/useAnimation.js');
require('../_util/setStyle.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('./_util/type.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var SubAccordion = function SubAccordion(props) {
  var _classNames, _classNames2;

  var content = props.content,
      _props$children = props.children,
      children = _props$children === void 0 ? content : _props$children,
      disabled = props.disabled,
      icon = props.icon,
      title = props.title,
      value = props.value,
      className = props.className,
      style = props.style,
      _props$level = props.level,
      level = _props$level === void 0 ? 1 : _props$level;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useState = React.useState(false),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      open = _useState2[0],
      setOpen = _useState2[1];

  var _useContext = React.useContext(menu_MenuContext.MenuContext),
      _useContext$expanded = _useContext.expanded,
      expanded = _useContext$expanded === void 0 ? [] : _useContext$expanded,
      onExpand = _useContext.onExpand,
      active = _useContext.active,
      expandType = _useContext.expandType;

  var isPopUp = expandType === "popup";
  var isExpand = expanded.includes(value) && !disabled && !isPopUp;

  var handleClick = function handleClick(e) {
    e.stopPropagation();
    onExpand(value, expanded);
  };

  var handleMouseEvent = function handleMouseEvent(type) {
    if (!isPopUp) return;
    if (type === "enter") setOpen(true);else if (type === "leave") setOpen(false);
  };

  var childrens = React__default["default"].Children.map(children, function (child) {
    return /*#__PURE__*/React__default["default"].cloneElement(child, {
      className: classNames__default["default"]("".concat(classPrefix, "-menu__item--plain"), "".concat(classPrefix, "-submenu__item"), "".concat(classPrefix, "-submenu__item--icon"))
    });
  });
  var childStyle = {
    maxHeight: isExpand || open && isPopUp ? menu__util_getSubMenuChildStyle.getSubMenuMaxHeight(children) : 0
  };
  var isOpen = React.useMemo(function () {
    if (disabled) return false;
    if (isPopUp) return open;
    return isExpand;
  }, [disabled, isPopUp, open, isExpand]);
  var menuPaddingLeft = menu__util_cacularPaddingLeft.cacularPaddingLeft(level - 1);
  var fakeArrowStyle = isPopUp && level > 1 ? {
    transform: "rotate(-90deg)"
  } : {};
  return /* @__PURE__ */React__default["default"].createElement("li", {
    className: classNames__default["default"]("".concat(classPrefix, "-submenu"), className, (_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-is-disabled"), disabled), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-is-opened"), isOpen), _classNames)),
    onClick: handleClick,
    style: style,
    onMouseEnter: function onMouseEnter() {
      return handleMouseEvent("enter");
    },
    onMouseLeave: function onMouseLeave() {
      return handleMouseEvent("leave");
    }
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-menu__item"), (_classNames2 = {}, defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-is-opened"), isOpen), defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-is-active"), menu__util_checkSubMenuChildrenActive["default"](children, active)), _classNames2))
  }, icon, " ", /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-menu__content")
  }, title), /* @__PURE__ */React__default["default"].createElement(common_FakeArrow["default"], {
    style: fakeArrowStyle,
    isActive: isOpen,
    disabled: disabled
  })), isPopUp ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-menu__popup"), "".concat(classPrefix, "-is-vertical"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-opened"), isOpen))
  }, /* @__PURE__ */React__default["default"].createElement("ul", {
    className: classNames__default["default"]("".concat(classPrefix, "-menu__popup-wrapper"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-opened"), isOpen)),
    key: "popup",
    style: childStyle
  }, childrens)) : /* @__PURE__ */React__default["default"].createElement("ul", {
    key: "normal",
    style: _objectSpread(_objectSpread({}, childStyle), {}, {
      "--padding-left": "".concat(menuPaddingLeft, "px")
    }),
    className: "".concat(classPrefix, "-menu__sub")
  }, childrens));
};

var SubTitleMenu = function SubTitleMenu(props) {
  var _popRef$current, _classNames6;

  var className = props.className,
      style = props.style,
      children = props.children,
      title = props.title,
      value = props.value,
      level = props.level;

  var _useContext2 = React.useContext(menu_MenuContext.MenuContext),
      active = _useContext2.active,
      onChange = _useContext2.onChange,
      expandType = _useContext2.expandType;

  var _useConfig2 = hooks_useConfig["default"](),
      classPrefix = _useConfig2.classPrefix;

  var _useState3 = React.useState(false),
      _useState4 = slicedToArray._slicedToArray(_useState3, 2),
      open = _useState4[0],
      setOpen = _useState4[1];

  var popRef = React.useRef();

  var handleClick = function handleClick() {
    return onChange(value);
  };

  var subMenuRef = React.useRef();
  _util_useRipple["default"](subMenuRef);
  var isPopUp = expandType === "popup";
  var isActive = menu__util_checkSubMenuChildrenActive["default"](children, active) || active === value;

  var handleMouseEvent = function handleMouseEvent(type) {
    if (!isPopUp) return;
    if (type === "enter") setOpen(true);else if (type === "leave") setOpen(false);
  };

  var fakeArrowStyle = level > 1 ? {
    transform: "rotate(-90deg)"
  } : {};
  var submenuMaxHeight = menu__util_getSubMenuChildStyle.getSubMenuMaxHeight(children);
  var popupMaxHeight = ((_popRef$current = popRef.current) === null || _popRef$current === void 0 ? void 0 : _popRef$current.getBoundingClientRect().height) || submenuMaxHeight;
  var showPopup = isPopUp && submenuMaxHeight > 0;
  return /* @__PURE__ */React__default["default"].createElement("li", {
    className: classNames__default["default"]("".concat(classPrefix, "-submenu"), className, defineProperty._defineProperty({}, "".concat(classPrefix, "-is-opened"), open)),
    onMouseEnter: function onMouseEnter() {
      return handleMouseEvent("enter");
    },
    onMouseLeave: function onMouseLeave() {
      return handleMouseEvent("leave");
    }
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    ref: subMenuRef,
    className: classNames__default["default"]("".concat(classPrefix, "-menu__item"), (_classNames6 = {}, defineProperty._defineProperty(_classNames6, "".concat(classPrefix, "-is-active"), isActive), defineProperty._defineProperty(_classNames6, "".concat(classPrefix, "-is-opened"), open), _classNames6)),
    onClick: handleClick,
    style: style
  }, /* @__PURE__ */React__default["default"].createElement("span", null, title), showPopup && /* @__PURE__ */React__default["default"].createElement(common_FakeArrow["default"], {
    style: fakeArrowStyle,
    isActive: open
  })), showPopup && /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-menu__popup"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-opened"), open)),
    style: {
      "--popup-max-height": "".concat(popupMaxHeight, "px")
    }
  }, /* @__PURE__ */React__default["default"].createElement("ul", {
    ref: popRef,
    className: classNames__default["default"]("".concat(classPrefix, "-menu__popup-wrapper"))
  }, children)));
};

var SubMenu = function SubMenu(props) {
  var _useContext3 = React.useContext(menu_MenuContext.MenuContext),
      mode = _useContext3.mode;

  var children = props.children,
      _props$level2 = props.level,
      level = _props$level2 === void 0 ? 1 : _props$level2;

  var changeItemLevel = function changeItemLevel(item) {
    if (menu__util_checkMenuType.checkIsSubMenu(item)) {
      return /*#__PURE__*/React__default["default"].cloneElement(item, {
        level: level + 1
      });
    }

    if (menu__util_checkMenuType.checkIsMenuGroup(item)) {
      var groupChildren = React__default["default"].Children.map(item.props.children, function (item2) {
        return changeItemLevel(item2);
      });
      return /*#__PURE__*/React__default["default"].cloneElement(item, {
        level: level + 1
      }, groupChildren);
    }

    return item;
  };

  var childElement = React__default["default"].Children.map(children, function (item) {
    return changeItemLevel(item);
  });
  if (mode === "accordion") return /* @__PURE__ */React__default["default"].createElement(SubAccordion, _objectSpread({}, props), childElement);
  if (mode === "title") return /* @__PURE__ */React__default["default"].createElement(SubTitleMenu, _objectSpread({}, props), childElement);
  return null;
};

SubMenu.displayName = "SubMenu";

exports["default"] = SubMenu;
