// tdesign-react/es/image/Image.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import { _ as _objectWithoutProperties } from '../_chunks/dep-8368bb87.js';
import React, { useRef, useState, useEffect, Fragment } from 'react';
import classNames from 'classnames';
import { ImageErrorIcon, ImageIcon } from 'tdesign-icons-react';
import observe from '../_common/js/utils/observe.js';
import useConfig from '../hooks/useConfig.js';
import { imageDefaultProps } from './defaultProps.js';
import { Space } from '../space/index.js';
import useGlobalIcon from '../hooks/useGlobalIcon.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-e54bb0dc.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-e237f0f2.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-ebf3a272.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../space/Space.js';
import 'react-is';
import '../space/defaultProps.js';
import '../space/style/css.js';
import '../space/type.js';

var _excluded = ["className", "src", "style", "alt", "fit", "position", "shape", "placeholder", "loading", "error", "overlayTrigger", "overlayContent", "lazy", "gallery", "onLoad", "onError"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var Image = function Image(props) {
  var className = props.className,
      src = props.src,
      style = props.style,
      alt = props.alt,
      fit = props.fit,
      position = props.position,
      shape = props.shape,
      placeholder = props.placeholder,
      loading = props.loading,
      error = props.error,
      overlayTrigger = props.overlayTrigger,
      overlayContent = props.overlayContent,
      lazy = props.lazy,
      gallery = props.gallery,
      onLoad = props.onLoad,
      onError = props.onError,
      rest = _objectWithoutProperties(props, _excluded);

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var imageRef = useRef(null);

  var _useGlobalIcon = useGlobalIcon({
    ImageErrorIcon: ImageErrorIcon,
    ImageIcon: ImageIcon
  }),
      ImageErrorIcon$1 = _useGlobalIcon.ImageErrorIcon,
      ImageIcon$1 = _useGlobalIcon.ImageIcon;

  var _useState = useState(!lazy),
      _useState2 = _slicedToArray(_useState, 2),
      shouldLoad = _useState2[0],
      setShouldLoad = _useState2[1];

  var handleLoadImage = function handleLoadImage() {
    setShouldLoad(true);
  };

  var _useState3 = useState(false),
      _useState4 = _slicedToArray(_useState3, 2),
      isLoaded = _useState4[0],
      setIsLoaded = _useState4[1];

  var handleLoad = function handleLoad() {
    setIsLoaded(true);
    onLoad === null || onLoad === void 0 ? void 0 : onLoad();
  };

  useEffect(function () {
    if (!lazy || !(imageRef !== null && imageRef !== void 0 && imageRef.current)) return;
    var observerRefValue = null;
    var io = observe(imageRef.current, null, handleLoadImage, 0);
    observerRefValue = imageRef.current;
    return function () {
      observerRefValue && io && io.unobserve(observerRefValue);
    };
  }, [lazy, imageRef]);

  var _useState5 = useState(false),
      _useState6 = _slicedToArray(_useState5, 2),
      hasError = _useState6[0],
      setHasError = _useState6[1];

  var handleError = function handleError() {
    setHasError(true);
    onError === null || onError === void 0 ? void 0 : onError();
  };

  var hasMouseEvent = overlayTrigger === "hover";

  var _useState7 = useState(!hasMouseEvent),
      _useState8 = _slicedToArray(_useState7, 2),
      shouldShowOverlay = _useState8[0],
      setShouldShowOverlay = _useState8[1];

  var handleToggleOverlay = function handleToggleOverlay(overlay) {
    setShouldShowOverlay(overlay);
  };

  var renderOverlay = function renderOverlay() {
    if (!overlayContent) return null;
    return /* @__PURE__ */React.createElement("div", {
      className: classNames("".concat(classPrefix, "-image__overlay-content"), !shouldShowOverlay && "".concat(classPrefix, "-image__overlay-content--hidden"))
    }, overlayContent);
  };

  var renderPlaceholder = function renderPlaceholder() {
    if (!placeholder) {
      return null;
    }

    return /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-image__placeholder")
    }, placeholder);
  };

  var renderGalleryShadow = function renderGalleryShadow() {
    if (!gallery) return null;
    return /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-image__gallery-shadow")
    });
  };

  return /* @__PURE__ */React.createElement("div", _objectSpread(_objectSpread({
    ref: imageRef,
    className: classNames("".concat(classPrefix, "-image__wrapper"), "".concat(classPrefix, "-image__wrapper--shape-").concat(shape), gallery && "".concat(classPrefix, "-image__wrapper--gallery"), hasMouseEvent && "".concat(classPrefix, "-image__wrapper--need-hover"), className),
    style: style
  }, hasMouseEvent ? {
    onMouseEnter: function onMouseEnter() {
      return handleToggleOverlay(true);
    },
    onMouseLeave: function onMouseLeave() {
      return handleToggleOverlay(false);
    }
  } : null), rest), renderPlaceholder(), renderGalleryShadow(), hasError || !shouldLoad ? /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-image")
  }) : /* @__PURE__ */React.createElement(Fragment, null, /* @__PURE__ */React.createElement("img", {
    src: src,
    onError: handleError,
    onLoad: handleLoad,
    className: classNames("".concat(classPrefix, "-image"), "".concat(classPrefix, "-image--fit-").concat(fit), "".concat(classPrefix, "-image--position-").concat(position)),
    alt: alt
  }), !isLoaded && /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-image__loading")
  }, loading || /* @__PURE__ */React.createElement(Space, {
    direction: "vertical",
    size: 8,
    align: "center"
  }, /* @__PURE__ */React.createElement(ImageIcon$1, {
    size: 24
  }), "\u56FE\u7247\u52A0\u8F7D\u4E2D"))), hasError && /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-image__error")
  }, error || /* @__PURE__ */React.createElement(Space, {
    direction: "vertical",
    size: 8,
    align: "center"
  }, /* @__PURE__ */React.createElement(ImageErrorIcon$1, {
    size: 24
  }), "\u56FE\u7247\u65E0\u6CD5\u663E\u793A")), renderOverlay());
};

Image.displayName = "Image";
Image.defaultProps = imageDefaultProps;

export { Image as default };
