// tdesign-react/cjs/select-input/useMultiple.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var isObject = require('../_chunks/dep-aab723b3.js');
var classNames = require('classnames');
var tagInput_index = require('../tag-input/index.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var hooks_useConfig = require('../hooks/useConfig.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-2a90f794.js');
require('../tag-input/TagInput.js');
require('tdesign-icons-react');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../hooks/useGlobalIcon.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../_util/useDragSorter.js');
require('../input/index.js');
require('../input/Input.js');
require('../_chunks/dep-1a0462cc.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../_common/js/utils/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../input/InputGroup.js');
require('../locale/LocalReceiver.js');
require('../input/defaultProps.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-09805e35.js');
require('../_util/noop.js');
require('../tag-input/useTagScroll.js');
require('../tag-input/useTagList.js');
require('lodash');
require('../tag/index.js');
require('../tag/Tag.js');
require('../tag/CheckTag.js');
require('../tag/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var DEFAULT_KEYS = {
  label: "label",
  key: "key",
  children: "children"
};
function useMultiple(props) {
  var value = props.value;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var tagInputRef = React.useRef();

  var _useControlled = hooks_useControlled["default"](props, "inputValue", props.onInputChange),
      _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
      tInputValue = _useControlled2[0],
      setTInputValue = _useControlled2[1];

  var iKeys = _objectSpread(_objectSpread({}, DEFAULT_KEYS), props.keys);

  var getTags = function getTags() {
    if (!(value instanceof Array)) {
      return isObject.isObject_1(value) ? [value[iKeys.label]] : [value];
    }

    return value.map(function (item) {
      return isObject.isObject_1(item) ? item[iKeys.label] : item;
    });
  };

  var tags = getTags();
  var tPlaceholder = !tags || !tags.length ? props.placeholder : "";

  var onTagInputChange = function onTagInputChange(val, context) {
    var _props$onTagChange;

    if (context.trigger === "tag-remove") {
      var _context$e;

      (_context$e = context.e) === null || _context$e === void 0 ? void 0 : _context$e.stopPropagation();
    }

    (_props$onTagChange = props.onTagChange) === null || _props$onTagChange === void 0 ? void 0 : _props$onTagChange.call(props, val, context);
  };

  var renderSelectMultiple = function renderSelectMultiple(p) {
    var _props$tagInputProps;

    return /* @__PURE__ */React__default["default"].createElement(tagInput_index.TagInput, _objectSpread(_objectSpread(_objectSpread({
      ref: tagInputRef
    }, p.commonInputProps), {}, {
      autoWidth: props.autoWidth,
      readonly: props.readonly,
      minCollapsedNum: props.minCollapsedNum,
      collapsedItems: props.collapsedItems,
      tag: props.tag,
      valueDisplay: props.valueDisplay,
      placeholder: tPlaceholder,
      value: tags,
      inputValue: tInputValue || "",
      onChange: onTagInputChange,
      onInputChange: function onInputChange(val, context) {
        if ((context === null || context === void 0 ? void 0 : context.trigger) === "enter") return;
        setTInputValue(val, {
          trigger: context.trigger,
          e: context.e
        });
      },
      tagProps: props.tagProps,
      onClear: p.onInnerClear,
      onBlur: function onBlur(val, context) {
        var _props$onBlur;

        (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 ? void 0 : _props$onBlur.call(props, props.value, _objectSpread(_objectSpread({}, context), {}, {
          tagInputValue: val
        }));
      },
      onFocus: function onFocus(val, context) {
        var _props$onFocus;

        (_props$onFocus = props.onFocus) === null || _props$onFocus === void 0 ? void 0 : _props$onFocus.call(props, props.value, _objectSpread(_objectSpread({}, context), {}, {
          tagInputValue: val
        }));
      }
    }, props.tagInputProps), {}, {
      inputProps: {
        readonly: !props.allowInput || props.readonly,
        inputClass: classNames__default["default"]((_props$tagInputProps = props.tagInputProps) === null || _props$tagInputProps === void 0 ? void 0 : _props$tagInputProps.className, defineProperty._defineProperty({}, "".concat(classPrefix, "-input--focused"), p.popupVisible))
      }
    }));
  };

  return {
    tags: tags,
    tPlaceholder: tPlaceholder,
    tagInputRef: tagInputRef,
    renderSelectMultiple: renderSelectMultiple
  };
}

exports["default"] = useMultiple;
