// tdesign-react/cjs/pagination/hooks/useBoundaryJumper.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-4eb93341.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var hooks_useConfig = require('../../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../../hooks/useGlobalIcon.js');
require('../../_chunks/dep-4344eec8.js');
require('../../_chunks/dep-fe984d70.js');
require('../../_chunks/dep-46cf36fd.js');
require('../../_chunks/dep-f5e18a2f.js');
require('../../_chunks/dep-53ba6729.js');
require('../../_chunks/dep-61a8a7b0.js');
require('../../_chunks/dep-2a90f794.js');
require('../../_chunks/dep-aab723b3.js');
require('../../_chunks/dep-384b291c.js');
require('../../_chunks/dep-a0b5d8f6.js');
require('../../_chunks/dep-12656997.js');
require('../../_chunks/dep-1fff9729.js');
require('../../_chunks/dep-bb60493d.js');
require('../../_chunks/dep-72020528.js');
require('../../_chunks/dep-ad854ba5.js');
require('../../_chunks/dep-eebdbd74.js');
require('../../_chunks/dep-8f18a7c4.js');
require('../../_chunks/dep-9cd0fde8.js');
require('../../_chunks/dep-d3ad6e52.js');
require('../../_chunks/dep-aafeb50a.js');
require('../../_chunks/dep-6fa7a9e9.js');
require('../../_chunks/dep-994ec160.js');
require('../../_chunks/dep-9429a38a.js');
require('../../_common/js/global-config/locale/zh_CN.js');
require('../../_common/js/global-config/default-config.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function useBoundaryJumper(props) {
  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
    PageLastIcon: tdesignIconsReact.PageLastIcon,
    PageFirstIcon: tdesignIconsReact.PageFirstIcon
  }),
      PageFirstIcon = _useGlobalIcon.PageFirstIcon,
      PageLastIcon = _useGlobalIcon.PageLastIcon;

  var name = "".concat(classPrefix, "-pagination");
  var showFirstAndLastPageBtn = props.showFirstAndLastPageBtn,
      disabled = props.disabled,
      current = props.current,
      pageCount = props.pageCount,
      changeCurrent = props.changeCurrent;
  var firstPageJumper = showFirstAndLastPageBtn && /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(name, "__btn"), "".concat(name, "__btn-first-page"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-disabled"), disabled || current === 1)),
    onClick: function onClick() {
      return changeCurrent(1);
    }
  }, /* @__PURE__ */React__default["default"].createElement(PageFirstIcon, null));
  var lastPageJumper = showFirstAndLastPageBtn && /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(name, "__btn"), "".concat(name, "__btn-first-page"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-disabled"), disabled || current === pageCount)),
    onClick: function onClick() {
      return changeCurrent(pageCount);
    }
  }, /* @__PURE__ */React__default["default"].createElement(PageLastIcon, null));
  return {
    firstPageJumper: firstPageJumper,
    lastPageJumper: lastPageJumper
  };
}

exports["default"] = useBoundaryJumper;
