// tdesign-react/lib/layout/Layout.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _typeof } from '../_chunks/dep-4e0da095.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _objectWithoutProperties } from '../_chunks/dep-8368bb87.js';
import React, { useState, useEffect } from 'react';
import classNames from 'classnames';
import useConfig from '../hooks/useConfig.js';
import Aside from './Aside.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';

var _excluded = ["className", "style", "children", "height"],
    _excluded2 = ["className", "style", "children", "height"],
    _excluded3 = ["className", "style", "children"],
    _excluded4 = ["direction", "className", "style", "children"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var Header = function Header(props) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var className = props.className,
      _props$style = props.style,
      style = _props$style === void 0 ? {} : _props$style,
      children = props.children,
      height = props.height,
      others = _objectWithoutProperties(props, _excluded);

  var renderHeight = isNaN(Number(height)) ? height : "".concat(height, "px");
  var headerClassNames = classNames("".concat(classPrefix, "-layout__header"), className);
  return /* @__PURE__ */React.createElement("header", _objectSpread({
    className: headerClassNames,
    style: _objectSpread({
      height: renderHeight
    }, style)
  }, others), children);
};

var Footer = function Footer(props) {
  var _useConfig2 = useConfig(),
      classPrefix = _useConfig2.classPrefix;

  var className = props.className,
      _props$style2 = props.style,
      style = _props$style2 === void 0 ? {} : _props$style2,
      children = props.children,
      height = props.height,
      others = _objectWithoutProperties(props, _excluded2);

  var renderHeight = isNaN(Number(height)) ? height : "".concat(height, "px");
  var footerClassNames = classNames("".concat(classPrefix, "-layout__footer"), className);
  return /* @__PURE__ */React.createElement("footer", _objectSpread({
    className: footerClassNames,
    style: _objectSpread({
      height: renderHeight
    }, style)
  }, others), children);
};

var Content = function Content(props) {
  var _useConfig3 = useConfig(),
      classPrefix = _useConfig3.classPrefix;

  var className = props.className,
      style = props.style,
      children = props.children,
      others = _objectWithoutProperties(props, _excluded3);

  var contentClassNames = classNames("".concat(classPrefix, "-layout__content"), className);
  return /* @__PURE__ */React.createElement("main", _objectSpread({
    className: contentClassNames,
    style: style
  }, others), children);
};

var Layout = function Layout(props) {
  var _classNames;

  var direction = props.direction,
      className = props.className,
      style = props.style,
      children = props.children,
      otherLayoutProps = _objectWithoutProperties(props, _excluded4);

  var _useState = useState([]),
      _useState2 = _slicedToArray(_useState, 2),
      asides = _useState2[0],
      setAsides = _useState2[1];

  useEffect(function () {
    React.Children.forEach(children, function (child) {
      if (!child || _typeof(child) !== "object") return;
      if (child.type === Aside) setAsides([child]);
    });
  }, [children]);

  var _useConfig4 = useConfig(),
      classPrefix = _useConfig4.classPrefix;

  var layoutClassNames = classNames("".concat(classPrefix, "-layout"), (_classNames = {}, _defineProperty(_classNames, "".concat(classPrefix, "-layout--with-sider"), !!asides.length), _defineProperty(_classNames, "".concat(classPrefix, "-layout__direction-").concat(direction), direction), _classNames), className);
  return /* @__PURE__ */React.createElement("div", _objectSpread({
    className: layoutClassNames,
    style: style
  }, otherLayoutProps), children);
};

Layout.Header = Header;
Layout.Content = Content;
Layout.Footer = Footer;
Layout.Aside = Aside;
Header.displayName = "Header";
Content.displayName = "Content";
Footer.displayName = "Footer";
Layout.displayName = "Layout";

export { Layout as default };
