// tdesign-react/cjs/slider/SliderHandleButton.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var classNames = require('classnames');
var tooltip_Tooltip = require('../tooltip/Tooltip.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-1a0462cc.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../hooks/useControlled.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-09805e35.js');
require('../_chunks/dep-1fff9729.js');
require('../_util/noop.js');
require('../_util/useAnimation.js');
require('../hooks/useConfig.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../common/Portal.js');
require('react-dom');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../popup/utils/ref.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../_util/useMutationObserver.js');
require('../_chunks/dep-db2b3323.js');
require('../_chunks/dep-52c5813c.js');
require('../_util/useWindowSize.js');
require('../popup/defaultProps.js');
require('../tooltip/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var SliderHandleButton = function SliderHandleButton(_ref) {
  var onChange = _ref.onChange,
      style = _ref.style,
      classPrefix = _ref.classPrefix,
      toolTipProps = _ref.toolTipProps,
      hideTips = _ref.hideTips;

  var _useState = React.useState(false),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      popupVisible = _useState2[0],
      setPopupVisible = _useState2[1];

  var _useState3 = React.useState(false),
      _useState4 = slicedToArray._slicedToArray(_useState3, 2),
      isDragging = _useState4[0],
      toggleIsDragging = _useState4[1];

  var onSliderDragging = function onSliderDragging(e) {
    toggleIsDragging(true);
    onChange(e);
  };

  var onSliderDraggingEnd = function onSliderDraggingEnd() {
    toggleIsDragging(false);
    window.removeEventListener("mousemove", onSliderDragging);
    window.removeEventListener("mouseup", onSliderDraggingEnd);
    window.removeEventListener("touchmove", onSliderDragging);
    window.removeEventListener("touchend", onSliderDraggingEnd);
  };

  var handleSliderMouseDown = function handleSliderMouseDown(e) {
    e.stopPropagation();
    setPopupVisible(true);
    window.addEventListener("mousemove", onSliderDragging);
    window.addEventListener("mouseup", onSliderDraggingEnd);
    window.addEventListener("touchmove", onSliderDragging);
    window.addEventListener("touchend", onSliderDraggingEnd);
  };

  var handleSliderEnter = function handleSliderEnter(event) {
    event.stopPropagation();
    setPopupVisible(true);
  };

  var handleSliderLeave = function handleSliderLeave(event) {
    event.stopPropagation();
    setPopupVisible(false);
  };

  var handleNode = /* @__PURE__ */React__default["default"].createElement("div", {
    style: style,
    className: "".concat(classPrefix, "-slider__button-wrapper"),
    onMouseDown: function onMouseDown(e) {
      return handleSliderMouseDown(e);
    },
    onMouseEnter: function onMouseEnter(e) {
      return handleSliderEnter(e);
    },
    onMouseLeave: function onMouseLeave(e) {
      return handleSliderLeave(e);
    }
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-slider__button"), defineProperty._defineProperty({}, "".concat(classPrefix, "-slider__button--dragging"), isDragging))
  }));
  return hideTips ? handleNode : /* @__PURE__ */React__default["default"].createElement(tooltip_Tooltip["default"], _objectSpread({
    visible: popupVisible,
    placement: "top"
  }, toolTipProps), handleNode);
};

exports["default"] = SliderHandleButton;
