// tdesign-react/cjs/select/base/Option.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-4eb93341.js');
var React = require('react');
var classNames = require('classnames');
var isNumber = require('../../_chunks/dep-364d4c00.js');
var isString = require('../../_chunks/dep-9a2dbbb1.js');
var get = require('../../_chunks/dep-65cb34d8.js');
var hooks_useConfig = require('../../hooks/useConfig.js');
var _util_useRipple = require('../../_util/useRipple.js');
require('../../_chunks/dep-61a8a7b0.js');
require('../../_chunks/dep-2a90f794.js');
require('../../_chunks/dep-1fff9729.js');
require('../../_chunks/dep-8f18a7c4.js');
require('../../_chunks/dep-8329ea23.js');
require('../../_chunks/dep-09805e35.js');
require('../../_chunks/dep-46cf36fd.js');
require('../../_chunks/dep-f5e18a2f.js');
require('../../_chunks/dep-53ba6729.js');
require('../../_chunks/dep-aab723b3.js');
require('../../_chunks/dep-384b291c.js');
require('../../_chunks/dep-afce8659.js');
require('../../_chunks/dep-4344eec8.js');
require('../../_chunks/dep-fe984d70.js');
require('../../_chunks/dep-a0b5d8f6.js');
require('../../_chunks/dep-12656997.js');
require('../../_chunks/dep-bb60493d.js');
require('../../_chunks/dep-72020528.js');
require('../../_chunks/dep-ad854ba5.js');
require('../../_chunks/dep-eebdbd74.js');
require('../../_chunks/dep-9cd0fde8.js');
require('../../_chunks/dep-d3ad6e52.js');
require('../../_chunks/dep-aafeb50a.js');
require('../../_chunks/dep-6fa7a9e9.js');
require('../../_chunks/dep-994ec160.js');
require('../../_chunks/dep-9429a38a.js');
require('../../_common/js/global-config/locale/zh_CN.js');
require('../../_common/js/global-config/default-config.js');
require('../../_util/useAnimation.js');
require('../../_util/setStyle.js');
require('../../_util/dom.js');
require('raf');
require('../../_util/easing.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var componentType = "select";

var Option = function Option(props) {
  var _classNames2;

  var propDisabled = props.disabled,
      propLabel = props.label,
      selectedValue = props.selectedValue,
      multiple = props.multiple,
      size = props.size,
      max = props.max,
      keys = props.keys,
      value = props.value,
      onSelect = props.onSelect,
      children = props.children,
      content = props.content,
      restData = props.restData,
      style = props.style,
      className = props.className;
  var selected;
  var label = propLabel || value;
  var disabled = propDisabled || multiple && Array.isArray(selectedValue) && max && selectedValue.length >= max;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var optionRef = React.useRef();
  _util_useRipple["default"](optionRef);

  if (!multiple) {
    selected = isNumber.isNumber_1(selectedValue) || isString.isString_1(selectedValue) ? value === selectedValue : value === get.get_1(selectedValue, (keys === null || keys === void 0 ? void 0 : keys.value) || "value");
  }

  if (multiple && Array.isArray(selectedValue)) {
    selected = selectedValue.some(function (item) {
      if (isNumber.isNumber_1(item) || isString.isString_1(item)) {
        return item === value;
      }

      return get.get_1(item, (keys === null || keys === void 0 ? void 0 : keys.value) || "value") === value;
    });
  }

  var handleSelect = function handleSelect(event) {
    if (!disabled) {
      onSelect(value, {
        label: String(label),
        selected: selected,
        event: event,
        restData: restData
      });
    }
  };

  var renderItem = function renderItem(children2) {
    if (multiple) {
      var _classNames;

      return /* @__PURE__ */React__default["default"].createElement("label", {
        className: classNames__default["default"]("".concat(classPrefix, "-checkbox"), (_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-is-disabled"), disabled), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-is-checked"), selected), _classNames))
      }, /* @__PURE__ */React__default["default"].createElement("input", {
        type: "checkbox",
        className: classNames__default["default"]("".concat(classPrefix, "-checkbox__former")),
        value: "",
        disabled: disabled && !selected,
        onClick: function onClick(e) {
          return e.stopPropagation();
        }
      }), /* @__PURE__ */React__default["default"].createElement("span", {
        className: classNames__default["default"]("".concat(classPrefix, "-checkbox__input"))
      }), /* @__PURE__ */React__default["default"].createElement("span", {
        className: classNames__default["default"]("".concat(classPrefix, "-checkbox__label"))
      }, children2 || content || label));
    }

    return /* @__PURE__ */React__default["default"].createElement("span", null, children2 || content || label);
  };

  return /* @__PURE__ */React__default["default"].createElement("li", {
    className: classNames__default["default"](className, "".concat(classPrefix, "-").concat(componentType, "-option"), (_classNames2 = {}, defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-is-disabled"), disabled), defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-is-selected"), selected), defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-size-s"), size === "small"), defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-size-l"), size === "large"), _classNames2)),
    key: value,
    onClick: handleSelect,
    ref: optionRef,
    style: style
  }, renderItem(children));
};

exports["default"] = Option;
