// tdesign-react/lib/_util/useMutationObserver.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { useRef, useEffect } from 'react';
import { d as debounce_1 } from '../_chunks/dep-05d10538.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-652c1652.js';
import '../_chunks/dep-b989427d.js';
import '../_chunks/dep-7f6754ac.js';

var DEFAULT_OPTIONS = {
  debounceTime: 0,
  config: {
    attributes: true,
    childList: true,
    characterData: true,
    subtree: true
  }
};
function useMutationObservable(targetEl, cb) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : DEFAULT_OPTIONS;
  var observeRef = useRef(null);
  useEffect(function () {
    if (!cb || typeof cb !== "function") return;
    var debounceTime = options.debounceTime;
    observeRef.current = new MutationObserver(debounceTime > 0 ? debounce_1(cb, debounceTime) : cb);
  }, [cb, options]);
  useEffect(function () {
    if (!targetEl || !(targetEl !== null && targetEl !== void 0 && targetEl.nodeType)) return;
    var config = options.config;

    try {
      observeRef.current.observe(targetEl, config);
    } catch (e) {
      console.error(e);
    }

    return function () {
      observeRef.current.disconnect();
    };
  }, [targetEl, options]);
}

export { useMutationObservable as default };
