// tdesign-react/lib/_util/useRipple.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { useMemo, useCallback, useEffect } from 'react';
import useConfig from '../hooks/useConfig.js';
import useAnimation from './useAnimation.js';
import setStyle from './setStyle.js';
import { canUseDocument } from './dom.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-1e32006c.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import 'raf';
import '../_chunks/dep-8d5556d7.js';
import './easing.js';

var period = 200;
var noneRippleBg = "rgba(0, 0, 0, 0)";
var defaultRippleColor = "rgba(0, 0, 0, 0.35)";

var getRippleColor = function getRippleColor(el, fixedRippleColor) {
  var _el$dataset;

  if (fixedRippleColor) {
    return fixedRippleColor;
  }

  if (el !== null && el !== void 0 && (_el$dataset = el.dataset) !== null && _el$dataset !== void 0 && _el$dataset.ripple) {
    var rippleColor = el.dataset.ripple;
    return rippleColor;
  }

  var cssVariable = getComputedStyle(el).getPropertyValue("--ripple-color");

  if (cssVariable) {
    return cssVariable;
  }

  return defaultRippleColor;
};

function useRipple(ref, fixedRippleColor) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useAnimation = useAnimation(),
      keepRipple = _useAnimation.keepRipple;

  var rippleContainer = useMemo(function () {
    if (!canUseDocument) return null;
    var container = document.createElement("div");
    container.className = "".concat(classPrefix, "-ripple");
    return container;
  }, [classPrefix]);
  var handleAddRipple = useCallback(function (e) {
    var el = ref === null || ref === void 0 ? void 0 : ref.current;
    var rippleColor = getRippleColor(el, fixedRippleColor);
    if (e.button !== 0 || !el || !keepRipple) return;
    if (el.classList.contains("".concat(classPrefix, "-is-active")) || el.classList.contains("".concat(classPrefix, "-is-disabled")) || el.classList.contains("".concat(classPrefix, "-is-checked")) || el.classList.contains("".concat(classPrefix, "-is-loading"))) return;
    var elStyle = getComputedStyle(el);
    var elBorder = parseInt(elStyle.borderWidth, 10);
    var border = elBorder > 0 ? elBorder : 0;
    var width = el.offsetWidth;
    var height = el.offsetHeight;

    if (rippleContainer.parentNode === null) {
      setStyle(rippleContainer, {
        position: "absolute",
        left: "".concat(0 - border, "px"),
        top: "".concat(0 - border, "px"),
        width: "".concat(width, "px"),
        height: "".concat(height, "px"),
        borderRadius: elStyle.borderRadius,
        pointerEvents: "none",
        overflow: "hidden"
      });
      el.appendChild(rippleContainer);
    }

    var ripple = document.createElement("div");
    ripple.className = "".concat(classPrefix, "-ripple__inner");
    setStyle(ripple, {
      marginTop: "0",
      marginLeft: "0",
      right: "".concat(width, "px"),
      width: "".concat(width + 20, "px"),
      height: "100%",
      transition: "transform ".concat(period, "ms cubic-bezier(.38, 0, .24, 1), background ").concat(period * 2, "ms linear"),
      transform: "skewX(-8deg)",
      pointerEvents: "none",
      position: "absolute",
      zIndex: 0,
      backgroundColor: rippleColor,
      opacity: "0.9"
    });
    var elMap = /* @__PURE__ */new WeakMap();

    for (var n = el.children.length, i = 0; i < n; ++i) {
      var child = el.children[i];

      if (child.style.zIndex === "" && child !== rippleContainer) {
        child.style.zIndex = "1";
        elMap.set(child, true);
      }
    }

    var initPosition = el.style.position ? el.style.position : getComputedStyle(el).position;

    if (initPosition === "" || initPosition === "static") {
      el.style.position = "relative";
    }

    rippleContainer.insertBefore(ripple, rippleContainer.firstChild);
    setTimeout(function () {
      ripple.style.transform = "translateX(".concat(width, "px)");
    }, 0);

    var handleClearRipple = function handleClearRipple() {
      var el2 = ref === null || ref === void 0 ? void 0 : ref.current;
      ripple.style.backgroundColor = noneRippleBg;
      if (!el2) return;
      el2.removeEventListener("pointerup", handleClearRipple, false);
      el2.removeEventListener("pointerleave", handleClearRipple, false);
      setTimeout(function () {
        ripple.remove();
        if (rippleContainer.children.length === 0) rippleContainer.remove();
      }, period * 2 + 100);
    };

    el.addEventListener("pointerup", handleClearRipple, false);
    el.addEventListener("pointerleave", handleClearRipple, false);
  }, [classPrefix, ref, fixedRippleColor, rippleContainer, keepRipple]);
  useEffect(function () {
    var el = ref === null || ref === void 0 ? void 0 : ref.current;
    if (!el) return;
    el.addEventListener("pointerdown", handleAddRipple, false);
    return function () {
      el.removeEventListener("pointerdown", handleAddRipple, false);
    };
  }, [handleAddRipple, fixedRippleColor, ref]);
}

export { useRipple as default };
