// tdesign-react/es/tooltip/Tooltip.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import { _ as _objectWithoutProperties } from '../_chunks/dep-8368bb87.js';
import React, { forwardRef, useState, useRef, cloneElement, useEffect, useImperativeHandle, isValidElement } from 'react';
import classNames from 'classnames';
import { Popup } from '../popup/index.js';
import useConfig from '../hooks/useConfig.js';
import { tooltipDefaultProps } from './defaultProps.js';
import '../_chunks/dep-7909fefa.js';
import '../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../hooks/useControlled.js';
import '../_chunks/dep-f54ff1e5.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-e1d4dd61.js';
import '../_chunks/dep-236ef238.js';
import '../_chunks/dep-c8bfa11e.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-199b1581.js';
import '../_chunks/dep-e237f0f2.js';
import '../_util/noop.js';
import '../_util/useAnimation.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-e54bb0dc.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-ebf3a272.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../common/Portal.js';
import 'react-dom';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-3c98169f.js';
import '../_util/easing.js';
import '../popup/hooks/useTrigger.js';
import 'react-is';
import '../popup/utils/ref.js';
import '../_util/composeRefs.js';
import '../popup/utils/transition.js';
import '../_util/useMutationObserver.js';
import '../_chunks/dep-236b0359.js';
import '../_chunks/dep-6d5c0a90.js';
import '../_util/useWindowSize.js';
import '../popup/defaultProps.js';
import '../popup/style/css.js';
import '../popup/type.js';

var _excluded = ["theme", "showArrow", "destroyOnClose", "overlayClassName", "children", "duration", "placement"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Tooltip = /*#__PURE__*/forwardRef(function (props, ref) {
  var theme = props.theme,
      _props$showArrow = props.showArrow,
      showArrow = _props$showArrow === void 0 ? true : _props$showArrow,
      _props$destroyOnClose = props.destroyOnClose,
      destroyOnClose = _props$destroyOnClose === void 0 ? true : _props$destroyOnClose,
      overlayClassName = props.overlayClassName,
      children = props.children,
      _props$duration = props.duration,
      duration = _props$duration === void 0 ? 0 : _props$duration,
      _props$placement = props.placement,
      placement = _props$placement === void 0 ? "top" : _props$placement,
      restProps = _objectWithoutProperties(props, _excluded);

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useState = useState(duration !== 0),
      _useState2 = _slicedToArray(_useState, 2),
      isTipShowed = _useState2[0],
      setTipshow = _useState2[1];

  var _useState3 = useState(false),
      _useState4 = _slicedToArray(_useState3, 2),
      timeup = _useState4[0],
      setTimeup = _useState4[1];

  var popupRef = useRef();
  var timerRef = useRef(null);

  var _useState5 = useState([0, 0]),
      _useState6 = _slicedToArray(_useState5, 2),
      offset = _useState6[0],
      setOffset = _useState6[1];

  var toolTipClass = classNames("".concat(classPrefix, "-tooltip"), _defineProperty({}, "".concat(classPrefix, "-tooltip--").concat(theme), theme), overlayClassName);
  var isPlacedByMouse = placement === "mouse";

  var setVisible = function setVisible(v) {
    if (duration !== 0) setTimeup(false);
    setTipshow(v);
  };

  var calculatePos = function calculatePos(e) {
    var rect = e.target.getBoundingClientRect();
    var x = e.clientX - rect.left;
    var y = e.clientY - rect.top;
    return {
      x: x,
      y: y
    };
  };

  var getTriggerChildren = function getTriggerChildren(children2) {
    var _children2$type, _children2$props;

    var displayName = (_children2$type = children2.type) === null || _children2$type === void 0 ? void 0 : _children2$type.displayName;

    if ((children2.type === "button" || displayName === "Button") && children2 !== null && children2 !== void 0 && (_children2$props = children2.props) !== null && _children2$props !== void 0 && _children2$props.disabled) {
      var _children2$props2, _children2$props2$sty;

      var displayStyle = (_children2$props2 = children2.props) !== null && _children2$props2 !== void 0 && (_children2$props2$sty = _children2$props2.style) !== null && _children2$props2$sty !== void 0 && _children2$props2$sty.display ? children2.props.style.display : "inline-block";
      var child = /*#__PURE__*/cloneElement(children2, {
        style: _objectSpread(_objectSpread({}, children2.props.style), {}, {
          pointerEvents: "none"
        })
      });
      return /* @__PURE__ */React.createElement("span", {
        style: {
          display: displayStyle,
          cursor: "not-allowed"
        }
      }, child);
    }

    return children2;
  };

  var handleShowTip = function handleShowTip(visible, _ref) {
    var e = _ref.e,
        trigger = _ref.trigger;

    if (duration === 0 || duration !== 0 && timeup) {
      if (visible && placement === "mouse" && (trigger === "trigger-element-hover" || trigger === "trigger-element-click")) {
        var _calculatePos = calculatePos(e),
            x = _calculatePos.x;

        setOffset([x, 0]);
      }

      setTipshow(visible);
    }
  };

  useEffect(function () {
    if (duration !== 0 && !timeup) {
      timerRef.current = window.setTimeout(function () {
        setTipshow(false);
        setTimeup(true);
      }, duration);
    }

    return function () {
      if (timerRef.current) window.clearTimeout(timerRef.current);
    };
  }, [duration, timeup]);
  useImperativeHandle(ref, function () {
    return _objectSpread({
      setVisible: setVisible
    }, popupRef.current || {});
  });
  return /* @__PURE__ */React.createElement(Popup, _objectSpread({
    ref: popupRef,
    destroyOnClose: destroyOnClose,
    showArrow: isPlacedByMouse ? false : showArrow,
    overlayClassName: toolTipClass,
    visible: isTipShowed,
    onVisibleChange: handleShowTip,
    popperOptions: {
      modifier: isPlacedByMouse ? [{
        name: "offset",
        options: {
          offset: offset
        }
      }] : []
    },
    placement: isPlacedByMouse ? "bottom-left" : placement
  }, restProps), /*#__PURE__*/isValidElement(children) ? getTriggerChildren(children) : children);
});
Tooltip.displayName = "Tooltip";
Tooltip.defaultProps = tooltipDefaultProps;

export { Tooltip as default };
