// tdesign-react/lib/pagination/hooks/usePrevNextJumper.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import React from 'react';
import classNames from 'classnames';
import { ChevronLeftIcon, ChevronRightIcon } from 'tdesign-icons-react';
import useConfig from '../../hooks/useConfig.js';
import useGlobalIcon from '../../hooks/useGlobalIcon.js';
import '../../_chunks/dep-918e41e2.js';
import '../../_chunks/dep-734a21a0.js';
import '../../_chunks/dep-68f7a306.js';
import '../../_chunks/dep-a74f6fff.js';
import '../../_chunks/dep-d23d346c.js';
import '../../_chunks/dep-9ea787ca.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-0e2dfd87.js';
import '../../_chunks/dep-f50ff346.js';
import '../../_chunks/dep-51592cde.js';
import '../../_chunks/dep-ae94c4e7.js';
import '../../_chunks/dep-7f6754ac.js';
import '../../_chunks/dep-2509667d.js';
import '../../_chunks/dep-b817c90e.js';
import '../../_chunks/dep-fe9825b6.js';
import '../../_chunks/dep-ebc05da7.js';
import '../../_chunks/dep-3160df29.js';
import '../../_chunks/dep-c91b3852.js';
import '../../_chunks/dep-9e44c34c.js';
import '../../_chunks/dep-0c62d47a.js';
import '../../_chunks/dep-e2fe68b7.js';
import '../../_chunks/dep-44a03722.js';
import '../../_chunks/dep-f5616354.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';

function usePrevNextJumper(props) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useGlobalIcon = useGlobalIcon({
    ChevronLeftIcon: ChevronLeftIcon,
    ChevronRightIcon: ChevronRightIcon
  }),
      ChevronLeftIcon$1 = _useGlobalIcon.ChevronLeftIcon,
      ChevronRightIcon$1 = _useGlobalIcon.ChevronRightIcon;

  var name = "".concat(classPrefix, "-pagination");
  var showPreviousAndNextBtn = props.showPreviousAndNextBtn,
      disabled = props.disabled,
      current = props.current,
      pageCount = props.pageCount,
      changeCurrent = props.changeCurrent;
  var prevJumper = showPreviousAndNextBtn && /* @__PURE__ */React.createElement("div", {
    className: classNames("".concat(name, "__btn"), "".concat(name, "__btn-prev"), _defineProperty({}, "".concat(classPrefix, "-is-disabled"), disabled || current === 1)),
    onClick: function onClick() {
      if (disabled || current === 1) return;
      changeCurrent(current - 1);
    }
  }, /* @__PURE__ */React.createElement(ChevronLeftIcon$1, null));
  var nextJumper = showPreviousAndNextBtn && /* @__PURE__ */React.createElement("div", {
    className: classNames("".concat(name, "__btn"), "".concat(name, "__btn-next"), _defineProperty({}, "".concat(classPrefix, "-is-disabled"), disabled || current === pageCount)),
    onClick: function onClick() {
      if (disabled || current === pageCount) return;
      changeCurrent(current + 1);
    }
  }, /* @__PURE__ */React.createElement(ChevronRightIcon$1, null));
  return {
    prevJumper: prevJumper,
    nextJumper: nextJumper
  };
}

export { usePrevNextJumper as default };
