// tdesign-react/es/popconfirm/Popcontent.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import React from 'react';
import { i as isString_1 } from '../_chunks/dep-3c98169f.js';
import classNames from 'classnames';
import { InfoCircleFilledIcon } from 'tdesign-icons-react';
import { Button } from '../button/index.js';
import noop from '../_util/noop.js';
import useConfig from '../hooks/useConfig.js';
import useGlobalIcon from '../hooks/useGlobalIcon.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-e237f0f2.js';
import '../button/Button.js';
import '../_chunks/dep-8368bb87.js';
import '../_util/useRipple.js';
import '../_util/useAnimation.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-e54bb0dc.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-ebf3a272.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../_util/setStyle.js';
import '../_util/dom.js';
import 'raf';
import '../_util/easing.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../_chunks/dep-9615f03f.js';
import '../_chunks/dep-7909fefa.js';
import '../common/Portal.js';
import 'react-dom';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-d1b4ab24.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import '../loading/style/css.js';
import '../loading/type.js';
import '../button/defaultProps.js';
import '../button/style/css.js';
import '../button/type.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var Popcontent = function Popcontent(props) {
  var content = props.content,
      cancelBtn = props.cancelBtn,
      confirmBtn = props.confirmBtn,
      icon = props.icon,
      theme = props.theme,
      _props$onCancel = props.onCancel,
      onCancel = _props$onCancel === void 0 ? noop : _props$onCancel,
      _props$onConfirm = props.onConfirm,
      onConfirm = _props$onConfirm === void 0 ? noop : _props$onConfirm,
      _props$onClose = props.onClose,
      onClose = _props$onClose === void 0 ? noop : _props$onClose;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useGlobalIcon = useGlobalIcon({
    InfoCircleFilledIcon: InfoCircleFilledIcon
  }),
      InfoCircleFilledIcon$1 = _useGlobalIcon.InfoCircleFilledIcon;

  var hideCancel = cancelBtn === null || cancelBtn === void 0;
  var hideConfirm = confirmBtn === null || confirmBtn === void 0;

  function renderIcon() {
    var color = "#0052D9";
    var defaultIcon = /* @__PURE__ */React.createElement(InfoCircleFilledIcon$1, null);

    switch (theme) {
      case "warning":
        color = "#FFAA00";
        break;

      case "danger":
        color = "#FF3E00";
        break;
    }

    var iconComponent = null;

    if ( /*#__PURE__*/React.isValidElement(icon)) {
      iconComponent = /*#__PURE__*/React.cloneElement(icon, _objectSpread({
        style: {
          color: color
        }
      }, icon.props));
    } else if (typeof icon === "function") {
      iconComponent = icon();
    } else if (defaultIcon) {
      iconComponent = /*#__PURE__*/React.cloneElement(defaultIcon, {
        style: {
          color: color
        }
      });
    }

    return iconComponent;
  }

  function renderCancel() {
    if ( /*#__PURE__*/React.isValidElement(cancelBtn)) {
      return /*#__PURE__*/React.cloneElement(cancelBtn, {
        onClick: function onClick(e) {
          var _cancelBtn$props, _cancelBtn$props$onCl;

          onClose({
            e: e,
            trigger: "cancel"
          });
          (_cancelBtn$props = cancelBtn.props) === null || _cancelBtn$props === void 0 ? void 0 : (_cancelBtn$props$onCl = _cancelBtn$props.onClick) === null || _cancelBtn$props$onCl === void 0 ? void 0 : _cancelBtn$props$onCl.call(_cancelBtn$props, e);
        }
      });
    }

    if (hideCancel) {
      return null;
    }

    return /* @__PURE__ */React.createElement(Button, {
      size: "small",
      theme: "default",
      variant: "base",
      onClick: function onClick(e) {
        onClose({
          e: e,
          trigger: "cancel"
        });
        onCancel({
          e: e
        });
      }
    }, isString_1(cancelBtn) && cancelBtn);
  }

  function renderConfirm() {
    if ( /*#__PURE__*/React.isValidElement(confirmBtn)) {
      return /*#__PURE__*/React.cloneElement(confirmBtn, {
        onClick: function onClick(e) {
          var _confirmBtn$props, _confirmBtn$props$onC;

          onClose({
            e: e,
            trigger: "confirm"
          });
          (_confirmBtn$props = confirmBtn.props) === null || _confirmBtn$props === void 0 ? void 0 : (_confirmBtn$props$onC = _confirmBtn$props.onClick) === null || _confirmBtn$props$onC === void 0 ? void 0 : _confirmBtn$props$onC.call(_confirmBtn$props, e);
        }
      });
    }

    if (hideConfirm) {
      return null;
    }

    return /* @__PURE__ */React.createElement(Button, {
      size: "small",
      theme: "primary",
      onClick: function onClick(e) {
        onClose({
          e: e,
          trigger: "confirm"
        });
        onConfirm({
          e: e
        });
      }
    }, isString_1(confirmBtn) && confirmBtn);
  }

  return /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-popconfirm__content")
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-popconfirm__body")
  }, renderIcon(), /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-popconfirm__inner")
  }, content)), /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-popconfirm__buttons")
  }, /* @__PURE__ */React.createElement("span", {
    className: classNames("".concat(classPrefix, "-popconfirm__cancel"))
  }, renderCancel()), /* @__PURE__ */React.createElement("span", {
    className: classNames("".concat(classPrefix, "-popconfirm__confirm"))
  }, renderConfirm())));
};

Popcontent.displayName = "Popcontent";

export { Popcontent as default };
