// tdesign-react/cjs/tree/Tree.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var reactTransitionGroup = require('react-transition-group');
var classNames = require('classnames');
var _util_usePersistFn = require('../_util/usePersistFn.js');
var tree_useTreeConfig = require('./useTreeConfig.js');
var tree_useControllable = require('./useControllable.js');
var tree_TreeItem = require('./TreeItem.js');
var tree_useStore = require('./useStore.js');
require('../_chunks/dep-2205decf.js');
require('../hooks/useConfig.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../locale/LocalReceiver.js');
require('../hooks/useControlled.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-09805e35.js');
require('../_util/noop.js');
require('tdesign-icons-react');
require('../loading/index.js');
require('../loading/Loading.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../_chunks/dep-1a0462cc.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');
require('../_util/useRipple.js');
require('../_util/useAnimation.js');
require('../_util/setStyle.js');
require('../hooks/useGlobalIcon.js');
require('../checkbox/index.js');
require('../_chunks/dep-3b82b90b.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../common/Check.js');
require('../_util/helper.js');
require('../_chunks/dep-0ebc045a.js');
require('../_chunks/dep-364d4c00.js');
require('../checkbox/defaultProps.js');
require('../_chunks/dep-3afb7a98.js');
require('../_chunks/dep-4578b559.js');
require('../_chunks/dep-4fb39763.js');
require('../_chunks/dep-07e5d807.js');
require('../_util/useUpdateEffect.js');
require('../_util/useLayoutEffect.js');
require('../_common/js/tree/tree-store.js');
require('../_chunks/dep-c3f475a2.js');
require('../_chunks/dep-f21aff2d.js');
require('../_chunks/dep-aa470c9c.js');
require('../_chunks/dep-1e8de681.js');
require('../_common/js/tree/tree-node.js');
require('../_chunks/dep-266b1890.js');
require('../_chunks/dep-65cb34d8.js');
require('../_chunks/dep-8329ea23.js');
require('../_common/js/tree/tree-node-model.js');
require('../_chunks/dep-b5af0182.js');
require('../_chunks/dep-55c5ba71.js');
require('../_chunks/dep-70fe2bd0.js');
require('../_common/js/log/log.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Tree = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _classNames;

  var _useTreeConfig = tree_useTreeConfig.useTreeConfig(),
      treeClassNames = _useTreeConfig.treeClassNames,
      transitionNames = _useTreeConfig.transitionNames,
      transitionClassNames = _useTreeConfig.transitionClassNames,
      transitionDuration = _useTreeConfig.transitionDuration,
      locale = _useTreeConfig.locale;

  var _useState = React.useState([]),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      visibleNodes = _useState2[0],
      setVisibleNodes = _useState2[1];

  var empty = props.empty,
      activable = props.activable,
      disabled = props.disabled,
      checkable = props.checkable,
      checkProps = props.checkProps,
      disableCheck = props.disableCheck,
      hover = props.hover,
      icon = props.icon,
      line = props.line,
      label = props.label,
      operations = props.operations,
      transition = props.transition,
      expandOnClickNode = props.expandOnClickNode,
      onClick = props.onClick;

  var _useControllable = tree_useControllable["default"](props),
      value = _useControllable.value,
      onChange = _useControllable.onChange,
      expanded = _useControllable.expanded,
      onExpand = _useControllable.onExpand,
      onActive = _useControllable.onActive,
      actived = _useControllable.actived;

  var emptyText = locale("empty");
  var store = tree_useStore.useStore(_objectSpread(_objectSpread({}, props), {}, {
    value: value,
    onChange: onChange,
    expanded: expanded,
    onExpand: onExpand,
    onActive: onActive,
    actived: actived
  }), function () {
    var nodes = store.getNodes();
    var newVisibleNodes = nodes.filter(function (node) {
      return node.visible;
    });
    setVisibleNodes(newVisibleNodes);
  });
  var setExpanded = _util_usePersistFn.usePersistFn(function (node, isExpanded, e) {
    var expanded2 = node.setExpanded(isExpanded);
    var treeNodeModel = node === null || node === void 0 ? void 0 : node.getModel();
    e && (onExpand === null || onExpand === void 0 ? void 0 : onExpand(expanded2, {
      node: treeNodeModel,
      e: e
    }));
    return expanded2;
  });
  var setActived = _util_usePersistFn.usePersistFn(function (node, isActived) {
    var actived2 = node.setActived(isActived);
    var treeNodeModel = node === null || node === void 0 ? void 0 : node.getModel();
    onActive === null || onActive === void 0 ? void 0 : onActive(actived2, {
      node: treeNodeModel
    });
    return actived2;
  });
  var setChecked = _util_usePersistFn.usePersistFn(function (node, isChecked) {
    var checked = node.setChecked(isChecked);
    var treeNodeModel = node === null || node === void 0 ? void 0 : node.getModel();
    onChange === null || onChange === void 0 ? void 0 : onChange(checked, {
      node: treeNodeModel
    });
    return checked;
  });

  var handleItemClick = function handleItemClick(node, options) {
    if (!node) {
      return;
    }

    var isDisabled = disabled || node.disabled;
    var expand = options.expand,
        active = options.active,
        event = options.event;
    if (expand) setExpanded(node, !node.isExpanded(), event);

    if (active && !isDisabled) {
      setActived(node, !node.isActived());
      var treeNodeModel = node === null || node === void 0 ? void 0 : node.getModel();
      onClick === null || onClick === void 0 ? void 0 : onClick({
        node: treeNodeModel,
        e: event
      });
    }
  };

  var handleChange = function handleChange(node) {
    if (!node || disabled || node.disabled) {
      return;
    }

    setChecked(node, !node.isChecked());
  };

  React.useImperativeHandle(ref, function () {
    return {
      store: store,
      appendTo: function appendTo(value2, newData) {
        var list = [];

        if (Array.isArray(newData)) {
          list = newData;
        } else {
          list = [newData];
        }

        list.forEach(function (item) {
          store.appendNodes(value2, item);
        });
      },
      getIndex: function getIndex(value2) {
        return store.getNodeIndex(value2);
      },
      getItem: function getItem(value2) {
        var node = store.getNode(value2);
        return node === null || node === void 0 ? void 0 : node.getModel();
      },
      getItems: function getItems(value2) {
        var nodes = store.getNodes(value2);
        return nodes.map(function (node) {
          return node.getModel();
        });
      },
      getParent: function getParent(value2) {
        var node = store.getParent(value2);
        return node === null || node === void 0 ? void 0 : node.getModel();
      },
      getParents: function getParents(value2) {
        var nodes = store.getParents(value2);
        return nodes.map(function (node) {
          return node.getModel();
        });
      },
      getPath: function getPath(value2) {
        var node = store.getNode(value2);
        var pathNodes = [];

        if (node) {
          pathNodes = node.getPath().map(function (node2) {
            return node2.getModel();
          });
        }

        return pathNodes;
      },
      insertAfter: function insertAfter(value2, newData) {
        return store.insertAfter(value2, newData);
      },
      insertBefore: function insertBefore(value2, newData) {
        return store.insertBefore(value2, newData);
      },
      remove: function remove(value2) {
        return store.remove(value2);
      },
      setItem: function setItem(value2, options) {
        var node = this.store.getNode(value2);
        var spec = options;

        if (node && spec) {
          if ("expanded" in options) {
            setExpanded(node, spec.expanded);
            delete spec.expanded;
          }

          if ("actived" in options) {
            setActived(node, spec.actived);
            delete spec.actived;
          }

          if ("checked" in options) {
            setChecked(node, spec.checked);
            delete spec.checked;
          }

          node.set(spec);
        }
      }
    };
  }, [store, setExpanded, setActived, setChecked]);
  var nodeList = React.useMemo(function () {
    return visibleNodes.map(function () {
      return /*#__PURE__*/React__default["default"].createRef();
    });
  }, [visibleNodes]);

  var renderEmpty = function renderEmpty() {
    var emptyView = empty || emptyText;

    if (empty instanceof Function) {
      emptyView = empty();
    }

    return emptyView;
  };

  var renderItems = function renderItems() {
    if (visibleNodes.length <= 0) {
      return renderEmpty();
    }

    return /* @__PURE__ */React__default["default"].createElement(reactTransitionGroup.TransitionGroup, {
      name: transitionNames.treeNode,
      className: treeClassNames.treeList
    }, visibleNodes.map(function (node, index) {
      return /* @__PURE__ */React__default["default"].createElement(reactTransitionGroup.CSSTransition, {
        nodeRef: nodeList[index],
        key: node.value,
        timeout: transitionDuration,
        classNames: transitionClassNames
      }, /* @__PURE__ */React__default["default"].createElement(tree_TreeItem["default"], {
        ref: nodeList[index],
        node: node,
        empty: empty,
        icon: icon,
        label: label,
        line: line,
        transition: transition,
        expandOnClickNode: expandOnClickNode,
        activable: activable,
        operations: operations,
        checkProps: checkProps,
        disableCheck: disableCheck,
        onClick: handleItemClick,
        onChange: handleChange
      }));
    }));
  };

  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](treeClassNames.tree, (_classNames = {}, defineProperty._defineProperty(_classNames, treeClassNames.disabled, disabled), defineProperty._defineProperty(_classNames, treeClassNames.treeHoverable, hover), defineProperty._defineProperty(_classNames, treeClassNames.treeCheckable, checkable), defineProperty._defineProperty(_classNames, treeClassNames.treeFx, transition), defineProperty._defineProperty(_classNames, treeClassNames.treeBlockNode, expandOnClickNode), _classNames))
  }, renderItems());
});
Tree.displayName = "Tree";
Tree.defaultProps = {
  data: [],
  empty: "",
  expandLevel: 0,
  icon: true,
  line: false,
  transition: true,
  lazy: true,
  valueMode: "onlyLeaf"
};

exports["default"] = Tree;
