// tdesign-react/cjs/tree/useControllable.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var hooks_useControlled = require('../hooks/useControlled.js');
require('../_chunks/dep-2205decf.js');
require('react');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-09805e35.js');
require('../_chunks/dep-1fff9729.js');
require('../_util/noop.js');

function useControllable(props) {
  var _useControlled = hooks_useControlled["default"](props, "value", props.onChange),
      _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
      value = _useControlled2[0],
      onChange = _useControlled2[1];

  var _useControlled3 = hooks_useControlled["default"](props, "expanded", props.onExpand),
      _useControlled4 = slicedToArray._slicedToArray(_useControlled3, 2),
      expanded = _useControlled4[0],
      onExpand = _useControlled4[1];

  var _useControlled5 = hooks_useControlled["default"](props, "actived", props.onActive),
      _useControlled6 = slicedToArray._slicedToArray(_useControlled5, 2),
      actived = _useControlled6[0],
      onActive = _useControlled6[1];

  return {
    value: value,
    onChange: onChange,
    expanded: expanded,
    onExpand: onExpand,
    actived: actived,
    onActive: onActive
  };
}

exports["default"] = useControllable;
