// tdesign-react/cjs/tree/useTreeConfig.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var hooks_useConfig = require('../hooks/useConfig.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-4eb93341.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');

function useTreeConfig() {
  var _useConfig = hooks_useConfig["default"](),
      prefix = _useConfig.classPrefix;

  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("tree"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      _locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  return React.useMemo(function () {
    var tree = "".concat(prefix, "-tree");
    var treeClassNames = {
      icon: "".concat(prefix, "-icon"),
      folderIcon: "".concat(prefix, "-folder-icon"),
      actived: "".concat(prefix, "-is-active"),
      disabled: "".concat(prefix, "-is-disabled"),
      treeIconRight: "".concat(prefix, "-icon-arrow-right"),
      treeIconDown: "".concat(prefix, "-icon-arrow-down"),
      tree: tree,
      treeFx: "".concat(tree, "--transition"),
      treeBlockNode: "".concat(tree, "--block-node"),
      treeEmpty: "".concat(tree, "__empty"),
      treeList: "".concat(tree, "__list"),
      treeNode: "".concat(tree, "__item"),
      treeNodeOpen: "".concat(tree, "__item--open"),
      treeHoverable: "".concat(tree, "--hoverable"),
      treeCheckable: "".concat(tree, "--checkable"),
      treeLabel: "".concat(tree, "__label"),
      treeLabelStrictly: "".concat(tree, "__label--strictly"),
      treeIcon: "".concat(tree, "__icon"),
      treeIconDefault: "".concat(tree, "__icon--default"),
      treeSpace: "".concat(tree, "__space"),
      treeOperations: "".concat(tree, "__operations"),
      line: "".concat(tree, "__line"),
      lineIsLeaf: "".concat(tree, "__line--leaf"),
      lineIsFirst: "".concat(tree, "__line--first"),
      treeNodeShow: "".concat(tree, "__item--show"),
      treeClickable: "".concat(tree, "__item--clickable"),
      loading: "".concat(prefix, "-icon-loading ").concat(prefix, "-icon-loading-blue"),
      toggleEnter: "".concat(tree, "__item--enter-active"),
      toggleLeave: "".concat(tree, "__item--leave-active")
    };
    var transitionNames = {
      treeNode: "".concat(prefix, "-tree-toggle")
    };
    var transitionClassNames = {
      enter: treeClassNames.toggleEnter,
      exit: treeClassNames.toggleLeave
    };
    return {
      treeClassNames: treeClassNames,
      transitionNames: transitionNames,
      transitionClassNames: transitionClassNames,
      transitionDuration: 300,
      locale: function locale(key) {
        return t(_locale[key]);
      }
    };
  }, [_locale, prefix, t]);
}

exports.useTreeConfig = useTreeConfig;
