// tdesign-react/cjs/card/Card.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var React = require('react');
var classNames = require('classnames');
var loading_index = require('../loading/index.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var _util_useCommonClassName = require('../_util/useCommonClassName.js');
var card_defaultProps = require('./defaultProps.js');
require('../loading/Loading.js');
require('../_chunks/dep-9c5d4f93.js');
require('../_chunks/dep-2205decf.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-1fff9729.js');
require('../_util/easing.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../_chunks/dep-1a0462cc.js');
require('../loading/defaultProps.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../loading/plugin.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var Card = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _classNames, _classNames2;

  var actions = props.actions,
      avatar = props.avatar,
      bordered = props.bordered,
      children = props.children,
      className = props.className,
      cover = props.cover,
      description = props.description,
      footer = props.footer,
      header = props.header,
      headerBordered = props.headerBordered,
      hoverShadow = props.hoverShadow,
      loading = props.loading,
      shadow = props.shadow,
      size = props.size,
      style = props.style,
      subtitle = props.subtitle,
      title = props.title,
      theme = props.theme,
      status = props.status;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var commonClassNames = _util_useCommonClassName["default"]();
  var isPoster2 = theme === "poster2";
  var cardClass = classNames__default["default"]("".concat(classPrefix, "-card"), className, (_classNames = {}, defineProperty._defineProperty(_classNames, commonClassNames.SIZE.small, size === "small"), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-card--bordered"), bordered), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "--shadow"), shadow), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-card--shadow-hover"), hoverShadow), _classNames));
  var showHeader = header || title || subtitle || description || avatar || actions && !isPoster2 || status && isPoster2;
  var headerClass = classNames__default["default"]((_classNames2 = {}, defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-card__header"), showHeader), defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-card__title--bordered"), headerBordered), _classNames2));
  var titleClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__title"), title));
  var subtitleClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__subtitle"), subtitle));
  var actionClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__actions"), actions));
  var footerClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__footer"), footer));
  var coverClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__cover"), cover));
  var avatarClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__avatar"), avatar));
  var bodyClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__body"), children));
  var descriptionClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__description"), description));
  var renderTitle = title ? /* @__PURE__ */React__default["default"].createElement("span", {
    className: titleClass
  }, title) : null;
  var renderSubtitle = subtitle ? /* @__PURE__ */React__default["default"].createElement("span", {
    className: subtitleClass
  }, subtitle) : null;
  var renderDescription = description ? /* @__PURE__ */React__default["default"].createElement("p", {
    className: descriptionClass
  }, description) : null;
  var renderAvatar = avatar && /* @__PURE__ */React__default["default"].createElement("div", {
    className: avatarClass
  }, avatar);
  var renderHeaderActions = actions && !isPoster2 && /* @__PURE__ */React__default["default"].createElement("div", {
    className: actionClass
  }, actions);
  var renderFooterActions = actions && isPoster2 && /* @__PURE__ */React__default["default"].createElement("div", {
    className: actionClass
  }, actions);
  var renderStatus = status && isPoster2 && /* @__PURE__ */React__default["default"].createElement("div", {
    className: actionClass
  }, status);

  var renderHeader = function renderHeader() {
    if (header) return /* @__PURE__ */React__default["default"].createElement("div", {
      className: headerClass
    }, header);
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: headerClass
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-card__header-wrapper")
    }, renderAvatar, /* @__PURE__ */React__default["default"].createElement("div", null, renderTitle, renderSubtitle, renderDescription)), renderHeaderActions, renderStatus);
  };

  var renderCover = cover ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: coverClass
  }, typeof cover === "string" ? /* @__PURE__ */React__default["default"].createElement("img", {
    src: cover,
    alt: ""
  }) : cover) : null;
  var renderChildren = children && /* @__PURE__ */React__default["default"].createElement("div", {
    className: bodyClass
  }, children);
  var renderFooter = footer && /* @__PURE__ */React__default["default"].createElement("div", {
    className: footerClass
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-card__footer-wrapper")
  }, footer), renderFooterActions);

  if (loading) {
    return /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, null, /* @__PURE__ */React__default["default"].createElement("div", {
      className: cardClass
    }));
  }

  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    className: cardClass,
    style: style
  }, showHeader ? renderHeader() : null, renderCover, renderChildren, renderFooter);
});
Card.displayName = "Card";
Card.defaultProps = card_defaultProps.cardDefaultProps;

exports["default"] = Card;
