// tdesign-react/cjs/pagination/Pagination.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var objectWithoutProperties = require('../_chunks/dep-1a0462cc.js');
var React = require('react');
var classNames = require('classnames');
var omit = require('../_chunks/dep-70fe2bd0.js');
var _util_noop = require('../_util/noop.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var select_index = require('../select/index.js');
var inputNumber_index = require('../input-number/index.js');
var inputAdornment_InputAdornment = require('../input-adornment/InputAdornment.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var pagination_hooks_useBoundaryJumper = require('./hooks/useBoundaryJumper.js');
var pagination_hooks_usePrevNextJumper = require('./hooks/usePrevNextJumper.js');
var pagination_hooks_usePageNumber = require('./hooks/usePageNumber.js');
var pagination_hooks_useTotal = require('./hooks/useTotal.js');
var pagination_validators = require('./validators.js');
var pagination_defaultProps = require('./defaultProps.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-09805e35.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-4578b559.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-4fb39763.js');
require('../_chunks/dep-07e5d807.js');
require('../_chunks/dep-8329ea23.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-1e8de681.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../select/base/Select.js');
require('../_chunks/dep-65cb34d8.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../select/util/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../select/base/OptionGroup.js');
require('../select/defaultProps.js');
require('../select/base/Option.js');
require('../_chunks/dep-364d4c00.js');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/useRipple.js');
require('../_util/useAnimation.js');
require('../_util/setStyle.js');
require('../_util/dom.js');
require('raf');
require('../_util/easing.js');
require('../common/FakeArrow.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');
require('../select-input/index.js');
require('../select-input/SelectInput.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../popup/utils/ref.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../_util/useMutationObserver.js');
require('../_chunks/dep-db2b3323.js');
require('../_chunks/dep-52c5813c.js');
require('../_util/useWindowSize.js');
require('../popup/defaultProps.js');
require('../select-input/useSingle.js');
require('../_chunks/dep-b5af0182.js');
require('../_chunks/dep-55c5ba71.js');
require('../input/index.js');
require('../input/Input.js');
require('tdesign-icons-react');
require('../hooks/useGlobalIcon.js');
require('../input/InputGroup.js');
require('../input/defaultProps.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('../_util/useDragSorter.js');
require('../tag-input/useTagScroll.js');
require('../tag-input/useTagList.js');
require('lodash');
require('../tag/index.js');
require('../tag/Tag.js');
require('../tag/CheckTag.js');
require('../tag/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/defaultProps.js');
require('../select-input/useOverlayInnerStyle.js');
require('../select-input/defaultProps.js');
require('../select/base/PopupContent.js');
require('../input-number/InputNumber.js');
require('../button/index.js');
require('../button/Button.js');
require('../button/defaultProps.js');
require('../input-number/useInputNumber.js');
require('../_util/useCommonClassName.js');
require('../_common/js/input-number/number.js');
require('../_common/js/input-number/large-number.js');
require('../_common/js/log/log.js');
require('../input-number/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["theme", "size", "total", "showPageSize", "showPageNumber", "showPreviousAndNextBtn", "showFirstAndLastPageBtn", "showJumper", "pageEllipsisMode", "disabled", "foldedMaxPageBtn", "maxPageBtn", "totalContent", "pageSizeOptions", "onChange", "onCurrentChange", "onPageSizeChange", "style", "className"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Option = select_index.Select.Option;
var Pagination = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _classNames;

  var theme = props.theme,
      size = props.size,
      total = props.total,
      showPageSize = props.showPageSize,
      showPageNumber = props.showPageNumber,
      showPreviousAndNextBtn = props.showPreviousAndNextBtn,
      showFirstAndLastPageBtn = props.showFirstAndLastPageBtn,
      showJumper = props.showJumper,
      pageEllipsisMode = props.pageEllipsisMode,
      disabled = props.disabled,
      foldedMaxPageBtn = props.foldedMaxPageBtn,
      maxPageBtn = props.maxPageBtn,
      totalContent = props.totalContent,
      pageSizeOptions = props.pageSizeOptions,
      _props$onChange = props.onChange,
      onChange = _props$onChange === void 0 ? _util_noop["default"] : _props$onChange,
      onCurrentChange = props.onCurrentChange,
      onPageSizeChange = props.onPageSizeChange,
      style = props.style,
      className = props.className,
      otherProps = objectWithoutProperties._objectWithoutProperties(props, _excluded);

  var restProps = omit.omit_1(otherProps, ["current", "pageSize", "defaultPageSize", "defaultCurrent"]);

  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("pagination"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var _useControlled = hooks_useControlled["default"](props, "pageSize", onPageSizeChange),
      _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
      pageSize = _useControlled2[0],
      setPageSize = _useControlled2[1];

  var _useControlled3 = hooks_useControlled["default"](props, "current", onCurrentChange),
      _useControlled4 = slicedToArray._slicedToArray(_useControlled3, 2),
      current = _useControlled4[0],
      setCurrent = _useControlled4[1];

  var _useState = React.useState(current),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      jumpValue = _useState2[0],
      setJumpValue = _useState2[1];

  var min = 1;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var name = "".concat(classPrefix, "-pagination");
  var pageCount = React.useMemo(function () {
    var calCount = Math.ceil(total / pageSize);
    return calCount > 0 ? calCount : 1;
  }, [pageSize, total]);

  var changeCurrent = function changeCurrent(_nextCurrent, _nextPageSize) {
    if (disabled) return;
    var nextCurrent = _nextCurrent;
    var nextPageSize = _nextPageSize;

    if (!nextPageSize && !pagination_validators.pageSizeValidator(nextPageSize)) {
      var _pageSizeOptions$;

      nextPageSize = pageSize !== null && pageSize !== void 0 ? pageSize : typeof pageSizeOptions[0] === "number" ? pageSizeOptions[0] : (_pageSizeOptions$ = pageSizeOptions[0]) === null || _pageSizeOptions$ === void 0 ? void 0 : _pageSizeOptions$.value;
    }

    if (nextCurrent < min) nextCurrent = min;
    if (nextCurrent > pageCount) nextCurrent = pageCount;
    setCurrent(nextCurrent, {
      current: nextCurrent,
      previous: current,
      pageSize: nextPageSize
    });
    setJumpValue(nextCurrent);
    onChange({
      current: nextCurrent,
      previous: current,
      pageSize: nextPageSize
    });
  };

  var changePageSize = function changePageSize(nextPageSize) {
    var nextCurrent = Math.min(current, Math.ceil(total / nextPageSize));
    var pageInfo = {
      current: nextCurrent,
      previous: current,
      pageSize: nextPageSize
    };
    setPageSize(nextPageSize, pageInfo);

    if (current !== nextCurrent) {
      changeCurrent(nextCurrent, nextPageSize);
    } else {
      onChange(pageInfo);
    }
  };

  var _useTotal = pagination_hooks_useTotal["default"]({
    totalContent: totalContent,
    pageSize: pageSize,
    current: current,
    total: total
  }),
      totalContrl = _useTotal.totalContrl;

  var _useBoundaryJumper = pagination_hooks_useBoundaryJumper["default"]({
    disabled: disabled,
    current: current,
    pageCount: pageCount,
    showFirstAndLastPageBtn: showFirstAndLastPageBtn,
    changeCurrent: changeCurrent
  }),
      firstPageJumper = _useBoundaryJumper.firstPageJumper,
      lastPageJumper = _useBoundaryJumper.lastPageJumper;

  var _usePrevNextJumper = pagination_hooks_usePrevNextJumper["default"]({
    disabled: disabled,
    current: current,
    pageCount: pageCount,
    showPreviousAndNextBtn: showPreviousAndNextBtn,
    changeCurrent: changeCurrent
  }),
      prevJumper = _usePrevNextJumper.prevJumper,
      nextJumper = _usePrevNextJumper.nextJumper;

  var _usePageNumber = pagination_hooks_usePageNumber["default"]({
    showPageNumber: showPageNumber,
    maxPageBtn: maxPageBtn,
    disabled: disabled,
    current: current,
    pageCount: pageCount,
    foldedMaxPageBtn: foldedMaxPageBtn,
    changeCurrent: changeCurrent,
    pageEllipsisMode: pageEllipsisMode
  }),
      pageNumberContrl = _usePageNumber.pageNumberContrl;

  var pageSizeContrl = showPageSize && pageSizeOptions.length ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(name, "__select")
  }, /* @__PURE__ */React__default["default"].createElement(select_index.Select, {
    autoWidth: true,
    size: size,
    value: pageSize,
    disabled: disabled,
    onChange: changePageSize
  }, pageSizeOptions.map(function (item) {
    return typeof item === "number" ? /* @__PURE__ */React__default["default"].createElement(Option, {
      key: item,
      label: t(locale.itemsPerPage, {
        size: item
      }),
      value: item
    }) : /* @__PURE__ */React__default["default"].createElement(Option, {
      key: item.value,
      label: item.label,
      value: item.value
    });
  }))) : null;
  var Jumper = /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(name, "__jump")
  }, t(locale.jumpTo), /* @__PURE__ */React__default["default"].createElement(inputAdornment_InputAdornment["default"], {
    append: "/ ".concat(pageCount, " ").concat(t(locale.page))
  }, /* @__PURE__ */React__default["default"].createElement(inputNumber_index.InputNumber, {
    className: "".concat(classPrefix, "-pagination__input"),
    min: min,
    size: size,
    theme: "normal",
    max: pageCount,
    disabled: disabled,
    value: jumpValue,
    onChange: function onChange(val) {
      return setJumpValue(val);
    },
    onBlur: function onBlur(val) {
      return changeCurrent(val);
    },
    onEnter: function onEnter(val) {
      return changeCurrent(val);
    },
    placeholder: ""
  })));
  return /* @__PURE__ */React__default["default"].createElement("div", _objectSpread({
    className: classNames__default["default"](name, className, (_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-size-s"), size === "small"), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-is-disabled"), disabled), _classNames)),
    style: style,
    ref: ref
  }, restProps), totalContrl, pageSizeContrl, firstPageJumper, prevJumper, theme === "default" && pageNumberContrl, theme === "simple" && Jumper, nextJumper, lastPageJumper, theme === "default" && showJumper && Jumper);
});
Pagination.displayName = "Pagination";
Pagination.defaultProps = pagination_defaultProps.paginationDefaultProps;

exports["default"] = Pagination;
