// tdesign-react/es/card/Card.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import React, { forwardRef } from 'react';
import classNames from 'classnames';
import { Loading } from '../loading/index.js';
import useConfig from '../hooks/useConfig.js';
import useCommonClassName from '../_util/useCommonClassName.js';
import { cardDefaultProps } from './defaultProps.js';
import '../loading/Loading.js';
import '../_chunks/dep-9615f03f.js';
import '../_chunks/dep-7909fefa.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-3c98169f.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-e237f0f2.js';
import '../_util/easing.js';
import '../common/Portal.js';
import 'react-dom';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-d1b4ab24.js';
import '../_chunks/dep-8368bb87.js';
import '../loading/defaultProps.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-e54bb0dc.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-ebf3a272.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../loading/plugin.js';
import '../loading/style/css.js';
import '../loading/type.js';

var Card = /*#__PURE__*/forwardRef(function (props, ref) {
  var _classNames, _classNames2;

  var actions = props.actions,
      avatar = props.avatar,
      bordered = props.bordered,
      children = props.children,
      className = props.className,
      cover = props.cover,
      description = props.description,
      footer = props.footer,
      header = props.header,
      headerBordered = props.headerBordered,
      hoverShadow = props.hoverShadow,
      loading = props.loading,
      shadow = props.shadow,
      size = props.size,
      style = props.style,
      subtitle = props.subtitle,
      title = props.title,
      theme = props.theme,
      status = props.status;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var commonClassNames = useCommonClassName();
  var isPoster2 = theme === "poster2";
  var cardClass = classNames("".concat(classPrefix, "-card"), className, (_classNames = {}, _defineProperty(_classNames, commonClassNames.SIZE.small, size === "small"), _defineProperty(_classNames, "".concat(classPrefix, "-card--bordered"), bordered), _defineProperty(_classNames, "".concat(classPrefix, "--shadow"), shadow), _defineProperty(_classNames, "".concat(classPrefix, "-card--shadow-hover"), hoverShadow), _classNames));
  var showHeader = header || title || subtitle || description || avatar || actions && !isPoster2 || status && isPoster2;
  var headerClass = classNames((_classNames2 = {}, _defineProperty(_classNames2, "".concat(classPrefix, "-card__header"), showHeader), _defineProperty(_classNames2, "".concat(classPrefix, "-card__title--bordered"), headerBordered), _classNames2));
  var titleClass = classNames(_defineProperty({}, "".concat(classPrefix, "-card__title"), title));
  var subtitleClass = classNames(_defineProperty({}, "".concat(classPrefix, "-card__subtitle"), subtitle));
  var actionClass = classNames(_defineProperty({}, "".concat(classPrefix, "-card__actions"), actions));
  var footerClass = classNames(_defineProperty({}, "".concat(classPrefix, "-card__footer"), footer));
  var coverClass = classNames(_defineProperty({}, "".concat(classPrefix, "-card__cover"), cover));
  var avatarClass = classNames(_defineProperty({}, "".concat(classPrefix, "-card__avatar"), avatar));
  var bodyClass = classNames(_defineProperty({}, "".concat(classPrefix, "-card__body"), children));
  var descriptionClass = classNames(_defineProperty({}, "".concat(classPrefix, "-card__description"), description));
  var renderTitle = title ? /* @__PURE__ */React.createElement("span", {
    className: titleClass
  }, title) : null;
  var renderSubtitle = subtitle ? /* @__PURE__ */React.createElement("span", {
    className: subtitleClass
  }, subtitle) : null;
  var renderDescription = description ? /* @__PURE__ */React.createElement("p", {
    className: descriptionClass
  }, description) : null;
  var renderAvatar = avatar && /* @__PURE__ */React.createElement("div", {
    className: avatarClass
  }, avatar);
  var renderHeaderActions = actions && !isPoster2 && /* @__PURE__ */React.createElement("div", {
    className: actionClass
  }, actions);
  var renderFooterActions = actions && isPoster2 && /* @__PURE__ */React.createElement("div", {
    className: actionClass
  }, actions);
  var renderStatus = status && isPoster2 && /* @__PURE__ */React.createElement("div", {
    className: actionClass
  }, status);

  var renderHeader = function renderHeader() {
    if (header) return /* @__PURE__ */React.createElement("div", {
      className: headerClass
    }, header);
    return /* @__PURE__ */React.createElement("div", {
      className: headerClass
    }, /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-card__header-wrapper")
    }, renderAvatar, /* @__PURE__ */React.createElement("div", null, renderTitle, renderSubtitle, renderDescription)), renderHeaderActions, renderStatus);
  };

  var renderCover = cover ? /* @__PURE__ */React.createElement("div", {
    className: coverClass
  }, typeof cover === "string" ? /* @__PURE__ */React.createElement("img", {
    src: cover,
    alt: ""
  }) : cover) : null;
  var renderChildren = children && /* @__PURE__ */React.createElement("div", {
    className: bodyClass
  }, children);
  var renderFooter = footer && /* @__PURE__ */React.createElement("div", {
    className: footerClass
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-card__footer-wrapper")
  }, footer), renderFooterActions);

  if (loading) {
    return /* @__PURE__ */React.createElement(Loading, null, /* @__PURE__ */React.createElement("div", {
      className: cardClass
    }));
  }

  return /* @__PURE__ */React.createElement("div", {
    ref: ref,
    className: cardClass,
    style: style
  }, showHeader ? renderHeader() : null, renderCover, renderChildren, renderFooter);
});
Card.displayName = "Card";
Card.defaultProps = cardDefaultProps;

export { Card as default };
