// tdesign-react/cjs/drawer/hooks/useDrag.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-4eb93341.js');
var slicedToArray = require('../../_chunks/dep-9c5d4f93.js');
var React = require('react');
require('../../_chunks/dep-2205decf.js');

var useDrag = function useDrag(placement, sizeDraggable) {
  var _useState = React.useState(null),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      dragSizeValue = _useState2[0],
      changeDragSizeValue = _useState2[1];

  var handleMousemove = function handleMousemove(e) {
    var x = e.x,
        y = e.y;

    if (sizeDraggable) {
      if (placement === "right") {
        changeDragSizeValue("".concat(document.documentElement.clientWidth - x + 8, "px"));
      }

      if (placement === "left") {
        changeDragSizeValue("".concat(x + 8, "px"));
      }

      if (placement === "top") {
        changeDragSizeValue("".concat(y + 8, "px"));
      }

      if (placement === "bottom") {
        changeDragSizeValue("".concat(document.documentElement.clientHeight - y + 8, "px"));
      }
    }
  };

  var handleMouseup = function handleMouseup() {
    document.removeEventListener("mouseup", handleMouseup, true);
    document.removeEventListener("mousemove", handleMousemove, true);
  };

  var draggableLineStyles = React.useMemo(function () {
    var _ref;

    var isHorizontal = ["right", "left"].includes(placement);
    var oppositeMap = {
      left: "right",
      right: "left",
      top: "bottom",
      bottom: "top"
    };
    return _ref = {
      zIndex: 1,
      position: "absolute",
      background: "transparent"
    }, defineProperty._defineProperty(_ref, oppositeMap[placement], 0), defineProperty._defineProperty(_ref, "width", isHorizontal ? "16px" : "100%"), defineProperty._defineProperty(_ref, "height", isHorizontal ? "100%" : "16px"), defineProperty._defineProperty(_ref, "cursor", isHorizontal ? "col-resize" : "row-resize"), _ref;
  }, [placement]);

  var enableDrag = function enableDrag() {
    document.addEventListener("mouseup", handleMouseup, true);
    document.addEventListener("mousemove", handleMousemove, true);
  };

  return {
    dragSizeValue: dragSizeValue,
    enableDrag: enableDrag,
    draggableLineStyles: draggableLineStyles
  };
};

exports["default"] = useDrag;
