// tdesign-react/cjs/switch/Switch.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var objectWithoutProperties = require('../_chunks/dep-1a0462cc.js');
var React = require('react');
var classNames = require('classnames');
var loading_index = require('../loading/index.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var _util_useCommonClassName = require('../_util/useCommonClassName.js');
var switch_defaultProps = require('./defaultProps.js');
var _common_js_log_log = require('../_common/js/log/log.js');
require('../_chunks/dep-2205decf.js');
require('../loading/Loading.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-1fff9729.js');
require('../_util/easing.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../loading/defaultProps.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../loading/plugin.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["className", "value", "defaultValue", "disabled", "loading", "size", "label", "customValue", "onChange"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Switch = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _classNames;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var className = props.className,
      value = props.value,
      defaultValue = props.defaultValue,
      disabled = props.disabled,
      loading = props.loading,
      size = props.size,
      label = props.label,
      customValue = props.customValue,
      onChange = props.onChange,
      restProps = objectWithoutProperties._objectWithoutProperties(props, _excluded);

  var _ref = customValue || [],
      _ref2 = slicedToArray._slicedToArray(_ref, 2),
      _ref2$ = _ref2[0],
      activeValue = _ref2$ === void 0 ? true : _ref2$,
      _ref2$2 = _ref2[1],
      inactiveValue = _ref2$2 === void 0 ? false : _ref2$2;

  var isControlled = typeof value !== "undefined";
  var initChecked = defaultValue === activeValue || value === activeValue;

  var _useState = React.useState(initChecked),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      innerChecked = _useState2[0],
      setInnerChecked = _useState2[1];

  function renderContent(checked) {
    if (typeof label === "function") return label({
      value: value
    });
    if (typeof label === "string") return label;

    if (Array.isArray(label)) {
      var _label = slicedToArray._slicedToArray(label, 2),
          _label$ = _label[0],
          activeContent = _label$ === void 0 ? "" : _label$,
          _label$2 = _label[1],
          inactiveContent = _label$2 === void 0 ? "" : _label$2;

      var content = checked ? activeContent : inactiveContent;
      if (typeof content === "function") return content();
      return content;
    }

    return null;
  }

  function onInternalClick() {
    if (disabled) return;
    !isControlled && setInnerChecked(!innerChecked);
    var changedValue = !innerChecked ? activeValue : inactiveValue;
    onChange === null || onChange === void 0 ? void 0 : onChange(changedValue);
  }

  React.useEffect(function () {
    if (Array.isArray(customValue) && !customValue.includes(value)) {
      _common_js_log_log["default"].error("Switch", "value is not in customValue: ".concat(JSON.stringify(customValue)));
    }

    isControlled && setInnerChecked(value === activeValue);
  }, [value, customValue, activeValue, isControlled]);

  var _useCommonClassName = _util_useCommonClassName["default"](),
      SIZE = _useCommonClassName.SIZE,
      STATUS = _useCommonClassName.STATUS;

  var switchClassName = classNames__default["default"]("".concat(classPrefix, "-switch"), className, (_classNames = {}, defineProperty._defineProperty(_classNames, STATUS.checked, innerChecked), defineProperty._defineProperty(_classNames, STATUS.disabled, disabled), defineProperty._defineProperty(_classNames, STATUS.loading, loading), _classNames), SIZE[size]);
  return /* @__PURE__ */React__default["default"].createElement("button", _objectSpread(_objectSpread({}, restProps), {}, {
    type: "button",
    role: "switch",
    disabled: disabled || loading,
    className: switchClassName,
    ref: ref,
    onClick: onInternalClick
  }), /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-switch__handle")
  }, loading && /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
    loading: true,
    size: "small"
  })), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-switch__content")
  }, renderContent(innerChecked)));
});
Switch.displayName = "Switch";
Switch.defaultProps = switch_defaultProps.switchDefaultProps;

exports["default"] = Switch;
