// tdesign-react/cjs/anchor/AnchorItem.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var objectWithoutProperties = require('../_chunks/dep-1a0462cc.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var anchor_AnchorContext = require('./AnchorContext.js');
var anchor_defaultProps = require('./defaultProps.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../_util/noop.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["href", "title", "target", "children", "className"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var AnchorItem = function AnchorItem(props) {
  var _useContext = React.useContext(anchor_AnchorContext.AnchorContext),
      onClick = _useContext.onClick,
      activeItem = _useContext.activeItem,
      registerItem = _useContext.registerItem,
      unregisterItem = _useContext.unregisterItem;

  var href = props.href,
      title = props.title,
      target = props.target,
      children = props.children,
      className = props.className,
      rest = objectWithoutProperties._objectWithoutProperties(props, _excluded);

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var titleAttr = typeof title === "string" ? title : null;

  var handleClick = function handleClick(e) {
    onClick({
      title: titleAttr,
      href: href
    }, e);
  };

  React.useEffect(function () {
    registerItem(href);
    return function () {
      return unregisterItem(href);
    };
  }, [href, registerItem, unregisterItem]);
  return /* @__PURE__ */React__default["default"].createElement("div", _objectSpread(_objectSpread({}, rest), {}, {
    className: classNames__default["default"]("".concat(classPrefix, "-anchor__item"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-active"), activeItem === href), className)
  }), /* @__PURE__ */React__default["default"].createElement("a", {
    href: href,
    className: classNames__default["default"]("".concat(classPrefix, "-anchor__item-link")),
    title: titleAttr,
    target: target,
    onClick: function onClick(e) {
      return handleClick(e);
    }
  }, title), children);
};

AnchorItem.displayName = "AnchorItem";
AnchorItem.defaultProps = anchor_defaultProps.anchorItemDefaultProps;

exports["default"] = AnchorItem;
