// tdesign-react/cjs/list/List.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var classNames = require('classnames');
var isString = require('../_chunks/dep-9a2dbbb1.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var _util_noop = require('../_util/noop.js');
var loading_index = require('../loading/index.js');
var list_ListItem = require('./ListItem.js');
var list_ListItemMeta = require('./ListItemMeta.js');
var list_defaultProps = require('./defaultProps.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('hoist-non-react-statics');
require('../loading/Loading.js');
require('../_util/dom.js');
require('raf');
require('../_util/easing.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../_chunks/dep-1a0462cc.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var List = _util_forwardRefWithStatics["default"](function (props, ref) {
  var _classNames, _classNames2;

  var header = props.header,
      footer = props.footer,
      asyncLoading = props.asyncLoading,
      size = props.size,
      split = props.split,
      stripe = props.stripe,
      layout = props.layout,
      children = props.children,
      className = props.className,
      _props$onLoadMore = props.onLoadMore,
      onLoadMore = _props$onLoadMore === void 0 ? _util_noop["default"] : _props$onLoadMore,
      _props$onScroll = props.onScroll,
      onScroll = _props$onScroll === void 0 ? _util_noop["default"] : _props$onScroll,
      style = props.style;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("list"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var handleClickLoad = function handleClickLoad(e) {
    if (asyncLoading === "load-more") {
      onLoadMore({
        e: e
      });
    }
  };

  var handleScroll = function handleScroll(event) {
    var currentTarget = event.currentTarget;
    var scrollTop = currentTarget.scrollTop,
        offsetHeight = currentTarget.offsetHeight,
        scrollHeight = currentTarget.scrollHeight;
    var scrollBottom = scrollHeight - scrollTop - offsetHeight;
    onScroll({
      e: event,
      scrollTop: scrollTop,
      scrollBottom: scrollBottom
    });
  };

  var loadElement = isString.isString_1(asyncLoading) ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-list__load"), (_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-list__load--loading"), asyncLoading === "loading"), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-list__load--load-more"), asyncLoading === "load-more"), _classNames)),
    onClick: handleClickLoad
  }, asyncLoading === "loading" && /* @__PURE__ */React__default["default"].createElement("div", null, /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
    loading: true
  }), /* @__PURE__ */React__default["default"].createElement("span", null, t(local.loadingText))), asyncLoading === "load-more" && /* @__PURE__ */React__default["default"].createElement("span", null, t(local.loadingMoreText))) : asyncLoading;
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    style: style,
    onScroll: handleScroll,
    className: classNames__default["default"]("".concat(classPrefix, "-list"), className, (_classNames2 = {}, defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-list--split"), split), defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-list--stripe"), stripe), defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-list--vertical-action"), layout === "vertical"), defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-size-s"), size === "small"), defineProperty._defineProperty(_classNames2, "".concat(classPrefix, "-size-l"), size === "large"), _classNames2))
  }, header && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-list__header")
  }, header), /* @__PURE__ */React__default["default"].createElement("ul", {
    className: "".concat(classPrefix, "-list__inner")
  }, children), asyncLoading && loadElement, footer && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-list__footer")
  }, footer));
}, {
  ListItem: list_ListItem["default"],
  ListItemMeta: list_ListItemMeta["default"]
});
List.displayName = "List";
List.defaultProps = list_defaultProps.listDefaultProps;

exports["default"] = List;
