// tdesign-react/esm/tree/useTreeConfig.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import { useMemo } from 'react';
import useConfig from '../hooks/useConfig.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-1e32006c.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-e54bb0dc.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-e237f0f2.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-ebf3a272.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';

function useTreeConfig() {
  var _useConfig = useConfig(),
      prefix = _useConfig.classPrefix;

  var _useLocaleReceiver = useLocaleReceiver("tree"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      _locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  return useMemo(function () {
    var tree = "".concat(prefix, "-tree");
    var treeClassNames = {
      icon: "".concat(prefix, "-icon"),
      folderIcon: "".concat(prefix, "-folder-icon"),
      actived: "".concat(prefix, "-is-active"),
      disabled: "".concat(prefix, "-is-disabled"),
      treeIconRight: "".concat(prefix, "-icon-arrow-right"),
      treeIconDown: "".concat(prefix, "-icon-arrow-down"),
      tree: tree,
      treeFx: "".concat(tree, "--transition"),
      treeBlockNode: "".concat(tree, "--block-node"),
      treeEmpty: "".concat(tree, "__empty"),
      treeList: "".concat(tree, "__list"),
      treeNode: "".concat(tree, "__item"),
      treeNodeOpen: "".concat(tree, "__item--open"),
      treeHoverable: "".concat(tree, "--hoverable"),
      treeCheckable: "".concat(tree, "--checkable"),
      treeLabel: "".concat(tree, "__label"),
      treeLabelStrictly: "".concat(tree, "__label--strictly"),
      treeIcon: "".concat(tree, "__icon"),
      treeIconDefault: "".concat(tree, "__icon--default"),
      treeSpace: "".concat(tree, "__space"),
      treeOperations: "".concat(tree, "__operations"),
      line: "".concat(tree, "__line"),
      lineIsLeaf: "".concat(tree, "__line--leaf"),
      lineIsFirst: "".concat(tree, "__line--first"),
      treeNodeShow: "".concat(tree, "__item--show"),
      treeClickable: "".concat(tree, "__item--clickable"),
      loading: "".concat(prefix, "-icon-loading ").concat(prefix, "-icon-loading-blue"),
      toggleEnter: "".concat(tree, "__item--enter-active"),
      toggleLeave: "".concat(tree, "__item--leave-active")
    };
    var transitionNames = {
      treeNode: "".concat(prefix, "-tree-toggle")
    };
    var transitionClassNames = {
      enter: treeClassNames.toggleEnter,
      exit: treeClassNames.toggleLeave
    };
    return {
      treeClassNames: treeClassNames,
      transitionNames: transitionNames,
      transitionClassNames: transitionClassNames,
      transitionDuration: 300,
      locale: function locale(key) {
        return t(_locale[key]);
      }
    };
  }, [_locale, prefix, t]);
}

export { useTreeConfig };
