// tdesign-react/cjs/calendar/Calendar.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../_chunks/dep-8052f095.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var defineProperty = require('../_chunks/dep-4eb93341.js');
var React = require('react');
var dayjs = require('dayjs');
var button_index = require('../button/index.js');
var select_index = require('../select/index.js');
var radio_index = require('../radio/index.js');
var tag_CheckTag = require('../tag/CheckTag.js');
var _util_noop = require('../_util/noop.js');
var calendar_hooks_usePrefixClass = require('./hooks/usePrefixClass.js');
var _util_useLayoutEffect = require('../_util/useLayoutEffect.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var calendar__util = require('./_util.js');
var calendar_CalendarCellComp = require('./CalendarCellComp.js');
var calendar_defaultProps = require('./defaultProps.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-606cfe2a.js');
require('../button/Button.js');
require('../_chunks/dep-1a0462cc.js');
require('classnames');
require('../hooks/useConfig.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../_util/useRipple.js');
require('../_util/useAnimation.js');
require('../_util/setStyle.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');
require('../button/defaultProps.js');
require('../select/base/Select.js');
require('../_chunks/dep-65cb34d8.js');
require('../_chunks/dep-8329ea23.js');
require('../_chunks/dep-09805e35.js');
require('../_chunks/dep-afce8659.js');
require('../hooks/useControlled.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../select/util/helper.js');
require('../select/base/OptionGroup.js');
require('../select/defaultProps.js');
require('../select/base/Option.js');
require('../_chunks/dep-364d4c00.js');
require('../common/FakeArrow.js');
require('../select-input/index.js');
require('../select-input/SelectInput.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../popup/utils/ref.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../_util/useMutationObserver.js');
require('../_chunks/dep-db2b3323.js');
require('../_chunks/dep-52c5813c.js');
require('../_util/useWindowSize.js');
require('../popup/defaultProps.js');
require('../select-input/useSingle.js');
require('../_chunks/dep-b5af0182.js');
require('../_chunks/dep-55c5ba71.js');
require('../_chunks/dep-1e8de681.js');
require('../_chunks/dep-07e5d807.js');
require('../input/index.js');
require('../input/Input.js');
require('tdesign-icons-react');
require('../hooks/useGlobalIcon.js');
require('../input/InputGroup.js');
require('../input/defaultProps.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('../_util/useDragSorter.js');
require('../tag-input/useTagScroll.js');
require('../tag-input/useTagList.js');
require('lodash');
require('../tag/index.js');
require('../tag/Tag.js');
require('../tag/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/defaultProps.js');
require('../select-input/useOverlayInnerStyle.js');
require('../select-input/defaultProps.js');
require('../select/base/PopupContent.js');
require('../_chunks/dep-4ab4e7b4.js');
require('../common/Check.js');
require('../_util/helper.js');
require('../_chunks/dep-0ebc045a.js');
require('../_util/useCommonClassName.js');
require('../radio/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var getDefaultControllerConfigData = function getDefaultControllerConfigData() {
  var visible = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
  return {
    visible: visible,
    disabled: false,
    mode: {
      visible: true,
      radioGroupProps: {}
    },
    year: {
      visible: true,
      selectProps: {
        popupProps: {
          overlayInnerStyle: {
            width: "110px"
          }
        }
      }
    },
    month: {
      visible: true,
      selectProps: {
        popupProps: {
          overlayInnerStyle: {
            width: "90px"
          }
        }
      }
    },
    weekend: {
      visible: true,
      showWeekendButtonProps: {},
      hideWeekendButtonProps: {}
    },
    current: {
      visible: true,
      currentDayButtonProps: {},
      currentMonthButtonProps: {}
    }
  };
};

var Calendar = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var className = props.className,
      style = props.style,
      modeFromProps = props.mode,
      valueFromProps = props.value,
      firstDayOfWeek = props.firstDayOfWeek,
      format = props.format,
      range = props.range,
      head = props.head,
      cell = props.cell,
      cellAppend = props.cellAppend,
      week = props.week,
      theme = props.theme,
      controllerConfig = props.controllerConfig,
      _props$isShowWeekendD = props.isShowWeekendDefault,
      isShowWeekendDefault = _props$isShowWeekendD === void 0 ? true : _props$isShowWeekendD,
      _props$preventCellCon = props.preventCellContextmenu,
      preventCellContextmenu = _props$preventCellCon === void 0 ? false : _props$preventCellCon,
      monthProps = props.month,
      yearProps = props.year,
      _props$onControllerCh = props.onControllerChange,
      onControllerChange = _props$onControllerCh === void 0 ? _util_noop["default"] : _props$onControllerCh,
      _props$onCellClick = props.onCellClick,
      onCellClick = _props$onCellClick === void 0 ? _util_noop["default"] : _props$onCellClick,
      _props$onCellDoubleCl = props.onCellDoubleClick,
      onCellDoubleClick = _props$onCellDoubleCl === void 0 ? _util_noop["default"] : _props$onCellDoubleCl,
      _props$onCellRightCli = props.onCellRightClick,
      onCellRightClick = _props$onCellRightCli === void 0 ? _util_noop["default"] : _props$onCellRightCli,
      _props$onMonthChange = props.onMonthChange,
      onMonthChange = _props$onMonthChange === void 0 ? _util_noop["default"] : _props$onMonthChange,
      fillWithZero = props.fillWithZero;
  var controllerConfigData = typeof controllerConfig === "boolean" ? getDefaultControllerConfigData(controllerConfig) : _objectSpread(_objectSpread({}, getDefaultControllerConfigData()), controllerConfig);

  if (typeof controllerConfig !== "boolean") {
    if (monthProps) {
      controllerConfigData.month.visible = true;
    }

    if (yearProps) {
      controllerConfigData.year.visible = true;
    }
  }

  var visible = controllerConfigData.visible,
      disabled = controllerConfigData.disabled,
      modeFromConfig = controllerConfigData.mode,
      yearFromConfig = controllerConfigData.year,
      monthFromConfig = controllerConfigData.month,
      weekend = controllerConfigData.weekend,
      current = controllerConfigData.current;
  var _modeFromConfig$visib = modeFromConfig.visible,
      visibleForMode = _modeFromConfig$visib === void 0 ? true : _modeFromConfig$visib,
      _modeFromConfig$radio = modeFromConfig.radioGroupProps,
      radioGroupPropsForMode = _modeFromConfig$radio === void 0 ? {} : _modeFromConfig$radio;
  var _yearFromConfig$visib = yearFromConfig.visible,
      visibleForYear = _yearFromConfig$visib === void 0 ? true : _yearFromConfig$visib,
      _yearFromConfig$selec = yearFromConfig.selectProps,
      selectPropsForYear = _yearFromConfig$selec === void 0 ? {} : _yearFromConfig$selec;
  var _monthFromConfig$visi = monthFromConfig.visible,
      visibleForMonth = _monthFromConfig$visi === void 0 ? true : _monthFromConfig$visi,
      _monthFromConfig$sele = monthFromConfig.selectProps,
      selectPropsForMonth = _monthFromConfig$sele === void 0 ? {} : _monthFromConfig$sele;
  var _weekend$visible = weekend.visible,
      visibleForWeekendToggle = _weekend$visible === void 0 ? true : _weekend$visible,
      _weekend$showWeekendB = weekend.showWeekendButtonProps,
      showWeekendButtonProps = _weekend$showWeekendB === void 0 ? {} : _weekend$showWeekendB,
      _weekend$hideWeekendB = weekend.hideWeekendButtonProps,
      hideWeekendButtonProps = _weekend$hideWeekendB === void 0 ? {} : _weekend$hideWeekendB;
  var _current$visible = current.visible,
      visibleForCurrent = _current$visible === void 0 ? true : _current$visible,
      _current$currentDayBu = current.currentDayButtonProps,
      currentDayButtonProps = _current$currentDayBu === void 0 ? {} : _current$currentDayBu,
      _current$currentMonth = current.currentMonthButtonProps,
      currentMonthButtonProps = _current$currentMonth === void 0 ? {} : _current$currentMonth;

  var _useState = React.useState("month"),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      mode = _useState2[0],
      setMode = _useState2[1];

  var _useState3 = React.useState(dayjs__default["default"](valueFromProps || dayjs__default["default"]().format("YYYY-MM-DD"))),
      _useState4 = slicedToArray._slicedToArray(_useState3, 2),
      value = _useState4[0],
      setValue = _useState4[1];

  var _useState5 = React.useState(yearProps ? Number(yearProps) : value.year()),
      _useState6 = slicedToArray._slicedToArray(_useState5, 2),
      year = _useState6[0],
      setYear = _useState6[1];

  var _useState7 = React.useState(monthProps ? Number(monthProps) : parseInt(value.format("M"), 10)),
      _useState8 = slicedToArray._slicedToArray(_useState7, 2),
      month = _useState8[0],
      setMonth = _useState8[1];

  var _useState9 = React.useState(isShowWeekendDefault),
      _useState10 = slicedToArray._slicedToArray(_useState9, 2),
      isShowWeekend = _useState10[0],
      setIsShowWeekend = _useState10[1];

  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("calendar"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var weekLabelList = t(local.week).split(",");
  var colHeaderList = React.useMemo(function () {
    if (mode === "year") return [];
    var weekTextArr = Array.isArray(week) && week.length >= 7 ? week : toConsumableArray._toConsumableArray(weekLabelList);
    var list = [];

    for (var i = firstDayOfWeek; i <= 7; i++) {
      if (!isShowWeekend && i > 5) {
        break;
      }

      list.push({
        day: i,
        text: weekTextArr[i - 1]
      });
    }

    if (firstDayOfWeek > 1) {
      for (var _i = 1; _i < firstDayOfWeek; _i++) {
        if (!isShowWeekend && _i > 5) {
          break;
        }

        list.push({
          day: _i,
          text: weekTextArr[_i - 1]
        });
      }
    }

    return list;
  }, [mode, firstDayOfWeek, isShowWeekend, week, weekLabelList]);
  var rangeFromTo = React.useMemo(function () {
    if (!range || range.length < 2) {
      return null;
    }

    var _range = slicedToArray._slicedToArray(range, 2),
        v1 = _range[0],
        v2 = _range[1];

    if (dayjs__default["default"](v1).isBefore(dayjs__default["default"](v2))) {
      return {
        from: v1,
        to: v2
      };
    }

    return {
      from: v2,
      to: v1
    };
  }, [range]);
  var controllerOptions = React.useMemo(function () {
    var dayJsFilterDate = dayjs__default["default"]("".concat(year, "-").concat(month));
    var re = {
      filterDate: dayJsFilterDate.toDate(),
      formattedFilterDate: dayJsFilterDate.format(format),
      mode: mode,
      isShowWeekend: isShowWeekend
    };
    return re;
  }, [isShowWeekend, mode, year, month, format]);

  var _useMemo = React.useMemo(function () {
    var isRangeValid = rangeFromTo && rangeFromTo.from && rangeFromTo.to;

    var checkMonthSelectorDisabled = function checkMonthSelectorDisabled(yearIn, monthIn) {
      if (isRangeValid) {
        var beginYear = dayjs__default["default"](rangeFromTo.from).year();
        var endYear = dayjs__default["default"](rangeFromTo.to).year();

        if (yearIn === beginYear) {
          var beginMon = parseInt(dayjs__default["default"](rangeFromTo.from).format("M"), 10);
          return monthIn < beginMon;
        }

        if (yearIn === endYear) {
          var endMon = parseInt(dayjs__default["default"](rangeFromTo.to).format("M"), 10);
          return monthIn > endMon;
        }
      }

      return false;
    };

    var yearList = [];
    var monthList2 = [];
    var yearBegin = isRangeValid ? dayjs__default["default"](rangeFromTo.from).year() : Math.max(calendar__util.minYear, year - 10);
    var yearEnd = isRangeValid ? dayjs__default["default"](rangeFromTo.to).year() : Math.max(calendar__util.minYear, year + 10);

    for (var i = yearBegin; i <= yearEnd; i++) {
      yearList.push({
        value: i,
        disabled: checkMonthSelectorDisabled(i, month)
      });
    }

    for (var _i2 = 1; _i2 <= 12; _i2++) {
      monthList2.push({
        value: _i2,
        disabled: checkMonthSelectorDisabled(year, _i2)
      });
    }

    return [yearList, monthList2];
  }, [rangeFromTo, year, month]),
      _useMemo2 = slicedToArray._slicedToArray(_useMemo, 2),
      yearSelectList = _useMemo2[0],
      monthSelectList = _useMemo2[1];

  var dateList = React.useMemo(function () {
    return calendar__util.createDateList(year, month, firstDayOfWeek, value, format);
  }, [year, month, firstDayOfWeek, format, value]);
  var monthList = React.useMemo(function () {
    return calendar__util.createMonthList(year, value, format);
  }, [year, value, format]);
  var createCalendarCell = React.useCallback(function (cellData) {
    return _objectSpread(_objectSpread({}, cellData), controllerOptions);
  }, [controllerOptions]);
  var toCurrent = React.useCallback(function () {
    var valueIn = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
    var now = dayjs__default["default"](valueIn).isValid() ? dayjs__default["default"](valueIn) : dayjs__default["default"](dayjs__default["default"]().format("YYYY-MM-DD"));
    setValue(now);
    setYear(now.year());
    setMonth(parseInt(now.format("M"), 10));
  }, []);
  React__default["default"].useImperativeHandle(ref, function () {
    return {
      toCurrent: toCurrent
    };
  }, [toCurrent]);

  var handleMonthChange = function handleMonthChange(newMonth) {
    setMonth(newMonth);
    onMonthChange({
      month: String(newMonth),
      year: String(year)
    });
  };

  var execCellEvent = React.useCallback(function (event, calendarCell, handleFunc) {
    if (handleFunc && typeof handleFunc === "function") {
      handleFunc({
        cell: createCalendarCell(calendarCell),
        e: event
      });
    }
  }, [createCalendarCell]);

  var clickCell = function clickCell(event, calendarCell) {
    setValue(dayjs__default["default"](calendarCell.formattedDate));
    execCellEvent(event, calendarCell, onCellClick);
  };

  var doubleClickCell = function doubleClickCell(event, calendarCell) {
    execCellEvent(event, calendarCell, onCellDoubleClick);
  };

  var rightClickCell = function rightClickCell(event, calendarCell) {
    if (preventCellContextmenu) event.preventDefault();
    execCellEvent(event, calendarCell, onCellRightClick);
  };

  React.useEffect(function () {
    toCurrent(valueFromProps);
  }, [valueFromProps, toCurrent]);
  React.useEffect(function () {
    if (monthProps) {
      setMonth(Number(monthProps));
    }
  }, [monthProps]);
  React.useEffect(function () {
    if (yearProps) {
      setYear(Number(yearProps));
    }
  }, [yearProps]);
  React.useEffect(function () {
    setMode(modeFromProps);
  }, [modeFromProps]);
  React.useEffect(function () {
    setIsShowWeekend(isShowWeekendDefault);
  }, [isShowWeekendDefault]);
  _util_useLayoutEffect["default"](function () {
    onControllerChange(controllerOptions);
  }, [controllerOptions, onControllerChange]);
  var prefixCls = calendar_hooks_usePrefixClass["default"]();
  var currentDate = dayjs__default["default"]().format("YYYY-MM-DD");
  var currentMonth = dayjs__default["default"]().format("YYYY-MM");
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls(calendar__util.blockName, [calendar__util.blockName, "", theme]).concat(" ", className),
    style: style
  }, visible && /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([calendar__util.blockName, "control"])
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([calendar__util.blockName, "title"])
  }, function () {
    if (head && typeof head === "function") return head(controllerOptions);
    if (head && typeof head !== "function") return head;
  }()), /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([calendar__util.blockName, "control-section"])
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([calendar__util.blockName, "control-section-cell"])
  }, visibleForYear && /* @__PURE__ */React__default["default"].createElement(select_index.Select, _objectSpread({
    autoWidth: true,
    size: calendar__util.controlSectionSize,
    value: year,
    disabled: disabled,
    options: yearSelectList.map(function (item) {
      return {
        label: t(local.yearSelection, {
          year: item.value
        }),
        value: item.value,
        disabled: item.disabled
      };
    }),
    onChange: function onChange(selectYear) {
      return setYear(selectYear);
    }
  }, selectPropsForYear))), /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([calendar__util.blockName, "control-section-cell"])
  }, visibleForMonth && mode === "month" && /* @__PURE__ */React__default["default"].createElement(select_index.Select, _objectSpread({
    autoWidth: true,
    size: calendar__util.controlSectionSize,
    value: month,
    options: monthSelectList.map(function (item) {
      return {
        label: t(local.monthSelection, {
          month: item.value
        }),
        value: item.value,
        disabled: item.disabled
      };
    }),
    disabled: disabled,
    onChange: handleMonthChange
  }, selectPropsForMonth)))), /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([calendar__util.blockName, "control-section"]),
    style: {
      height: "auto"
    }
  }, visibleForMode && /* @__PURE__ */React__default["default"].createElement(radio_index.Radio.Group, _objectSpread({
    variant: "default-filled",
    size: calendar__util.controlSectionSize,
    value: mode,
    disabled: disabled,
    onChange: function onChange(value2) {
      return setMode(value2);
    }
  }, radioGroupPropsForMode), /* @__PURE__ */React__default["default"].createElement(radio_index.Radio.Button, {
    value: "month"
  }, t(local.monthRadio)), /* @__PURE__ */React__default["default"].createElement(radio_index.Radio.Button, {
    value: "year"
  }, t(local.yearRadio)))), mode === "month" && theme === "full" && visibleForWeekendToggle && /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([calendar__util.blockName, "control-section"])
  }, /* @__PURE__ */React__default["default"].createElement(tag_CheckTag["default"], _objectSpread({
    className: "t-calendar__control-tag",
    checked: !isShowWeekend,
    disabled: disabled,
    size: calendar__util.controlSectionSize,
    onClick: function onClick() {
      setIsShowWeekend(!isShowWeekend);
    }
  }, isShowWeekend ? hideWeekendButtonProps : showWeekendButtonProps), "".concat(isShowWeekend ? t(local.hideWeekend) : t(local.showWeekend)))), theme === "full" && visibleForCurrent && /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([calendar__util.blockName, "control-section"])
  }, /* @__PURE__ */React__default["default"].createElement(button_index.Button, _objectSpread({
    size: calendar__util.controlSectionSize,
    disabled: disabled,
    onClick: function onClick() {
      toCurrent();
    }
  }, mode === "year" ? currentMonthButtonProps : currentDayButtonProps), mode === "year" ? t(local.thisMonth) : t(local.today)))), /* @__PURE__ */React__default["default"].createElement("div", {
    className: prefixCls([calendar__util.blockName, "panel"], [calendar__util.blockName, "panel--".concat(mode)])
  }, /* @__PURE__ */React__default["default"].createElement("table", {
    className: prefixCls([calendar__util.blockName, "table"])
  }, colHeaderList.length > 0 && /* @__PURE__ */React__default["default"].createElement("thead", {
    className: prefixCls([calendar__util.blockName, "table-head"])
  }, /* @__PURE__ */React__default["default"].createElement("tr", {
    className: prefixCls([calendar__util.blockName, "table-head-row"])
  }, colHeaderList.map(function (item) {
    return /* @__PURE__ */React__default["default"].createElement("th", {
      key: item.day,
      className: prefixCls([calendar__util.blockName, "table-head-cell"])
    }, week && typeof week === "function" ? week({
      day: item.day
    }) : item.text);
  }))), mode === "month" && /* @__PURE__ */React__default["default"].createElement("tbody", {
    className: prefixCls([calendar__util.blockName, "table-body"])
  }, dateList.map(function (dateRow, dateRowIndex) {
    return /* @__PURE__ */React__default["default"].createElement("tr", {
      key: String(dateRowIndex),
      className: prefixCls([calendar__util.blockName, "table-body-row"])
    }, dateRow.map(function (dateCell, dateCellIndex) {
      if (!isShowWeekend && [6, 7].indexOf(dateCell.day) >= 0) return null;
      var isNow = dateCell.formattedDate === currentDate;
      return /* @__PURE__ */React__default["default"].createElement(calendar_CalendarCellComp["default"], {
        key: dateCellIndex,
        mode: mode,
        theme: theme,
        cell: cell,
        cellData: dateCell,
        cellAppend: cellAppend,
        fillWithZero: fillWithZero,
        isCurrent: dateCell.isCurrent,
        isNow: isNow,
        isDisabled: dateCell.belongTo !== 0,
        createCalendarCell: createCalendarCell,
        onCellClick: function onCellClick(event) {
          return clickCell(event, dateCell);
        },
        onCellDoubleClick: function onCellDoubleClick(event) {
          return doubleClickCell(event, dateCell);
        },
        onCellRightClick: function onCellRightClick(event) {
          return rightClickCell(event, dateCell);
        }
      });
    }));
  })), mode === "year" && /* @__PURE__ */React__default["default"].createElement("tbody", {
    className: prefixCls([calendar__util.blockName, "table-body"])
  }, monthList.map(function (monthRow, monthRowIndex) {
    return /* @__PURE__ */React__default["default"].createElement("tr", {
      key: String(monthRowIndex),
      className: prefixCls([calendar__util.blockName, "table-body-row"])
    }, monthRow.map(function (monthCell, monthCellIndex) {
      var isNow = monthCell.formattedDate.startsWith(currentMonth);
      return /* @__PURE__ */React__default["default"].createElement(calendar_CalendarCellComp["default"], {
        key: monthCellIndex,
        mode: mode,
        theme: theme,
        cell: cell,
        cellData: monthCell,
        cellAppend: cellAppend,
        fillWithZero: fillWithZero,
        isCurrent: monthCell.isCurrent,
        isNow: isNow,
        createCalendarCell: createCalendarCell,
        onCellClick: function onCellClick(event) {
          return clickCell(event, monthCell);
        },
        onCellDoubleClick: function onCellDoubleClick(event) {
          return doubleClickCell(event, monthCell);
        },
        onCellRightClick: function onCellRightClick(event) {
          return rightClickCell(event, monthCell);
        }
      });
    }));
  })))));
});
Calendar.displayName = "Calendar";
Calendar.defaultProps = calendar_defaultProps.calendarDefaultProps;

exports["default"] = Calendar;
