// tdesign-react/lib/table/hooks/useAsyncLoading.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import React from 'react';
import { i as isString_1 } from '../../_chunks/dep-8d5556d7.js';
import classNames from 'classnames';
import { Loading } from '../../loading/index.js';
import useClassName from './useClassName.js';
import { useLocaleReceiver } from '../../locale/LocalReceiver.js';
import '../../_chunks/dep-7909fefa.js';
import '../../_chunks/dep-9ea787ca.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-3160df29.js';
import '../../_chunks/dep-7f6754ac.js';
import '../../loading/Loading.js';
import '../../_util/dom.js';
import 'raf';
import '../../_util/easing.js';
import '../../hooks/useConfig.js';
import '../../_chunks/dep-918e41e2.js';
import '../../_chunks/dep-734a21a0.js';
import '../../_chunks/dep-68f7a306.js';
import '../../_chunks/dep-a74f6fff.js';
import '../../_chunks/dep-d23d346c.js';
import '../../_chunks/dep-0e2dfd87.js';
import '../../_chunks/dep-f50ff346.js';
import '../../_chunks/dep-51592cde.js';
import '../../_chunks/dep-ae94c4e7.js';
import '../../_chunks/dep-2509667d.js';
import '../../_chunks/dep-b817c90e.js';
import '../../_chunks/dep-fe9825b6.js';
import '../../_chunks/dep-ebc05da7.js';
import '../../_chunks/dep-c91b3852.js';
import '../../_chunks/dep-9e44c34c.js';
import '../../_chunks/dep-0c62d47a.js';
import '../../_chunks/dep-e2fe68b7.js';
import '../../_chunks/dep-44a03722.js';
import '../../_chunks/dep-f5616354.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../loading/gradient.js';
import '../../_common/js/loading/circle-adapter.js';
import '../../_common/js/utils/set-style.js';
import '../../_common/js/utils/helper.js';
import '../../_chunks/dep-59b9253f.js';
import '../../_chunks/dep-d1b4ab24.js';
import '../../_chunks/dep-8368bb87.js';
import '../../loading/defaultProps.js';
import '../../loading/plugin.js';

function useAsyncLoading(props) {
  var _ref;

  var _useLocaleReceiver = useLocaleReceiver("table"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 1),
      local = _useLocaleReceiver2[0];

  var _useClassName = useClassName(),
      isLoadingClass = _useClassName.isLoadingClass,
      isLoadMoreClass = _useClassName.isLoadMoreClass,
      asyncLoadingClass = _useClassName.asyncLoadingClass;

  var classes = [asyncLoadingClass, (_ref = {}, _defineProperty(_ref, isLoadingClass, props.asyncLoading === "loading"), _defineProperty(_ref, isLoadMoreClass, props.asyncLoading === "load-more"), _ref)];

  function onLoadClick() {
    var _props$onAsyncLoading;

    if (typeof props.asyncLoading !== "string") return;
    var status = props.asyncLoading === "loading" ? "loading" : "load-more";
    (_props$onAsyncLoading = props.onAsyncLoadingClick) === null || _props$onAsyncLoading === void 0 ? void 0 : _props$onAsyncLoading.call(props, {
      status: status
    });
  }

  function renderAsyncLoading() {
    var asyncLoadingNode = props.asyncLoading;

    if (isString_1(asyncLoadingNode)) {
      var asyncLoading = props.asyncLoading;
      var loadingText = {
        "load-more": local.loadingMoreText,
        loading: local.loadingText
      }[String(asyncLoading)];
      return /* @__PURE__ */React.createElement("div", {
        className: classNames(classes),
        onClick: onLoadClick
      }, /* @__PURE__ */React.createElement(Loading, {
        loading: true,
        size: "small",
        text: loadingText,
        indicator: asyncLoading === "loading"
      }));
    }

    if (asyncLoadingNode) {
      return /* @__PURE__ */React.createElement("div", {
        className: classNames(classes),
        onClick: onLoadClick
      }, asyncLoadingNode);
    }

    return null;
  }

  return {
    renderAsyncLoading: renderAsyncLoading
  };
}

export { useAsyncLoading as default };
