// tdesign-react/lib/table/hooks/useSorter.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _toConsumableArray } from '../../_chunks/dep-59b9253f.js';
import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import React, { useState } from 'react';
import { i as isFunction_1 } from '../../_chunks/dep-d23d346c.js';
import SorterButton from '../SorterButton.js';
import useControlled from '../../hooks/useControlled.js';
import '../../_chunks/dep-7909fefa.js';
import '../../_chunks/dep-d1b4ab24.js';
import '../../_chunks/dep-9ea787ca.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-0e2dfd87.js';
import 'tdesign-icons-react';
import 'classnames';
import '../../hooks/useGlobalIcon.js';
import '../../hooks/useConfig.js';
import '../../_chunks/dep-918e41e2.js';
import '../../_chunks/dep-734a21a0.js';
import '../../_chunks/dep-68f7a306.js';
import '../../_chunks/dep-a74f6fff.js';
import '../../_chunks/dep-f50ff346.js';
import '../../_chunks/dep-51592cde.js';
import '../../_chunks/dep-ae94c4e7.js';
import '../../_chunks/dep-7f6754ac.js';
import '../../_chunks/dep-2509667d.js';
import '../../_chunks/dep-b817c90e.js';
import '../../_chunks/dep-fe9825b6.js';
import '../../_chunks/dep-ebc05da7.js';
import '../../_chunks/dep-3160df29.js';
import '../../_chunks/dep-c91b3852.js';
import '../../_chunks/dep-9e44c34c.js';
import '../../_chunks/dep-0c62d47a.js';
import '../../_chunks/dep-e2fe68b7.js';
import '../../_chunks/dep-44a03722.js';
import '../../_chunks/dep-f5616354.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import './useClassName.js';
import '../../tooltip/index.js';
import '../../tooltip/Tooltip.js';
import '../../_chunks/dep-8368bb87.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../../_util/useAnimation.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../_util/dom.js';
import 'raf';
import '../../_chunks/dep-8d5556d7.js';
import '../../_util/easing.js';
import '../../popup/hooks/useTrigger.js';
import 'react-is';
import '../../popup/utils/ref.js';
import '../../_util/composeRefs.js';
import '../../popup/utils/transition.js';
import '../../_util/useMutationObserver.js';
import '../../_chunks/dep-05d10538.js';
import '../../_chunks/dep-652c1652.js';
import '../../_chunks/dep-b989427d.js';
import '../../_util/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../_chunks/dep-12260644.js';
import '../../_chunks/dep-ec84085d.js';
import '../../_chunks/dep-66273dbc.js';
import '../../_chunks/dep-2ac048c0.js';
import '../../_util/noop.js';
import '../../tooltip/defaultProps.js';
import '../../locale/LocalReceiver.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function useSorter(props) {
  var _useState = useState(),
      _useState2 = _slicedToArray(_useState, 2),
      originalData = _useState2[0],
      setOriginalData = _useState2[1];

  var _useControlled = useControlled(props, "sort", props.onSortChange),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      tSortInfo = _useControlled2[0],
      setTSortInfo = _useControlled2[1];

  var _useControlled3 = useControlled(props, "data", props.onDataChange),
      _useControlled4 = _slicedToArray(_useControlled3, 2),
      tData = _useControlled4[0],
      setTData = _useControlled4[1];

  var sorterFuncMap = getSorterFuncMap(props.columns);

  var sortArray = function () {
    var sort = tSortInfo;
    if (!sort) return [];
    return Array.isArray(sort) ? sort : [sort];
  }();

  var sortMap = function () {
    var sortMap2 = {};
    sortArray.forEach(function (info, index) {
      var sortBy = info.sortBy;
      sortMap2[sortBy] = _objectSpread({
        index: index
      }, info);
    });
    return sortMap2;
  }();

  function getSorterFuncMap(columns) {
    var map = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    for (var i = 0, len = columns.length; i < len; i++) {
      var _col$children;

      var col = columns[i];

      if (isFunction_1(col.sorter)) {
        map[col.colKey] = col.sorter;
      }

      if ((_col$children = col.children) !== null && _col$children !== void 0 && _col$children.length) {
        getSorterFuncMap(col.children, map);
      }
    }

    return map;
  }

  function handleDataSort(sortInfo) {
    var sort = sortInfo;
    if (!sorterFuncMap || !Object.keys(sorterFuncMap).length) return;

    if (!originalData) {
      setOriginalData(tData);
    }

    var isEmptyArraySort = !sort || sort instanceof Array && !sort.length;
    var isEmptyObjectSort = !(sort instanceof Array) && !(sort !== null && sort !== void 0 && sort.sortBy);

    if (isEmptyArraySort || isEmptyObjectSort) {
      setTData(originalData, {
        trigger: "sort"
      });
      return originalData;
    }

    var formattedSort = sort instanceof Array ? sort : [sort];
    var newData = tData.slice().sort(function (a, b) {
      var sortResult = 0;

      for (var i = 0, len = formattedSort.length; i < len; i++) {
        var item = formattedSort[i];
        var sortFunc = sorterFuncMap[item.sortBy];

        if (sortResult === 0 && sortFunc) {
          sortResult = item.descending ? sortFunc(b, a) : sortFunc(a, b);
        } else {
          break;
        }
      }

      return sortResult;
    });
    if (JSON.stringify(newData) === JSON.stringify(tData)) return;
    setTData(newData, {
      trigger: "sort"
    });
    return newData;
  }

  function handleSortHeaderClick(col, p) {
    var _props$onChange;

    var sortInfo;

    if (props.multipleSort) {
      sortInfo = getMultipleNextSort(col, p);
    } else {
      var sort = tSortInfo instanceof Array ? tSortInfo[0] : tSortInfo;
      sortInfo = getSingleNextSort(col, sort, p);
    }

    var newData = handleDataSort(sortInfo);
    var currentData = newData || tData;
    var currentDataSource = currentData;
    setTSortInfo(sortInfo, {
      currentDataSource: currentDataSource,
      col: col
    });
    (_props$onChange = props.onChange) === null || _props$onChange === void 0 ? void 0 : _props$onChange.call(props, {
      sorter: sortInfo
    }, {
      currentData: currentData,
      trigger: "sorter"
    });
  }

  function getSortOrder(descending) {
    if (descending === void 0) return;
    return descending ? "desc" : "asc";
  }

  function getSingleNextSort(col, sortInfo, p) {
    if (sortInfo && sortInfo.sortBy === col.colKey && sortInfo.descending === p.descending) {
      return void 0;
    }

    return {
      sortBy: col.colKey,
      descending: p.descending
    };
  }

  function getMultipleNextSort(col, p) {
    var sort = tSortInfo;
    if (!(sort instanceof Array)) return;
    var colKey = col.colKey;

    var result = _toConsumableArray(sort);

    for (var i = 0, len = sort.length; i < len; i++) {
      if (sort[i].sortBy === colKey) {
        var next = getSingleNextSort(col, sort[i], p);
        next ? result[i] = next : result.splice(i, 1);
        return result;
      }
    }

    result.push({
      sortBy: colKey,
      descending: p.descending
    });
    return result;
  }

  function renderSortIcon(_ref) {
    var _sortMap$col$colKey;

    var col = _ref.col;
    if (!col.sorter) return null;
    var sorterButtonsProps = {
      sortType: col.sortType,
      sortOrder: getSortOrder((_sortMap$col$colKey = sortMap[col.colKey]) === null || _sortMap$col$colKey === void 0 ? void 0 : _sortMap$col$colKey.descending),
      sortIcon: props.sortIcon,
      hideSortTips: props.hideSortTips
    };
    return /* @__PURE__ */React.createElement(SorterButton, _objectSpread(_objectSpread({
      key: "sorter-button-".concat(col.colKey)
    }, sorterButtonsProps), {}, {
      onSortIconClick: function onSortIconClick(_, p) {
        return handleSortHeaderClick(col, p);
      }
    }));
  }

  return {
    renderSortIcon: renderSortIcon
  };
}

export { useSorter as default };
