// tdesign-react/lib/form/hooks/useInstance.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _objectWithoutProperties } from '../../_chunks/dep-8368bb87.js';
import { _ as _asyncToGenerator, r as regenerator } from '../../_chunks/dep-99cf9839.js';
import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import { _ as _toConsumableArray } from '../../_chunks/dep-59b9253f.js';
import { i as isEmpty_1 } from '../../_chunks/dep-7efee863.js';
import { i as isFunction_1 } from '../../_chunks/dep-d23d346c.js';
import { m as merge_1 } from '../../_chunks/dep-918e41e2.js';
import useConfig from '../../hooks/useConfig.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-7909fefa.js';
import '../../_chunks/dep-d1b4ab24.js';
import '../../_chunks/dep-6be1e491.js';
import '../../_chunks/dep-51592cde.js';
import '../../_chunks/dep-9ea787ca.js';
import '../../_chunks/dep-ae94c4e7.js';
import '../../_chunks/dep-7f6754ac.js';
import '../../_chunks/dep-2509667d.js';
import '../../_chunks/dep-a74f6fff.js';
import '../../_chunks/dep-0e2dfd87.js';
import '../../_chunks/dep-ebc05da7.js';
import '../../_chunks/dep-3160df29.js';
import '../../_chunks/dep-9e44c34c.js';
import 'react';
import '../../_chunks/dep-734a21a0.js';
import '../../_chunks/dep-68f7a306.js';
import '../../_chunks/dep-f50ff346.js';
import '../../_chunks/dep-b817c90e.js';
import '../../_chunks/dep-fe9825b6.js';
import '../../_chunks/dep-c91b3852.js';
import '../../_chunks/dep-0c62d47a.js';
import '../../_chunks/dep-e2fe68b7.js';
import '../../_chunks/dep-44a03722.js';
import '../../_chunks/dep-f5616354.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';

var _excluded = ["name"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function getMapValue(name, formMapRef) {
  var mapKeys = _toConsumableArray(formMapRef.current.keys());

  var key = mapKeys.find(function (key2) {
    return String(key2) === String(name);
  });
  return formMapRef.current.get(key);
}

function travelMapFromObject(obj, formMapRef, callback) {
  var _iterator = _createForOfIteratorHelper(formMapRef.current.entries()),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var _step$value = _slicedToArray(_step.value, 2),
          mapName = _step$value[0],
          formItemRef = _step$value[1];

      if (Array.isArray(mapName)) {
        var symbol = Symbol("name");
        var fieldValue = null;

        for (var i = 0; i < mapName.length; i++) {
          var item = mapName[i];

          if (Reflect.has(fieldValue || obj, item)) {
            fieldValue = Reflect.get(fieldValue || obj, item);
          } else {
            fieldValue = symbol;
            break;
          }
        }

        if (fieldValue !== symbol) {
          callback(formItemRef, fieldValue);
        }
      } else if (Reflect.has(obj, mapName)) {
        callback(formItemRef, obj[mapName]);
      }
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }
}

function needValidate(name, fields) {
  if (!fields || !Array.isArray(fields)) return true;
  return fields.indexOf(name) !== -1;
}

function formatValidateResult(validateResultList) {
  var result = validateResultList.reduce(function (r, err) {
    return Object.assign(r || {}, err);
  }, {});
  Object.keys(result).forEach(function (key) {
    if (result[key] === true) {
      delete result[key];
    } else {
      result[key] = result[key].filter(function (fr) {
        return fr.result === false;
      });
    }

    if (result[key] && key.includes(",")) {
      var keyList = key.split(",");
      var fieldValue = keyList.reduceRight(function (prev, curr) {
        return _defineProperty({}, curr, prev);
      }, result[key]);
      merge_1(result, fieldValue);
      delete result[key];
    }
  });
  return isEmpty_1(result) ? true : result;
}

function useInstance(props, formRef, formMapRef) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var FORM_ITEM_CLASS_PREFIX = "".concat(classPrefix, "-form-item__");
  var scrollToFirstError = props.scrollToFirstError,
      _props$preventSubmitD = props.preventSubmitDefault,
      preventSubmitDefault = _props$preventSubmitD === void 0 ? true : _props$preventSubmitD,
      onSubmit = props.onSubmit;

  function getFirstError(r) {
    var _r$firstKey$;

    if (r === true) return;

    var _Object$keys = Object.keys(r),
        _Object$keys2 = _slicedToArray(_Object$keys, 1),
        firstKey = _Object$keys2[0];

    if (scrollToFirstError) {
      scrollTo(".".concat(FORM_ITEM_CLASS_PREFIX + firstKey));
    }

    return (_r$firstKey$ = r[firstKey][0]) === null || _r$firstKey$ === void 0 ? void 0 : _r$firstKey$.message;
  }

  function scrollTo(selector) {
    var _formRef$current$quer, _formRef$current;

    var dom = (_formRef$current$quer = (_formRef$current = formRef.current).querySelector) === null || _formRef$current$quer === void 0 ? void 0 : _formRef$current$quer.call(_formRef$current, selector);
    var behavior = scrollToFirstError;
    dom && dom.scrollIntoView({
      behavior: behavior
    });
  }

  function submit(e) {
    if (preventSubmitDefault) {
      var _e$preventDefault, _e$stopPropagation;

      e === null || e === void 0 ? void 0 : (_e$preventDefault = e.preventDefault) === null || _e$preventDefault === void 0 ? void 0 : _e$preventDefault.call(e);
      e === null || e === void 0 ? void 0 : (_e$stopPropagation = e.stopPropagation) === null || _e$stopPropagation === void 0 ? void 0 : _e$stopPropagation.call(e);
    }

    validate().then(function (r) {
      var firstError = getFirstError(r);
      onSubmit === null || onSubmit === void 0 ? void 0 : onSubmit({
        validateResult: r,
        firstError: firstError,
        e: e
      });
    });
  }

  function validate(_x) {
    return _validate.apply(this, arguments);
  }

  function _validate() {
    _validate = _asyncToGenerator( /*#__PURE__*/regenerator.mark(function _callee(param) {
      var _ref4, fields, _ref4$trigger, trigger, showErrorMessage, list, validateList;

      return regenerator.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _ref4 = param || {}, fields = _ref4.fields, _ref4$trigger = _ref4.trigger, trigger = _ref4$trigger === void 0 ? "all" : _ref4$trigger, showErrorMessage = _ref4.showErrorMessage;
              list = _toConsumableArray(formMapRef.current.values()).filter(function (formItemRef) {
                var _formItemRef$current15, _formItemRef$current16;

                return isFunction_1(formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current15 = formItemRef.current) === null || _formItemRef$current15 === void 0 ? void 0 : _formItemRef$current15.validate) && needValidate(formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current16 = formItemRef.current) === null || _formItemRef$current16 === void 0 ? void 0 : _formItemRef$current16.name, fields);
              }).map(function (formItemRef) {
                return formItemRef === null || formItemRef === void 0 ? void 0 : formItemRef.current.validate(trigger, showErrorMessage);
              });
              _context.next = 4;
              return Promise.all(list);

            case 4:
              validateList = _context.sent;
              return _context.abrupt("return", formatValidateResult(validateList));

            case 6:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));
    return _validate.apply(this, arguments);
  }

  function validateOnly(_x2) {
    return _validateOnly.apply(this, arguments);
  }

  function _validateOnly() {
    _validateOnly = _asyncToGenerator( /*#__PURE__*/regenerator.mark(function _callee2(param) {
      var _ref5, fields, _ref5$trigger, trigger, list, validateList;

      return regenerator.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _ref5 = param || {}, fields = _ref5.fields, _ref5$trigger = _ref5.trigger, trigger = _ref5$trigger === void 0 ? "all" : _ref5$trigger;
              list = _toConsumableArray(formMapRef.current.values()).filter(function (formItemRef) {
                var _formItemRef$current17, _formItemRef$current18;

                return isFunction_1(formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current17 = formItemRef.current) === null || _formItemRef$current17 === void 0 ? void 0 : _formItemRef$current17.validateOnly) && needValidate(formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current18 = formItemRef.current) === null || _formItemRef$current18 === void 0 ? void 0 : _formItemRef$current18.name, fields);
              }).map(function (formItemRef) {
                var _formItemRef$current$6, _formItemRef$current19;

                return formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current$6 = (_formItemRef$current19 = formItemRef.current).validateOnly) === null || _formItemRef$current$6 === void 0 ? void 0 : _formItemRef$current$6.call(_formItemRef$current19, trigger);
              });
              _context2.next = 4;
              return Promise.all(list);

            case 4:
              validateList = _context2.sent;
              return _context2.abrupt("return", formatValidateResult(validateList));

            case 6:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2);
    }));
    return _validateOnly.apply(this, arguments);
  }

  function getFieldValue(name) {
    var _formItemRef$current$, _formItemRef$current;

    if (!name) return null;
    var formItemRef = getMapValue(name, formMapRef);
    return formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current$ = (_formItemRef$current = formItemRef.current).getValue) === null || _formItemRef$current$ === void 0 ? void 0 : _formItemRef$current$.call(_formItemRef$current);
  }

  function getFieldsValue(nameList) {
    var fieldsValue = {};

    if (nameList === true) {
      var _iterator2 = _createForOfIteratorHelper(formMapRef.current.entries()),
          _step2;

      try {
        for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
          var _step2$value = _slicedToArray(_step2.value, 2),
              name = _step2$value[0],
              formItemRef = _step2$value[1];

          if (Array.isArray(name)) {
            var _formItemRef$current$2, _formItemRef$current2;

            var fieldValue = name.reduceRight(function (prev, curr) {
              return _defineProperty({}, curr, prev);
            }, formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current$2 = (_formItemRef$current2 = formItemRef.current).getValue) === null || _formItemRef$current$2 === void 0 ? void 0 : _formItemRef$current$2.call(_formItemRef$current2));
            merge_1(fieldsValue, fieldValue);
          } else {
            var _formItemRef$current$3, _formItemRef$current3;

            fieldsValue[name] = formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current$3 = (_formItemRef$current3 = formItemRef.current).getValue) === null || _formItemRef$current$3 === void 0 ? void 0 : _formItemRef$current$3.call(_formItemRef$current3);
          }
        }
      } catch (err) {
        _iterator2.e(err);
      } finally {
        _iterator2.f();
      }
    } else {
      if (!Array.isArray(nameList)) throw new Error("getFieldsValue \u53C2\u6570\u9700\u8981 Array \u7C7B\u578B");
      nameList.forEach(function (name) {
        var formItemRef = getMapValue(name, formMapRef);

        if (Array.isArray(name)) {
          var _formItemRef$current$4, _formItemRef$current4;

          var _fieldValue = name.reduceRight(function (prev, curr) {
            return _defineProperty({}, curr, prev);
          }, formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current$4 = (_formItemRef$current4 = formItemRef.current).getValue) === null || _formItemRef$current$4 === void 0 ? void 0 : _formItemRef$current$4.call(_formItemRef$current4));

          merge_1(fieldsValue, _fieldValue);
        } else {
          var _formItemRef$current$5, _formItemRef$current5;

          formItemRef && (fieldsValue[name] = formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current$5 = (_formItemRef$current5 = formItemRef.current).getValue) === null || _formItemRef$current$5 === void 0 ? void 0 : _formItemRef$current$5.call(_formItemRef$current5));
        }
      });
    }

    return fieldsValue;
  }

  function setFieldsValue() {
    var fields = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    travelMapFromObject(fields, formMapRef, function (formItemRef, fieldValue) {
      var _formItemRef$current6, _formItemRef$current7;

      formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current6 = formItemRef.current) === null || _formItemRef$current6 === void 0 ? void 0 : (_formItemRef$current7 = _formItemRef$current6.setValue) === null || _formItemRef$current7 === void 0 ? void 0 : _formItemRef$current7.call(_formItemRef$current6, fieldValue);
    });
  }

  function setFields() {
    var fields = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
    if (!Array.isArray(fields)) throw new Error("setFields \u53C2\u6570\u9700\u8981 Array \u7C7B\u578B");
    fields.forEach(function (field) {
      var _formItemRef$current8;

      var name = field.name,
          restFields = _objectWithoutProperties(field, _excluded);

      var formItemRef = getMapValue(name, formMapRef);
      formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current8 = formItemRef.current) === null || _formItemRef$current8 === void 0 ? void 0 : _formItemRef$current8.setField(_objectSpread({}, restFields));
    });
  }

  function reset(params) {
    if (typeof params === "undefined") {
      _toConsumableArray(formMapRef.current.values()).forEach(function (formItemRef) {
        var _formItemRef$current9;

        formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current9 = formItemRef.current) === null || _formItemRef$current9 === void 0 ? void 0 : _formItemRef$current9.resetField();
      });
    } else {
      var _params$type = params.type,
          type = _params$type === void 0 ? "initial" : _params$type,
          _params$fields = params.fields,
          fields = _params$fields === void 0 ? [] : _params$fields;
      fields.forEach(function (name) {
        var _formItemRef$current10;

        var formItemRef = getMapValue(name, formMapRef);
        formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current10 = formItemRef.current) === null || _formItemRef$current10 === void 0 ? void 0 : _formItemRef$current10.resetField(type);
      });
    }
  }

  function clearValidate(fields) {
    if (typeof fields === "undefined") {
      _toConsumableArray(formMapRef.current.values()).forEach(function (formItemRef) {
        var _formItemRef$current11;

        formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current11 = formItemRef.current) === null || _formItemRef$current11 === void 0 ? void 0 : _formItemRef$current11.resetValidate();
      });
    } else {
      if (!Array.isArray(fields)) throw new Error("clearValidate \u53C2\u6570\u9700\u8981 Array \u7C7B\u578B");
      fields.forEach(function (name) {
        var _formItemRef$current12;

        var formItemRef = getMapValue(name, formMapRef);
        formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current12 = formItemRef.current) === null || _formItemRef$current12 === void 0 ? void 0 : _formItemRef$current12.resetValidate();
      });
    }
  }

  function setValidateMessage(message) {
    travelMapFromObject(message, formMapRef, function (formItemRef, fieldValue) {
      var _formItemRef$current13, _formItemRef$current14;

      formItemRef === null || formItemRef === void 0 ? void 0 : (_formItemRef$current13 = formItemRef.current) === null || _formItemRef$current13 === void 0 ? void 0 : (_formItemRef$current14 = _formItemRef$current13.setValidateMessage) === null || _formItemRef$current14 === void 0 ? void 0 : _formItemRef$current14.call(_formItemRef$current13, fieldValue);
    });
  }

  return {
    submit: submit,
    reset: reset,
    validate: validate,
    validateOnly: validateOnly,
    clearValidate: clearValidate,
    setFields: setFields,
    setFieldsValue: setFieldsValue,
    setValidateMessage: setValidateMessage,
    getFieldValue: getFieldValue,
    getFieldsValue: getFieldsValue,
    currentElement: formRef.current
  };
}

export { useInstance as default };
