// tdesign-react/lib/popconfirm/Popconfirm.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import React, { forwardRef } from 'react';
import classNames from 'classnames';
import { Popup } from '../popup/index.js';
import useConfig from '../hooks/useConfig.js';
import useControlled from '../hooks/useControlled.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import Popcontent from './Popcontent.js';
import { popconfirmDefaultProps } from './defaultProps.js';
import '../_chunks/dep-7909fefa.js';
import '../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../_util/useAnimation.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../common/Portal.js';
import 'react-dom';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-8d5556d7.js';
import '../_util/easing.js';
import '../popup/hooks/useTrigger.js';
import 'react-is';
import '../popup/utils/ref.js';
import '../_util/composeRefs.js';
import '../popup/utils/transition.js';
import '../_util/useMutationObserver.js';
import '../_chunks/dep-05d10538.js';
import '../_chunks/dep-652c1652.js';
import '../_chunks/dep-b989427d.js';
import '../_util/useWindowSize.js';
import '../popup/defaultProps.js';
import '../_chunks/dep-12260644.js';
import '../_chunks/dep-ec84085d.js';
import '../_chunks/dep-66273dbc.js';
import '../_chunks/dep-2ac048c0.js';
import '../_util/noop.js';
import 'tdesign-icons-react';
import '../button/index.js';
import '../button/Button.js';
import '../_chunks/dep-8368bb87.js';
import '../_util/useRipple.js';
import '../_util/setStyle.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-d1b4ab24.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import '../button/defaultProps.js';
import '../hooks/useGlobalIcon.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Popconfirm = /*#__PURE__*/forwardRef(function (props, ref) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useLocaleReceiver = useLocaleReceiver("popconfirm"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var cancelContent = typeof local.cancel === "string" ? local.cancel : local.cancel.content;
  var confirmContent = typeof local.confirm === "string" ? local.confirm : local.confirm.content;
  var _props$cancelBtn = props.cancelBtn,
      cancelBtn = _props$cancelBtn === void 0 ? t(cancelContent) : _props$cancelBtn,
      _props$confirmBtn = props.confirmBtn,
      confirmBtn = _props$confirmBtn === void 0 ? t(confirmContent) : _props$confirmBtn;

  var _useControlled = useControlled(props, "visible", props.onVisibleChange),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      visible = _useControlled2[0],
      setVisible = _useControlled2[1];

  return /* @__PURE__ */React.createElement(Popup, _objectSpread(_objectSpread({
    ref: ref
  }, props), {}, {
    visible: visible,
    trigger: "click",
    onVisibleChange: function onVisibleChange(visible2) {
      return setVisible(visible2);
    },
    overlayClassName: classNames("".concat(classPrefix, "-popconfirm")),
    content: /* @__PURE__ */React.createElement(Popcontent, _objectSpread(_objectSpread({
      cancelBtn: cancelBtn,
      confirmBtn: confirmBtn
    }, props), {}, {
      onClose: function onClose(context) {
        return setVisible(false, context);
      }
    }))
  }, props.popupProps));
});
Popconfirm.displayName = "Popconfirm";
Popconfirm.defaultProps = popconfirmDefaultProps;

export { Popconfirm as default };
