// tdesign-react/lib/upload/themes/CustomFile.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import classNames from 'classnames';
import React from 'react';
import useDrag from '../hooks/useDrag.js';
import '../../_chunks/dep-9615f03f.js';
import '../../_chunks/dep-7909fefa.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var CustomFile = function CustomFile(props) {
  var classPrefix = props.classPrefix,
      displayFiles = props.displayFiles;
  var drag = useDrag(props.dragEvents);
  var dragActive = drag.dragActive;
  var dragEvents = props.draggable ? {
    onDrop: drag.handleDrop,
    onDragEnter: drag.handleDragenter,
    onDragOver: drag.handleDragover,
    onDragLeave: drag.handleDragleave
  } : {};

  var renderDragContent = function renderDragContent() {
    var _props$dragContent, _props$trigger;

    return /* @__PURE__ */React.createElement("div", _objectSpread(_objectSpread({
      className: "".concat(classPrefix, "-upload__dragger ").concat(classPrefix, "-upload__dragger-center")
    }, dragEvents), {}, {
      onClick: props.triggerUpload
    }), /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-upload__trigger")
    }, ((_props$dragContent = props.dragContent) === null || _props$dragContent === void 0 ? void 0 : _props$dragContent.call(props, {
      dragActive: dragActive,
      files: displayFiles
    })) || ((_props$trigger = props.trigger) === null || _props$trigger === void 0 ? void 0 : _props$trigger.call(props, {
      dragActive: dragActive,
      files: displayFiles
    })) || props.childrenNode));
  };

  return /* @__PURE__ */React.createElement(React.Fragment, null, props.draggable ? renderDragContent() : /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-upload__trigger"),
    onClick: props.triggerUpload
  }, props.childrenNode || props.children), props.tips && /* @__PURE__ */React.createElement("small", {
    className: classNames(props.tipsClasses)
  }, props.tips));
};

CustomFile.displayName = "CustomFile";

export { CustomFile as default };
