// tdesign-react/lib/table/SorterButton.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import React from 'react';
import { ChevronDownIcon } from 'tdesign-icons-react';
import classNames from 'classnames';
import useGlobalIcon from '../hooks/useGlobalIcon.js';
import useClassName from './hooks/useClassName.js';
import { Tooltip } from '../tooltip/index.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import '../_chunks/dep-7909fefa.js';
import '../hooks/useConfig.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../tooltip/Tooltip.js';
import '../_chunks/dep-8368bb87.js';
import '../popup/index.js';
import '../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../hooks/useControlled.js';
import '../_chunks/dep-12260644.js';
import '../_chunks/dep-ec84085d.js';
import '../_chunks/dep-66273dbc.js';
import '../_chunks/dep-2ac048c0.js';
import '../_chunks/dep-b989427d.js';
import '../_util/noop.js';
import '../_util/useAnimation.js';
import '../common/Portal.js';
import 'react-dom';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-8d5556d7.js';
import '../_util/easing.js';
import '../popup/hooks/useTrigger.js';
import 'react-is';
import '../popup/utils/ref.js';
import '../_util/composeRefs.js';
import '../popup/utils/transition.js';
import '../_util/useMutationObserver.js';
import '../_chunks/dep-05d10538.js';
import '../_chunks/dep-652c1652.js';
import '../_util/useWindowSize.js';
import '../popup/defaultProps.js';
import '../tooltip/defaultProps.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function SorterButton(props) {
  var _props$sortType = props.sortType,
      sortType = _props$sortType === void 0 ? "all" : _props$sortType;

  var _useGlobalIcon = useGlobalIcon({
    ChevronDownIcon: ChevronDownIcon
  }),
      ChevronDownIcon$1 = _useGlobalIcon.ChevronDownIcon;

  var _useClassName = useClassName(),
      tableSortClasses = _useClassName.tableSortClasses,
      negativeRotate180 = _useClassName.negativeRotate180;

  var _useLocaleReceiver = useLocaleReceiver("table"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var allowSortTypes = sortType === "all" ? ["asc", "desc"] : [sortType];

  var onSortIconClick = function onSortIconClick(e, direction) {
    props === null || props === void 0 ? void 0 : props.onSortIconClick(e, {
      descending: direction === "desc"
    });
  };

  function getSortIcon(direction, activeClass) {
    var defaultIcon = t(locale.sortIcon) || /* @__PURE__ */React.createElement(ChevronDownIcon$1, null);
    var icon = props.sortIcon || defaultIcon;
    var sortClassName = [activeClass, tableSortClasses.sortIcon, tableSortClasses.iconDirection[direction], _defineProperty({}, negativeRotate180, direction === "asc")];
    return /* @__PURE__ */React.createElement("span", {
      key: direction,
      className: classNames(sortClassName),
      onClick: function onClick(e) {
        return onSortIconClick(e, direction);
      }
    }, icon);
  }

  var classes = [tableSortClasses.trigger, _defineProperty({}, tableSortClasses.doubleIcon, allowSortTypes.length > 1)];
  var tooltips = {
    asc: locale.sortAscendingOperationText,
    desc: locale.sortDescendingOperationText
  };
  var sortButton = allowSortTypes.map(function (direction) {
    var _props$hideSortTips;

    var activeClass = direction === props.sortOrder ? tableSortClasses.iconActive : tableSortClasses.iconDefault;
    var cancelTips = locale.sortCancelOperationText;
    var tips = direction === props.sortOrder ? cancelTips : tooltips[direction];
    if ((_props$hideSortTips = props.hideSortTips) !== null && _props$hideSortTips !== void 0 ? _props$hideSortTips : locale.hideSortTips) return getSortIcon(direction, activeClass);
    return /* @__PURE__ */React.createElement(Tooltip, _objectSpread(_objectSpread({
      content: tips,
      key: direction,
      placement: "right"
    }, props.tooltipProps), {}, {
      showArrow: false
    }), getSortIcon(direction, activeClass));
  });
  return /* @__PURE__ */React.createElement("div", {
    className: classNames(classes)
  }, sortButton);
}
SorterButton.displayName = "SorterButton";

export { SorterButton as default };
