// tdesign-react/lib/table/THead.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _typeof } from '../_chunks/dep-4e0da095.js';
import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import React, { useRef } from 'react';
import { i as isFunction_1 } from '../_chunks/dep-d23d346c.js';
import classNames from 'classnames';
import { getColumnFixedStyles } from './hooks/useFixed.js';
import useClassName from './hooks/useClassName.js';
import useConfig from '../hooks/useConfig.js';
import { renderTitle } from './hooks/useTableHeader.js';
import Ellipsis from './Ellipsis.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-9615f03f.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-1f300d2c.js';
import '../_chunks/dep-e103618e.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-b989427d.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-2ac048c0.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-d1b4ab24.js';
import '../_chunks/dep-8368bb87.js';
import '../_common/js/log/log.js';
import '../_common/js/utils/getScrollbarWidth.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-8d5556d7.js';
import '../_util/easing.js';
import '../hooks/useDebounce.js';
import 'lodash';
import '../_util/usePersistFn.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import './hooks/useMultiHeader.js';
import '../tooltip/index.js';
import '../tooltip/Tooltip.js';
import '../popup/index.js';
import '../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../hooks/useControlled.js';
import '../_chunks/dep-12260644.js';
import '../_chunks/dep-ec84085d.js';
import '../_chunks/dep-66273dbc.js';
import '../_util/noop.js';
import '../_util/useAnimation.js';
import '../common/Portal.js';
import 'react-dom';
import '../popup/hooks/useTrigger.js';
import 'react-is';
import '../popup/utils/ref.js';
import '../_util/composeRefs.js';
import '../popup/utils/transition.js';
import '../_util/useMutationObserver.js';
import '../_chunks/dep-05d10538.js';
import '../_chunks/dep-652c1652.js';
import '../_util/useWindowSize.js';
import '../popup/defaultProps.js';
import '../tooltip/defaultProps.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function THead(props) {
  var _ref;

  var columnResizeParams = props.columnResizeParams;
  var theadRef = useRef(null);
  var classnames = useClassName();
  var tableHeaderClasses = classnames.tableHeaderClasses,
      tableBaseClass = classnames.tableBaseClass;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var theadClasses = [tableHeaderClasses.header, (_ref = {}, _defineProperty(_ref, tableHeaderClasses.fixed, props.isFixedHeader), _defineProperty(_ref, tableBaseClass.bordered, props.bordered && props.isMultipleHeader), _defineProperty(_ref, tableHeaderClasses.multipleHeader, props.isMultipleHeader), _ref)];

  var renderThNodeList = function renderThNodeList(rowAndColFixedPosition, thWidthList) {
    var thBorderMap = /* @__PURE__ */new Map();
    var thRowspanAndColspan = props.spansAndLeafNodes.rowspanAndColspanMap;
    return props.thList.map(function (row, rowIndex) {
      var thRow = row.map(function (col, index) {
        var _ref2, _col$children;

        var rowspanAndColspan = thRowspanAndColspan.get(col);

        if (index === 0 && rowspanAndColspan.rowspan > 1) {
          for (var j = rowIndex + 1; j < rowIndex + rowspanAndColspan.rowspan; j++) {
            thBorderMap.set(props.thList[j][0], true);
          }
        }

        var thStyles = getColumnFixedStyles(col, index, rowAndColFixedPosition, classnames.tableColFixedClasses);
        var colParams = {
          col: col,
          colIndex: index,
          row: {},
          rowIndex: -1
        };
        var customClasses = isFunction_1(col.className) ? col.className(_objectSpread(_objectSpread({}, colParams), {}, {
          type: "th"
        })) : col.className;
        var thClasses = [thStyles.classes, customClasses, (_ref2 = {}, _defineProperty(_ref2, tableHeaderClasses.thBordered, thBorderMap.get(col)), _defineProperty(_ref2, "".concat(classPrefix, "-table__th-").concat(col.colKey), col.colKey), _defineProperty(_ref2, classnames.tdAlignClasses[col.align], col.align && col.align !== "left"), _ref2)];
        var withoutChildren = !((_col$children = col.children) !== null && _col$children !== void 0 && _col$children.length);
        var width = withoutChildren && thWidthList !== null && thWidthList !== void 0 && thWidthList[col.colKey] ? "".concat(thWidthList === null || thWidthList === void 0 ? void 0 : thWidthList[col.colKey], "px") : void 0;

        var styles = _objectSpread(_objectSpread({}, thStyles.style || {}), {}, {
          width: width
        });

        var innerTh = renderTitle(col, index);
        if (!col.colKey) return null;
        var resizeColumnListener = props.resizable ? {
          onMouseDown: function onMouseDown(e) {
            var _columnResizeParams$o;

            return columnResizeParams === null || columnResizeParams === void 0 ? void 0 : (_columnResizeParams$o = columnResizeParams.onColumnMousedown) === null || _columnResizeParams$o === void 0 ? void 0 : _columnResizeParams$o.call(columnResizeParams, e, col);
          },
          onMouseMove: function onMouseMove(e) {
            var _columnResizeParams$o2;

            return columnResizeParams === null || columnResizeParams === void 0 ? void 0 : (_columnResizeParams$o2 = columnResizeParams.onColumnMouseover) === null || _columnResizeParams$o2 === void 0 ? void 0 : _columnResizeParams$o2.call(columnResizeParams, e);
          }
        } : {};
        var content = isFunction_1(col.ellipsisTitle) ? col.ellipsisTitle({
          col: col,
          colIndex: index
        }) : void 0;
        return /* @__PURE__ */React.createElement("th", _objectSpread(_objectSpread({
          key: col.colKey,
          "data-colkey": col.colKey,
          className: classNames(thClasses),
          style: styles
        }, {
          rowSpan: rowspanAndColspan.rowspan,
          colSpan: rowspanAndColspan.colspan
        }), resizeColumnListener), /* @__PURE__ */React.createElement("div", {
          className: tableBaseClass.thCellInner
        }, col.ellipsis && col.ellipsisTitle !== false && col.ellipsisTitle !== null ? /* @__PURE__ */React.createElement(Ellipsis, {
          placement: "bottom",
          attach: theadRef.current ? function () {
            return theadRef.current.parentNode.parentNode;
          } : void 0,
          popupContent: content,
          tooltipProps: _typeof(col.ellipsisTitle) === "object" ? col.ellipsisTitle : void 0
        }, innerTh) : innerTh));
      });
      return /* @__PURE__ */React.createElement("tr", {
        key: rowIndex
      }, thRow);
    });
  };

  return /* @__PURE__ */React.createElement("thead", {
    ref: theadRef,
    className: classNames(theadClasses)
  }, renderThNodeList(props.rowAndColFixedPosition, props.thWidthList));
}

export { THead as default };
