// tdesign-react/cjs/dropdown/DropdownMenu.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var hooks_useConfig = require('../hooks/useConfig.js');
var divider_index = require('../divider/index.js');
var dropdown_DropdownItem = require('./DropdownItem.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../divider/Divider.js');
require('../_chunks/dep-1a0462cc.js');
require('../divider/defaultProps.js');
require('../_util/useRipple.js');
require('../_util/useAnimation.js');
require('../_util/setStyle.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('./defaultProps.js');
require('../_util/helper.js');
require('../_chunks/dep-9c5d4f93.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-0ebc045a.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-09805e35.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var DropdownMenu = function DropdownMenu(props) {
  var _props$options = props.options,
      options = _props$options === void 0 ? [] : _props$options,
      _props$maxHeight = props.maxHeight,
      maxHeight = _props$maxHeight === void 0 ? 300 : _props$maxHeight,
      _props$minColumnWidth = props.minColumnWidth,
      minColumnWidth = _props$minColumnWidth === void 0 ? 10 : _props$minColumnWidth,
      _props$maxColumnWidth = props.maxColumnWidth,
      maxColumnWidth = _props$maxColumnWidth === void 0 ? 160 : _props$maxColumnWidth,
      direction = props.direction;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var dropdownClass = "".concat(classPrefix, "-dropdown");
  var dropdownMenuClass = "".concat(dropdownClass, "__menu");

  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
    ChevronRightIcon: tdesignIconsReact.ChevronRightIcon,
    ChevronLeftIcon: tdesignIconsReact.ChevronLeftIcon
  }),
      ChevronRightIcon = _useGlobalIcon.ChevronRightIcon,
      ChevronLeftIcon = _useGlobalIcon.ChevronLeftIcon;

  var handleItemClick = function handleItemClick(options2) {
    var _data$onClick, _props$onClick;

    var data = options2.data,
        context = options2.context;
    data === null || data === void 0 ? void 0 : (_data$onClick = data.onClick) === null || _data$onClick === void 0 ? void 0 : _data$onClick.call(data, data, context);
    (_props$onClick = props.onClick) === null || _props$onClick === void 0 ? void 0 : _props$onClick.call(props, data, context);
  };

  var renderOptions = function renderOptions(data) {
    var _data$forEach;

    var arr = [];
    var renderContent;
    (_data$forEach = data.forEach) === null || _data$forEach === void 0 ? void 0 : _data$forEach.call(data, function (menu, idx) {
      var optionItem = _objectSpread({}, menu);

      if (optionItem.children) {
        var _classNames;

        optionItem.children = renderOptions(optionItem.children);
        renderContent = /* @__PURE__ */React__default["default"].createElement("div", {
          key: idx
        }, /* @__PURE__ */React__default["default"].createElement(dropdown_DropdownItem["default"], {
          className: classNames__default["default"]("".concat(dropdownClass, "__item"), "".concat(dropdownClass, "__item--suffix")),
          value: optionItem.value,
          theme: optionItem.theme,
          active: optionItem.active,
          prefixIcon: optionItem.prefixIcon,
          disabled: optionItem.disabled,
          minColumnWidth: minColumnWidth,
          maxColumnWidth: maxColumnWidth,
          isSubmenu: true
        }, /* @__PURE__ */React__default["default"].createElement("div", {
          className: "".concat(dropdownClass, "__item-content")
        }, direction === "right" ? /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("span", {
          className: "".concat(dropdownClass, "__item-text")
        }, optionItem.content), /* @__PURE__ */React__default["default"].createElement(ChevronRightIcon, {
          className: "".concat(dropdownClass, "__item-direction"),
          size: "16"
        })) : /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement(ChevronLeftIcon, {
          className: "".concat(dropdownClass, "__item-direction"),
          size: "16"
        }), /* @__PURE__ */React__default["default"].createElement("span", {
          className: "".concat(dropdownClass, "__item-text")
        }, optionItem.content))), /* @__PURE__ */React__default["default"].createElement("div", {
          className: classNames__default["default"]("".concat(dropdownClass, "__submenu"), (_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(dropdownClass, "__submenu--disabled"), optionItem.disabled), defineProperty._defineProperty(_classNames, "".concat(dropdownClass, "__submenu--").concat(direction), direction), _classNames)),
          style: {
            top: "".concat(idx * 30, "px")
          }
        }, /* @__PURE__ */React__default["default"].createElement("ul", null, optionItem.children))), optionItem.divider ? /* @__PURE__ */React__default["default"].createElement(divider_index.Divider, null) : null);
      } else {
        renderContent = /* @__PURE__ */React__default["default"].createElement("div", {
          key: idx
        }, /* @__PURE__ */React__default["default"].createElement(dropdown_DropdownItem["default"], {
          className: "".concat(dropdownClass, "__item"),
          value: optionItem.value,
          theme: optionItem.theme,
          active: optionItem.active,
          prefixIcon: optionItem.prefixIcon,
          disabled: optionItem.disabled,
          minColumnWidth: minColumnWidth,
          maxColumnWidth: maxColumnWidth,
          onClick: optionItem.disabled || optionItem.children ? function () {
            return null;
          } : function (value, context) {
            return handleItemClick({
              data: optionItem,
              context: context
            });
          }
        }, /* @__PURE__ */React__default["default"].createElement("span", {
          className: "".concat(dropdownClass, "-text")
        }, optionItem.content)), optionItem.divider ? /* @__PURE__ */React__default["default"].createElement(divider_index.Divider, null) : null);
      }

      arr.push(renderContent);
    });
    return arr;
  };

  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](dropdownMenuClass, "".concat(dropdownMenuClass, "--").concat(direction)),
    style: {
      maxHeight: "".concat(maxHeight, "px")
    }
  }, renderOptions(options));
};

DropdownMenu.displayName = "DropdownMenu";

exports["default"] = DropdownMenu;
