// tdesign-react/es/table/hooks/useRowSelect.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _toConsumableArray } from '../../_chunks/dep-59b9253f.js';
import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import React, { useState, useEffect } from 'react';
import { i as intersection_1 } from '../../_chunks/dep-6120e322.js';
import { g as get_1 } from '../../_chunks/dep-ddb0f1c1.js';
import { i as isFunction_1 } from '../../_chunks/dep-e430e2ee.js';
import useControlled from '../../hooks/useControlled.js';
import { isRowSelectedDisabled } from '../utils.js';
import { Checkbox } from '../../checkbox/index.js';
import { Radio } from '../../radio/index.js';
import '../../_common/js/log/index.js';
import log from '../../_common/js/log/log.js';
import '../../_chunks/dep-7909fefa.js';
import '../../_chunks/dep-d1b4ab24.js';
import '../../_chunks/dep-1cce5a60.js';
import '../../_chunks/dep-c8bfa11e.js';
import '../../_chunks/dep-4a64abb1.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-264b0e55.js';
import '../../_chunks/dep-199b1581.js';
import '../../_chunks/dep-e237f0f2.js';
import '../../_chunks/dep-6011038b.js';
import '../../_chunks/dep-4e2181de.js';
import '../../_chunks/dep-c197ce4f.js';
import '../../_chunks/dep-5a2778c1.js';
import '../../_chunks/dep-e54bb0dc.js';
import '../../_chunks/dep-b2626e28.js';
import '../../_chunks/dep-a2b97df8.js';
import '../../_chunks/dep-4ed2dce5.js';
import '../../_chunks/dep-7ff8b18e.js';
import '../../_chunks/dep-fdd1d1a6.js';
import '../../_chunks/dep-2afd805c.js';
import '../../_chunks/dep-ebf3a272.js';
import '../../_chunks/dep-96aef4a9.js';
import '../../_chunks/dep-93a10d8c.js';
import '../../_chunks/dep-f54ff1e5.js';
import '../../_chunks/dep-e1d4dd61.js';
import '../../_chunks/dep-236ef238.js';
import '../../_util/noop.js';
import '../../_chunks/dep-66738e15.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../common/Check.js';
import '../../_chunks/dep-8368bb87.js';
import 'classnames';
import '../../_util/helper.js';
import '../../_chunks/dep-fc5ba3d5.js';
import '../../hooks/useConfig.js';
import '../../_chunks/dep-13a70a46.js';
import '../../_chunks/dep-dc1ce7ed.js';
import '../../_chunks/dep-7aeeb170.js';
import '../../_chunks/dep-c8db35c0.js';
import '../../_chunks/dep-557a1b0f.js';
import '../../_chunks/dep-f5693bf6.js';
import '../../_chunks/dep-6a2482c6.js';
import '../../_chunks/dep-a980ac05.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import '../../_chunks/dep-a5f8b6ec.js';
import '../../checkbox/defaultProps.js';
import '../../checkbox/style/css.js';
import '../../checkbox/type.js';
import '../../_chunks/dep-5bd899df.js';
import '../../_util/useCommonClassName.js';
import '../../radio/defaultProps.js';
import '../../_util/useMutationObserver.js';
import '../../_chunks/dep-236b0359.js';
import '../../_chunks/dep-6d5c0a90.js';
import '../../radio/style/css.js';
import '../../radio/type.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var selectedRowDataMap = /* @__PURE__ */new Map();
function useRowSelect(props, tableSelectedClasses) {
  var selectedRowKeys = props.selectedRowKeys,
      columns = props.columns,
      data = props.data,
      rowKey = props.rowKey,
      indeterminateSelectedRowKeys = props.indeterminateSelectedRowKeys;

  var _useState = useState(),
      _useState2 = _slicedToArray(_useState, 2),
      selectedRowClassNames = _useState2[0],
      setSelectedRowClassNames = _useState2[1];

  var _useControlled = useControlled(props, "selectedRowKeys", props.onSelectChange, {
    defaultSelectedRowKeys: props.defaultSelectedRowKeys || []
  }),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      tSelectedRowKeys = _useControlled2[0],
      setTSelectedRowKeys = _useControlled2[1];

  var selectColumn = columns.find(function (_ref) {
    var type = _ref.type;
    return ["multiple", "single"].includes(type);
  });
  var canSelectedRows = data.filter(function (row, rowIndex) {
    return !isDisabled(row, rowIndex);
  });
  var intersectionKeys = intersection_1(tSelectedRowKeys, canSelectedRows.map(function (t) {
    return get_1(t, rowKey || "id");
  }));
  useEffect(function () {
    if (!selectColumn && (!tSelectedRowKeys || !tSelectedRowKeys.length)) return;

    var disabledRowFunc = function disabledRowFunc(p) {
      return selectColumn.disabled(p) ? tableSelectedClasses.disabled : "";
    };

    var disabledRowClass = selectColumn !== null && selectColumn !== void 0 && selectColumn.disabled ? disabledRowFunc : void 0;
    var selected = new Set(tSelectedRowKeys);

    var selectedRowClassFunc = function selectedRowClassFunc(_ref2) {
      var row = _ref2.row;
      var rowId = get_1(row, rowKey || "id");
      return selected.has(rowId) ? tableSelectedClasses.selected : "";
    };

    var selectedRowClass = selected.size ? selectedRowClassFunc : void 0;
    setSelectedRowClassNames([disabledRowClass, selectedRowClass]);
  }, [data, columns, tSelectedRowKeys, selectColumn, rowKey]);

  function isDisabled(row, rowIndex) {
    return isRowSelectedDisabled(selectColumn, row, rowIndex);
  }

  function getSelectedHeader() {
    return function () {
      var isIndeterminate = intersectionKeys.length > 0 && intersectionKeys.length < canSelectedRows.length;
      var isChecked = intersectionKeys.length !== 0 && canSelectedRows.length !== 0 && intersectionKeys.length === canSelectedRows.length;
      return /* @__PURE__ */React.createElement(Checkbox, {
        checked: isChecked,
        indeterminate: isIndeterminate,
        disabled: !canSelectedRows.length,
        onChange: handleSelectAll
      });
    };
  }

  function renderSelectCell(p) {
    var column = p.col,
        _p$row = p.row,
        row = _p$row === void 0 ? {} : _p$row,
        rowIndex = p.rowIndex;
    var checked = tSelectedRowKeys.includes(get_1(row, rowKey || "id"));
    var disabled = typeof column.disabled === "function" ? column.disabled({
      row: row,
      rowIndex: rowIndex
    }) : column.disabled;
    var checkProps = isFunction_1(column.checkProps) ? column.checkProps({
      row: row,
      rowIndex: rowIndex
    }) : column.checkProps;

    var selectBoxProps = _objectSpread(_objectSpread({
      checked: checked,
      disabled: disabled
    }, checkProps), {}, {
      onChange: function onChange() {
        handleSelectChange(row);
      }
    });

    var onCheckClick = function onCheckClick(e) {
      e === null || e === void 0 ? void 0 : e.stopPropagation();
    };

    if (column.type === "single") return /* @__PURE__ */React.createElement(Radio, _objectSpread(_objectSpread({}, selectBoxProps), {}, {
      onClick: onCheckClick
    }));

    if (column.type === "multiple") {
      var isIndeterminate = indeterminateSelectedRowKeys !== null && indeterminateSelectedRowKeys !== void 0 && indeterminateSelectedRowKeys.length ? indeterminateSelectedRowKeys.includes(get_1(row, rowKey)) : false;
      return /* @__PURE__ */React.createElement(Checkbox, _objectSpread(_objectSpread({
        indeterminate: isIndeterminate
      }, selectBoxProps), {}, {
        onClick: onCheckClick
      }));
    }

    return null;
  }

  function handleSelectChange() {
    var row = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var selectedRowKeys2 = _toConsumableArray(tSelectedRowKeys);

    var reRowKey = rowKey || "id";
    var id = get_1(row, reRowKey);
    selectedRowDataMap.set(id, row);
    var selectedRowIndex = selectedRowKeys2.indexOf(id);
    var isExisted = selectedRowIndex !== -1;

    if (selectColumn.type === "multiple") {
      isExisted ? selectedRowKeys2.splice(selectedRowIndex, 1) : selectedRowKeys2.push(id);
    } else if (selectColumn.type === "single") {
      selectedRowKeys2 = !isExisted ? [id] : [];
    } else {
      log.warn("Table", "`column.type` must be one of `multiple` and `single`");
      return;
    }

    setTSelectedRowKeys(selectedRowKeys2, {
      selectedRowData: selectedRowKeys2.map(function (t) {
        return selectedRowDataMap.get(t);
      }),
      currentRowKey: id,
      currentRowData: row,
      type: isExisted ? "uncheck" : "check"
    });
  }

  function handleSelectAll(checked) {
    var reRowKey = rowKey || "id";
    var canSelectedRowKeys = canSelectedRows.map(function (record) {
      return get_1(record, reRowKey);
    });
    var disabledSelectedRowKeys = (selectedRowKeys === null || selectedRowKeys === void 0 ? void 0 : selectedRowKeys.filter(function (id) {
      return !canSelectedRowKeys.includes(id);
    })) || [];
    var allIds = checked ? [].concat(_toConsumableArray(disabledSelectedRowKeys), _toConsumableArray(canSelectedRowKeys)) : _toConsumableArray(disabledSelectedRowKeys);
    setTSelectedRowKeys(allIds, {
      selectedRowData: checked ? allIds.map(function (t) {
        return selectedRowDataMap.get(t);
      }) : [],
      type: checked ? "check" : "uncheck",
      currentRowKey: "CHECK_ALL_BOX"
    });
  }

  function formatToRowSelectColumn(col) {
    var isSelection = ["multiple", "single"].includes(col.type);
    if (!isSelection) return col;
    return _objectSpread(_objectSpread({}, col), {}, {
      width: col.width || 64,
      className: tableSelectedClasses.checkCell,
      cell: function cell(p) {
        return renderSelectCell(p);
      },
      title: col.type === "multiple" ? getSelectedHeader() : ""
    });
  }

  useEffect(function () {
    for (var i = 0, len = data.length; i < len; i++) {
      selectedRowDataMap.set(get_1(data[i], rowKey || "id"), data[i]);
    }
  }, [data, rowKey]);
  return {
    selectedRowClassNames: selectedRowClassNames,
    formatToRowSelectColumn: formatToRowSelectColumn
  };
}

export { useRowSelect as default };
