// tdesign-react/es/table/hooks/useRowExpand.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import { _ as _toConsumableArray } from '../../_chunks/dep-59b9253f.js';
import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import React from 'react';
import { ChevronRightCircleIcon } from 'tdesign-icons-react';
import { g as get_1 } from '../../_chunks/dep-ddb0f1c1.js';
import { i as isFunction_1 } from '../../_chunks/dep-e430e2ee.js';
import classNames from 'classnames';
import useClassName from './useClassName.js';
import useControlled from '../../hooks/useControlled.js';
import useGlobalIcon from '../../hooks/useGlobalIcon.js';
import { useLocaleReceiver } from '../../locale/LocalReceiver.js';
import '../../_chunks/dep-7909fefa.js';
import '../../_chunks/dep-d1b4ab24.js';
import '../../_chunks/dep-1cce5a60.js';
import '../../_chunks/dep-93a10d8c.js';
import '../../_chunks/dep-264b0e55.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-199b1581.js';
import '../../_chunks/dep-4a64abb1.js';
import '../../_chunks/dep-e237f0f2.js';
import '../../_chunks/dep-4e2181de.js';
import '../../_chunks/dep-c197ce4f.js';
import '../../_chunks/dep-5a2778c1.js';
import '../../_chunks/dep-e54bb0dc.js';
import '../../_chunks/dep-b2626e28.js';
import '../../_chunks/dep-a2b97df8.js';
import '../../_chunks/dep-c8bfa11e.js';
import '../../hooks/useConfig.js';
import '../../_chunks/dep-13a70a46.js';
import '../../_chunks/dep-dc1ce7ed.js';
import '../../_chunks/dep-7aeeb170.js';
import '../../_chunks/dep-2afd805c.js';
import '../../_chunks/dep-4ed2dce5.js';
import '../../_chunks/dep-c8db35c0.js';
import '../../_chunks/dep-96aef4a9.js';
import '../../_chunks/dep-557a1b0f.js';
import '../../_chunks/dep-f5693bf6.js';
import '../../_chunks/dep-fdd1d1a6.js';
import '../../_chunks/dep-7ff8b18e.js';
import '../../_chunks/dep-ebf3a272.js';
import '../../_chunks/dep-6a2482c6.js';
import '../../_chunks/dep-a980ac05.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import '../../_chunks/dep-f54ff1e5.js';
import '../../_chunks/dep-e1d4dd61.js';
import '../../_chunks/dep-236ef238.js';
import '../../_util/noop.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';

function useRowExpand(props) {
  var _props$columns, _props$columns$;

  var expandIcon = props.expandIcon;

  var _useGlobalIcon = useGlobalIcon({
    ChevronRightCircleIcon: ChevronRightCircleIcon
  }),
      ChevronRightCircleIcon$1 = _useGlobalIcon.ChevronRightCircleIcon;

  var _useLocaleReceiver = useLocaleReceiver("table"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 1),
      locale = _useLocaleReceiver2[0];

  var _useClassName = useClassName(),
      tableExpandClasses = _useClassName.tableExpandClasses,
      positiveRotate90 = _useClassName.positiveRotate90,
      tableFullRowClasses = _useClassName.tableFullRowClasses;

  var _useControlled = useControlled(props, "expandedRowKeys", props.onExpandChange, {
    defaultExpandedRowKeys: props.defaultExpandedRowKeys || []
  }),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      tExpandedRowKeys = _useControlled2[0],
      setTExpandedRowKeys = _useControlled2[1];

  var showExpandedRow = Boolean(props.expandedRow);
  var showExpandIconColumn = props.expandIcon !== false && showExpandedRow;
  var isFirstColumnFixed = ((_props$columns = props.columns) === null || _props$columns === void 0 ? void 0 : (_props$columns$ = _props$columns[0]) === null || _props$columns$ === void 0 ? void 0 : _props$columns$.fixed) === "left";

  var onToggleExpand = function onToggleExpand(e, row) {
    props.expandOnRowClick && e.stopPropagation();
    var currentId = get_1(row, props.rowKey || "id");
    var index = tExpandedRowKeys.indexOf(currentId);

    var newKeys = _toConsumableArray(tExpandedRowKeys);

    index !== -1 ? newKeys.splice(index, 1) : newKeys.push(currentId);
    setTExpandedRowKeys(newKeys, {
      expandedRowData: props.data.filter(function (t) {
        return newKeys.includes(get_1(t, props.rowKey || "id"));
      }),
      currentRowData: row
    });
  };

  var renderExpandIcon = function renderExpandIcon(p, expandIcon2) {
    var row = p.row,
        rowIndex = p.rowIndex;
    var currentId = get_1(row, props.rowKey || "id");
    var expanded = tExpandedRowKeys.includes(currentId);
    var defaultIcon = locale.expandIcon || /* @__PURE__ */React.createElement(ChevronRightCircleIcon$1, null);
    var icon = defaultIcon;

    if (expandIcon2 === false || expandIcon2 === null) {
      icon = null;
    } else if (isFunction_1(expandIcon2)) {
      icon = expandIcon2({
        row: row,
        index: rowIndex
      });
    }

    var classes = [tableExpandClasses.iconBox, tableExpandClasses[expanded ? "expanded" : "collapsed"], _defineProperty({}, positiveRotate90, expanded)];
    return /* @__PURE__ */React.createElement("span", {
      className: classNames(classes),
      onClick: function onClick(e) {
        return onToggleExpand(e, row);
      }
    }, icon);
  };

  var getExpandColumn = function getExpandColumn() {
    var expandCol = {
      colKey: "__EXPAND_ROW_ICON_COLUMN__",
      width: 64,
      className: tableExpandClasses.iconCell,
      fixed: isFirstColumnFixed ? "left" : void 0,
      cell: function cell(p) {
        return renderExpandIcon(p, expandIcon);
      }
    };
    return expandCol;
  };

  var renderExpandedRow = function renderExpandedRow(p) {
    var rowId = get_1(p.row, props.rowKey || "id");
    if (!tExpandedRowKeys.includes(rowId)) return null;
    var isFixedLeft = p.isWidthOverflow && props.columns.find(function (item) {
      return item.fixed === "left";
    });
    return /* @__PURE__ */React.createElement("tr", {
      key: "expand_".concat(rowId),
      className: classNames([tableExpandClasses.row, _defineProperty({}, tableFullRowClasses.base, isFixedLeft)])
    }, /* @__PURE__ */React.createElement("td", {
      colSpan: p.columns.length
    }, /* @__PURE__ */React.createElement("div", {
      className: classNames([tableExpandClasses.rowInner, _defineProperty({}, tableFullRowClasses.innerFullRow, isFixedLeft)]),
      style: isFixedLeft ? {
        width: "".concat(p.tableWidth, "px")
      } : {}
    }, /* @__PURE__ */React.createElement("div", {
      className: tableFullRowClasses.innerFullElement
    }, props.expandedRow(p)))));
  };

  var onInnerExpandRowClick = function onInnerExpandRowClick(p) {
    onToggleExpand(p.e, p.row);
  };

  return {
    showExpandedRow: showExpandedRow,
    showExpandIconColumn: showExpandIconColumn,
    getExpandColumn: getExpandColumn,
    renderExpandedRow: renderExpandedRow,
    onInnerExpandRowClick: onInnerExpandRowClick
  };
}

export { useRowExpand as default };
