// tdesign-react/es/table/hooks/useTableHeader.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _typeof } from '../../_chunks/dep-4e0da095.js';
import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import React, { useMemo } from 'react';
import { i as isFunction_1 } from '../../_chunks/dep-e430e2ee.js';
import classNames from 'classnames';
import { getThRowspanAndColspan, getThList } from './useMultiHeader.js';
import useClassName from './useClassName.js';
import Ellipsis from '../Ellipsis.js';
import '../../_chunks/dep-7909fefa.js';
import '../../_chunks/dep-1cce5a60.js';
import '../../_chunks/dep-4a64abb1.js';
import '../../_chunks/dep-e54bb0dc.js';
import '../../_chunks/dep-b2626e28.js';
import '../../hooks/useConfig.js';
import '../../_chunks/dep-13a70a46.js';
import '../../_chunks/dep-dc1ce7ed.js';
import '../../_chunks/dep-4e2181de.js';
import '../../_chunks/dep-c197ce4f.js';
import '../../_chunks/dep-5a2778c1.js';
import '../../_chunks/dep-a2b97df8.js';
import '../../_chunks/dep-7aeeb170.js';
import '../../_chunks/dep-2afd805c.js';
import '../../_chunks/dep-e237f0f2.js';
import '../../_chunks/dep-4ed2dce5.js';
import '../../_chunks/dep-c8db35c0.js';
import '../../_chunks/dep-96aef4a9.js';
import '../../_chunks/dep-557a1b0f.js';
import '../../_chunks/dep-264b0e55.js';
import '../../_chunks/dep-f5693bf6.js';
import '../../_chunks/dep-fdd1d1a6.js';
import '../../_chunks/dep-7ff8b18e.js';
import '../../_chunks/dep-ebf3a272.js';
import '../../_chunks/dep-6a2482c6.js';
import '../../_chunks/dep-a980ac05.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import '../../_util/dom.js';
import 'raf';
import '../../_chunks/dep-3c98169f.js';
import '../../_util/easing.js';
import '../../tooltip/index.js';
import '../../tooltip/Tooltip.js';
import '../../_chunks/dep-8368bb87.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../../hooks/useControlled.js';
import '../../_chunks/dep-f54ff1e5.js';
import '../../_chunks/dep-e1d4dd61.js';
import '../../_chunks/dep-236ef238.js';
import '../../_chunks/dep-c8bfa11e.js';
import '../../_chunks/dep-199b1581.js';
import '../../_util/noop.js';
import '../../_util/useAnimation.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../popup/hooks/useTrigger.js';
import 'react-is';
import '../../popup/utils/ref.js';
import '../../_util/composeRefs.js';
import '../../popup/utils/transition.js';
import '../../_util/useMutationObserver.js';
import '../../_chunks/dep-236b0359.js';
import '../../_chunks/dep-6d5c0a90.js';
import '../../_util/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../popup/style/css.js';
import '../../popup/type.js';
import '../../tooltip/defaultProps.js';
import '../../tooltip/style/css.js';
import '../../tooltip/type.js';
import '../../hooks/useDebounce.js';
import 'lodash';
import '../../_util/usePersistFn.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function renderTitle(col, index) {
  var params = {
    col: col,
    colIndex: index
  };

  if (isFunction_1(col.title)) {
    return col.title(params);
  }

  if (isFunction_1(col.render)) {
    return col.render(_objectSpread(_objectSpread({}, params), {}, {
      row: {},
      rowIndex: -1,
      type: "title"
    }));
  }

  return col.title;
}
function useTableHeader(_ref) {
  var columns = _ref.columns;

  var _useClassName = useClassName(),
      tableSortClasses = _useClassName.tableSortClasses,
      tableFilterClasses = _useClassName.tableFilterClasses;

  var spansAndLeafNodes = useMemo(function () {
    return getThRowspanAndColspan(columns);
  }, [columns]);
  var thList = useMemo(function () {
    return getThList(columns);
  }, [columns]);
  var isMultipleHeader = useMemo(function () {
    return thList.length > 1;
  }, [thList]);

  var renderTitleWidthIcon = function renderTitleWidthIcon(_ref2, col, colIndex, ellipsisTitle, attach) {
    var _classes;

    var _ref3 = _slicedToArray(_ref2, 3),
        title = _ref3[0],
        sortIcon = _ref3[1],
        filterIcon = _ref3[2];

    var classes = (_classes = {}, _defineProperty(_classes, tableSortClasses.sortable, !!sortIcon), _defineProperty(_classes, tableFilterClasses.filterable, !!filterIcon), _classes);
    var content = isFunction_1(ellipsisTitle) ? ellipsisTitle({
      col: col,
      colIndex: colIndex
    }) : void 0;
    return /* @__PURE__ */React.createElement("div", {
      className: classNames(classes)
    }, /* @__PURE__ */React.createElement("div", {
      className: tableSortClasses.title
    }, col.ellipsis && ellipsisTitle !== false && ellipsisTitle !== null ? /* @__PURE__ */React.createElement(Ellipsis, {
      placement: "bottom",
      attach: attach ? function () {
        return attach;
      } : void 0,
      popupContent: content,
      tooltipProps: _typeof(ellipsisTitle) === "object" ? ellipsisTitle : void 0
    }, title) : /* @__PURE__ */React.createElement("div", null, title), Boolean(sortIcon || filterIcon) && /* @__PURE__ */React.createElement("div", {
      className: tableFilterClasses.iconWrap
    }, sortIcon, filterIcon)));
  };

  return {
    thList: thList,
    isMultipleHeader: isMultipleHeader,
    spansAndLeafNodes: spansAndLeafNodes,
    renderTitleWidthIcon: renderTitleWidthIcon
  };
}

export { useTableHeader as default, renderTitle };
