// tdesign-react/cjs/tree-select/TreeSelect.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var tree_index = require('../tree/index.js');
var selectInput_SelectInput = require('../select-input/SelectInput.js');
var _util_usePersistFn = require('../_util/usePersistFn.js');
var _util_useSwitch = require('../_util/useSwitch.js');
var _util_noop = require('../_util/noop.js');
var treeSelect_useTreeSelectUtils = require('./useTreeSelectUtils.js');
var treeSelect_SelectArrow = require('./SelectArrow.js');
var treeSelect_useTreeSelectPassthoughProps = require('./useTreeSelectPassthoughProps.js');
var treeSelect_useTreeSelectLocale = require('./useTreeSelectLocale.js');
var treeSelect_defaultProps = require('./defaultProps.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-09805e35.js');
require('../tree/Tree.js');
require('react-transition-group');
require('../tree/useTreeConfig.js');
require('../locale/LocalReceiver.js');
require('../tree/useControllable.js');
require('../tree/TreeItem.js');
require('tdesign-icons-react');
require('../loading/index.js');
require('../loading/Loading.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../_chunks/dep-1a0462cc.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');
require('../_util/useRipple.js');
require('../_util/useAnimation.js');
require('../_util/setStyle.js');
require('../hooks/useGlobalIcon.js');
require('../checkbox/index.js');
require('../_chunks/dep-3b82b90b.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../common/Check.js');
require('../_util/helper.js');
require('../_chunks/dep-0ebc045a.js');
require('../_chunks/dep-364d4c00.js');
require('../checkbox/defaultProps.js');
require('../tree/useStore.js');
require('../_chunks/dep-3afb7a98.js');
require('../_chunks/dep-4578b559.js');
require('../_chunks/dep-4fb39763.js');
require('../_chunks/dep-07e5d807.js');
require('../_util/useUpdateEffect.js');
require('../_util/useLayoutEffect.js');
require('../_common/js/tree/tree-store.js');
require('../_chunks/dep-c3f475a2.js');
require('../_chunks/dep-f21aff2d.js');
require('../_chunks/dep-aa470c9c.js');
require('../_chunks/dep-1e8de681.js');
require('../_common/js/tree/tree-node.js');
require('../_chunks/dep-266b1890.js');
require('../_chunks/dep-65cb34d8.js');
require('../_chunks/dep-8329ea23.js');
require('../_common/js/tree/tree-node-model.js');
require('../_chunks/dep-b5af0182.js');
require('../_chunks/dep-55c5ba71.js');
require('../_chunks/dep-70fe2bd0.js');
require('../_common/js/log/log.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-popper');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../popup/utils/ref.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../_util/useMutationObserver.js');
require('../_chunks/dep-db2b3323.js');
require('../_chunks/dep-52c5813c.js');
require('../_util/useWindowSize.js');
require('../popup/defaultProps.js');
require('../select-input/useSingle.js');
require('../input/index.js');
require('../input/Input.js');
require('../input/InputGroup.js');
require('../input/defaultProps.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('../_util/useDragSorter.js');
require('../tag-input/useTagScroll.js');
require('../tag-input/useTagList.js');
require('lodash');
require('../tag/index.js');
require('../tag/Tag.js');
require('../tag/CheckTag.js');
require('../tag/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/defaultProps.js');
require('../select-input/useOverlayInnerStyle.js');
require('../select-input/defaultProps.js');
require('../common/FakeArrow.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var useMergeFn = function useMergeFn() {
  for (var _len = arguments.length, fns = new Array(_len), _key = 0; _key < _len; _key++) {
    fns[_key] = arguments[_key];
  }

  return _util_usePersistFn.usePersistFn(function () {
    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    return fns.forEach(function (fn) {
      return fn === null || fn === void 0 ? void 0 : fn.apply(void 0, args);
    });
  });
};

var TreeSelect = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useTreeSelectLocale = treeSelect_useTreeSelectLocale.useTreeSelectLocale(props),
      placeholder = _useTreeSelectLocale.placeholder,
      empty = _useTreeSelectLocale.empty,
      loadingItem = _useTreeSelectLocale.loadingItem;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var className = props.className,
      onInputChange = props.onInputChange,
      readonly = props.readonly,
      disabled = props.disabled,
      multiple = props.multiple,
      prefixIcon = props.prefixIcon,
      loading = props.loading,
      size = props.size,
      max = props.max,
      data = props.data,
      _props$filter = props.filter,
      filter = _props$filter === void 0 ? function (text, option) {
    return option.label.includes(text);
  } : _props$filter,
      rawFilterable = props.filterable,
      onClear = props.onClear,
      valueDisplay = props.valueDisplay,
      treeProps = props.treeProps,
      inputProps = props.inputProps,
      onBlur = props.onBlur,
      onFocus = props.onFocus,
      onSearch = props.onSearch,
      onRemove = props.onRemove;
  var selectInputProps = treeSelect_useTreeSelectPassthoughProps.useTreeSelectPassThroughProps(props);

  var _useControlled = hooks_useControlled["default"](props, "value", props.onChange),
      _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
      value = _useControlled2[0],
      onChange = _useControlled2[1];

  var _useControlled3 = hooks_useControlled["default"](props, "popupVisible", props.onPopupVisibleChange),
      _useControlled4 = slicedToArray._slicedToArray(_useControlled3, 2),
      popupVisible = _useControlled4[0],
      setPopupVisible = _useControlled4[1];

  var _useSwitch = _util_useSwitch["default"](),
      _useSwitch2 = slicedToArray._slicedToArray(_useSwitch, 2),
      hover = _useSwitch2[0],
      hoverAction = _useSwitch2[1];

  var _useControlled5 = hooks_useControlled["default"](props, "inputValue", onInputChange),
      _useControlled6 = slicedToArray._slicedToArray(_useControlled5, 2),
      filterInput = _useControlled6[0],
      setFilterInput = _useControlled6[1];

  var treeRef = React.useRef();

  var _useTreeSelectUtils = treeSelect_useTreeSelectUtils.useTreeSelectUtils(props, treeRef),
      normalizeValue = _useTreeSelectUtils.normalizeValue,
      formatValue = _useTreeSelectUtils.formatValue,
      getNodeItem = _useTreeSelectUtils.getNodeItem;

  var filterable = rawFilterable || !!props.filter;
  var normalizedValue = React.useMemo(function () {
    var calcValue = Array.isArray(value) ? value : [value];
    return calcValue.reduce(function (result, value2) {
      var normalized = normalizeValue(value2);
      normalized.value && result.push(normalized);
      return result;
    }, []);
  }, [normalizeValue, value, data]);
  var internalInputValue = React.useMemo(function () {
    if (multiple) return normalizedValue;
    return filterable && popupVisible ? filterInput : normalizedValue[0] || "";
  }, [multiple, normalizedValue, filterable, popupVisible, filterInput]);
  var inputPlaceholader = React.useMemo(function () {
    if (filterable && !multiple && popupVisible && normalizedValue.length) {
      return typeof normalizedValue[0].label === "string" ? normalizedValue[0].label : String(normalizedValue[0].value);
    }

    return placeholder;
  }, [filterable, multiple, popupVisible, normalizedValue, placeholder]);
  var showLoading = !disabled && loading;
  var showFakePlaceholder = multiple && !filterable && !normalizedValue.length;
  var handleFilter = React.useCallback(function (node) {
    return filterable ? filter(filterInput, node) : true;
  }, [filter, filterInput, filterable]);
  var handleSingleChange = _util_usePersistFn.usePersistFn(function (value2, context) {
    var $value = value2.length ? value2[0] : null;
    onChange(formatValue($value, context.node.label), _objectSpread(_objectSpread({}, context), {}, {
      trigger: $value === null ? "uncheck" : "check"
    }));
    setPopupVisible(false, {
      trigger: "trigger-element-click"
    });
  });
  var handleMultiChange = _util_usePersistFn.usePersistFn(function (value2, context) {
    (max === 0 || value2.length <= max) && onChange(value2.map(function (value3) {
      var _getNodeItem;

      return formatValue(value3, (_getNodeItem = getNodeItem(value3)) === null || _getNodeItem === void 0 ? void 0 : _getNodeItem.label);
    }), _objectSpread(_objectSpread({}, context), {}, {
      trigger: value2.length > normalizedValue.length ? "check" : "uncheck"
    }));
  });
  var handleClear = _util_usePersistFn.usePersistFn(function (ctx) {
    ctx.e.stopPropagation();
    onChange(multiple ? [] : formatValue(null), {
      node: null,
      trigger: "clear",
      e: ctx.e
    });
    onClear === null || onClear === void 0 ? void 0 : onClear(ctx);
    setPopupVisible(false, {
      trigger: "trigger-element-click"
    });
  });
  var handleRemove = _util_usePersistFn.usePersistFn(function (index, e) {
    var node = getNodeItem(normalizedValue[index].value);
    onChange(normalizedValue.filter(function (value2, i) {
      return i !== index;
    }).map(function (_ref) {
      var value2 = _ref.value,
          label = _ref.label;
      return formatValue(value2, label);
    }), {
      node: node,
      trigger: "tag-remove",
      e: e
    });
    onRemove === null || onRemove === void 0 ? void 0 : onRemove({
      value: node.value,
      data: {
        value: node.value,
        label: node.label
      },
      e: e
    });
  });
  var handleTagChange = _util_usePersistFn.usePersistFn(function (tags, ctx) {
    switch (ctx.trigger) {
      case "clear":
        handleClear({
          e: ctx.e
        });
        break;

      case "tag-remove":
        handleRemove(ctx.index, ctx.e);
        break;

      case "backspace":
        handleRemove(ctx.index);
    }
  });
  var handleBlur = _util_usePersistFn.usePersistFn(function (v, ctx) {
    onBlur === null || onBlur === void 0 ? void 0 : onBlur({
      value: multiple ? normalizedValue : normalizedValue[0],
      e: ctx.e
    });
  });
  var handleFocus = _util_usePersistFn.usePersistFn(function (v, ctx) {
    onFocus === null || onFocus === void 0 ? void 0 : onFocus({
      value: multiple ? normalizedValue : normalizedValue[0],
      e: ctx.e
    });
  });
  var handleEnter = _util_usePersistFn.usePersistFn(function (text) {
    onSearch === null || onSearch === void 0 ? void 0 : onSearch(text);
  });
  var handleFilterChange = _util_usePersistFn.usePersistFn(function (value2) {
    return setFilterInput(value2);
  });
  React.useEffect(function () {
    popupVisible && setFilterInput("");
  }, [popupVisible]);
  React.useEffect(function () {
    setFilterInput("");
  }, [value]);

  var renderTree = function renderTree() {
    if (readonly) return empty;
    if (showLoading) return loadingItem;
    return /* @__PURE__ */React__default["default"].createElement(tree_index.Tree, _objectSpread(_objectSpread({
      ref: treeRef,
      hover: true,
      transition: true,
      expandAll: true,
      filter: handleFilter,
      data: data,
      disabled: disabled,
      empty: empty,
      allowFoldNodeOnFilter: true
    }, multiple ? {
      checkable: true,
      onChange: handleMultiChange,
      value: normalizedValue.map(function (_ref2) {
        var value2 = _ref2.value;
        return value2;
      })
    } : {
      activable: true,
      actived: normalizedValue.map(function (_ref3) {
        var value2 = _ref3.value;
        return value2;
      }),
      onActive: handleSingleChange
    }), treeProps));
  };

  var renderCollapsedItems = React.useMemo(function () {
    return props.collapsedItems ? function () {
      return props.collapsedItems({
        value: normalizedValue,
        collapsedSelectedItems: normalizedValue.slice(props.minCollapsedNum, normalizedValue.length),
        count: normalizedValue.length - props.minCollapsedNum
      });
    } : null;
  }, [normalizedValue, props]);

  var renderLabel = function renderLabel() {
    return showFakePlaceholder ? /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, prefixIcon, /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-tree-select--placeholder")
    }, placeholder)) : prefixIcon;
  };

  var normalizedValueDisplay = function normalizedValueDisplay() {
    if (typeof valueDisplay === "string") return valueDisplay;
    if (multiple) return function (_ref4) {
      var onClose = _ref4.onClose;
      return valueDisplay({
        value: normalizedValue,
        onClose: onClose
      });
    };
    return normalizedValue.length ? valueDisplay({
      value: normalizedValue[0],
      onClose: _util_noop["default"]
    }) : "";
  };

  return /* @__PURE__ */React__default["default"].createElement(selectInput_SelectInput["default"], _objectSpread(_objectSpread(_objectSpread({
    status: props.status,
    tips: props.tips
  }, props.selectInputProps), selectInputProps), {}, {
    ref: ref,
    className: classNames__default["default"]("".concat(classPrefix, "-tree-select"), defineProperty._defineProperty({}, "".concat(classPrefix, "-tree-select--without-input"), multiple && !filterable), className),
    value: internalInputValue,
    inputValue: filterInput,
    panel: renderTree(),
    allowInput: multiple || filterable,
    inputProps: _objectSpread(_objectSpread({}, inputProps), {}, {
      size: size
    }),
    tagInputProps: {
      size: size,
      excessTagsDisplayType: "break-line",
      inputProps: inputProps,
      tagProps: props.tagProps
    },
    placeholder: inputPlaceholader,
    popupVisible: popupVisible && !disabled,
    onInputChange: handleFilterChange,
    onPopupVisibleChange: useMergeFn(setPopupVisible),
    onFocus: useMergeFn(handleFocus, function () {
      return setPopupVisible(true, {
        trigger: "trigger-element-click"
      });
    }),
    onBlur: useMergeFn(handleBlur),
    onClear: handleClear,
    onTagChange: handleTagChange,
    onEnter: handleEnter,
    onMouseenter: hoverAction.on,
    onMouseleave: hoverAction.off,
    suffixIcon: readonly ? null : /* @__PURE__ */React__default["default"].createElement(treeSelect_SelectArrow.SelectArrow, {
      isActive: popupVisible,
      isHighlight: hover || popupVisible,
      disabled: disabled
    }),
    collapsedItems: renderCollapsedItems,
    label: renderLabel(),
    valueDisplay: valueDisplay && normalizedValueDisplay()
  }));
});
TreeSelect.displayName = "TreeSelect";
TreeSelect.defaultProps = treeSelect_defaultProps.treeSelectDefaultProps;

exports["default"] = TreeSelect;
