// tdesign-react/lib/select/base/Option.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import React, { useRef } from 'react';
import classNames from 'classnames';
import { i as isNumber_1 } from '../../_chunks/dep-31043569.js';
import { i as isString_1 } from '../../_chunks/dep-8d5556d7.js';
import { g as get_1 } from '../../_chunks/dep-1f300d2c.js';
import useConfig from '../../hooks/useConfig.js';
import useRipple from '../../_util/useRipple.js';
import '../../_chunks/dep-9ea787ca.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-7f6754ac.js';
import '../../_chunks/dep-3160df29.js';
import '../../_chunks/dep-e103618e.js';
import '../../_chunks/dep-b989427d.js';
import '../../_chunks/dep-68f7a306.js';
import '../../_chunks/dep-a74f6fff.js';
import '../../_chunks/dep-d23d346c.js';
import '../../_chunks/dep-0e2dfd87.js';
import '../../_chunks/dep-f50ff346.js';
import '../../_chunks/dep-2ac048c0.js';
import '../../_chunks/dep-918e41e2.js';
import '../../_chunks/dep-734a21a0.js';
import '../../_chunks/dep-51592cde.js';
import '../../_chunks/dep-ae94c4e7.js';
import '../../_chunks/dep-2509667d.js';
import '../../_chunks/dep-b817c90e.js';
import '../../_chunks/dep-fe9825b6.js';
import '../../_chunks/dep-ebc05da7.js';
import '../../_chunks/dep-c91b3852.js';
import '../../_chunks/dep-9e44c34c.js';
import '../../_chunks/dep-0c62d47a.js';
import '../../_chunks/dep-e2fe68b7.js';
import '../../_chunks/dep-44a03722.js';
import '../../_chunks/dep-f5616354.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import '../../_util/useAnimation.js';
import '../../_util/setStyle.js';
import '../../_util/dom.js';
import 'raf';
import '../../_util/easing.js';

var componentType = "select";

var Option = function Option(props) {
  var _classNames2;

  var propDisabled = props.disabled,
      propLabel = props.label,
      selectedValue = props.selectedValue,
      multiple = props.multiple,
      size = props.size,
      max = props.max,
      keys = props.keys,
      value = props.value,
      onSelect = props.onSelect,
      children = props.children,
      content = props.content,
      restData = props.restData,
      style = props.style,
      className = props.className;
  var selected;
  var label = propLabel || value;
  var disabled = propDisabled || multiple && Array.isArray(selectedValue) && max && selectedValue.length >= max;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var optionRef = useRef();
  useRipple(optionRef);

  if (!multiple) {
    selected = isNumber_1(selectedValue) || isString_1(selectedValue) ? value === selectedValue : value === get_1(selectedValue, (keys === null || keys === void 0 ? void 0 : keys.value) || "value");
  }

  if (multiple && Array.isArray(selectedValue)) {
    selected = selectedValue.some(function (item) {
      if (isNumber_1(item) || isString_1(item)) {
        return item === value;
      }

      return get_1(item, (keys === null || keys === void 0 ? void 0 : keys.value) || "value") === value;
    });
  }

  var handleSelect = function handleSelect(event) {
    if (!disabled) {
      onSelect(value, {
        label: String(label),
        selected: selected,
        event: event,
        restData: restData
      });
    }
  };

  var renderItem = function renderItem(children2) {
    if (multiple) {
      var _classNames;

      return /* @__PURE__ */React.createElement("label", {
        className: classNames("".concat(classPrefix, "-checkbox"), (_classNames = {}, _defineProperty(_classNames, "".concat(classPrefix, "-is-disabled"), disabled), _defineProperty(_classNames, "".concat(classPrefix, "-is-checked"), selected), _classNames))
      }, /* @__PURE__ */React.createElement("input", {
        type: "checkbox",
        className: classNames("".concat(classPrefix, "-checkbox__former")),
        value: "",
        disabled: disabled && !selected,
        onClick: function onClick(e) {
          return e.stopPropagation();
        }
      }), /* @__PURE__ */React.createElement("span", {
        className: classNames("".concat(classPrefix, "-checkbox__input"))
      }), /* @__PURE__ */React.createElement("span", {
        className: classNames("".concat(classPrefix, "-checkbox__label"))
      }, children2 || content || label));
    }

    return /* @__PURE__ */React.createElement("span", null, children2 || content || label);
  };

  return /* @__PURE__ */React.createElement("li", {
    className: classNames(className, "".concat(classPrefix, "-").concat(componentType, "-option"), (_classNames2 = {}, _defineProperty(_classNames2, "".concat(classPrefix, "-is-disabled"), disabled), _defineProperty(_classNames2, "".concat(classPrefix, "-is-selected"), selected), _defineProperty(_classNames2, "".concat(classPrefix, "-size-s"), size === "small"), _defineProperty(_classNames2, "".concat(classPrefix, "-size-l"), size === "large"), _classNames2)),
    key: value,
    onClick: handleSelect,
    ref: optionRef,
    style: style
  }, renderItem(children));
};

export { Option as default };
