// tdesign-react/esm/date-picker/hooks/useSingleValue.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import { useState, useEffect } from 'react';
import dayjs from 'dayjs';
import { getDefaultFormat, formatTime, formatDate, isValidDate } from '../../_common/js/date-picker/format.js';
import useControlled from '../../hooks/useControlled.js';
import { extractTimeFormat } from '../../_common/js/date-picker/utils.js';
import '../../_common/js/log/index.js';
import log from '../../_common/js/log/log.js';
import '../../_chunks/dep-7909fefa.js';
import '../../_chunks/dep-1e32006c.js';
import '../../_chunks/dep-1cce5a60.js';
import '../../_chunks/dep-76ad5996.js';
import '../../_chunks/dep-6d5c0a90.js';
import '../../_chunks/dep-e54bb0dc.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-b2626e28.js';
import '../../_chunks/dep-199b1581.js';
import '../../_chunks/dep-4a64abb1.js';
import '../../_chunks/dep-e237f0f2.js';
import '../../_chunks/dep-236ef238.js';
import '../../_chunks/dep-a980ac05.js';
import '../../_chunks/dep-a2b97df8.js';
import '../../_chunks/dep-fdd1d1a6.js';
import '../../_chunks/dep-e430e2ee.js';
import '../../_chunks/dep-5a2778c1.js';
import '../../_chunks/dep-2afd805c.js';
import '../../_chunks/dep-f5693bf6.js';
import '../../_chunks/dep-f54ff1e5.js';
import '../../_chunks/dep-e1d4dd61.js';
import '../../_chunks/dep-c8bfa11e.js';
import '../../_chunks/dep-264b0e55.js';
import '../../_util/noop.js';

function useSingleValue(props) {
  var _useControlled = useControlled(props, "value", props.onChange),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      value = _useControlled2[0],
      onChange = _useControlled2[1];

  var _getDefaultFormat = getDefaultFormat({
    mode: props.mode,
    format: props.format,
    valueType: props.valueType,
    enableTimePicker: props.enableTimePicker
  }),
      format = _getDefaultFormat.format,
      valueType = _getDefaultFormat.valueType,
      timeFormat = _getDefaultFormat.timeFormat;

  if (props.enableTimePicker) {
    if (!extractTimeFormat(format)) log.error("DatePicker", "format: ".concat(format, " \u4E0D\u89C4\u8303\uFF0C\u5305\u542B\u65F6\u95F4\u9009\u62E9\u5FC5\u987B\u8981\u6709\u65F6\u95F4\u683C\u5F0F\u5316 HH:mm:ss"));
    if (!extractTimeFormat(valueType) && valueType !== "time-stamp") log.error("DatePicker", "valueType: ".concat(valueType, " \u4E0D\u89C4\u8303\uFF0C\u5305\u542B\u65F6\u95F4\u9009\u62E9\u5FC5\u987B\u8981\u6709\u65F6\u95F4\u683C\u5F0F\u5316 HH:mm:ss"));
  }

  var _useState = useState(formatTime(value, timeFormat)),
      _useState2 = _slicedToArray(_useState, 2),
      time = _useState2[0],
      setTime = _useState2[1];

  var _useState3 = useState(dayjs(value).month() || new Date().getMonth()),
      _useState4 = _slicedToArray(_useState3, 2),
      month = _useState4[0],
      setMonth = _useState4[1];

  var _useState5 = useState(dayjs(value).year() || new Date().getFullYear()),
      _useState6 = _slicedToArray(_useState5, 2),
      year = _useState6[0],
      setYear = _useState6[1];

  var _useState7 = useState(formatDate(value, {
    format: format,
    targetFormat: format
  })),
      _useState8 = _slicedToArray(_useState7, 2),
      cacheValue = _useState8[0],
      setCacheValue = _useState8[1];

  useEffect(function () {
    if (!value) {
      setCacheValue("");
      return;
    }

    if (!isValidDate(value, valueType)) return;
    setCacheValue(formatDate(value, {
      format: format,
      targetFormat: format
    }));
    setTime(formatTime(value, timeFormat));
  }, [value]);
  return {
    year: year,
    month: month,
    value: value,
    time: time,
    cacheValue: cacheValue,
    onChange: onChange,
    setYear: setYear,
    setMonth: setMonth,
    setTime: setTime,
    setCacheValue: setCacheValue
  };
}

export { useSingleValue as default };
