// tdesign-react/esm/image-viewer/ImageViewer.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import React, { useState } from 'react';
import { createPortal } from 'react-dom';
import { i as isFunction_1 } from '../_chunks/dep-e430e2ee.js';
import { I as ImageModal } from '../_chunks/dep-4b069274.js';
import { imageViewerDefaultProps } from './defaultProps.js';
import useImageScale from './hooks/useImageScale.js';
import useList from './hooks/useList.js';
import useViewerScale from './hooks/useViewerScale.js';
import useControlled from '../hooks/useControlled.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-e54bb0dc.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-8368bb87.js';
import '../_chunks/dep-1e32006c.js';
import '../_chunks/dep-264b0e55.js';
import 'tdesign-icons-react';
import 'classnames';
import '../tooltip/index.js';
import '../tooltip/Tooltip.js';
import '../popup/index.js';
import '../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../_util/useAnimation.js';
import '../hooks/useConfig.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-e237f0f2.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-ebf3a272.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../common/Portal.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-3c98169f.js';
import '../_util/easing.js';
import '../popup/hooks/useTrigger.js';
import 'react-is';
import '../popup/utils/ref.js';
import '../_util/composeRefs.js';
import '../popup/utils/transition.js';
import '../_util/useMutationObserver.js';
import '../_chunks/dep-236b0359.js';
import '../_chunks/dep-6d5c0a90.js';
import '../_chunks/dep-199b1581.js';
import '../_util/useWindowSize.js';
import '../popup/defaultProps.js';
import '../_chunks/dep-f54ff1e5.js';
import '../_chunks/dep-e1d4dd61.js';
import '../_chunks/dep-236ef238.js';
import '../_chunks/dep-c8bfa11e.js';
import '../_util/noop.js';
import '../popup/style/index.js';
import '../popup/type.js';
import '../tooltip/defaultProps.js';
import '../tooltip/style/index.js';
import '../tooltip/type.js';
import './utils.js';
import '../dialog/index.js';
import '../dialog/Dialog.js';
import '../locale/LocalReceiver.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';
import '../button/index.js';
import '../button/Button.js';
import '../_util/useRipple.js';
import '../_util/setStyle.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-d1b4ab24.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import '../loading/style/index.js';
import '../loading/type.js';
import '../button/defaultProps.js';
import '../button/style/index.js';
import '../button/type.js';
import '../dialog/RenderDialog.js';
import '../_util/useLayoutEffect.js';
import '../_util/useDialogEsc.js';
import '../dialog/defaultProps.js';
import '../_util/useSetState.js';
import '../hooks/useGlobalIcon.js';
import '../dialog/plugin.js';
import '../_common/js/log/index.js';
import '../_common/js/log/log.js';
import '../dialog/style/index.js';
import '../dialog/type.js';
import './hooks/useMirror.js';
import './hooks/usePosition.js';
import './hooks/useIndex.js';
import './hooks/useRotate.js';
import './hooks/useScale.js';
import './hooks/useIconMap.js';

var ImageViewer = function ImageViewer(props) {
  var mode = props.mode,
      trigger = props.trigger,
      images = props.images,
      title = props.title,
      imageScaleD = props.imageScale,
      viewerScaleD = props.viewerScale;

  var _useControlled = useControlled(props, "visible", function (visible2, context) {
    isFunction_1(props.onClose) && props.onClose(context);
  }),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      visible = _useControlled2[0],
      setVisible = _useControlled2[1];

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      visibled = _useState2[0],
      setVisibled = _useState2[1];

  var list = useList(images);
  var imageScale = useImageScale(imageScaleD);
  var viewerScale = useViewerScale(viewerScaleD);
  var isMini = mode === "modeless";

  var close = function close(context) {
    setVisible(false, context);
    setTimeout(function () {
      return setVisibled(false);
    }, 196);
  };

  var open = function open() {
    if (!images) return;
    setVisible(true, null);
    setVisibled(true);
  };

  var uiImage = isFunction_1(trigger) ? trigger({
    open: open,
    close: close,
    onOpen: open,
    onClose: close
  }) : trigger;
  return /* @__PURE__ */React.createElement(React.Fragment, null, uiImage, visibled && /*#__PURE__*/createPortal( /* @__PURE__ */React.createElement(ImageModal, {
    title: title,
    visible: visible,
    images: list,
    isMini: isMini,
    imageScale: imageScale,
    viewerScale: viewerScale,
    zIndex: props.zIndex,
    defaultIndex: props.defaultIndex,
    index: props.index,
    onIndexChange: props.onIndexChange,
    draggable: props.draggable,
    closeOnOverlay: props.closeOnOverlay,
    closeBtn: props.closeBtn,
    showOverlay: props.showOverlay,
    onClose: close,
    onOpen: open
  }), document.body));
};

ImageViewer.displayName = "ImageViewer";
ImageViewer.defaultProps = imageViewerDefaultProps;

export { ImageViewer as default };
