// tdesign-react/cjs/transfer/utils.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var isEmpty = require('../_chunks/dep-ca4cc26d.js');
var isFunction = require('../_chunks/dep-53ba6729.js');
var isString = require('../_chunks/dep-9a2dbbb1.js');
var React = require('react');
require('../_chunks/dep-4fb39763.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-d3ad6e52.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var getDefaultValue = function getDefaultValue(value, defaultValue) {
  if (isEmpty.isEmpty_1(value)) {
    return Array.isArray(defaultValue) ? defaultValue : [defaultValue, defaultValue];
  }

  if (Array.isArray(value)) return value;
  return [defaultValue, defaultValue];
};
var getJSX = function getJSX(value) {
  if (isString.isString_1(value)) return value;
  if (isFunction.isFunction_1(value)) return value();
  if ( /*#__PURE__*/React__default["default"].isValidElement(value)) return value;
  return value;
};
var getLeafNodes = function getLeafNodes(nodes) {
  var resData = [];

  var map = function map(nodes2) {
    nodes2.forEach(function (child) {
      if (child.children && child.children.length > 0) {
        return map(child.children);
      }

      resData.push(child);
    });
  };

  map(nodes);
  return resData;
};
var filterCheckedTreeNodes = function filterCheckedTreeNodes(nodes, checkeds) {
  return nodes.filter(function (item) {
    return !checkeds.includes(item.value);
  }).map(function (item) {
    var newItem = _objectSpread({}, item);

    if (item.children && item.children.length > 0) {
      newItem.children = filterCheckedTreeNodes(item.children, checkeds);
    }

    return newItem;
  }).filter(function (item) {
    return item.children && item.children.length !== 0 || !item.children;
  });
};
var getTargetNodes = function getTargetNodes(sourceNodes, data) {
  var source = getLeafNodes(sourceNodes).map(function (item) {
    return item.value;
  });
  return filterCheckedTreeNodes(data, source);
};

exports.filterCheckedTreeNodes = filterCheckedTreeNodes;
exports.getDefaultValue = getDefaultValue;
exports.getJSX = getJSX;
exports.getLeafNodes = getLeafNodes;
exports.getTargetNodes = getTargetNodes;
