// tdesign-react/es/jumper/Jumper.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import React, { forwardRef, useMemo } from 'react';
import classNames from 'classnames';
import { i as isObject_1 } from '../_chunks/dep-e54bb0dc.js';
import { RoundIcon, ChevronUpIcon, ChevronDownIcon, ChevronLeftIcon, ChevronRightIcon } from 'tdesign-icons-react';
import useConfig from '../hooks/useConfig.js';
import useGlobalIcon from '../hooks/useGlobalIcon.js';
import { Button } from '../button/index.js';
import { jumperDefaultProps } from './defaultProps.js';
import noop from '../_util/noop.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-e237f0f2.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-ebf3a272.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../button/Button.js';
import '../_chunks/dep-8368bb87.js';
import '../_util/useRipple.js';
import '../_util/useAnimation.js';
import '../_util/setStyle.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-3c98169f.js';
import '../_util/easing.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../_chunks/dep-9615f03f.js';
import '../_chunks/dep-7909fefa.js';
import '../common/Portal.js';
import 'react-dom';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-d1b4ab24.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import '../loading/style/css.js';
import '../loading/type.js';
import '../button/defaultProps.js';
import '../button/style/css.js';
import '../button/type.js';

var Jumper = /*#__PURE__*/forwardRef(function (props, ref) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useGlobalIcon = useGlobalIcon({
    RoundIcon: RoundIcon,
    ChevronUpIcon: ChevronUpIcon,
    ChevronDownIcon: ChevronDownIcon,
    ChevronLeftIcon: ChevronLeftIcon,
    ChevronRightIcon: ChevronRightIcon
  }),
      RoundIcon$1 = _useGlobalIcon.RoundIcon,
      ChevronUpIcon$1 = _useGlobalIcon.ChevronUpIcon,
      ChevronDownIcon$1 = _useGlobalIcon.ChevronDownIcon,
      ChevronLeftIcon$1 = _useGlobalIcon.ChevronLeftIcon,
      ChevronRightIcon$1 = _useGlobalIcon.ChevronRightIcon;

  var variant = props.variant,
      tips = props.tips,
      showCurrent = props.showCurrent,
      disabled = props.disabled,
      layout = props.layout,
      size = props.size,
      _props$onChange = props.onChange,
      onChange = _props$onChange === void 0 ? noop : _props$onChange,
      className = props.className,
      style = props.style;
  var titleConfig = useMemo(function () {
    if (isObject_1(tips)) return tips;
    if (tips === true) return {
      prev: "\u4E0A\u4E00\u9875",
      current: "\u5F53\u524D",
      next: "\u4E0B\u4E00\u9875"
    };
    return {};
  }, [tips]);
  var disabledConfig = useMemo(function () {
    if (isObject_1(disabled)) return disabled;
    if (disabled === true) return {
      prev: true,
      current: true,
      next: true
    };
    return {
      prev: false,
      current: false,
      next: false
    };
  }, [disabled]);
  return /* @__PURE__ */React.createElement("div", {
    className: classNames("".concat(classPrefix, "-jumper"), className, _defineProperty({}, "".concat(classPrefix, "-jumper--outline"), variant === "outline")),
    ref: ref,
    style: style
  }, /* @__PURE__ */React.createElement(Button, {
    title: titleConfig.prev,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "prev"
      });
    },
    icon: layout === "horizontal" ? /* @__PURE__ */React.createElement(ChevronLeftIcon$1, null) : /* @__PURE__ */React.createElement(ChevronUpIcon$1, null),
    className: "".concat(classPrefix, "-jumper__prev"),
    disabled: disabledConfig.prev
  }), showCurrent && /* @__PURE__ */React.createElement(Button, {
    title: titleConfig.current,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "current"
      });
    },
    icon: /* @__PURE__ */React.createElement(RoundIcon$1, null),
    className: "".concat(classPrefix, "-jumper__current"),
    disabled: disabledConfig.current
  }), /* @__PURE__ */React.createElement(Button, {
    title: titleConfig.next,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "next"
      });
    },
    icon: layout === "horizontal" ? /* @__PURE__ */React.createElement(ChevronRightIcon$1, null) : /* @__PURE__ */React.createElement(ChevronDownIcon$1, null),
    className: "".concat(classPrefix, "-jumper__next"),
    disabled: disabledConfig.next
  }));
});
Jumper.displayName = "Jumper";
Jumper.defaultProps = jumperDefaultProps;

export { Jumper as default };
