// tdesign-react/lib/date-picker/hooks/useRangeValue.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import { useState, useEffect } from 'react';
import dayjs from 'dayjs';
import useControlled from '../../hooks/useControlled.js';
import { getDefaultFormat, isValidDate, formatDate, formatTime, parseToDayjs } from '../../_common/js/date-picker/format.js';
import { extractTimeFormat } from '../../_common/js/date-picker/utils.js';
import log from '../../_common/js/log/log.js';
import '../../_chunks/dep-7909fefa.js';
import '../../_chunks/dep-12260644.js';
import '../../_chunks/dep-ec84085d.js';
import '../../_chunks/dep-66273dbc.js';
import '../../_chunks/dep-2ac048c0.js';
import '../../_chunks/dep-9ea787ca.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-3160df29.js';
import '../../_chunks/dep-b989427d.js';
import '../../_chunks/dep-7f6754ac.js';
import '../../_util/noop.js';
import '../../_chunks/dep-1e32006c.js';
import '../../_chunks/dep-f4afdfc0.js';
import '../../_chunks/dep-652c1652.js';
import '../../_chunks/dep-0e2dfd87.js';
import '../../_chunks/dep-f5616354.js';
import '../../_chunks/dep-f50ff346.js';
import '../../_chunks/dep-9e44c34c.js';
import '../../_chunks/dep-d23d346c.js';
import '../../_chunks/dep-ae94c4e7.js';
import '../../_chunks/dep-c91b3852.js';

var PARTIAL_MAP = {
  first: "start",
  second: "end"
};
function initYearMonthTime(_ref) {
  var value = _ref.value,
      _ref$mode = _ref.mode,
      mode = _ref$mode === void 0 ? "date" : _ref$mode,
      format = _ref.format,
      _ref$timeFormat = _ref.timeFormat,
      timeFormat = _ref$timeFormat === void 0 ? "HH:mm:ss" : _ref$timeFormat,
      enableTimePicker = _ref.enableTimePicker;
  var defaultYearMonthTime = {
    year: [dayjs().year(), dayjs().year()],
    month: [dayjs().month(), dayjs().month()],
    time: [dayjs().format(timeFormat), dayjs().format(timeFormat)]
  };

  if (mode === "year") {
    defaultYearMonthTime.year[1] += 10;
  } else if (mode === "month" || mode === "quarter") {
    defaultYearMonthTime.year[1] += 1;
  } else if ((mode === "date" || mode === "week") && !enableTimePicker) {
    defaultYearMonthTime.month[1] += 1;
  }

  if (!value || !Array.isArray(value) || !value.length) {
    return defaultYearMonthTime;
  }

  return {
    year: value.map(function (v) {
      return parseToDayjs(v, format).year();
    }),
    month: value.map(function (v) {
      return parseToDayjs(v, format).month();
    }),
    time: value.map(function (v) {
      return parseToDayjs(v, format).format(timeFormat);
    })
  };
}
function useRange(props) {
  var _useControlled = useControlled(props, "value", props.onChange),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      value = _useControlled2[0],
      onChange = _useControlled2[1];

  var _getDefaultFormat = getDefaultFormat({
    mode: props.mode,
    format: props.format,
    valueType: props.valueType,
    enableTimePicker: props.enableTimePicker
  }),
      format = _getDefaultFormat.format,
      valueType = _getDefaultFormat.valueType,
      timeFormat = _getDefaultFormat.timeFormat;

  if (props.enableTimePicker) {
    if (!extractTimeFormat(format)) log.error("DatePicker", "format: ".concat(format, " \u4E0D\u89C4\u8303\uFF0C\u5305\u542B\u65F6\u95F4\u9009\u62E9\u5FC5\u987B\u8981\u6709\u65F6\u95F4\u683C\u5F0F\u5316 HH:mm:ss"));
    if (!extractTimeFormat(valueType) && valueType !== "time-stamp") log.error("DatePicker", "valueType: ".concat(valueType, " \u4E0D\u89C4\u8303\uFF0C\u5305\u542B\u65F6\u95F4\u9009\u62E9\u5FC5\u987B\u8981\u6709\u65F6\u95F4\u683C\u5F0F\u5316 HH:mm:ss"));
  }

  if (!Array.isArray(value)) {
    log.error("DatePicker", "typeof value: ".concat(value, " must be Array!"));
  } else if (!isValidDate(value, valueType)) {
    log.error("DatePicker", "value: ".concat(value, " is invalid dateTime! Check whether the value is consistent with format: ").concat(format));
  }

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isFirstValueSelected = _useState2[0],
      setIsFirstValueSelected = _useState2[1];

  var _useState3 = useState(initYearMonthTime({
    value: value,
    mode: props.mode,
    format: format,
    timeFormat: timeFormat
  }).time),
      _useState4 = _slicedToArray(_useState3, 2),
      time = _useState4[0],
      setTime = _useState4[1];

  var _useState5 = useState(initYearMonthTime({
    value: value,
    mode: props.mode,
    format: format,
    enableTimePicker: props.enableTimePicker
  }).month),
      _useState6 = _slicedToArray(_useState5, 2),
      month = _useState6[0],
      setMonth = _useState6[1];

  var _useState7 = useState(initYearMonthTime({
    value: value,
    mode: props.mode,
    format: format
  }).year),
      _useState8 = _slicedToArray(_useState7, 2),
      year = _useState8[0],
      setYear = _useState8[1];

  var _useState9 = useState(formatDate(value, {
    format: format,
    targetFormat: format
  })),
      _useState10 = _slicedToArray(_useState9, 2),
      cacheValue = _useState10[0],
      setCacheValue = _useState10[1];

  useEffect(function () {
    if (!value) {
      setCacheValue([]);
      return;
    }

    if (!isValidDate(value, valueType)) return;
    setCacheValue(formatDate(value, {
      format: format,
      targetFormat: format
    }));
    setTime(formatTime(value, timeFormat));
  }, [value]);
  return {
    value: value,
    onChange: onChange,
    year: year,
    setYear: setYear,
    month: month,
    setMonth: setMonth,
    time: time,
    setTime: setTime,
    isFirstValueSelected: isFirstValueSelected,
    setIsFirstValueSelected: setIsFirstValueSelected,
    cacheValue: cacheValue,
    setCacheValue: setCacheValue
  };
}

export { PARTIAL_MAP, useRange as default, initYearMonthTime };
