// tdesign-react/cjs/table/FilterController.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var isEmpty = require('../_chunks/dep-ca4cc26d.js');
var classNames = require('classnames');
var popup_index = require('../popup/index.js');
var checkbox_index = require('../checkbox/index.js');
var radio_index = require('../radio/index.js');
var input_index = require('../input/index.js');
var button_index = require('../button/index.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var _common_js_log_log = require('../_common/js/log/log.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-4fb39763.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-d3ad6e52.js');
require('../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../hooks/useControlled.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-09805e35.js');
require('../_util/noop.js');
require('../_util/useAnimation.js');
require('../hooks/useConfig.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../common/Portal.js');
require('react-dom');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../popup/utils/ref.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../_util/useMutationObserver.js');
require('../_chunks/dep-db2b3323.js');
require('../_chunks/dep-52c5813c.js');
require('../_util/useWindowSize.js');
require('../popup/defaultProps.js');
require('../_chunks/dep-3b82b90b.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../common/Check.js');
require('../_chunks/dep-1a0462cc.js');
require('../_util/helper.js');
require('../_chunks/dep-0ebc045a.js');
require('../_chunks/dep-364d4c00.js');
require('../checkbox/defaultProps.js');
require('../_chunks/dep-4ab4e7b4.js');
require('../_util/useCommonClassName.js');
require('../radio/defaultProps.js');
require('../input/Input.js');
require('../_common/js/utils/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../input/InputGroup.js');
require('../input/defaultProps.js');
require('../button/Button.js');
require('../_util/useRipple.js');
require('../_util/setStyle.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../loading/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');
require('../button/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var CheckboxGroup = checkbox_index.Checkbox.Group;
var RadioGroup = radio_index.Radio.Group;
function TableFilterController(props) {
  var tFilterValue = props.tFilterValue,
      innerFilterValue = props.innerFilterValue,
      tableFilterClasses = props.tableFilterClasses,
      isFocusClass = props.isFocusClass,
      column = props.column;

  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
    FilterIcon: tdesignIconsReact.FilterIcon
  }),
      FilterIcon = _useGlobalIcon.FilterIcon;

  var triggerElementRef = React.useRef(null);

  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("table"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var _useState = React.useState(false),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      filterPopupVisible = _useState2[0],
      setFilterPopupVisible = _useState2[1];

  var onFilterPopupVisibleChange = function onFilterPopupVisibleChange(visible) {
    var _props$onVisibleChang;

    setFilterPopupVisible(visible);
    (_props$onVisibleChang = props.onVisibleChange) === null || _props$onVisibleChang === void 0 ? void 0 : _props$onVisibleChang.call(props, visible);
  };

  var getFilterContent = function getFilterContent(column2) {
    var _column2$filter, _column2$filter2, _column2$filter3, _column2$filter4, _column2$filter5;

    var types = ["single", "multiple", "input"];

    if (column2.type && !types.includes(column2.filter.type)) {
      _common_js_log_log["default"].error("Table", "TDesign Table Error: column.filter.type must be the following: ".concat(JSON.stringify(types)));
      return;
    }

    var Component = {
      single: RadioGroup,
      multiple: CheckboxGroup,
      input: input_index.Input
    }[column2.filter.type];
    if (!Component && !(column2 !== null && column2 !== void 0 && (_column2$filter = column2.filter) !== null && _column2$filter !== void 0 && _column2$filter.component)) return;

    var filterComponentProps = _objectSpread(_objectSpread({
      options: ["single", "multiple"].includes(column2.filter.type) ? (_column2$filter2 = column2.filter) === null || _column2$filter2 === void 0 ? void 0 : _column2$filter2.list : void 0
    }, ((_column2$filter3 = column2.filter) === null || _column2$filter3 === void 0 ? void 0 : _column2$filter3.props) || {}), {}, {
      value: innerFilterValue === null || innerFilterValue === void 0 ? void 0 : innerFilterValue[column2.colKey],
      onChange: function onChange(val) {
        var _props$onInnerFilterC;

        (_props$onInnerFilterC = props.onInnerFilterChange) === null || _props$onInnerFilterC === void 0 ? void 0 : _props$onInnerFilterC.call(props, val, column2);
      }
    });

    if ((_column2$filter4 = column2.filter) !== null && _column2$filter4 !== void 0 && _column2$filter4.confirmEvents) {
      column2.filter.confirmEvents.forEach(function (event) {
        filterComponentProps[event] = function () {
          var _props$onConfirm;

          setFilterPopupVisible(false);
          (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 ? void 0 : _props$onConfirm.call(props, column2);
        };
      });
    }

    var FilterComponent = (column2 === null || column2 === void 0 ? void 0 : (_column2$filter5 = column2.filter) === null || _column2$filter5 === void 0 ? void 0 : _column2$filter5.component) || Component;
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: tableFilterClasses.contentInner
    }, /* @__PURE__ */React__default["default"].createElement(FilterComponent, _objectSpread({
      value: innerFilterValue === null || innerFilterValue === void 0 ? void 0 : innerFilterValue[column2.colKey]
    }, filterComponentProps)));
  };

  var getBottomButtons = function getBottomButtons(column2) {
    if (!column2.filter.showConfirmAndReset) return;
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: tableFilterClasses.bottomButtons
    }, /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
      theme: "default",
      size: "small",
      onClick: function onClick() {
        var _props$onReset;

        setFilterPopupVisible(false);
        (_props$onReset = props.onReset) === null || _props$onReset === void 0 ? void 0 : _props$onReset.call(props, column2);
      }
    }, locale.resetText), /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
      theme: "primary",
      size: "small",
      onClick: function onClick() {
        var _props$onConfirm2;

        setFilterPopupVisible(false);
        (_props$onConfirm2 = props.onConfirm) === null || _props$onConfirm2 === void 0 ? void 0 : _props$onConfirm2.call(props, column2);
      }
    }, locale.confirmText));
  };

  if (!column.filter || column.filter && !Object.keys(column.filter).length) return null;
  var defaultFilterIcon = t(locale.filterIcon) || /* @__PURE__ */React__default["default"].createElement(FilterIcon, null);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]([tableFilterClasses.icon, defineProperty._defineProperty({}, isFocusClass, !isEmpty.isEmpty_1(tFilterValue === null || tFilterValue === void 0 ? void 0 : tFilterValue[column.colKey]))])
  }, /* @__PURE__ */React__default["default"].createElement(popup_index.Popup, {
    visible: filterPopupVisible,
    destroyOnClose: true,
    trigger: "click",
    placement: "bottom-right",
    showArrow: true,
    overlayClassName: tableFilterClasses.popup,
    onVisibleChange: function onVisibleChange(val) {
      return onFilterPopupVisibleChange(val);
    },
    content: /* @__PURE__ */React__default["default"].createElement("div", {
      className: tableFilterClasses.popupContent
    }, getFilterContent(column), getBottomButtons(column))
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    ref: triggerElementRef
  }, props.filterIcon || defaultFilterIcon)));
}

exports["default"] = TableFilterController;
