// tdesign-react/cjs/table/PrimaryTable.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var React = require('react');
var get = require('../_chunks/dep-65cb34d8.js');
var classNames = require('classnames');
var table_BaseTable = require('./BaseTable.js');
var table_hooks_useColumnController = require('./hooks/useColumnController.js');
var table_hooks_useRowExpand = require('./hooks/useRowExpand.js');
var table_hooks_useTableHeader = require('./hooks/useTableHeader.js');
var table_hooks_useRowSelect = require('./hooks/useRowSelect.js');
var table_hooks_useSorter = require('./hooks/useSorter.js');
var table_hooks_useFilter = require('./hooks/useFilter.js');
var table_hooks_useDragSort = require('./hooks/useDragSort.js');
var table_hooks_useAsyncLoading = require('./hooks/useAsyncLoading.js');
var table_hooks_useClassName = require('./hooks/useClassName.js');
var table_EditableCell = require('./EditableCell.js');
var table_hooks_useEditableRow = require('./hooks/useEditableRow.js');
require('../_chunks/dep-8329ea23.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-09805e35.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-9c5d4f93.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-b5af0182.js');
require('../_chunks/dep-55c5ba71.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1e8de681.js');
require('../_chunks/dep-07e5d807.js');
require('../_chunks/dep-6fa7a9e9.js');
require('./TBody.js');
require('../_chunks/dep-0ebc045a.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('./TR.js');
require('./utils.js');
require('./hooks/useFixed.js');
require('../_common/js/utils/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../_chunks/dep-1a0462cc.js');
require('../_common/js/log/log.js');
require('../_common/js/utils/getScrollbarWidth.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../hooks/useDebounce.js');
require('lodash');
require('../_util/usePersistFn.js');
require('./Ellipsis.js');
require('../tooltip/index.js');
require('../tooltip/Tooltip.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_util/useAnimation.js');
require('../hooks/useConfig.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../common/Portal.js');
require('react-dom');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../popup/utils/ref.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../_util/useMutationObserver.js');
require('../_chunks/dep-db2b3323.js');
require('../_chunks/dep-52c5813c.js');
require('../_util/useWindowSize.js');
require('../popup/defaultProps.js');
require('../tooltip/defaultProps.js');
require('./hooks/useLazyLoad.js');
require('../_common/js/utils/observe.js');
require('./hooks/useRowspanAndColspan.js');
require('../locale/LocalReceiver.js');
require('../affix/index.js');
require('../affix/Affix.js');
require('../affix/defaultProps.js');
require('./THead.js');
require('./TFoot.js');
require('./hooks/useColumnResize.js');
require('../_chunks/dep-364d4c00.js');
require('../_common/js/table/set-column-width-by-drag.js');
require('../_common/js/table/utils.js');
require('../_common/js/table/recalculate-column-width.js');
require('./hooks/useAffix.js');
require('./hooks/usePagination.js');
require('../pagination/index.js');
require('../pagination/Pagination.js');
require('../_chunks/dep-70fe2bd0.js');
require('../_chunks/dep-4578b559.js');
require('../_chunks/dep-4fb39763.js');
require('../select/index.js');
require('../select/base/Select.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../select/util/helper.js');
require('../select/base/OptionGroup.js');
require('../select/defaultProps.js');
require('../select/base/Option.js');
require('../_util/useRipple.js');
require('../_util/setStyle.js');
require('../common/FakeArrow.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../loading/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');
require('../select-input/index.js');
require('../select-input/SelectInput.js');
require('../select-input/useSingle.js');
require('../input/index.js');
require('../input/Input.js');
require('tdesign-icons-react');
require('../hooks/useGlobalIcon.js');
require('../input/InputGroup.js');
require('../input/defaultProps.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('../_util/useDragSorter.js');
require('../tag-input/useTagScroll.js');
require('../tag-input/useTagList.js');
require('../tag/index.js');
require('../tag/Tag.js');
require('../tag/CheckTag.js');
require('../tag/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/defaultProps.js');
require('../select-input/useOverlayInnerStyle.js');
require('../select-input/defaultProps.js');
require('../select/base/PopupContent.js');
require('../input-number/index.js');
require('../input-number/InputNumber.js');
require('../button/index.js');
require('../button/Button.js');
require('../button/defaultProps.js');
require('../input-number/useInputNumber.js');
require('../_util/useCommonClassName.js');
require('../_common/js/input-number/number.js');
require('../_common/js/input-number/large-number.js');
require('../input-number/defaultProps.js');
require('../input-adornment/InputAdornment.js');
require('../pagination/hooks/useBoundaryJumper.js');
require('../pagination/hooks/usePrevNextJumper.js');
require('../pagination/hooks/usePageNumber.js');
require('../pagination/hooks/useTotal.js');
require('../pagination/validators.js');
require('../pagination/defaultProps.js');
require('./hooks/useStyle.js');
require('../hooks/useCommonClassName.js');
require('./hooks/useMultiHeader.js');
require('../_chunks/dep-b2e55cea.js');
require('../_chunks/dep-aa470c9c.js');
require('../checkbox/index.js');
require('../_chunks/dep-3b82b90b.js');
require('../common/Check.js');
require('../_util/helper.js');
require('../checkbox/defaultProps.js');
require('../dialog/plugin.js');
require('../dialog/Dialog.js');
require('../dialog/RenderDialog.js');
require('../_util/useLayoutEffect.js');
require('../_util/useDialogEsc.js');
require('../dialog/defaultProps.js');
require('../_util/useSetState.js');
require('../radio/index.js');
require('../_chunks/dep-4ab4e7b4.js');
require('../radio/defaultProps.js');
require('./SorterButton.js');
require('./FilterController.js');
require('../_chunks/dep-ca4cc26d.js');
require('sortablejs');
require('../_common/js/utils/swapDragArrayElement.js');
require('../form/formModel.js');
require('../_chunks/dep-266b1890.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var PrimaryTable = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _primaryTableClasses;

  var columns = props.columns,
      columnController = props.columnController,
      editableRowKeys = props.editableRowKeys,
      style = props.style,
      className = props.className;
  var primaryTableRef = React.useRef(null);

  var _useClassName = table_hooks_useClassName["default"](),
      tableDraggableClasses = _useClassName.tableDraggableClasses,
      tableBaseClass = _useClassName.tableBaseClass,
      tableSelectedClasses = _useClassName.tableSelectedClasses;

  var _useColumnController = table_hooks_useColumnController["default"](props),
      tDisplayColumns = _useColumnController.tDisplayColumns,
      renderColumnController = _useColumnController.renderColumnController;

  var _useRowExpand = table_hooks_useRowExpand["default"](props),
      showExpandedRow = _useRowExpand.showExpandedRow,
      showExpandIconColumn = _useRowExpand.showExpandIconColumn,
      getExpandColumn = _useRowExpand.getExpandColumn,
      renderExpandedRow = _useRowExpand.renderExpandedRow,
      onInnerExpandRowClick = _useRowExpand.onInnerExpandRowClick;

  var _useSorter = table_hooks_useSorter["default"](props),
      renderSortIcon = _useSorter.renderSortIcon;

  var _useRowSelect = table_hooks_useRowSelect["default"](props, tableSelectedClasses),
      formatToRowSelectColumn = _useRowSelect.formatToRowSelectColumn,
      selectedRowClassNames = _useRowSelect.selectedRowClassNames;

  var _useFilter = table_hooks_useFilter["default"](props, primaryTableRef),
      hasEmptyCondition = _useFilter.hasEmptyCondition,
      isTableOverflowHidden = _useFilter.isTableOverflowHidden,
      renderFilterIcon = _useFilter.renderFilterIcon,
      renderFirstFilterRow = _useFilter.renderFirstFilterRow;

  var _useDragSort = table_hooks_useDragSort["default"](props, primaryTableRef),
      isRowHandlerDraggable = _useDragSort.isRowHandlerDraggable,
      isRowDraggable = _useDragSort.isRowDraggable,
      isColDraggable = _useDragSort.isColDraggable,
      setDragSortColumns = _useDragSort.setDragSortColumns;

  var _useTableHeader = table_hooks_useTableHeader["default"]({
    columns: props.columns
  }),
      renderTitleWidthIcon = _useTableHeader.renderTitleWidthIcon;

  var _useAsyncLoading = table_hooks_useAsyncLoading["default"](props),
      renderAsyncLoading = _useAsyncLoading.renderAsyncLoading;

  var primaryTableClasses = (_primaryTableClasses = {}, defineProperty._defineProperty(_primaryTableClasses, tableDraggableClasses.colDraggable, isColDraggable), defineProperty._defineProperty(_primaryTableClasses, tableDraggableClasses.rowHandlerDraggable, isRowHandlerDraggable), defineProperty._defineProperty(_primaryTableClasses, tableDraggableClasses.rowDraggable, isRowDraggable), defineProperty._defineProperty(_primaryTableClasses, tableBaseClass.overflowVisible, isTableOverflowHidden === false), defineProperty._defineProperty(_primaryTableClasses, tableBaseClass.tableRowEdit, editableRowKeys), _primaryTableClasses);

  var _useEditableRow = table_hooks_useEditableRow.useEditableRow(props),
      errorListMap = _useEditableRow.errorListMap,
      editableKeysMap = _useEditableRow.editableKeysMap,
      validateRowData = _useEditableRow.validateRowData,
      validateTableData = _useEditableRow.validateTableData,
      onRuleChange = _useEditableRow.onRuleChange,
      clearValidateData = _useEditableRow.clearValidateData;

  var tRowClassNames = function () {
    var tClassNames = [props.rowClassName, selectedRowClassNames];
    return tClassNames.filter(function (v) {
      return v;
    });
  }();

  var tRowAttributes = function () {
    var tAttributes = [props.rowAttributes];

    if (isRowHandlerDraggable || isRowDraggable) {
      tAttributes.push(function (_ref) {
        var row = _ref.row;
        return {
          "data-id": get.get_1(row, props.rowKey || "id")
        };
      });
    }

    return tAttributes.filter(function (v) {
      return v;
    });
  }();

  React.useImperativeHandle(ref, function () {
    return _objectSpread({
      validateRowData: validateRowData,
      validateTableData: validateTableData,
      clearValidateData: clearValidateData
    }, primaryTableRef.current);
  });

  var getColumns = function getColumns(columns2) {
    var arr = [];

    var _loop = function _loop(i, len) {
      var _item$children, _item$edit, _item$children2, _item$children3;

      var item = _objectSpread({}, columns2[i]);

      var isDisplayColumn = ((_item$children = item.children) === null || _item$children === void 0 ? void 0 : _item$children.length) || (tDisplayColumns === null || tDisplayColumns === void 0 ? void 0 : tDisplayColumns.includes(item.colKey));
      if (!isDisplayColumn && props.columnController && tDisplayColumns) return "continue";
      item = formatToRowSelectColumn(item);

      if (item.sorter || item.filter) {
        var titleContent = table_hooks_useTableHeader.renderTitle(item, i);
        var _item = item,
            ellipsisTitle = _item.ellipsisTitle;

        item.title = function (p) {
          var _primaryTableRef$curr;

          var sortIcon = item.sorter ? renderSortIcon(p) : null;
          var filterIcon = item.filter ? renderFilterIcon(p) : null;
          var attach = (_primaryTableRef$curr = primaryTableRef.current) === null || _primaryTableRef$curr === void 0 ? void 0 : _primaryTableRef$curr.tableContentRef;
          return renderTitleWidthIcon([titleContent, sortIcon, filterIcon], p.col, p.colIndex, ellipsisTitle, attach);
        };

        item.ellipsisTitle = false;
      }

      if ((_item$edit = item.edit) !== null && _item$edit !== void 0 && _item$edit.component) {
        var oldCell = item.cell;

        item.cell = function (p) {
          var cellProps = _objectSpread(_objectSpread({}, p), {}, {
            oldCell: oldCell,
            tableBaseClass: tableBaseClass,
            cellEmptyContent: props.cellEmptyContent,
            onChange: props.onRowEdit,
            onValidate: props.onRowValidate,
            onRuleChange: onRuleChange
          });

          if (props.editableRowKeys) {
            var rowValue = get.get_1(p.row, props.rowKey || "id");
            cellProps.editable = editableKeysMap[rowValue] || false;
            var key = [rowValue, p.col.colKey].join("__");
            var errorList = errorListMap[key];
            errorList && (cellProps.errors = errorList);
          }

          if (props.editableCellState) {
            cellProps.readonly = !props.editableCellState(p);
          }

          return /* @__PURE__ */React__default["default"].createElement(table_EditableCell["default"], _objectSpread({}, cellProps));
        };
      }

      if ((_item$children2 = item.children) !== null && _item$children2 !== void 0 && _item$children2.length) {
        item.children = getColumns(item.children);
      }

      if (!item.children || (_item$children3 = item.children) !== null && _item$children3 !== void 0 && _item$children3.length) {
        arr.push(item);
      }
    };

    for (var i = 0, len = columns2.length; i < len; i++) {
      var _ret = _loop(i);

      if (_ret === "continue") continue;
    }

    return arr;
  };

  var tColumns = function () {
    var cols = getColumns(columns);

    if (showExpandIconColumn) {
      cols.unshift(getExpandColumn());
    }

    return cols;
  }();

  var onInnerPageChange = function onInnerPageChange(pageInfo, newData) {
    var _props$onPageChange, _props$onChange;

    (_props$onPageChange = props.onPageChange) === null || _props$onPageChange === void 0 ? void 0 : _props$onPageChange.call(props, pageInfo, newData);
    var changeParams = [{
      pagination: pageInfo
    }, {
      trigger: "pagination",
      currentData: newData
    }];
    (_props$onChange = props.onChange) === null || _props$onChange === void 0 ? void 0 : _props$onChange.call.apply(_props$onChange, [props].concat(changeParams));
  };

  function formatNode(api, renderInnerNode, condition, extra) {
    if (!condition) return props[api];
    var innerNode = renderInnerNode();
    var propsNode = props[api];
    if (innerNode && !propsNode) return innerNode;
    if (propsNode && !innerNode) return propsNode;

    if (innerNode && propsNode) {
      return extra !== null && extra !== void 0 && extra.reverse ? /* @__PURE__ */React__default["default"].createElement("div", null, innerNode, propsNode) : /* @__PURE__ */React__default["default"].createElement("div", null, propsNode, innerNode);
    }

    return null;
  }

  var isColumnController = !!(columnController && Object.keys(columnController).length);
  var placement = isColumnController ? columnController.placement || "top-right" : "";
  var isBottomController = isColumnController && (placement === null || placement === void 0 ? void 0 : placement.indexOf("bottom")) !== -1;
  var topContent = formatNode("topContent", renderColumnController, isColumnController && !isBottomController);
  var bottomContent = formatNode("bottomContent", renderColumnController, isBottomController, {
    reverse: true
  });
  var firstFullRow = formatNode("firstFullRow", renderFirstFilterRow, !hasEmptyCondition);
  var lastFullRow = formatNode("lastFullRow", renderAsyncLoading, !!props.asyncLoading);

  var baseTableProps = _objectSpread(_objectSpread({}, props), {}, {
    rowClassName: tRowClassNames,
    rowAttributes: tRowAttributes,
    columns: tColumns,
    topContent: topContent,
    bottomContent: bottomContent,
    firstFullRow: firstFullRow,
    lastFullRow: lastFullRow,
    onPageChange: onInnerPageChange,
    renderExpandedRow: showExpandedRow ? renderExpandedRow : void 0
  });

  if (props.expandOnRowClick) {
    baseTableProps.onRowClick = onInnerExpandRowClick;
  }

  return /* @__PURE__ */React__default["default"].createElement(table_BaseTable["default"], _objectSpread(_objectSpread({
    ref: primaryTableRef
  }, baseTableProps), {}, {
    className: classNames__default["default"](primaryTableClasses, className),
    style: style,
    onLeafColumnsChange: setDragSortColumns
  }));
});
PrimaryTable.displayName = "PrimaryTable";

exports.BASE_TABLE_ALL_EVENTS = table_BaseTable.BASE_TABLE_ALL_EVENTS;
exports["default"] = PrimaryTable;
