// tdesign-react/cjs/table/SorterButton.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var classNames = require('classnames');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var table_hooks_useClassName = require('./hooks/useClassName.js');
var tooltip_index = require('../tooltip/index.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
require('../_chunks/dep-2205decf.js');
require('../hooks/useConfig.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../tooltip/Tooltip.js');
require('../_chunks/dep-1a0462cc.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../hooks/useControlled.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-09805e35.js');
require('../_util/noop.js');
require('../_util/useAnimation.js');
require('../common/Portal.js');
require('react-dom');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../popup/utils/ref.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../_util/useMutationObserver.js');
require('../_chunks/dep-db2b3323.js');
require('../_chunks/dep-52c5813c.js');
require('../_util/useWindowSize.js');
require('../popup/defaultProps.js');
require('../tooltip/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function SorterButton(props) {
  var _props$sortType = props.sortType,
      sortType = _props$sortType === void 0 ? "all" : _props$sortType;

  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
    ChevronDownIcon: tdesignIconsReact.ChevronDownIcon
  }),
      ChevronDownIcon = _useGlobalIcon.ChevronDownIcon;

  var _useClassName = table_hooks_useClassName["default"](),
      tableSortClasses = _useClassName.tableSortClasses,
      negativeRotate180 = _useClassName.negativeRotate180;

  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("table"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var allowSortTypes = sortType === "all" ? ["asc", "desc"] : [sortType];

  var onSortIconClick = function onSortIconClick(e, direction) {
    props === null || props === void 0 ? void 0 : props.onSortIconClick(e, {
      descending: direction === "desc"
    });
  };

  function getSortIcon(direction, activeClass) {
    var defaultIcon = t(locale.sortIcon) || /* @__PURE__ */React__default["default"].createElement(ChevronDownIcon, null);
    var icon = props.sortIcon || defaultIcon;
    var sortClassName = [activeClass, tableSortClasses.sortIcon, tableSortClasses.iconDirection[direction], defineProperty._defineProperty({}, negativeRotate180, direction === "asc")];
    return /* @__PURE__ */React__default["default"].createElement("span", {
      key: direction,
      className: classNames__default["default"](sortClassName),
      onClick: function onClick(e) {
        return onSortIconClick(e, direction);
      }
    }, icon);
  }

  var classes = [tableSortClasses.trigger, defineProperty._defineProperty({}, tableSortClasses.doubleIcon, allowSortTypes.length > 1)];
  var tooltips = {
    asc: locale.sortAscendingOperationText,
    desc: locale.sortDescendingOperationText
  };
  var sortButton = allowSortTypes.map(function (direction) {
    var _props$hideSortTips;

    var activeClass = direction === props.sortOrder ? tableSortClasses.iconActive : tableSortClasses.iconDefault;
    var cancelTips = locale.sortCancelOperationText;
    var tips = direction === props.sortOrder ? cancelTips : tooltips[direction];
    if ((_props$hideSortTips = props.hideSortTips) !== null && _props$hideSortTips !== void 0 ? _props$hideSortTips : locale.hideSortTips) return getSortIcon(direction, activeClass);
    return /* @__PURE__ */React__default["default"].createElement(tooltip_index.Tooltip, _objectSpread(_objectSpread({
      content: tips,
      key: direction,
      placement: "right"
    }, props.tooltipProps), {}, {
      showArrow: false
    }), getSortIcon(direction, activeClass));
  });
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](classes)
  }, sortButton);
}
SorterButton.displayName = "SorterButton";

exports["default"] = SorterButton;
