// tdesign-react/cjs/table/TBody.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var camelCase = require('../_chunks/dep-0ebc045a.js');
var get = require('../_chunks/dep-65cb34d8.js');
var pick = require('../_chunks/dep-b5af0182.js');
var classNames = require('classnames');
var table_TR = require('./TR.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var table_hooks_useClassName = require('./hooks/useClassName.js');
var table_hooks_useRowspanAndColspan = require('./hooks/useRowspanAndColspan.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-09805e35.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-8329ea23.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-55c5ba71.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1e8de681.js');
require('../_chunks/dep-07e5d807.js');
require('../_chunks/dep-6fa7a9e9.js');
require('./utils.js');
require('./hooks/useFixed.js');
require('../_common/js/utils/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../_chunks/dep-1a0462cc.js');
require('../_common/js/log/log.js');
require('../_common/js/utils/getScrollbarWidth.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../hooks/useDebounce.js');
require('lodash');
require('../_util/usePersistFn.js');
require('./Ellipsis.js');
require('../tooltip/index.js');
require('../tooltip/Tooltip.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_util/useAnimation.js');
require('../hooks/useConfig.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../common/Portal.js');
require('react-dom');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../popup/utils/ref.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../_util/useMutationObserver.js');
require('../_chunks/dep-db2b3323.js');
require('../_chunks/dep-52c5813c.js');
require('../_util/useWindowSize.js');
require('../popup/defaultProps.js');
require('../tooltip/defaultProps.js');
require('./hooks/useLazyLoad.js');
require('../_common/js/utils/observe.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var ROW_AND_TD_LISTENERS = table_TR.ROW_LISTENERS.concat("cell-click");
var extendTableProps = ["rowKey", "rowClassName", "rowAttributes", "loading", "empty", "fixedRows", "firstFullRow", "lastFullRow", "rowspanAndColspan", "scroll", "cellEmptyContent", "onCellClick", "onPageChange", "onRowClick", "onRowDblclick", "onRowMouseover", "onRowMousedown", "onRowMouseenter", "onRowMouseleave", "onRowMouseup", "onScroll", "onScrollX", "onScrollY"];
function TBody(props) {
  var data = props.data,
      columns = props.columns,
      rowKey = props.rowKey,
      firstFullRow = props.firstFullRow,
      lastFullRow = props.lastFullRow;

  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("table"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      global = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var _useClassName = table_hooks_useClassName["default"](),
      tableFullRowClasses = _useClassName.tableFullRowClasses,
      tableBaseClass = _useClassName.tableBaseClass;

  var _useRowspanAndColspan = table_hooks_useRowspanAndColspan["default"](data, columns, rowKey, props.rowspanAndColspan),
      skipSpansMap = _useRowspanAndColspan.skipSpansMap;

  var tbodyClasses = React.useMemo(function () {
    return [tableBaseClass.body];
  }, [tableBaseClass.body]);
  var hasFullRowConfig = React.useMemo(function () {
    return firstFullRow || lastFullRow;
  }, [firstFullRow, lastFullRow]);

  var renderEmpty = function renderEmpty(columns2) {
    return /* @__PURE__ */React__default["default"].createElement("tr", {
      className: classNames__default["default"]([tableBaseClass.emptyRow, defineProperty._defineProperty({}, tableFullRowClasses.base, props.isWidthOverflow)])
    }, /* @__PURE__ */React__default["default"].createElement("td", {
      colSpan: columns2.length
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"]([tableBaseClass.empty, defineProperty._defineProperty({}, tableFullRowClasses.innerFullRow, props.isWidthOverflow)]),
      style: props.isWidthOverflow ? {
        width: "".concat(props.tableWidth, "px")
      } : {}
    }, props.empty || t(global.empty))));
  };

  var getFullRow = function getFullRow(columnLength2, type) {
    var tType = camelCase.camelCase_1(type);
    var fullRowNode = {
      "first-full-row": firstFullRow,
      "last-full-row": lastFullRow
    }[type];
    if (!fullRowNode) return null;
    var isFixedToLeft = props.isWidthOverflow && columns.find(function (col) {
      return col.fixed === "left";
    });
    var classes = [tableFullRowClasses.base, tableFullRowClasses[tType]];
    return /* @__PURE__ */React__default["default"].createElement("tr", {
      className: classNames__default["default"](classes)
    }, /* @__PURE__ */React__default["default"].createElement("td", {
      colSpan: columnLength2
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"](defineProperty._defineProperty({}, tableFullRowClasses.innerFullRow, isFixedToLeft)),
      style: isFixedToLeft ? {
        width: "".concat(props.tableWidth, "px")
      } : {}
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: tableFullRowClasses.innerFullElement
    }, fullRowNode))));
  };

  var columnLength = columns.length;
  var dataLength = data.length;
  var trNodeList = [];
  var properties = ["rowAndColFixedPosition", "scroll", "tableElm", "tableContentElm", "trs", "bufferSize", "isVirtual", "rowHeight", "scrollType"];
  data === null || data === void 0 ? void 0 : data.forEach(function (row, rowIndex) {
    var trProps = _objectSpread(_objectSpread({}, pick.pick_1(props, table_TR.TABLE_PROPS)), {}, {
      rowKey: props.rowKey || "id",
      row: row,
      columns: columns,
      rowIndex: rowIndex,
      dataLength: dataLength,
      skipSpansMap: skipSpansMap
    }, pick.pick_1(props, properties));

    if (props.onCellClick) {
      trProps.onCellClick = props.onCellClick;
    }

    var trNode = /* @__PURE__ */React__default["default"].createElement(table_TR["default"], _objectSpread(_objectSpread({
      key: get.get_1(row, props.rowKey || "id") || rowIndex
    }, trProps), {}, {
      onRowMounted: props.handleRowMounted
    }));
    trNodeList.push(trNode);

    if (props.renderExpandedRow) {
      var p = {
        row: row,
        index: rowIndex,
        columns: columns,
        tableWidth: props.tableWidth,
        isWidthOverflow: props.isWidthOverflow
      };
      var expandedContent = props.renderExpandedRow(p);
      expandedContent && trNodeList.push(expandedContent);
    }
  });
  var list = /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, getFullRow(columnLength, "first-full-row"), trNodeList, getFullRow(columnLength, "last-full-row"));
  var isEmpty = !(data !== null && data !== void 0 && data.length) && !props.loading && !hasFullRowConfig;
  var translate = "translate(0, ".concat(props.translateY, "px)");
  var posStyle = {
    transform: translate,
    "-ms-transform": translate,
    "-moz-transform": translate,
    "-webkit-transform": translate
  };
  return /* @__PURE__ */React__default["default"].createElement("tbody", {
    className: classNames__default["default"](tbodyClasses),
    style: props.isVirtual && _objectSpread({}, posStyle)
  }, isEmpty ? renderEmpty(columns) : list);
}
TBody.displayName = "TBody";

exports.ROW_AND_TD_LISTENERS = ROW_AND_TD_LISTENERS;
exports["default"] = TBody;
exports.extendTableProps = extendTableProps;
