// tdesign-react/cjs/color-picker/components/panel/hue.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var colorPicker_components_panel_slider = require('./slider.js');
require('../../../_chunks/dep-4eb93341.js');
require('classnames');
require('../../const.js');
require('../../../_util/useDrag.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var HUESlider = function HUESlider(props) {
  var color = props.color,
      baseClassName = props.baseClassName,
      disabled = props.disabled,
      onChange = props.onChange;
  return /* @__PURE__ */React__default["default"].createElement(colorPicker_components_panel_slider["default"], {
    disabled: disabled,
    baseClassName: baseClassName,
    className: "".concat(baseClassName, "__hue"),
    color: color,
    value: color.hue,
    type: "hue",
    onChange: onChange
  });
};

var HUESlider$1 = /*#__PURE__*/React__default["default"].memo(HUESlider);

exports["default"] = HUESlider$1;
