// tdesign-react/cjs/color-picker/components/panel/slider.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../_chunks/dep-4eb93341.js');
var React = require('react');
var classNames = require('classnames');
var colorPicker_const = require('../../const.js');
var _util_useDrag = require('../../../_util/useDrag.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var ColorSlider = function ColorSlider(props) {
  var color = props.color,
      _props$className = props.className,
      className = _props$className === void 0 ? "" : _props$className,
      _props$value = props.value,
      value = _props$value === void 0 ? 0 : _props$value,
      _props$railStyle = props.railStyle,
      railStyle = _props$railStyle === void 0 ? {} : _props$railStyle,
      _props$maxValue = props.maxValue,
      maxValue = _props$maxValue === void 0 ? 360 : _props$maxValue,
      baseClassName = props.baseClassName,
      disabled = props.disabled,
      onChange = props.onChange,
      type = props.type;
  var panelRef = React.useRef(null);
  var thumbRef = React.useRef(null);
  var isMovedRef = React.useRef(false);
  var panelRectRef = React.useRef({
    width: colorPicker_const.SLIDER_DEFAULT_WIDTH
  });

  var styles = function styles() {
    var width = panelRectRef.current.width;

    if (!width) {
      return;
    }

    var left = Math.round(Number(value) / Number(maxValue) * width);
    return {
      left: "".concat(left, "px"),
      color: color.rgb
    };
  };

  var handleDrag = function handleDrag(coordinate, isEnded) {
    if (disabled) {
      return;
    }

    var width = panelRectRef.current.width;
    var x = coordinate.x;
    var value2 = Math.round(x / width * Number(maxValue) * 100) / 100;
    isMovedRef.current = true;
    onChange(value2, isEnded);
  };

  var handleDragEnd = function handleDragEnd(coordinate) {
    if (disabled || !isMovedRef.current) {
      return;
    }

    handleDrag(coordinate, true);
    isMovedRef.current = false;
  };

  _util_useDrag["default"](panelRef, {
    start: function start(coordinate) {
      panelRectRef.current.width = panelRef.current.offsetWidth;
      isMovedRef.current = false;
      handleDrag(coordinate);
    },
    end: function end(coordinate) {
      handleDragEnd(coordinate);
    },
    drag: function drag(coordinate) {
      handleDrag(coordinate);
    }
  });
  React.useEffect(function () {
    panelRectRef.current.width = panelRef.current.offsetWidth || colorPicker_const.SLIDER_DEFAULT_WIDTH;
  }, []);
  var paddingStyle = {
    background: "linear-gradient(90deg, rgba(0,0,0,.0) 0%, rgba(0,0,0,.0) 93%, ".concat(props.color.rgb, " 93%, ").concat(props.color.rgb, " 100%)")
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(baseClassName, "__slider-wrapper"), "".concat(baseClassName, "__slider-wrapper--").concat(type, "-type"))
  }, type === "alpha" && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(baseClassName, "__slider-padding"),
    style: paddingStyle
  }), /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(baseClassName, "__slider"), className),
    ref: panelRef
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(baseClassName, "__rail"),
    style: railStyle
  }), /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(baseClassName, "__thumb"),
    role: "slider",
    tabIndex: 0,
    ref: thumbRef,
    style: _objectSpread({}, styles())
  })));
};

var ColorSlider$1 = /*#__PURE__*/React__default["default"].memo(ColorSlider);

exports["default"] = ColorSlider$1;
