// tdesign-react/cjs/_util/useRipple.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var hooks_useConfig = require('../hooks/useConfig.js');
var _util_useAnimation = require('./useAnimation.js');
var _util_setStyle = require('./setStyle.js');
var _util_dom = require('./dom.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-4eb93341.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('./easing.js');

var period = 200;
var noneRippleBg = "rgba(0, 0, 0, 0)";
var defaultRippleColor = "rgba(0, 0, 0, 0.35)";

var getRippleColor = function getRippleColor(el, fixedRippleColor) {
  var _el$dataset;

  if (fixedRippleColor) {
    return fixedRippleColor;
  }

  if (el !== null && el !== void 0 && (_el$dataset = el.dataset) !== null && _el$dataset !== void 0 && _el$dataset.ripple) {
    var rippleColor = el.dataset.ripple;
    return rippleColor;
  }

  var cssVariable = getComputedStyle(el).getPropertyValue("--ripple-color");

  if (cssVariable) {
    return cssVariable;
  }

  return defaultRippleColor;
};

function useRipple(ref, fixedRippleColor) {
  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useAnimation = _util_useAnimation["default"](),
      keepRipple = _useAnimation.keepRipple;

  var rippleContainer = React.useMemo(function () {
    if (!_util_dom.canUseDocument) return null;
    var container = document.createElement("div");
    container.className = "".concat(classPrefix, "-ripple");
    return container;
  }, [classPrefix]);
  var handleAddRipple = React.useCallback(function (e) {
    var el = ref === null || ref === void 0 ? void 0 : ref.current;
    var rippleColor = getRippleColor(el, fixedRippleColor);
    if (e.button !== 0 || !el || !keepRipple) return;
    if (el.classList.contains("".concat(classPrefix, "-is-active")) || el.classList.contains("".concat(classPrefix, "-is-disabled")) || el.classList.contains("".concat(classPrefix, "-is-checked")) || el.classList.contains("".concat(classPrefix, "-is-loading"))) return;
    var elStyle = getComputedStyle(el);
    var elBorder = parseInt(elStyle.borderWidth, 10);
    var border = elBorder > 0 ? elBorder : 0;
    var width = el.offsetWidth;
    var height = el.offsetHeight;

    if (rippleContainer.parentNode === null) {
      _util_setStyle["default"](rippleContainer, {
        position: "absolute",
        left: "".concat(0 - border, "px"),
        top: "".concat(0 - border, "px"),
        width: "".concat(width, "px"),
        height: "".concat(height, "px"),
        borderRadius: elStyle.borderRadius,
        pointerEvents: "none",
        overflow: "hidden"
      });
      el.appendChild(rippleContainer);
    }

    var ripple = document.createElement("div");
    ripple.className = "".concat(classPrefix, "-ripple__inner");
    _util_setStyle["default"](ripple, {
      marginTop: "0",
      marginLeft: "0",
      right: "".concat(width, "px"),
      width: "".concat(width + 20, "px"),
      height: "100%",
      transition: "transform ".concat(period, "ms cubic-bezier(.38, 0, .24, 1), background ").concat(period * 2, "ms linear"),
      transform: "skewX(-8deg)",
      pointerEvents: "none",
      position: "absolute",
      zIndex: 0,
      backgroundColor: rippleColor,
      opacity: "0.9"
    });
    var elMap = /* @__PURE__ */new WeakMap();

    for (var n = el.children.length, i = 0; i < n; ++i) {
      var child = el.children[i];

      if (child.style.zIndex === "" && child !== rippleContainer) {
        child.style.zIndex = "1";
        elMap.set(child, true);
      }
    }

    var initPosition = el.style.position ? el.style.position : getComputedStyle(el).position;

    if (initPosition === "" || initPosition === "static") {
      el.style.position = "relative";
    }

    rippleContainer.insertBefore(ripple, rippleContainer.firstChild);
    setTimeout(function () {
      ripple.style.transform = "translateX(".concat(width, "px)");
    }, 0);

    var handleClearRipple = function handleClearRipple() {
      var el2 = ref === null || ref === void 0 ? void 0 : ref.current;
      ripple.style.backgroundColor = noneRippleBg;
      if (!el2) return;
      el2.removeEventListener("pointerup", handleClearRipple, false);
      el2.removeEventListener("pointerleave", handleClearRipple, false);
      setTimeout(function () {
        ripple.remove();
        if (rippleContainer.children.length === 0) rippleContainer.remove();
      }, period * 2 + 100);
    };

    el.addEventListener("pointerup", handleClearRipple, false);
    el.addEventListener("pointerleave", handleClearRipple, false);
  }, [classPrefix, ref, fixedRippleColor, rippleContainer, keepRipple]);
  React.useEffect(function () {
    var el = ref === null || ref === void 0 ? void 0 : ref.current;
    if (!el) return;
    el.addEventListener("pointerdown", handleAddRipple, false);
    return function () {
      el.removeEventListener("pointerdown", handleAddRipple, false);
    };
  }, [handleAddRipple, fixedRippleColor, ref]);
}

exports["default"] = useRipple;
