// tdesign-react/esm/_common/js/tree/tree-node-model.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../../_chunks/dep-1e32006c.js';
import { p as pick_1 } from '../../../_chunks/dep-c776db26.js';
import { o as omit_1 } from '../../../_chunks/dep-27f96743.js';
import log from '../log/log.js';
import '../../../_chunks/dep-1cce5a60.js';
import '../../../_chunks/dep-93a10d8c.js';
import '../../../_chunks/dep-264b0e55.js';
import '../../../_chunks/dep-4e0da095.js';
import '../../../_chunks/dep-199b1581.js';
import '../../../_chunks/dep-4a64abb1.js';
import '../../../_chunks/dep-e237f0f2.js';
import '../../../_chunks/dep-4e2181de.js';
import '../../../_chunks/dep-c197ce4f.js';
import '../../../_chunks/dep-e430e2ee.js';
import '../../../_chunks/dep-e54bb0dc.js';
import '../../../_chunks/dep-b2626e28.js';
import '../../../_chunks/dep-5a2778c1.js';
import '../../../_chunks/dep-a2b97df8.js';
import '../../../_chunks/dep-c8bfa11e.js';
import '../../../_chunks/dep-afbd242f.js';
import '../../../_chunks/dep-c8db35c0.js';
import '../../../_chunks/dep-96aef4a9.js';
import '../../../_chunks/dep-f5693bf6.js';
import '../../../_chunks/dep-557a1b0f.js';
import '../../../_chunks/dep-2afd805c.js';
import '../../../_chunks/dep-3576b6cd.js';
import '../../../_chunks/dep-aadcae4d.js';
import '../../../_chunks/dep-ebf3a272.js';
import '../../../_chunks/dep-1a36642a.js';
import '../../../_chunks/dep-dc1ce7ed.js';
import '../../../_chunks/dep-7aeeb170.js';
import '../../../_chunks/dep-4ed2dce5.js';
import '../../../_chunks/dep-fdd1d1a6.js';
import '../../../_chunks/dep-614f9a08.js';
import '../../../_chunks/dep-236ef238.js';
import '../../../_chunks/dep-6a2482c6.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function getExposedProps(node) {
  var props = pick_1(node, ["value", "label", "data", "actived", "expanded", "checked", "indeterminate", "loading"]);
  return props;
}

function createNodeModel(node) {
  var props = getExposedProps(node);

  var model = _objectSpread(_objectSpread({}, props), {}, {
    getLevel: function getLevel() {
      return node.getLevel();
    },
    getIndex: function getIndex() {
      return node.getIndex();
    },
    isFirst: function isFirst() {
      return node.isFirst();
    },
    isLast: function isLast() {
      return node.isLast();
    },
    isLeaf: function isLeaf() {
      return node.isLeaf();
    },
    insertBefore: function insertBefore(newData) {
      return node.insertBefore(newData);
    },
    insertAfter: function insertAfter(newData) {
      return node.insertAfter(newData);
    },
    appendData: function appendData(data) {
      return node.append(data);
    },
    getPath: function getPath() {
      var nodes = node.getPath();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getParent: function getParent() {
      var _node$parent;

      return (_node$parent = node.parent) === null || _node$parent === void 0 ? void 0 : _node$parent.getModel();
    },
    getParents: function getParents() {
      var nodes = node.getParents();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getRoot: function getRoot() {
      var root = node.getRoot();
      return root === null || root === void 0 ? void 0 : root.getModel();
    },
    getSiblings: function getSiblings() {
      var nodes = node.getSiblings();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getChildren: function getChildren(deep) {
      var childrenModel = false;
      var children = node.children;

      if (Array.isArray(children)) {
        if (children.length > 0) {
          if (deep) {
            var nodes = node.walk();
            nodes.shift();
            childrenModel = nodes.map(function (item) {
              return item.getModel();
            });
          } else {
            childrenModel = children.map(function (item) {
              return item.getModel();
            });
          }
        } else {
          childrenModel = false;
        }
      } else if (typeof children === "boolean") {
        childrenModel = children;
      }

      return childrenModel;
    },
    remove: function remove(value) {
      if (!value) {
        node.remove();
        return;
      }

      var tree = node.tree;
      var targetNode = tree.getNode(value);

      if (!targetNode) {
        log.warnOnce("Tree", "`".concat(value, "` is not exist"));
        return;
      }

      var parents = targetNode.getParents();
      var parentValues = parents.map(function (pnode) {
        return pnode.value;
      });

      if (parentValues.indexOf(node.value) < 0) {
        log.warnOnce("Tree", "`".concat(value, "` is not a childNode of current node"));
        return;
      }

      targetNode.remove();
    },
    setData: function setData(data) {
      var _data = omit_1(data, ["children", "value", "label"]);

      var keys = node.tree.config.keys;
      var dataValue = data[(keys === null || keys === void 0 ? void 0 : keys.value) || "value"];
      var dataLabel = data[(keys === null || keys === void 0 ? void 0 : keys.label) || "label"];
      if (dataValue !== void 0) _data.value = dataValue;
      if (dataLabel !== void 0) _data.label = dataLabel;
      Object.assign(node.data, _data);
      Object.assign(node, _data);
    }
  });

  return model;
}
function updateNodeModel(model, node) {
  var props = getExposedProps(node);
  Object.assign(model, props);
}

export { createNodeModel, updateNodeModel };
