// tdesign-react/cjs/table/hooks/useColumnController.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-4eb93341.js');
var slicedToArray = require('../../_chunks/dep-9c5d4f93.js');
var toConsumableArray = require('../../_chunks/dep-8052f095.js');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var intersection = require('../../_chunks/dep-b2e55cea.js');
var classNames = require('classnames');
var checkbox_index = require('../../checkbox/index.js');
var dialog_plugin = require('../../dialog/plugin.js');
var table_hooks_useTableHeader = require('./useTableHeader.js');
var hooks_useConfig = require('../../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../../hooks/useGlobalIcon.js');
var hooks_useControlled = require('../../hooks/useControlled.js');
var table_utils = require('../utils.js');
var button_index = require('../../button/index.js');
require('../../_chunks/dep-2205decf.js');
require('../../_chunks/dep-606cfe2a.js');
require('../../_chunks/dep-afce8659.js');
require('../../_chunks/dep-61a8a7b0.js');
require('../../_chunks/dep-2a90f794.js');
require('../../_chunks/dep-8f18a7c4.js');
require('../../_chunks/dep-09805e35.js');
require('../../_chunks/dep-1fff9729.js');
require('../../_chunks/dep-aa470c9c.js');
require('../../_chunks/dep-46cf36fd.js');
require('../../_chunks/dep-f5e18a2f.js');
require('../../_chunks/dep-53ba6729.js');
require('../../_chunks/dep-aab723b3.js');
require('../../_chunks/dep-384b291c.js');
require('../../_chunks/dep-bb60493d.js');
require('../../_chunks/dep-aafeb50a.js');
require('../../_chunks/dep-d3ad6e52.js');
require('../../_chunks/dep-12656997.js');
require('../../_chunks/dep-6fa7a9e9.js');
require('../../_chunks/dep-ad854ba5.js');
require('../../_chunks/dep-3b82b90b.js');
require('../../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../../common/Check.js');
require('../../_chunks/dep-1a0462cc.js');
require('../../_util/helper.js');
require('../../_chunks/dep-0ebc045a.js');
require('../../_chunks/dep-c76c2a08.js');
require('../../_chunks/dep-dc6c808e.js');
require('../../_chunks/dep-76deafcc.js');
require('../../_chunks/dep-4344eec8.js');
require('../../_chunks/dep-fe984d70.js');
require('../../_chunks/dep-a0b5d8f6.js');
require('../../_chunks/dep-72020528.js');
require('../../_chunks/dep-eebdbd74.js');
require('../../_chunks/dep-9cd0fde8.js');
require('../../_chunks/dep-994ec160.js');
require('../../_chunks/dep-9429a38a.js');
require('../../_common/js/global-config/locale/zh_CN.js');
require('../../_common/js/global-config/default-config.js');
require('../../_util/noop.js');
require('../../_chunks/dep-364d4c00.js');
require('../../checkbox/defaultProps.js');
require('react-dom');
require('../../dialog/Dialog.js');
require('../../_chunks/dep-9a2dbbb1.js');
require('../../locale/LocalReceiver.js');
require('../../dialog/RenderDialog.js');
require('react-transition-group');
require('../../common/Portal.js');
require('../../_util/dom.js');
require('raf');
require('../../_util/easing.js');
require('../../_util/useLayoutEffect.js');
require('../../_util/useDialogEsc.js');
require('../../dialog/defaultProps.js');
require('../../_util/useSetState.js');
require('../../button/Button.js');
require('../../_util/useRipple.js');
require('../../_util/useAnimation.js');
require('../../_util/setStyle.js');
require('../../loading/index.js');
require('../../loading/Loading.js');
require('../../loading/gradient.js');
require('../../_common/js/loading/circle-adapter.js');
require('../../_common/js/utils/set-style.js');
require('../../_common/js/utils/helper.js');
require('../../loading/defaultProps.js');
require('../../loading/plugin.js');
require('../../button/defaultProps.js');
require('../../_common/js/log/log.js');
require('./useMultiHeader.js');
require('./useClassName.js');
require('../Ellipsis.js');
require('../../tooltip/index.js');
require('../../tooltip/Tooltip.js');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-popper');
require('../../popup/hooks/useTrigger.js');
require('react-is');
require('../../popup/utils/ref.js');
require('../../_util/composeRefs.js');
require('../../popup/utils/transition.js');
require('../../_util/useMutationObserver.js');
require('../../_chunks/dep-db2b3323.js');
require('../../_chunks/dep-52c5813c.js');
require('../../_util/useWindowSize.js');
require('../../popup/defaultProps.js');
require('../../tooltip/defaultProps.js');
require('../../hooks/useDebounce.js');
require('lodash');
require('../../_util/usePersistFn.js');
require('../../_chunks/dep-65cb34d8.js');
require('../../_chunks/dep-8329ea23.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var CheckboxGroup = checkbox_index.Checkbox.Group;
function getColumnKeys(columns) {
  var keys = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : /* @__PURE__ */new Set();

  for (var i = 0, len = columns.length; i < len; i++) {
    var _col$children;

    var col = columns[i];
    col.colKey && keys.add(col.colKey);

    if ((_col$children = col.children) !== null && _col$children !== void 0 && _col$children.length) {
      getColumnKeys(col.children, keys);
    }
  }

  return keys;
}
function useColumnController(props) {
  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix,
      table = _useConfig.table;

  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
    SettingIcon: tdesignIconsReact.SettingIcon
  }),
      SettingIcon = _useGlobalIcon.SettingIcon;

  var columns = props.columns,
      columnController = props.columnController,
      displayColumns = props.displayColumns,
      columnControllerVisible = props.columnControllerVisible;
  var dialogInstance = React.useRef();

  var enabledColKeys = function () {
    var arr = ((columnController === null || columnController === void 0 ? void 0 : columnController.fields) || toConsumableArray._toConsumableArray(getColumnKeys(columns))).filter(function (v) {
      return v;
    });

    return new Set(arr);
  }();

  var keys = toConsumableArray._toConsumableArray(getColumnKeys(columns));

  var _useControlled = hooks_useControlled["default"](props, "displayColumns", props.onDisplayColumnsChange),
      _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
      tDisplayColumns = _useControlled2[0],
      setTDisplayColumns = _useControlled2[1];

  var defaultColumnCheckboxKeys = displayColumns || props.defaultDisplayColumns || keys;
  var columnCheckboxKeys = React.useRef(defaultColumnCheckboxKeys);
  React.useEffect(function () {
    var _dialogInstance$curre;

    columnCheckboxKeys.current = toConsumableArray._toConsumableArray(displayColumns || props.defaultDisplayColumns || keys);
    (_dialogInstance$curre = dialogInstance.current) === null || _dialogInstance$curre === void 0 ? void 0 : _dialogInstance$curre.update({
      body: getDialogContent()
    });
  }, [displayColumns]);

  function getCheckboxOptions(columns2) {
    var arr = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];

    for (var i = 0, len = columns2.length; i < len; i++) {
      var _item$children;

      var item = columns2[i];

      if (item.colKey) {
        arr.push({
          label: table_hooks_useTableHeader.renderTitle(item, i),
          value: item.colKey,
          disabled: !enabledColKeys.has(item.colKey)
        });
      }

      if ((_item$children = item.children) !== null && _item$children !== void 0 && _item$children.length) {
        getCheckboxOptions(item.children, arr);
      }
    }

    return arr;
  }

  var handleCheckChange = function handleCheckChange(val, ctx) {
    var _props$onColumnChange;

    columnCheckboxKeys.current = val;
    var params = {
      columns: val,
      type: ctx.type,
      currentColumn: table_utils.getCurrentRowByKey(columns, String(ctx.current)),
      e: ctx.e
    };
    (_props$onColumnChange = props.onColumnChange) === null || _props$onColumnChange === void 0 ? void 0 : _props$onColumnChange.call(props, params);
    dialogInstance.current.update({
      body: getDialogContent()
    });
  };

  var handleClickAllShowColumns = function handleClickAllShowColumns(checked, ctx) {
    if (checked) {
      var _props$onColumnChange2;

      var checkboxOptions = getCheckboxOptions(columns);
      var newData = (checkboxOptions === null || checkboxOptions === void 0 ? void 0 : checkboxOptions.map(function (t) {
        return t.value;
      })) || [];
      columnCheckboxKeys.current = newData;
      (_props$onColumnChange2 = props.onColumnChange) === null || _props$onColumnChange2 === void 0 ? void 0 : _props$onColumnChange2.call(props, {
        type: "check",
        columns: newData,
        e: ctx.e
      });
    } else {
      var _props$onColumnChange3;

      var disabledColKeys = getCheckboxOptions(columns).filter(function (t) {
        return t.disabled;
      }).map(function (t) {
        return t.value;
      });
      columnCheckboxKeys.current = disabledColKeys;
      (_props$onColumnChange3 = props.onColumnChange) === null || _props$onColumnChange3 === void 0 ? void 0 : _props$onColumnChange3.call(props, {
        type: "uncheck",
        columns: disabledColKeys,
        e: ctx.e
      });
    }

    dialogInstance.current.update({
      body: getDialogContent()
    });
  };

  function getDialogContent() {
    var checkboxOptions = getCheckboxOptions(columns);
    var intersectionChecked = intersection.intersection_1(columnCheckboxKeys.current, toConsumableArray._toConsumableArray(enabledColKeys));
    var widthMode = (columnController === null || columnController === void 0 ? void 0 : columnController.displayType) === "fixed-width" ? "fixed" : "auto";
    var checkedLength = intersectionChecked.length;
    var isCheckedAll = checkedLength === enabledColKeys.size;
    var isIndeterminate = checkedLength > 0 && checkedLength < enabledColKeys.size;
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"](["".concat(classPrefix, "-table__column-controller"), "".concat(classPrefix, "-table__column-controller--").concat(widthMode)])
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-table__column-controller-body")
    }, /* @__PURE__ */React__default["default"].createElement("p", {
      className: "".concat(classPrefix, "-table__column-controller-desc")
    }, table.columnConfigDescriptionText), /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-table__column-controller-block")
    }, /* @__PURE__ */React__default["default"].createElement(checkbox_index.Checkbox, {
      indeterminate: isIndeterminate,
      checked: isCheckedAll,
      onChange: handleClickAllShowColumns
    }, table.selectAllText)), /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-table__column-controller-block")
    }, /* @__PURE__ */React__default["default"].createElement(CheckboxGroup, _objectSpread(_objectSpread({
      options: checkboxOptions
    }, (columnController === null || columnController === void 0 ? void 0 : columnController.checkboxProps) || {}), {}, {
      value: columnCheckboxKeys.current,
      onChange: handleCheckChange
    })))));
  }

  var handleToggleColumnController = function handleToggleColumnController() {
    dialogInstance.current = dialog_plugin.DialogPlugin.confirm(_objectSpread({
      header: table.columnConfigTitleText,
      body: getDialogContent(),
      confirmBtn: table.confirmText,
      cancelBtn: table.cancelText,
      width: 612,
      onConfirm: function onConfirm() {
        setTDisplayColumns(toConsumableArray._toConsumableArray(columnCheckboxKeys.current));

        if (columnControllerVisible === void 0) {
          dialogInstance.current.hide();
        } else {
          var _props$onColumnContro;

          (_props$onColumnContro = props.onColumnControllerVisibleChange) === null || _props$onColumnContro === void 0 ? void 0 : _props$onColumnContro.call(props, false, {
            trigger: "cancel"
          });
        }
      },
      onClose: function onClose() {
        if (columnControllerVisible === void 0) {
          dialogInstance.current.hide();
        } else {
          var _props$onColumnContro2;

          (_props$onColumnContro2 = props.onColumnControllerVisibleChange) === null || _props$onColumnContro2 === void 0 ? void 0 : _props$onColumnContro2.call(props, false, {
            trigger: "confirm"
          });
        }
      }
    }, (columnController === null || columnController === void 0 ? void 0 : columnController.dialogProps) || {}));
  };

  React.useEffect(function () {
    if (columnControllerVisible === void 0) return;

    if (dialogInstance.current) {
      columnControllerVisible ? dialogInstance.current.show() : dialogInstance.current.hide();
    } else {
      columnControllerVisible && handleToggleColumnController();
    }
  }, [columnControllerVisible, dialogInstance]);

  var renderColumnController = function renderColumnController() {
    var _props$columnControll;

    var isColumnController = !!(columnController && Object.keys(columnController).length);
    var placement = isColumnController ? columnController.placement || "top-right" : "";
    if (isColumnController && columnController.hideTriggerButton) return null;
    var classes = ["".concat(classPrefix, "-table__column-controller-trigger"), defineProperty._defineProperty({}, "".concat(classPrefix, "-align-").concat(placement), !!placement)];
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"](classes)
    }, /* @__PURE__ */React__default["default"].createElement(button_index.Button, _objectSpread({
      theme: "default",
      variant: "outline",
      onClick: handleToggleColumnController,
      content: table.columnConfigButtonText,
      icon: /* @__PURE__ */React__default["default"].createElement(SettingIcon, null)
    }, (_props$columnControll = props.columnController) === null || _props$columnControll === void 0 ? void 0 : _props$columnControll.buttonProps)));
  };

  return {
    tDisplayColumns: tDisplayColumns,
    renderColumnController: renderColumnController
  };
}

exports["default"] = useColumnController;
exports.getColumnKeys = getColumnKeys;
