// tdesign-react/cjs/table/hooks/useTreeSelect.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-4eb93341.js');
var slicedToArray = require('../../_chunks/dep-9c5d4f93.js');
var toConsumableArray = require('../../_chunks/dep-8052f095.js');
var React = require('react');
var get = require('../../_chunks/dep-65cb34d8.js');
var intersection = require('../../_chunks/dep-b2e55cea.js');
var hooks_useControlled = require('../../hooks/useControlled.js');
require('../../_chunks/dep-2205decf.js');
require('../../_chunks/dep-606cfe2a.js');
require('../../_chunks/dep-8329ea23.js');
require('../../_chunks/dep-8f18a7c4.js');
require('../../_chunks/dep-2a90f794.js');
require('../../_chunks/dep-09805e35.js');
require('../../_chunks/dep-61a8a7b0.js');
require('../../_chunks/dep-1fff9729.js');
require('../../_chunks/dep-46cf36fd.js');
require('../../_chunks/dep-f5e18a2f.js');
require('../../_chunks/dep-53ba6729.js');
require('../../_chunks/dep-aab723b3.js');
require('../../_chunks/dep-384b291c.js');
require('../../_chunks/dep-afce8659.js');
require('../../_chunks/dep-aa470c9c.js');
require('../../_chunks/dep-bb60493d.js');
require('../../_chunks/dep-aafeb50a.js');
require('../../_chunks/dep-d3ad6e52.js');
require('../../_chunks/dep-12656997.js');
require('../../_chunks/dep-6fa7a9e9.js');
require('../../_chunks/dep-ad854ba5.js');
require('../../_chunks/dep-c76c2a08.js');
require('../../_chunks/dep-dc6c808e.js');
require('../../_chunks/dep-76deafcc.js');
require('../../_util/noop.js');

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var childrenMap = /* @__PURE__ */new Map();
function getChildrenData(treeDataMap, data, keys, r) {
  if (childrenMap.get(data)) return childrenMap.get(data);
  var result = r || {
    allChildren: [],
    allChildrenKeys: [],
    leafNodeKeys: []
  };
  var children = get.get_1(data, keys.childrenKey);
  if (!children || !children.length) return result;
  var selectableChildren = children.filter(function (item) {
    var _treeDataMap$get;

    return !((_treeDataMap$get = treeDataMap.get(get.get_1(item, keys.rowKey))) !== null && _treeDataMap$get !== void 0 && _treeDataMap$get.disabled);
  });
  result.allChildren = toConsumableArray._toConsumableArray(new Set(result.allChildren.concat(selectableChildren)));

  for (var i = 0, len = children.length; i < len; i++) {
    var tItem = children[i];
    var c = get.get_1(tItem, keys.childrenKey);

    if (c !== null && c !== void 0 && c.length) {
      var nextLevelData = getChildrenData(treeDataMap, tItem, keys, result);
      result.allChildren = toConsumableArray._toConsumableArray(new Set(result.allChildren.concat(nextLevelData.allChildren)));
    }
  }

  for (var _i = 0, _len = result.allChildren.length; _i < _len; _i++) {
    var item = result.allChildren[_i];
    var children2 = get.get_1(item, keys.childrenKey);
    var rowValue = get.get_1(item, keys.rowKey);
    result.allChildrenKeys.push(rowValue);

    if (!children2 || !children2.length) {
      result.leafNodeKeys.push(rowValue);
    }
  }

  result.allChildrenKeys = toConsumableArray._toConsumableArray(new Set(result.allChildrenKeys));
  result.leafNodeKeys = toConsumableArray._toConsumableArray(new Set(result.leafNodeKeys));
  return result;
}
function removeChildrenKeys(p, r) {
  var selectedRowKeys = p.selectedRowKeys,
      removeKeys = p.removeKeys;
  var result = r || {
    data: [],
    keys: []
  };

  for (var i = 0, len = selectedRowKeys.length; i < len; i++) {
    var key = selectedRowKeys[i];

    if (!removeKeys.includes(key)) {
      result.keys.push(key);
    }
  }

  return result;
}
function getRowDataByKeys(p) {
  var treeDataMap = p.treeDataMap,
      selectedRowKeys = p.selectedRowKeys;
  var result = [];

  for (var i = 0, len = selectedRowKeys.length; i < len; i++) {
    var key = selectedRowKeys[i];
    result.push(treeDataMap.get(key));
  }

  return result;
}
function useTreeSelect(props, treeDataMap) {
  var tree = props.tree,
      rowKey = props.rowKey,
      data = props.data,
      indeterminateSelectedRowKeys = props.indeterminateSelectedRowKeys;

  var _useState = React.useState([]),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      tIndeterminateSelectedRowKeys = _useState2[0],
      setTIndeterminateSelectedRowKeys = _useState2[1];

  var _useControlled = hooks_useControlled["default"](props, "selectedRowKeys", props.onSelectChange, {
    defaultSelectedRowKeys: props.defaultSelectedRowKeys || []
  }),
      _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
      tSelectedRowKeys = _useControlled2[0],
      setTSelectedRowKeys = _useControlled2[1];

  var rowDataKeys = React.useMemo(function () {
    return {
      rowKey: rowKey || "id",
      childrenKey: (tree === null || tree === void 0 ? void 0 : tree.childrenKey) || "children"
    };
  }, [rowKey, tree === null || tree === void 0 ? void 0 : tree.childrenKey]);
  React.useEffect(function () {
    if (!tree || !treeDataMap.size || tree.checkStrictly) return;
    updateIndeterminateState();
  }, [tSelectedRowKeys, data, tree, treeDataMap]);

  function updateIndeterminateState() {
    if (!tree || tree.checkStrictly) return;

    if (!tSelectedRowKeys.length) {
      setTIndeterminateSelectedRowKeys([]);
      return;
    }

    var keys = [];
    var parentMap = {};

    for (var i = 0, len = tSelectedRowKeys.length; i < len; i++) {
      var rowValue = tSelectedRowKeys[i];
      var state = treeDataMap.get(rowValue);
      var children = get.get_1(state.row, rowDataKeys.childrenKey);

      if (!children || !children.length) {
        var parentTmp = state.parent;

        while (parentTmp) {
          if (!parentMap[parentTmp.id]) {
            parentMap[parentTmp.id] = [];
          }

          parentMap[parentTmp.id].push(state.row);
          var checkedLength = parentMap[parentTmp.id].length;

          var _getChildrenData = getChildrenData(treeDataMap, parentTmp.row, rowDataKeys),
              allChildrenKeys = _getChildrenData.allChildrenKeys;

          var parentTmpIndex = keys.indexOf(parentTmp.id);
          var selectedIndex = tSelectedRowKeys.indexOf(parentTmp.id);

          if (checkedLength > 0 && checkedLength < allChildrenKeys.length && selectedIndex === -1) {
            parentTmpIndex === -1 && keys.push(parentTmp.id);
          } else {
            parentTmpIndex !== -1 && keys.splice(parentTmpIndex, 1);
          }

          parentTmp = parentTmp.parent;
        }
      }
    }

    setTIndeterminateSelectedRowKeys(keys);
  }

  function updateParentCheckedState(selectedKeys, currentRowKey, type) {
    if (!tree || tree.checkStrictly) return;

    var keys = toConsumableArray._toConsumableArray(selectedKeys);

    var state = treeDataMap.get(currentRowKey);
    var parentTmp = state.parent;

    while (parentTmp) {
      var _getChildrenData2 = getChildrenData(treeDataMap, parentTmp.row, rowDataKeys),
          leafNodeKeys = _getChildrenData2.leafNodeKeys;

      var checkedChildrenKeys = intersection.intersection_1(leafNodeKeys, selectedKeys);
      var selectedIndex = keys.indexOf(parentTmp.id);

      if (type === "uncheck") {
        selectedIndex !== -1 && keys.splice(selectedIndex, 1);
      } else if (checkedChildrenKeys.length === leafNodeKeys.length) {
        selectedIndex === -1 && keys.push(parentTmp.id);
      }

      parentTmp = parentTmp.parent;
    }

    return keys;
  }

  function onInnerSelectChange(rowKeys, extraData) {
    if (!tree || tree.checkStrictly) {
      setTSelectedRowKeys(rowKeys, extraData);
      return;
    }

    if (extraData.currentRowKey === "CHECK_ALL_BOX") {
      handleSelectAll(extraData);
    } else {
      handleSelect(rowKeys, extraData);
    }
  }

  function handleSelectAll(extraData) {
    var newRowKeys = [];
    var newRowData = [];

    if ((extraData === null || extraData === void 0 ? void 0 : extraData.type) === "check") {
      var arr = toConsumableArray._toConsumableArray(treeDataMap.values());

      for (var i = 0, len = arr.length; i < len; i++) {
        var item = arr[i];

        if (!(item !== null && item !== void 0 && item.disabled)) {
          newRowData.push(item.row);
          newRowKeys.push(get.get_1(item.row, rowDataKeys.rowKey));
        }
      }
    }

    var newExtraData = _objectSpread(_objectSpread({}, extraData), {}, {
      selectedRowData: newRowData || []
    });

    setTSelectedRowKeys(newRowKeys, newExtraData);
  }

  function handleSelect(rowKeys, extraData) {
    var newRowKeys = toConsumableArray._toConsumableArray(rowKeys);

    if (tree.checkStrictly === false) {
      if (extraData.type === "check") {
        var result = getChildrenData(treeDataMap, extraData.currentRowData, rowDataKeys);
        var allChildrenKeys = result.allChildrenKeys;
        childrenMap.set(extraData.currentRowData, result);
        newRowKeys = toConsumableArray._toConsumableArray(new Set(newRowKeys.concat(allChildrenKeys)));
      } else if (extraData.type === "uncheck") {
        var children = getChildrenData(treeDataMap, extraData.currentRowData, rowDataKeys);

        var _result = removeChildrenKeys({
          selectedRowKeys: rowKeys,
          removeKeys: children.allChildrenKeys
        });

        newRowKeys = _result.keys;
      }
    }

    newRowKeys = updateParentCheckedState(newRowKeys, extraData.currentRowKey, extraData.type);
    var newRowData = getRowDataByKeys({
      treeDataMap: treeDataMap,
      selectedRowKeys: newRowKeys
    });

    var newExtraData = _objectSpread(_objectSpread({}, extraData), {}, {
      selectedRowData: newRowData
    });

    setTSelectedRowKeys(newRowKeys, newExtraData);
  }

  return {
    tIndeterminateSelectedRowKeys: indeterminateSelectedRowKeys || tIndeterminateSelectedRowKeys,
    onInnerSelectChange: onInnerSelectChange
  };
}

exports.childrenMap = childrenMap;
exports["default"] = useTreeSelect;
exports.getChildrenData = getChildrenData;
exports.getRowDataByKeys = getRowDataByKeys;
exports.removeChildrenKeys = removeChildrenKeys;
