// tdesign-react/cjs/jumper/Jumper.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var React = require('react');
var classNames = require('classnames');
var isObject = require('../_chunks/dep-aab723b3.js');
var tdesignIconsReact = require('tdesign-icons-react');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var button_index = require('../button/index.js');
var jumper_defaultProps = require('./defaultProps.js');
var _util_noop = require('../_util/noop.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../button/Button.js');
require('../_chunks/dep-1a0462cc.js');
require('../_util/useRipple.js');
require('../_util/useAnimation.js');
require('../_util/setStyle.js');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../_chunks/dep-9c5d4f93.js');
require('../_chunks/dep-2205decf.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');
require('../button/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var Jumper = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
    RoundIcon: tdesignIconsReact.RoundIcon,
    ChevronUpIcon: tdesignIconsReact.ChevronUpIcon,
    ChevronDownIcon: tdesignIconsReact.ChevronDownIcon,
    ChevronLeftIcon: tdesignIconsReact.ChevronLeftIcon,
    ChevronRightIcon: tdesignIconsReact.ChevronRightIcon
  }),
      RoundIcon = _useGlobalIcon.RoundIcon,
      ChevronUpIcon = _useGlobalIcon.ChevronUpIcon,
      ChevronDownIcon = _useGlobalIcon.ChevronDownIcon,
      ChevronLeftIcon = _useGlobalIcon.ChevronLeftIcon,
      ChevronRightIcon = _useGlobalIcon.ChevronRightIcon;

  var variant = props.variant,
      tips = props.tips,
      showCurrent = props.showCurrent,
      disabled = props.disabled,
      layout = props.layout,
      size = props.size,
      _props$onChange = props.onChange,
      onChange = _props$onChange === void 0 ? _util_noop["default"] : _props$onChange,
      className = props.className,
      style = props.style;
  var titleConfig = React.useMemo(function () {
    if (isObject.isObject_1(tips)) return tips;
    if (tips === true) return {
      prev: "\u4E0A\u4E00\u9875",
      current: "\u5F53\u524D",
      next: "\u4E0B\u4E00\u9875"
    };
    return {};
  }, [tips]);
  var disabledConfig = React.useMemo(function () {
    if (isObject.isObject_1(disabled)) return disabled;
    if (disabled === true) return {
      prev: true,
      current: true,
      next: true
    };
    return {
      prev: false,
      current: false,
      next: false
    };
  }, [disabled]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-jumper"), className, defineProperty._defineProperty({}, "".concat(classPrefix, "-jumper--outline"), variant === "outline")),
    ref: ref,
    style: style
  }, /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    title: titleConfig.prev,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "prev"
      });
    },
    icon: layout === "horizontal" ? /* @__PURE__ */React__default["default"].createElement(ChevronLeftIcon, null) : /* @__PURE__ */React__default["default"].createElement(ChevronUpIcon, null),
    className: "".concat(classPrefix, "-jumper__prev"),
    disabled: disabledConfig.prev
  }), showCurrent && /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    title: titleConfig.current,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "current"
      });
    },
    icon: /* @__PURE__ */React__default["default"].createElement(RoundIcon, null),
    className: "".concat(classPrefix, "-jumper__current"),
    disabled: disabledConfig.current
  }), /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    title: titleConfig.next,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "next"
      });
    },
    icon: layout === "horizontal" ? /* @__PURE__ */React__default["default"].createElement(ChevronRightIcon, null) : /* @__PURE__ */React__default["default"].createElement(ChevronDownIcon, null),
    className: "".concat(classPrefix, "-jumper__next"),
    disabled: disabledConfig.next
  }));
});
Jumper.displayName = "Jumper";
Jumper.defaultProps = jumper_defaultProps.jumperDefaultProps;

exports["default"] = Jumper;
