// tdesign-react/cjs/time-picker/TimePicker.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var classNames = require('classnames');
var dayjs = require('dayjs');
var toInteger = require('../_chunks/dep-e4c3e633.js');
var tdesignIconsReact = require('tdesign-icons-react');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var _util_noop = require('../_util/noop.js');
var selectInput_index = require('../select-input/index.js');
var timePicker_TimeRangePicker = require('./TimeRangePicker.js');
var timePicker_panel_TimePickerPanel = require('./panel/TimePickerPanel.js');
var timePicker_hooks_useTimePickerTextConfig = require('./hooks/useTimePickerTextConfig.js');
var _common_js_timePicker_utils = require('../_common/js/time-picker/utils.js');
var _common_js_timePicker_const = require('../_common/js/time-picker/const.js');
var timePicker_defaultProps = require('./defaultProps.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-52c5813c.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-09805e35.js');
require('../_chunks/dep-1fff9729.js');
require('hoist-non-react-statics');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../select-input/SelectInput.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../_util/useAnimation.js');
require('../common/Portal.js');
require('react-dom');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../popup/utils/ref.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../_util/useMutationObserver.js');
require('../_chunks/dep-db2b3323.js');
require('../_util/useWindowSize.js');
require('../popup/defaultProps.js');
require('../select-input/useSingle.js');
require('../_chunks/dep-b5af0182.js');
require('../_chunks/dep-8329ea23.js');
require('../_chunks/dep-55c5ba71.js');
require('../_chunks/dep-1e8de681.js');
require('../_chunks/dep-07e5d807.js');
require('../input/index.js');
require('../input/Input.js');
require('../_chunks/dep-1a0462cc.js');
require('../_common/js/utils/helper.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../input/InputGroup.js');
require('../locale/LocalReceiver.js');
require('../input/defaultProps.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../loading/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('../_util/useDragSorter.js');
require('../tag-input/useTagScroll.js');
require('../tag-input/useTagList.js');
require('lodash');
require('../tag/index.js');
require('../tag/Tag.js');
require('../tag/CheckTag.js');
require('../tag/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/defaultProps.js');
require('../select-input/useOverlayInnerStyle.js');
require('../select-input/defaultProps.js');
require('../range-input/index.js');
require('../range-input/RangeInput.js');
require('../range-input/defaultProps.js');
require('../range-input/RangeInputPopup.js');
require('./panel/SinglePanel.js');
require('../hooks/useDebounce.js');
require('../_util/usePersistFn.js');
require('../hooks/usePropsRef.js');
require('../button/index.js');
require('../button/Button.js');
require('../_util/useRipple.js');
require('../_util/setStyle.js');
require('../button/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
dayjs__default["default"].extend(toInteger.customParseFormat);
var TimePicker = _util_forwardRefWithStatics["default"](function (props, ref) {
  var TEXT_CONFIG = timePicker_hooks_useTimePickerTextConfig.useTimePickerTextConfig();
  var allowInput = props.allowInput,
      className = props.className,
      clearable = props.clearable,
      disabled = props.disabled,
      style = props.style,
      _props$format = props.format,
      format = _props$format === void 0 ? _common_js_timePicker_const.DEFAULT_FORMAT : _props$format,
      _props$hideDisabledTi = props.hideDisabledTime,
      hideDisabledTime = _props$hideDisabledTi === void 0 ? true : _props$hideDisabledTi,
      _props$steps = props.steps,
      steps = _props$steps === void 0 ? _common_js_timePicker_const.DEFAULT_STEPS : _props$steps,
      _props$placeholder = props.placeholder,
      placeholder = _props$placeholder === void 0 ? TEXT_CONFIG.placeholder : _props$placeholder,
      disableTime = props.disableTime,
      _props$onBlur = props.onBlur,
      onBlur = _props$onBlur === void 0 ? _util_noop["default"] : _props$onBlur,
      _props$onClose = props.onClose,
      onClose = _props$onClose === void 0 ? _util_noop["default"] : _props$onClose,
      _props$onFocus = props.onFocus,
      onFocus = _props$onFocus === void 0 ? _util_noop["default"] : _props$onFocus,
      _props$onOpen = props.onOpen,
      onOpen = _props$onOpen === void 0 ? _util_noop["default"] : _props$onOpen;

  var _useControlled = hooks_useControlled["default"](props, "value", props.onChange),
      _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
      value = _useControlled2[0],
      onChange = _useControlled2[1];

  var _useState = React.useState(false),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      isPanelShowed = _useState2[0],
      setPanelShow = _useState2[1];

  var _useState3 = React.useState(""),
      _useState4 = slicedToArray._slicedToArray(_useState3, 2),
      currentValue = _useState4[0],
      setCurrentValue = _useState4[1];

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
    TimeIcon: tdesignIconsReact.TimeIcon
  }),
      TimeIcon = _useGlobalIcon.TimeIcon;

  var name = "".concat(classPrefix, "-time-picker");
  var inputClasses = classNames__default["default"]("".concat(name, "__group"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-focused"), isPanelShowed));

  var handleShowPopup = function handleShowPopup(visible, context) {
    setPanelShow(visible);
    visible ? onOpen(context) : onClose(context);
  };

  var handleClear = function handleClear(context) {
    var e = context.e;
    e.stopPropagation();
    setCurrentValue("");
    onChange(null);
  };

  var handleInputChange = function handleInputChange(value2) {
    setCurrentValue(value2);
  };

  var handleInputBlur = function handleInputBlur(value2, _ref) {
    var e = _ref.e;

    if (allowInput) {
      var isValidTime = _common_js_timePicker_utils.validateInputValue(currentValue, format);

      if (isValidTime) {
        onChange(_common_js_timePicker_utils.formatInputValue(currentValue, format));
      }
    }

    onBlur({
      value: value2,
      e: e
    });
  };

  var handleClickConfirm = function handleClickConfirm() {
    var isValidTime = _common_js_timePicker_utils.validateInputValue(currentValue, format);
    if (isValidTime) onChange(currentValue);
    setPanelShow(false);
  };

  React.useEffect(function () {
    setCurrentValue(isPanelShowed ? value !== null && value !== void 0 ? value : "" : "");
  }, [isPanelShowed, value]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](name, className),
    ref: ref,
    style: style
  }, /* @__PURE__ */React__default["default"].createElement(selectInput_index.SelectInput, {
    onFocus: onFocus,
    onClear: handleClear,
    disabled: disabled,
    clearable: clearable,
    allowInput: allowInput,
    className: inputClasses,
    suffixIcon: /* @__PURE__ */React__default["default"].createElement(TimeIcon, null),
    popupVisible: isPanelShowed,
    onInputChange: handleInputChange,
    onBlur: handleInputBlur,
    onPopupVisibleChange: handleShowPopup,
    placeholder: !value ? placeholder : void 0,
    value: isPanelShowed ? currentValue : value !== null && value !== void 0 ? value : void 0,
    inputValue: isPanelShowed ? currentValue : value !== null && value !== void 0 ? value : void 0,
    inputProps: props.inputProps,
    popupProps: _objectSpread({
      overlayInnerStyle: {
        width: "auto"
      }
    }, props.popupProps),
    panel: /* @__PURE__ */React__default["default"].createElement(timePicker_panel_TimePickerPanel["default"], {
      steps: steps,
      format: format,
      value: currentValue,
      isFooterDisplay: true,
      isShowPanel: isPanelShowed,
      disableTime: disableTime,
      onChange: setCurrentValue,
      hideDisabledTime: hideDisabledTime,
      handleConfirmClick: handleClickConfirm
    })
  }));
}, {
  TimeRangePicker: timePicker_TimeRangePicker["default"],
  TimePickerPanel: timePicker_panel_TimePickerPanel["default"]
});
TimePicker.displayName = "TimePicker";
TimePicker.defaultProps = timePicker_defaultProps.timePickerDefaultProps;

exports["default"] = TimePicker;
