// tdesign-react/lib/breadcrumb/Breadcrumb.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _objectWithoutProperties } from '../_chunks/dep-8368bb87.js';
import React from 'react';
import classNames from 'classnames';
import useConfig from '../hooks/useConfig.js';
import forwardRefWithStatics from '../_util/forwardRefWithStatics.js';
import BreadcrumbItem from './BreadcrumbItem.js';
import { BreadcrumbContext } from './BreadcrumbContext.js';
import { breadcrumbDefaultProps } from './defaultProps.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import 'hoist-non-react-statics';
import 'tdesign-icons-react';
import '../hooks/useGlobalIcon.js';
import '../_util/useCommonClassName.js';

var _excluded = ["children", "options", "separator", "maxItemWidth", "className"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Breadcrumb = forwardRefWithStatics(function (props, ref) {
  var children = props.children,
      options = props.options,
      separator = props.separator,
      maxItemWidth = props.maxItemWidth,
      className = props.className,
      restProps = _objectWithoutProperties(props, _excluded);

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var content = children;

  if (options && options.length) {
    content = options.map(function (option, index) {
      return /* @__PURE__ */React.createElement(BreadcrumbItem, {
        key: index,
        maxWidth: option.maxWidth,
        disabled: option.disabled,
        href: option.href,
        target: option.target,
        to: option.to,
        router: option.router,
        replace: option.replace,
        separator: separator,
        maxItemWidth: maxItemWidth
      }, option.content || option.children);
    });
  }

  return /* @__PURE__ */React.createElement(BreadcrumbContext.Provider, {
    value: {
      maxItemWidthInContext: maxItemWidth,
      separator: separator
    }
  }, /* @__PURE__ */React.createElement("div", _objectSpread({
    ref: ref,
    className: classNames("".concat(classPrefix, "-breadcrumb"), className)
  }, restProps), content));
}, {
  BreadcrumbItem: BreadcrumbItem
});
Breadcrumb.displayName = "Breadcrumb";
Breadcrumb.defaultProps = breadcrumbDefaultProps;

export { Breadcrumb as default };
