// tdesign-react/esm/select/base/Select.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import { _ as _objectWithoutProperties } from '../../_chunks/dep-8368bb87.js';
import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import React, { useRef, useState, useEffect, useMemo, useCallback } from 'react';
import classNames from 'classnames';
import { i as isFunction_1 } from '../../_chunks/dep-e430e2ee.js';
import { g as get_1 } from '../../_chunks/dep-ddb0f1c1.js';
import useControlled from '../../hooks/useControlled.js';
import { useLocaleReceiver } from '../../locale/LocalReceiver.js';
import useConfig from '../../hooks/useConfig.js';
import forwardRefWithStatics from '../../_util/forwardRefWithStatics.js';
import { getValueToOption, getSelectValueArr } from '../util/helper.js';
import noop from '../../_util/noop.js';
import FakeArrow from '../../common/FakeArrow.js';
import { Loading } from '../../loading/index.js';
import { SelectInput } from '../../select-input/index.js';
import Option from './Option.js';
import OptionGroup from './OptionGroup.js';
import PopupContent from './PopupContent.js';
import { Tag } from '../../tag/index.js';
import { selectDefaultProps } from '../defaultProps.js';
import '../../_chunks/dep-7909fefa.js';
import '../../_chunks/dep-1cce5a60.js';
import '../../_chunks/dep-4a64abb1.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-e54bb0dc.js';
import '../../_chunks/dep-b2626e28.js';
import '../../_chunks/dep-93a10d8c.js';
import '../../_chunks/dep-264b0e55.js';
import '../../_chunks/dep-199b1581.js';
import '../../_chunks/dep-e237f0f2.js';
import '../../_chunks/dep-4e2181de.js';
import '../../_chunks/dep-c197ce4f.js';
import '../../_chunks/dep-5a2778c1.js';
import '../../_chunks/dep-a2b97df8.js';
import '../../_chunks/dep-c8bfa11e.js';
import '../../_chunks/dep-f54ff1e5.js';
import '../../_chunks/dep-e1d4dd61.js';
import '../../_chunks/dep-236ef238.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../_chunks/dep-13a70a46.js';
import '../../_chunks/dep-dc1ce7ed.js';
import '../../_chunks/dep-7aeeb170.js';
import '../../_chunks/dep-2afd805c.js';
import '../../_chunks/dep-4ed2dce5.js';
import '../../_chunks/dep-c8db35c0.js';
import '../../_chunks/dep-96aef4a9.js';
import '../../_chunks/dep-557a1b0f.js';
import '../../_chunks/dep-f5693bf6.js';
import '../../_chunks/dep-fdd1d1a6.js';
import '../../_chunks/dep-7ff8b18e.js';
import '../../_chunks/dep-ebf3a272.js';
import '../../_chunks/dep-6a2482c6.js';
import '../../_chunks/dep-a980ac05.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import 'hoist-non-react-statics';
import '../../_chunks/dep-59b9253f.js';
import '../../_chunks/dep-d1b4ab24.js';
import '../../loading/Loading.js';
import '../../_util/dom.js';
import 'raf';
import '../../_chunks/dep-3c98169f.js';
import '../../_util/easing.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../loading/gradient.js';
import '../../_common/js/loading/circle-adapter.js';
import '../../_common/js/utils/set-style.js';
import '../../_common/js/utils/helper.js';
import '../../loading/defaultProps.js';
import '../../loading/plugin.js';
import '../../loading/style/index.js';
import '../../loading/type.js';
import '../../select-input/SelectInput.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../../_util/useAnimation.js';
import '../../popup/hooks/useTrigger.js';
import 'react-is';
import '../../popup/utils/ref.js';
import '../../_util/composeRefs.js';
import '../../popup/utils/transition.js';
import '../../_util/useMutationObserver.js';
import '../../_chunks/dep-236b0359.js';
import '../../_chunks/dep-6d5c0a90.js';
import '../../_util/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../popup/style/index.js';
import '../../popup/type.js';
import '../../select-input/useSingle.js';
import '../../_chunks/dep-c776db26.js';
import '../../_chunks/dep-afbd242f.js';
import '../../_chunks/dep-3576b6cd.js';
import '../../_chunks/dep-aadcae4d.js';
import '../../input/index.js';
import '../../input/Input.js';
import 'tdesign-icons-react';
import '../../hooks/useGlobalIcon.js';
import '../../input/InputGroup.js';
import '../../input/defaultProps.js';
import '../../input/style/index.js';
import '../../input/type.js';
import '../../select-input/useMultiple.js';
import '../../tag-input/index.js';
import '../../tag-input/TagInput.js';
import '../../_util/useDragSorter.js';
import '../../tag-input/useTagScroll.js';
import '../../tag-input/useTagList.js';
import 'lodash';
import '../../tag-input/useHover.js';
import '../../tag-input/defaultProps.js';
import '../../tag/Tag.js';
import '../../tag/CheckTag.js';
import '../../tag/defaultProps.js';
import '../../tag/style/index.js';
import '../../tag/type.js';
import '../../tag-input/style/index.js';
import '../../tag-input/type.js';
import '../../select-input/useOverlayInnerStyle.js';
import '../../select-input/defaultProps.js';
import '../../select-input/style/index.js';
import '../../select-input/type.js';
import '../../_chunks/dep-a5f8b6ec.js';
import '../../_util/useRipple.js';
import '../../_util/setStyle.js';

var _excluded = ["overlayClassName"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Select = forwardRefWithStatics(function (props, ref) {
  var _useLocaleReceiver = useLocaleReceiver("select"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var emptyText = t(local.loadingText);
  var readonly = props.readonly,
      bordered = props.bordered,
      borderless = props.borderless,
      autoWidth = props.autoWidth,
      creatable = props.creatable,
      filter = props.filter,
      _props$loadingText = props.loadingText,
      loadingText = _props$loadingText === void 0 ? emptyText : _props$loadingText,
      max = props.max,
      popupProps = props.popupProps,
      reserveKeyword = props.reserveKeyword,
      className = props.className,
      style = props.style,
      disabled = props.disabled,
      size = props.size,
      multiple = props.multiple,
      placeholder = props.placeholder,
      clearable = props.clearable,
      prefixIcon = props.prefixIcon,
      options = props.options,
      filterable = props.filterable,
      loading = props.loading,
      onFocus = props.onFocus,
      onBlur = props.onBlur,
      _props$onClear = props.onClear,
      onClear = _props$onClear === void 0 ? noop : _props$onClear,
      onCreate = props.onCreate,
      onRemove = props.onRemove,
      onSearch = props.onSearch,
      empty = props.empty,
      valueType = props.valueType,
      keys = props.keys,
      children = props.children,
      collapsedItems = props.collapsedItems,
      minCollapsedNum = props.minCollapsedNum,
      valueDisplay = props.valueDisplay,
      onEnter = props.onEnter,
      onVisibleChange = props.onVisibleChange,
      showArrow = props.showArrow,
      inputProps = props.inputProps,
      panelBottomContent = props.panelBottomContent,
      panelTopContent = props.panelTopContent,
      selectInputProps = props.selectInputProps,
      tagInputProps = props.tagInputProps,
      tagProps = props.tagProps;
  var selectPopupRef = useRef();

  var _useControlled = useControlled(props, "value", props.onChange),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      value = _useControlled2[0],
      onChange = _useControlled2[1];

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _ref = popupProps || {},
      overlayClassName = _ref.overlayClassName,
      restPopupProps = _objectWithoutProperties(_ref, _excluded);

  var name = "".concat(classPrefix, "-select");

  var _useControlled3 = useControlled(props, "popupVisible", props.onPopupVisibleChange),
      _useControlled4 = _slicedToArray(_useControlled3, 2),
      showPopup = _useControlled4[0],
      setShowPopup = _useControlled4[1];

  var _useControlled5 = useControlled(props, "inputValue", props.onInputChange),
      _useControlled6 = _slicedToArray(_useControlled5, 2),
      inputValue = _useControlled6[0],
      onInputChange = _useControlled6[1];

  var _useState = useState([]),
      _useState2 = _slicedToArray(_useState, 2),
      currentOptions = _useState2[0],
      setCurrentOptions = _useState2[1];

  var _useState3 = useState([]),
      _useState4 = _slicedToArray(_useState3, 2),
      tmpPropOptions = _useState4[0],
      setTmpPropOptions = _useState4[1];

  var _useState5 = useState({}),
      _useState6 = _slicedToArray(_useState5, 2),
      valueToOption = _useState6[0],
      setValueToOption = _useState6[1];

  var _useState7 = useState([]),
      _useState8 = _slicedToArray(_useState7, 2),
      selectedOptions = _useState8[0],
      setSelectedOptions = _useState8[1];

  useEffect(function () {
    if (keys) {
      var transformedOptions = options === null || options === void 0 ? void 0 : options.map(function (option) {
        return _objectSpread(_objectSpread({}, option), {}, {
          value: get_1(option, (keys === null || keys === void 0 ? void 0 : keys.value) || "value"),
          label: get_1(option, (keys === null || keys === void 0 ? void 0 : keys.label) || "label")
        });
      });
      setCurrentOptions(transformedOptions);
      setTmpPropOptions(transformedOptions);
    } else {
      setCurrentOptions(options);
      setTmpPropOptions(options);
    }

    setValueToOption(getValueToOption(children, options, keys) || {});
  }, [options, keys, children]);
  useEffect(function () {
    setSelectedOptions(function (oldSelectedOptions) {
      var valueKey = (keys === null || keys === void 0 ? void 0 : keys.value) || "value";
      var labelKey = (keys === null || keys === void 0 ? void 0 : keys.label) || "label";

      if (Array.isArray(value)) {
        return value.map(function (item) {
          if (valueType === "value") {
            var _ref2;

            return valueToOption[item] || oldSelectedOptions.find(function (option) {
              return get_1(option, valueKey) === item;
            }) || (_ref2 = {}, _defineProperty(_ref2, valueKey, item), _defineProperty(_ref2, labelKey, item), _ref2);
          }

          return item;
        }).filter(Boolean);
      }

      if (value !== void 0 && value !== null) {
        if (valueType === "value") {
          var _ref3;

          return [valueToOption[value] || oldSelectedOptions.find(function (option) {
            return get_1(option, valueKey) === value;
          }) || (_ref3 = {}, _defineProperty(_ref3, valueKey, value), _defineProperty(_ref3, labelKey, value), _ref3)].filter(Boolean);
        }

        return [value];
      }

      return [];
    });
  }, [value, keys, valueType, valueToOption]);
  var selectedLabel = useMemo(function () {
    if (multiple) {
      return selectedOptions.map(function (selectedOption) {
        return get_1(selectedOption || {}, (keys === null || keys === void 0 ? void 0 : keys.label) || "label") || "";
      });
    }

    return get_1(selectedOptions[0] || {}, (keys === null || keys === void 0 ? void 0 : keys.label) || "label") || void 0;
  }, [selectedOptions, keys, multiple]);

  var handleShowPopup = function handleShowPopup(visible, ctx) {
    if (disabled) return;
    setShowPopup(visible, ctx);
    onVisibleChange === null || onVisibleChange === void 0 ? void 0 : onVisibleChange(visible);
    visible && onInputChange("");
  };

  var onTagChange = function onTagChange(_currentTags, context) {
    var trigger = context.trigger,
        index = context.index,
        item = context.item,
        e = context.e;

    if (trigger === "backspace") {
      e.stopPropagation();
      var closest = -1;
      var len = index;

      while (len >= 0) {
        var _selectedOptions$len;

        if (!((_selectedOptions$len = selectedOptions[len]) !== null && _selectedOptions$len !== void 0 && _selectedOptions$len.disabled)) {
          closest = len;
          break;
        }

        len -= 1;
      }

      if (closest < 0) {
        return;
      }

      var values = getSelectValueArr(value, value[closest], true, valueType, keys);
      onChange(values, {
        e: e,
        trigger: trigger,
        selectedOptions: []
      });
      return;
    }

    if (trigger === "clear") {
      e.stopPropagation();
      onChange([], {
        e: e,
        trigger: trigger,
        selectedOptions: []
      });
      return;
    }

    if (trigger === "tag-remove") {
      e.stopPropagation();

      var _values = getSelectValueArr(value, value[index], true, valueType, keys);

      onChange(_values, {
        e: e,
        trigger: trigger,
        selectedOptions: []
      });

      if (isFunction_1(onRemove)) {
        onRemove({
          value: value[index],
          data: {
            label: item,
            value: value[index]
          },
          e: e
        });
      }
    }
  };

  var handleChange = function handleChange(value2, context) {
    if (multiple) {
      !reserveKeyword && onInputChange("", {
        trigger: "clear"
      });
    }

    if (creatable && isFunction_1(onCreate)) {
      if (options.filter(function (option) {
        return option.value === value2;
      }).length === 0) {
        onCreate(value2);
      }
    }

    onChange === null || onChange === void 0 ? void 0 : onChange(value2, _objectSpread(_objectSpread({}, context), {}, {
      selectedOptions: []
    }));
  };

  var handleFilter = function handleFilter(value2) {
    var filteredOptions = [];

    if (!value2) {
      setCurrentOptions(tmpPropOptions);
      return;
    }

    if (filter && isFunction_1(filter)) {
      if (Array.isArray(tmpPropOptions)) {
        filteredOptions = tmpPropOptions.filter(function (option) {
          return filter(value2, option);
        });
      }
    } else if (Array.isArray(tmpPropOptions)) {
      var upperValue = value2.toUpperCase();
      filteredOptions = tmpPropOptions.filter(function (option) {
        return ((option === null || option === void 0 ? void 0 : option.label) || "").toUpperCase().includes(upperValue);
      });
    }

    if (creatable) {
      filteredOptions = filteredOptions.concat([{
        label: value2,
        value: value2
      }]);
    }

    setCurrentOptions(filteredOptions);
  };

  var handleInputChange = function handleInputChange(value2) {
    onInputChange(value2);
    if (value2 === void 0) return;

    if (isFunction_1(onSearch)) {
      onSearch(value2);
      return;
    }
  };

  var onClearValue = function onClearValue(context) {
    context.e.stopPropagation();

    if (Array.isArray(value)) {
      onChange([], context);
    } else {
      onChange(null, context);
    }

    onInputChange(void 0);
    onClear(context);
  };

  useEffect(function () {
    if (typeof inputValue !== "undefined") {
      handleFilter(String(inputValue));
    }
  }, [inputValue]);

  var renderSuffixIcon = function renderSuffixIcon() {
    if (loading) {
      return /* @__PURE__ */React.createElement(Loading, {
        className: classNames("".concat(name, "__right-icon"), "".concat(name, "__active-icon")),
        loading: true,
        size: "small"
      });
    }

    return showArrow && /* @__PURE__ */React.createElement(FakeArrow, {
      overlayClassName: "".concat(name, "__right-icon"),
      isActive: showPopup,
      disabled: disabled
    });
  };

  var renderContent = function renderContent() {
    var popupContentProps = {
      onChange: handleChange,
      value: value,
      className: className,
      size: size,
      multiple: multiple,
      showPopup: showPopup,
      setShowPopup: function setShowPopup(show) {
        return handleShowPopup(show, {});
      },
      options: currentOptions,
      empty: empty,
      max: max,
      loadingText: loadingText,
      loading: loading,
      valueType: valueType,
      keys: keys,
      panelBottomContent: panelBottomContent,
      panelTopContent: panelTopContent
    };
    return /* @__PURE__ */React.createElement(PopupContent, _objectSpread(_objectSpread({}, popupContentProps), {}, {
      ref: selectPopupRef
    }), children);
  };

  var renderValueDisplay = function renderValueDisplay() {
    if (!valueDisplay) {
      if (!multiple) {
        if (typeof selectedLabel !== "string") {
          return selectedLabel;
        }

        return "";
      }

      return function (_ref4) {
        var val = _ref4.value;
        return val.slice(0, minCollapsedNum ? minCollapsedNum : val.length).map(function (v, key) {
          var filterOption = options === null || options === void 0 ? void 0 : options.find(function (option) {
            return option.label === v;
          });
          return /* @__PURE__ */React.createElement(Tag, {
            key: key,
            onClose: function onClose(_ref5) {
              var e = _ref5.e;
              e.stopPropagation();
              var values = getSelectValueArr(value, value[key], true, valueType, keys);
              onChange(values, null);
              return;
            },
            closable: !(filterOption !== null && filterOption !== void 0 && filterOption.disabled)
          }, v);
        });
      };
    }

    if (typeof valueDisplay === "string") {
      return valueDisplay;
    }

    if (multiple) {
      return function (_ref6) {
        var onClose = _ref6.onClose;
        return valueDisplay({
          value: selectedLabel,
          onClose: onClose
        });
      };
    }

    return selectedLabel.length ? valueDisplay({
      value: selectedLabel[0],
      onClose: noop
    }) : "";
  };

  var renderCollapsedItems = useMemo(function () {
    return collapsedItems ? function () {
      return collapsedItems({
        value: selectedLabel,
        collapsedSelectedItems: selectedLabel.slice(minCollapsedNum, selectedLabel.length),
        count: selectedLabel.length - minCollapsedNum
      });
    } : null;
  }, [selectedLabel, collapsedItems, minCollapsedNum]);
  var updateScrollTop = useCallback(function (content) {
    if (!(selectPopupRef !== null && selectPopupRef !== void 0 && selectPopupRef.current)) {
      return;
    }

    var firstSelectedNode = (selectPopupRef === null || selectPopupRef === void 0 ? void 0 : selectPopupRef.current).querySelector(".".concat(classPrefix, "-is-selected"));

    if (firstSelectedNode && content) {
      var _getComputedStyle = getComputedStyle(firstSelectedNode),
          paddingBottom = _getComputedStyle.paddingBottom;

      var _getComputedStyle2 = getComputedStyle(content),
          marginBottom = _getComputedStyle2.marginBottom;

      var elementBottomHeight = parseInt(paddingBottom, 10) + parseInt(marginBottom, 10);
      var updateValue = firstSelectedNode.offsetTop - content.offsetTop - (content.clientHeight - firstSelectedNode.clientHeight) + elementBottomHeight;
      content.scrollTop = updateValue;
    }
  }, [classPrefix]);
  var onMouseEnter = props.onMouseEnter,
      onMouseLeave = props.onMouseLeave;
  return /* @__PURE__ */React.createElement("div", {
    className: classNames("".concat(name, "__wrap"), className),
    style: style,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave
  }, /* @__PURE__ */React.createElement(SelectInput, _objectSpread({
    autoWidth: !(style !== null && style !== void 0 && style.width) && autoWidth,
    ref: ref,
    className: name,
    readonly: readonly,
    allowInput: (filterable !== null && filterable !== void 0 ? filterable : local.filterable) || isFunction_1(filter),
    multiple: multiple,
    value: selectedLabel,
    valueDisplay: renderValueDisplay(),
    clearable: clearable,
    disabled: disabled,
    status: props.status,
    tips: props.tips,
    borderless: borderless || !bordered,
    label: prefixIcon,
    suffixIcon: renderSuffixIcon(),
    panel: renderContent(),
    placeholder: !multiple && showPopup && selectedLabel ? selectedLabel : placeholder || t(local.placeholder),
    inputValue: inputValue,
    tagInputProps: _objectSpread({}, tagInputProps),
    tagProps: tagProps,
    inputProps: _objectSpread({
      size: size
    }, inputProps),
    minCollapsedNum: minCollapsedNum,
    collapsedItems: renderCollapsedItems,
    popupProps: _objectSpread({
      overlayClassName: ["".concat(name, "__dropdown"), overlayClassName]
    }, restPopupProps),
    popupVisible: showPopup,
    onPopupVisibleChange: handleShowPopup,
    onTagChange: onTagChange,
    onInputChange: handleInputChange,
    onFocus: onFocus,
    onEnter: onEnter,
    onBlur: onBlur,
    onClear: function onClear(context) {
      onClearValue(context);
    },
    updateScrollTop: updateScrollTop
  }, selectInputProps)));
}, {
  Option: Option,
  OptionGroup: OptionGroup
});
Select.displayName = "Select";
Select.defaultProps = selectDefaultProps;

export { Select as default };
