// tdesign-react/es/tag/Tag.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _objectWithoutProperties } from '../_chunks/dep-8368bb87.js';
import React from 'react';
import classNames from 'classnames';
import { CloseIcon } from 'tdesign-icons-react';
import noop from '../_util/noop.js';
import forwardRefWithStatics from '../_util/forwardRefWithStatics.js';
import useConfig from '../hooks/useConfig.js';
import useGlobalIcon from '../hooks/useGlobalIcon.js';
import CheckTag from './CheckTag.js';
import { tagDefaultProps } from './defaultProps.js';
import 'hoist-non-react-statics';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-e54bb0dc.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-e237f0f2.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-ebf3a272.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../_chunks/dep-9615f03f.js';
import '../_chunks/dep-7909fefa.js';
import '../hooks/useControlled.js';
import '../_chunks/dep-f54ff1e5.js';
import '../_chunks/dep-e1d4dd61.js';
import '../_chunks/dep-236ef238.js';
import '../_chunks/dep-c8bfa11e.js';
import '../_chunks/dep-199b1581.js';

var _excluded = ["theme", "size", "shape", "variant", "closable", "maxWidth", "icon", "content", "onClick", "onClose", "className", "style", "disabled", "children"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Tag = forwardRefWithStatics(function (props, ref) {
  var _classNames;

  var theme = props.theme,
      size = props.size,
      shape = props.shape,
      variant = props.variant,
      closable = props.closable,
      maxWidth = props.maxWidth,
      icon = props.icon,
      content = props.content,
      _props$onClick = props.onClick,
      _onClick = _props$onClick === void 0 ? noop : _props$onClick,
      _props$onClose = props.onClose,
      onClose = _props$onClose === void 0 ? noop : _props$onClose,
      className = props.className,
      style = props.style,
      disabled = props.disabled,
      children = props.children,
      otherTagProps = _objectWithoutProperties(props, _excluded);

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useGlobalIcon = useGlobalIcon({
    CloseIcon: CloseIcon
  }),
      CloseIcon$1 = _useGlobalIcon.CloseIcon;

  var tagClassPrefix = "".concat(classPrefix, "-tag");
  var sizeMap = {
    large: "".concat(classPrefix, "-size-l"),
    small: "".concat(classPrefix, "-size-s")
  };
  var tagClassNames = classNames(tagClassPrefix, "".concat(tagClassPrefix, "--").concat(theme), "".concat(tagClassPrefix, "--").concat(variant), "".concat(tagClassPrefix, "--").concat(size), "".concat(tagClassPrefix, "--").concat(shape), (_classNames = {}, _defineProperty(_classNames, "".concat(tagClassPrefix, "--ellipsis"), !!maxWidth), _defineProperty(_classNames, "".concat(tagClassPrefix, "--disabled"), disabled), _classNames), sizeMap[size], className);
  var deleteIcon = /* @__PURE__ */React.createElement(CloseIcon$1, {
    onClick: function onClick(e) {
      if (disabled) return;
      onClose({
        e: e
      });
    },
    className: "".concat(tagClassPrefix, "__icon-close")
  });
  var tag = /* @__PURE__ */React.createElement("span", _objectSpread({
    ref: ref,
    className: tagClassNames,
    onClick: function onClick(e) {
      if (disabled) return;

      _onClick({
        e: e
      });
    },
    style: maxWidth ? _objectSpread({
      maxWidth: typeof maxWidth === "number" ? "".concat(maxWidth, "px") : maxWidth
    }, style) : style
  }, otherTagProps), icon, maxWidth ? /* @__PURE__ */React.createElement("span", {
    className: "".concat(tagClassPrefix, "--text")
  }, children || content) : children || content, closable && deleteIcon);
  return tag;
}, {
  CheckTag: CheckTag
});
Tag.displayName = "Tag";
Tag.defaultProps = tagDefaultProps;

export { Tag as default };
