// tdesign-react/cjs/input/Input.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _typeof = require('../_chunks/dep-2a90f794.js');
var defineProperty = require('../_chunks/dep-4eb93341.js');
var objectWithoutProperties = require('../_chunks/dep-1a0462cc.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var isFunction = require('../_chunks/dep-53ba6729.js');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var _common_js_utils_helper = require('../_common/js/utils/helper.js');
var input_InputGroup = require('./InputGroup.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var input_defaultProps = require('./defaultProps.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-aab723b3.js');
require('hoist-non-react-statics');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../_chunks/dep-8052f095.js');
require('../_chunks/dep-606cfe2a.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-afce8659.js');
require('../_chunks/dep-09805e35.js');
require('../_util/noop.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["type", "autoWidth", "placeholder", "disabled", "status", "size", "className", "inputClass", "style", "prefixIcon", "suffixIcon", "clearable", "tips", "align", "maxlength", "maxcharacter", "showClearIconOnEmpty", "autofocus", "autocomplete", "readonly", "label", "suffix", "showInput", "keepWrapperWidth", "format", "onClick", "onClear", "onEnter", "onKeydown", "onKeyup", "onKeypress", "onFocus", "onBlur", "onPaste", "onMouseenter", "onMouseleave", "onWheel", "onCompositionstart", "onCompositionend", "onChange"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var renderIcon = function renderIcon(classPrefix, type, icon) {
  var result = null;
  if (icon) result = icon;
  if (typeof icon === "function") result = icon();
  var iconClassName = icon ? "".concat(classPrefix, "-input__").concat(type, "-icon") : "";

  if (result) {
    result = /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-input__").concat(type, " ").concat(iconClassName)
    }, result);
  }

  return result;
};

var Input = _util_forwardRefWithStatics["default"](function (props, ref) {
  var _classNames;

  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("input"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
    BrowseIcon: tdesignIconsReact.BrowseIcon,
    BrowseOffIcon: tdesignIconsReact.BrowseOffIcon,
    CloseCircleFilledIcon: tdesignIconsReact.CloseCircleFilledIcon
  }),
      BrowseIcon = _useGlobalIcon.BrowseIcon,
      BrowseOffIcon = _useGlobalIcon.BrowseOffIcon,
      CloseCircleFilledIcon = _useGlobalIcon.CloseCircleFilledIcon;

  var type = props.type,
      autoWidth = props.autoWidth,
      _props$placeholder = props.placeholder,
      placeholder = _props$placeholder === void 0 ? t(local.placeholder) : _props$placeholder,
      disabled = props.disabled,
      status = props.status,
      size = props.size,
      className = props.className,
      inputClass = props.inputClass,
      style = props.style,
      prefixIcon = props.prefixIcon,
      suffixIcon = props.suffixIcon,
      clearable = props.clearable,
      tips = props.tips,
      align = props.align,
      maxlength = props.maxlength,
      maxcharacter = props.maxcharacter,
      showClearIconOnEmpty = props.showClearIconOnEmpty,
      autofocus = props.autofocus,
      autocomplete = props.autocomplete,
      readonly = props.readonly,
      label = props.label,
      suffix = props.suffix,
      _props$showInput = props.showInput,
      showInput = _props$showInput === void 0 ? true : _props$showInput,
      keepWrapperWidth = props.keepWrapperWidth,
      format = props.format,
      _onClick = props.onClick,
      onClear = props.onClear,
      onEnter = props.onEnter,
      onKeydown = props.onKeydown,
      onKeyup = props.onKeyup,
      onKeypress = props.onKeypress,
      onFocus = props.onFocus,
      onBlur = props.onBlur,
      onPaste = props.onPaste,
      onMouseenter = props.onMouseenter,
      onMouseleave = props.onMouseleave,
      onWheel = props.onWheel,
      onCompositionstart = props.onCompositionstart,
      onCompositionend = props.onCompositionend,
      onChangeFromProps = props.onChange,
      restProps = objectWithoutProperties._objectWithoutProperties(props, _excluded);

  var _useControlled = hooks_useControlled["default"](props, "value", onChangeFromProps),
      _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
      value = _useControlled2[0],
      onChange = _useControlled2[1];

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var composingRef = React.useRef(false);
  var inputRef = React.useRef();
  var inputPreRef = React.useRef();
  var wrapperRef = React.useRef();

  var _useState = React.useState(false),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      isHover = _useState2[0],
      toggleIsHover = _useState2[1];

  var _useState3 = React.useState(false),
      _useState4 = slicedToArray._slicedToArray(_useState3, 2),
      isFocused = _useState4[0],
      toggleIsFocused = _useState4[1];

  var _useState5 = React.useState(type),
      _useState6 = slicedToArray._slicedToArray(_useState5, 2),
      renderType = _useState6[0],
      setRenderType = _useState6[1];

  var _useState7 = React.useState(""),
      _useState8 = slicedToArray._slicedToArray(_useState7, 2),
      composingValue = _useState8[0],
      setComposingValue = _useState8[1];

  var isShowClearIcon = (clearable && value && !disabled || showClearIconOnEmpty) && isHover;
  var prefixIconContent = renderIcon(classPrefix, "prefix", prefixIcon);
  var suffixIconNew = suffixIcon;
  if (isShowClearIcon) suffixIconNew = /* @__PURE__ */React__default["default"].createElement(CloseCircleFilledIcon, {
    className: "".concat(classPrefix, "-input__suffix-clear"),
    onClick: handleClear
  });

  if (type === "password" && typeof suffixIcon === "undefined") {
    if (renderType === "password") {
      suffixIconNew = /* @__PURE__ */React__default["default"].createElement(BrowseOffIcon, {
        className: "".concat(classPrefix, "-input__suffix-clear"),
        onClick: togglePasswordVisible
      });
    } else if (renderType === "text") {
      suffixIconNew = /* @__PURE__ */React__default["default"].createElement(BrowseIcon, {
        className: "".concat(classPrefix, "-input__suffix-clear"),
        onClick: togglePasswordVisible
      });
    }
  }

  var suffixIconContent = renderIcon(classPrefix, "suffix", suffixIconNew);
  var labelContent = isFunction.isFunction_1(label) ? label() : label;
  var suffixContent = isFunction.isFunction_1(suffix) ? suffix() : suffix;
  React.useEffect(function () {
    var _inputPreRef$current, _inputPreRef$current2;

    if (!autoWidth) return;
    if (((_inputPreRef$current = inputPreRef.current) === null || _inputPreRef$current === void 0 ? void 0 : _inputPreRef$current.offsetWidth) === 0) return;
    if (inputRef.current) inputRef.current.style.width = "".concat((_inputPreRef$current2 = inputPreRef.current) === null || _inputPreRef$current2 === void 0 ? void 0 : _inputPreRef$current2.offsetWidth, "px");
  }, [autoWidth, value, placeholder, inputRef]);
  React.useEffect(function () {
    setRenderType(type);
  }, [type]);
  var renderInput = /* @__PURE__ */React__default["default"].createElement("input", {
    ref: inputRef,
    placeholder: placeholder,
    type: renderType,
    className: "".concat(classPrefix, "-input__inner"),
    value: composingRef.current ? composingValue : value !== null && value !== void 0 ? value : "",
    readOnly: readonly,
    disabled: disabled,
    autoComplete: autocomplete !== null && autocomplete !== void 0 ? autocomplete : local.autocomplete || void 0,
    autoFocus: autofocus,
    onChange: handleChange,
    onKeyDown: handleKeyDown,
    onKeyUp: handleKeyUp,
    onKeyPress: handleKeyPress,
    onCompositionStart: handleCompositionStart,
    onCompositionEnd: handleCompositionEnd,
    onFocus: handleFocus,
    onBlur: handleBlur,
    onPaste: handlePaste
  });
  var renderInputNode = /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](inputClass, "".concat(classPrefix, "-input"), (_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-is-readonly"), readonly), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-is-disabled"), disabled), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-is-focused"), isFocused), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-size-s"), size === "small"), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-size-l"), size === "large"), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-size-m"), size === "medium"), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-align-").concat(align), align), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-is-").concat(status), status), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-input--prefix"), prefixIcon || labelContent), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-input--suffix"), suffixIconContent || suffixContent), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-input--focused"), isFocused), _classNames)),
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    onWheel: handleWheel,
    onClick: function onClick(e) {
      return _onClick === null || _onClick === void 0 ? void 0 : _onClick({
        e: e
      });
    }
  }, prefixIconContent, labelContent ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-input__prefix")
  }, labelContent) : null, showInput && renderInput, autoWidth && /* @__PURE__ */React__default["default"].createElement("span", {
    ref: inputPreRef,
    className: "".concat(classPrefix, "-input__input-pre")
  }, value || props.placeholder), suffixContent ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-input__suffix")
  }, suffixContent) : null, suffixIconContent);

  function togglePasswordVisible() {
    var toggleType = renderType === "password" ? "text" : "password";
    setRenderType(toggleType);
  }

  function handleChange(e) {
    var value2 = e.currentTarget.value;

    if (composingRef.current) {
      setComposingValue(value2);
    } else {
      value2 = _common_js_utils_helper.limitUnicodeMaxLength(value2, maxlength);

      if (typeof maxcharacter === "number" && maxcharacter >= 0) {
        var stringInfo = _common_js_utils_helper.getCharacterLength(value2, maxcharacter);
        value2 = _typeof._typeof(stringInfo) === "object" && stringInfo.characters;
      }

      setComposingValue(value2);
      onChange(value2, {
        e: e
      });
    }
  }

  function handleClear(e) {
    onChange === null || onChange === void 0 ? void 0 : onChange("", {
      e: e
    });
    onClear === null || onClear === void 0 ? void 0 : onClear({
      e: e
    });
  }

  function handleKeyDown(e) {
    var key = e.key,
        value2 = e.currentTarget.value;
    key === "Enter" && (onEnter === null || onEnter === void 0 ? void 0 : onEnter(value2, {
      e: e
    }));
    onKeydown === null || onKeydown === void 0 ? void 0 : onKeydown(value2, {
      e: e
    });
  }

  function handleKeyUp(e) {
    var value2 = e.currentTarget.value;
    onKeyup === null || onKeyup === void 0 ? void 0 : onKeyup(value2, {
      e: e
    });
  }

  function handleKeyPress(e) {
    var value2 = e.currentTarget.value;
    onKeypress === null || onKeypress === void 0 ? void 0 : onKeypress(value2, {
      e: e
    });
  }

  function handleCompositionStart(e) {
    composingRef.current = true;
    var value2 = e.currentTarget.value;
    onCompositionstart === null || onCompositionstart === void 0 ? void 0 : onCompositionstart(value2, {
      e: e
    });
  }

  function handleCompositionEnd(e) {
    var value2 = e.currentTarget.value;

    if (composingRef.current) {
      composingRef.current = false;
      handleChange(e);
    }

    onCompositionend === null || onCompositionend === void 0 ? void 0 : onCompositionend(value2, {
      e: e
    });
  }

  function handleFocus(e) {
    if (readonly) return;
    var value2 = e.currentTarget.value;
    onFocus === null || onFocus === void 0 ? void 0 : onFocus(value2, {
      e: e
    });
    toggleIsFocused(true);
  }

  function handleBlur(e) {
    if (readonly) return;
    var value2 = e.currentTarget.value;
    format && onChange(format(value2), {
      e: e
    });
    onBlur === null || onBlur === void 0 ? void 0 : onBlur(value2, {
      e: e
    });
    toggleIsFocused(false);
  }

  function handlePaste(e) {
    var clipData = e.clipboardData;
    var pasteValue = clipData === null || clipData === void 0 ? void 0 : clipData.getData("text/plain");
    onPaste === null || onPaste === void 0 ? void 0 : onPaste({
      e: e,
      pasteValue: pasteValue
    });
  }

  function handleMouseEnter(e) {
    toggleIsHover(true);
    onMouseenter === null || onMouseenter === void 0 ? void 0 : onMouseenter({
      e: e
    });
  }

  function handleMouseLeave(e) {
    toggleIsHover(false);
    onMouseleave === null || onMouseleave === void 0 ? void 0 : onMouseleave({
      e: e
    });
  }

  function handleWheel(e) {
    onWheel === null || onWheel === void 0 ? void 0 : onWheel({
      e: e
    });
  }

  React.useImperativeHandle(ref, function () {
    return {
      currentElement: wrapperRef.current,
      inputElement: inputRef.current,
      focus: function focus() {
        var _inputRef$current;

        return (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
      },
      blur: function blur() {
        var _inputRef$current2;

        return (_inputRef$current2 = inputRef.current) === null || _inputRef$current2 === void 0 ? void 0 : _inputRef$current2.blur();
      },
      select: function select() {
        var _inputRef$current3;

        return (_inputRef$current3 = inputRef.current) === null || _inputRef$current3 === void 0 ? void 0 : _inputRef$current3.select();
      }
    };
  });
  return /* @__PURE__ */React__default["default"].createElement("div", _objectSpread({
    ref: wrapperRef,
    style: style,
    className: classNames__default["default"]("".concat(classPrefix, "-input__wrap"), className, defineProperty._defineProperty({}, "".concat(classPrefix, "-input--auto-width"), autoWidth && !keepWrapperWidth))
  }, restProps), renderInputNode, tips && /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-input__tips"), "".concat(classPrefix, "-input__tips--").concat(status || "normal"))
  }, tips));
}, {
  Group: input_InputGroup["default"]
});
Input.displayName = "Input";
Input.defaultProps = input_defaultProps.inputDefaultProps;

exports["default"] = Input;
