// tdesign-react/cjs/_common/js/tree/tree-node-model.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../_chunks/dep-4eb93341.js');
var pick = require('../../../_chunks/dep-b5af0182.js');
var omit = require('../../../_chunks/dep-70fe2bd0.js');
var _common_js_log_log = require('../log/log.js');
require('../../../_chunks/dep-8329ea23.js');
require('../../../_chunks/dep-8f18a7c4.js');
require('../../../_chunks/dep-2a90f794.js');
require('../../../_chunks/dep-09805e35.js');
require('../../../_chunks/dep-61a8a7b0.js');
require('../../../_chunks/dep-1fff9729.js');
require('../../../_chunks/dep-46cf36fd.js');
require('../../../_chunks/dep-f5e18a2f.js');
require('../../../_chunks/dep-53ba6729.js');
require('../../../_chunks/dep-aab723b3.js');
require('../../../_chunks/dep-384b291c.js');
require('../../../_chunks/dep-afce8659.js');
require('../../../_chunks/dep-55c5ba71.js');
require('../../../_chunks/dep-72020528.js');
require('../../../_chunks/dep-ad854ba5.js');
require('../../../_chunks/dep-9cd0fde8.js');
require('../../../_chunks/dep-eebdbd74.js');
require('../../../_chunks/dep-12656997.js');
require('../../../_chunks/dep-1e8de681.js');
require('../../../_chunks/dep-07e5d807.js');
require('../../../_chunks/dep-6fa7a9e9.js');
require('../../../_chunks/dep-4578b559.js');
require('../../../_chunks/dep-fe984d70.js');
require('../../../_chunks/dep-a0b5d8f6.js');
require('../../../_chunks/dep-bb60493d.js');
require('../../../_chunks/dep-d3ad6e52.js');
require('../../../_chunks/dep-4fb39763.js');
require('../../../_chunks/dep-76deafcc.js');
require('../../../_chunks/dep-994ec160.js');

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function getExposedProps(node) {
  var props = pick.pick_1(node, ["value", "label", "data", "actived", "expanded", "checked", "indeterminate", "loading"]);
  return props;
}

function createNodeModel(node) {
  var props = getExposedProps(node);

  var model = _objectSpread(_objectSpread({}, props), {}, {
    getLevel: function getLevel() {
      return node.getLevel();
    },
    getIndex: function getIndex() {
      return node.getIndex();
    },
    isFirst: function isFirst() {
      return node.isFirst();
    },
    isLast: function isLast() {
      return node.isLast();
    },
    isLeaf: function isLeaf() {
      return node.isLeaf();
    },
    insertBefore: function insertBefore(newData) {
      return node.insertBefore(newData);
    },
    insertAfter: function insertAfter(newData) {
      return node.insertAfter(newData);
    },
    appendData: function appendData(data) {
      return node.append(data);
    },
    getPath: function getPath() {
      var nodes = node.getPath();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getParent: function getParent() {
      var _node$parent;

      return (_node$parent = node.parent) === null || _node$parent === void 0 ? void 0 : _node$parent.getModel();
    },
    getParents: function getParents() {
      var nodes = node.getParents();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getRoot: function getRoot() {
      var root = node.getRoot();
      return root === null || root === void 0 ? void 0 : root.getModel();
    },
    getSiblings: function getSiblings() {
      var nodes = node.getSiblings();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getChildren: function getChildren(deep) {
      var childrenModel = false;
      var children = node.children;

      if (Array.isArray(children)) {
        if (children.length > 0) {
          if (deep) {
            var nodes = node.walk();
            nodes.shift();
            childrenModel = nodes.map(function (item) {
              return item.getModel();
            });
          } else {
            childrenModel = children.map(function (item) {
              return item.getModel();
            });
          }
        } else {
          childrenModel = false;
        }
      } else if (typeof children === "boolean") {
        childrenModel = children;
      }

      return childrenModel;
    },
    remove: function remove(value) {
      if (!value) {
        node.remove();
        return;
      }

      var tree = node.tree;
      var targetNode = tree.getNode(value);

      if (!targetNode) {
        _common_js_log_log["default"].warnOnce("Tree", "`".concat(value, "` is not exist"));
        return;
      }

      var parents = targetNode.getParents();
      var parentValues = parents.map(function (pnode) {
        return pnode.value;
      });

      if (parentValues.indexOf(node.value) < 0) {
        _common_js_log_log["default"].warnOnce("Tree", "`".concat(value, "` is not a childNode of current node"));
        return;
      }

      targetNode.remove();
    },
    setData: function setData(data) {
      var _data = omit.omit_1(data, ["children", "value", "label"]);

      var keys = node.tree.config.keys;
      var dataValue = data[(keys === null || keys === void 0 ? void 0 : keys.value) || "value"];
      var dataLabel = data[(keys === null || keys === void 0 ? void 0 : keys.label) || "label"];
      if (dataValue !== void 0) _data.value = dataValue;
      if (dataLabel !== void 0) _data.label = dataLabel;
      Object.assign(node.data, _data);
      Object.assign(node, _data);
    }
  });

  return model;
}
function updateNodeModel(model, node) {
  var props = getExposedProps(node);
  Object.assign(model, props);
}

exports.createNodeModel = createNodeModel;
exports.updateNodeModel = updateNodeModel;
