// tdesign-react/cjs/_chunks/dep-4ab4e7b4.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

var defineProperty = require('./dep-4eb93341.js');
var React = require('react');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var common_Check = require('../common/Check.js');
var slicedToArray = require('./dep-9c5d4f93.js');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var _util_useCommonClassName = require('../_util/useCommonClassName.js');
var radio_defaultProps = require('../radio/defaultProps.js');
var _util_useMutationObserver = require('../_util/useMutationObserver.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys$1(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread$1(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys$1(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys$1(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var RadioGroup = function RadioGroup(props) {
  var _classNames;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var disabled = props.disabled,
      children = props.children,
      onChange = props.onChange,
      size = props.size,
      variant = props.variant,
      _props$options = props.options,
      options = _props$options === void 0 ? [] : _props$options,
      className = props.className,
      style = props.style;

  var _useControlled = hooks_useControlled["default"](props, "value", onChange),
      _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
      internalValue = _useControlled2[0],
      setInternalValue = _useControlled2[1];

  var _useState = React.useState({}),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      barStyle = _useState2[0],
      setBarStyle = _useState2[1];

  var groupRef = React.useRef(null);
  var checkedRadioCls = ".".concat(classPrefix, "-radio-button.").concat(classPrefix, "-is-checked");

  var _useCommonClassName = _util_useCommonClassName["default"](),
      sizeMap = _useCommonClassName.SIZE;

  var context = {
    inject: function inject(checkProps) {
      if (typeof checkProps.checked !== "undefined") {
        return checkProps;
      }

      var checkValue = checkProps.value;
      return _objectSpread$1(_objectSpread$1({}, checkProps), {}, {
        checked: internalValue === checkProps.value,
        disabled: checkProps.disabled || disabled,
        onChange: function onChange(checked, _ref) {
          var e = _ref.e;

          if (typeof checkProps.onChange === "function") {
            checkProps.onChange(checked, {
              e: e
            });
          }

          setInternalValue(checked ? checkValue : void 0, {
            e: e
          });
        }
      });
    }
  };

  var calcBarStyle = function calcBarStyle() {
    var _groupRef$current$que, _groupRef$current;

    if (!variant.includes("filled")) return;
    var checkedRadio = (_groupRef$current$que = (_groupRef$current = groupRef.current).querySelector) === null || _groupRef$current$que === void 0 ? void 0 : _groupRef$current$que.call(_groupRef$current, checkedRadioCls);
    if (!checkedRadio) return setBarStyle({
      width: 0
    });
    var offsetWidth = checkedRadio.offsetWidth,
        offsetLeft = checkedRadio.offsetLeft;
    setBarStyle({
      width: "".concat(offsetWidth, "px"),
      left: "".concat(offsetLeft, "px")
    });
  };

  React.useEffect(function () {
    calcBarStyle();
  }, [groupRef.current, internalValue]);
  _util_useMutationObserver["default"](groupRef.current, calcBarStyle);

  var renderBlock = function renderBlock() {
    if (!variant.includes("filled")) return null;
    return /* @__PURE__ */React__default["default"].createElement("div", {
      style: barStyle,
      className: "".concat(classPrefix, "-radio-group__bg-block")
    });
  };

  var renderOptions = function renderOptions() {
    return options.map(function (item) {
      if (typeof item === "string" || typeof item === "number") {
        return /* @__PURE__ */React__default["default"].createElement(Radio, {
          value: item,
          key: item
        }, item);
      }

      return /* @__PURE__ */React__default["default"].createElement(Radio, {
        value: item.value,
        key: item.value,
        disabled: item.disabled
      }, item.label);
    });
  };

  return /* @__PURE__ */React__default["default"].createElement(common_Check.CheckContext.Provider, {
    value: context
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    ref: groupRef,
    style: style,
    className: classNames__default["default"]("".concat(classPrefix, "-radio-group"), sizeMap[size], className, (_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-radio-group__outline"), variant === "outline"), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-radio-group--filled"), variant.includes("filled")), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-radio-group--primary-filled"), variant === "primary-filled"), _classNames))
  }, children || renderOptions(), renderBlock()));
};

RadioGroup.displayName = "RadioGroup";
RadioGroup.defaultProps = radio_defaultProps.radioGroupDefaultProps;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Radio = _util_forwardRefWithStatics["default"](function (props, ref) {
  return /* @__PURE__ */React__default["default"].createElement(common_Check["default"], _objectSpread({
    ref: ref,
    type: "radio"
  }, props));
}, {
  Group: RadioGroup,
  Button: /*#__PURE__*/React.forwardRef(function (props, ref) {
    return /* @__PURE__ */React__default["default"].createElement(common_Check["default"], _objectSpread({
      ref: ref,
      type: "radio-button"
    }, props));
  })
});
Radio.displayName = "Radio";
Radio.defaultProps = radio_defaultProps.radioDefaultProps;

exports.Radio = Radio;
exports.RadioGroup = RadioGroup;
