// tdesign-react/cjs/_chunks/dep-54aceb71.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

var defineProperty = require('./dep-4eb93341.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var toConsumableArray = require('./dep-8052f095.js');
var slicedToArray = require('./dep-9c5d4f93.js');
var ReactDOM = require('react-dom');
var hooks_useConfig = require('../hooks/useConfig.js');
var _util_noop = require('../_util/noop.js');
var _util_renderTNode = require('../_util/renderTNode.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var notification_defaultProps = require('../notification/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var ReactDOM__default = /*#__PURE__*/_interopDefaultLegacy(ReactDOM);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var seed = 0;
var listMap = /* @__PURE__ */new Map();
var NotificationRemoveContext = /*#__PURE__*/React__default["default"].createContext(_util_noop["default"]);
var NotificationList = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var placement = props.placement,
      zIndex = props.zIndex;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useState = React.useState([]),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      list = _useState2[0],
      setList = _useState2[1];

  var remove = function remove(key) {
    setList(function (oldList) {
      var index = oldList.findIndex(function (item) {
        return item.key === key;
      });

      if (index !== -1) {
        var tempList = toConsumableArray._toConsumableArray(oldList);

        tempList.splice(index, 1);
        return toConsumableArray._toConsumableArray(tempList);
      }

      return oldList;
    });
  };

  var calOffset = function calOffset(offset) {
    return isNaN(Number(offset)) ? offset : "".concat(offset, "px");
  };

  var push = function push(theme, options) {
    var key = String(seed += 1);

    var _ref = toConsumableArray._toConsumableArray(options.offset),
        horizontal = _ref[0],
        vertical = _ref[1];

    var horizontalOffset = calOffset(horizontal);
    var verticalOffset = calOffset(vertical);
    var style = {
      marginTop: verticalOffset,
      marginBottom: verticalOffset,
      marginLeft: horizontalOffset,
      marginRight: horizontalOffset
    };
    var ref2 = /*#__PURE__*/React__default["default"].createRef();
    setList(function (oldList) {
      return [].concat(toConsumableArray._toConsumableArray(oldList), [_objectSpread(_objectSpread({}, options), {}, {
        key: key,
        theme: theme,
        style: style,
        ref: ref2,
        id: key
      })]);
    });
    return Promise.resolve(ref2.current);
  };

  var removeAll = function removeAll() {
    setList([]);
  };

  React.useImperativeHandle(ref, function () {
    return {
      push: push,
      remove: remove,
      removeAll: removeAll
    };
  });
  return /* @__PURE__ */React__default["default"].createElement(NotificationRemoveContext.Provider, {
    value: remove
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-notification__show--").concat(placement),
    style: {
      zIndex: zIndex
    }
  }, list.map(function (props2) {
    var _props2$onDurationEnd = props2.onDurationEnd,
        _onDurationEnd = _props2$onDurationEnd === void 0 ? _util_noop["default"] : _props2$onDurationEnd,
        _props2$onCloseBtnCli = props2.onCloseBtnClick,
        _onCloseBtnClick = _props2$onCloseBtnCli === void 0 ? _util_noop["default"] : _props2$onCloseBtnCli;

    return /* @__PURE__ */React__default["default"].createElement(Notification, _objectSpread(_objectSpread({
      ref: props2.ref,
      key: props2.key
    }, props2), {}, {
      onDurationEnd: function onDurationEnd() {
        remove(props2.key);

        _onDurationEnd();
      },
      onCloseBtnClick: function onCloseBtnClick(e) {
        remove(props2.key);

        _onCloseBtnClick(e);
      }
    }));
  })));
});
var fetchListInstance = function fetchListInstance(placement, attach, zIndex) {
  return new Promise(function (resolve) {
    if (listMap.has(placement)) {
      resolve(listMap.get(placement));
    } else {
      var hasExec = false;
      ReactDOM__default["default"].render( /* @__PURE__ */React__default["default"].createElement(NotificationList, {
        attach: attach,
        placement: placement,
        zIndex: Number(zIndex),
        ref: function ref(instance) {
          if (!hasExec) {
            hasExec = true;
            listMap.set(placement, instance);
            resolve(instance);
          }
        }
      }), attach);
    }
  });
};

var Notification = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var title = props.title,
      content = props.content,
      theme = props.theme,
      icon = props.icon,
      closeBtn = props.closeBtn,
      footer = props.footer,
      duration = props.duration,
      _props$onCloseBtnClic = props.onCloseBtnClick,
      onCloseBtnClick = _props$onCloseBtnClic === void 0 ? _util_noop["default"] : _props$onCloseBtnClic,
      _props$onDurationEnd = props.onDurationEnd,
      onDurationEnd = _props$onDurationEnd === void 0 ? _util_noop["default"] : _props$onDurationEnd,
      style = props.style,
      className = props.className,
      id = props.id;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var baseClassPrefix = "".concat(classPrefix, "-notification");

  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
    CloseIcon: tdesignIconsReact.CloseIcon,
    InfoCircleFilledIcon: tdesignIconsReact.InfoCircleFilledIcon,
    CheckCircleFilledIcon: tdesignIconsReact.CheckCircleFilledIcon
  }),
      CloseIcon = _useGlobalIcon.CloseIcon,
      InfoCircleFilledIcon = _useGlobalIcon.InfoCircleFilledIcon,
      CheckCircleFilledIcon = _useGlobalIcon.CheckCircleFilledIcon;

  var remove = React.useContext(NotificationRemoveContext);
  React__default["default"].useImperativeHandle(ref, function () {
    return {
      close: function close() {
        return remove(id);
      }
    };
  });
  React__default["default"].useEffect(function () {
    var timer;

    if (duration > 0) {
      timer = setTimeout(function () {
        clearTimeout(timer);
        onDurationEnd();
      }, duration);
    }

    return function () {
      if (timer) clearTimeout(timer);
    };
  }, []);

  var renderIcon = function renderIcon() {
    if (typeof icon === "boolean" && !icon) return null;

    var IconWrapper = function IconWrapper(_ref) {
      var children = _ref.children;
      return /* @__PURE__ */React__default["default"].createElement("div", {
        className: "".concat(baseClassPrefix, "__icon")
      }, children);
    };

    if ( /*#__PURE__*/React__default["default"].isValidElement(icon)) {
      return /* @__PURE__ */React__default["default"].createElement(IconWrapper, null, icon);
    }

    if (theme && theme === "success") {
      return /* @__PURE__ */React__default["default"].createElement(IconWrapper, null, /* @__PURE__ */React__default["default"].createElement(CheckCircleFilledIcon, {
        className: "".concat(classPrefix, "-is-success")
      }));
    }

    if (theme && ["info", "warning", "error"].indexOf(theme) >= 0) {
      return /* @__PURE__ */React__default["default"].createElement(IconWrapper, null, /* @__PURE__ */React__default["default"].createElement(InfoCircleFilledIcon, {
        className: "".concat(classPrefix, "-is-").concat(theme)
      }));
    }

    return null;
  };

  var renderCloseBtn = function renderCloseBtn() {
    if (typeof closeBtn === "boolean") {
      return closeBtn && /* @__PURE__ */React__default["default"].createElement(CloseIcon, {
        className: "".concat(baseClassPrefix, "-icon-close"),
        onClick: function onClick(e) {
          onCloseBtnClick({
            e: e
          });
        }
      });
    }

    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(baseClassPrefix, "-close"),
      onClick: function onClick(e) {
        onCloseBtnClick({
          e: e
        });
      }
    }, _util_renderTNode["default"](closeBtn));
  };

  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](className, baseClassPrefix, defineProperty._defineProperty({}, "".concat(baseClassPrefix, "-is-").concat(theme), theme)),
    style: style
  }, renderIcon(), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(baseClassPrefix, "__main")
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(baseClassPrefix, "__title__wrap")
  }, /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(baseClassPrefix, "__title")
  }, title), renderCloseBtn()), content && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(baseClassPrefix, "__content")
  }, _util_renderTNode["default"](content)), footer && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(baseClassPrefix, "__detail")
  }, _util_renderTNode["default"](footer))));
});
Notification.displayName = "Notification";
Notification.defaultProps = notification_defaultProps.notificationDefaultProps;

exports.Notification = Notification;
exports.NotificationRemoveContext = NotificationRemoveContext;
exports.fetchListInstance = fetchListInstance;
exports.listMap = listMap;
