// tdesign-react/es/table/FilterController.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import React, { useRef, useState } from 'react';
import { FilterIcon } from 'tdesign-icons-react';
import { i as isEmpty_1 } from '../_chunks/dep-ea78983d.js';
import classNames from 'classnames';
import { Popup } from '../popup/index.js';
import { Checkbox } from '../checkbox/index.js';
import { Radio } from '../radio/index.js';
import { Input } from '../input/index.js';
import { Button } from '../button/index.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import useGlobalIcon from '../hooks/useGlobalIcon.js';
import '../_common/js/log/index.js';
import log from '../_common/js/log/log.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-614f9a08.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-e237f0f2.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-e54bb0dc.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../hooks/useControlled.js';
import '../_chunks/dep-f54ff1e5.js';
import '../_chunks/dep-e1d4dd61.js';
import '../_chunks/dep-236ef238.js';
import '../_chunks/dep-c8bfa11e.js';
import '../_chunks/dep-199b1581.js';
import '../_util/noop.js';
import '../_util/useAnimation.js';
import '../hooks/useConfig.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-ebf3a272.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../common/Portal.js';
import 'react-dom';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-3c98169f.js';
import '../_util/easing.js';
import '../popup/hooks/useTrigger.js';
import 'react-is';
import '../popup/utils/ref.js';
import '../_util/composeRefs.js';
import '../popup/utils/transition.js';
import '../_util/useMutationObserver.js';
import '../_chunks/dep-236b0359.js';
import '../_chunks/dep-6d5c0a90.js';
import '../_util/useWindowSize.js';
import '../popup/defaultProps.js';
import '../popup/style/css.js';
import '../popup/type.js';
import '../_chunks/dep-66738e15.js';
import '../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../common/Check.js';
import '../_chunks/dep-8368bb87.js';
import '../_util/helper.js';
import '../_chunks/dep-fc5ba3d5.js';
import '../_chunks/dep-a5f8b6ec.js';
import '../checkbox/defaultProps.js';
import '../checkbox/style/css.js';
import '../checkbox/type.js';
import '../_chunks/dep-5bd899df.js';
import '../_util/useCommonClassName.js';
import '../radio/defaultProps.js';
import '../radio/style/css.js';
import '../radio/type.js';
import '../input/Input.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-d1b4ab24.js';
import '../input/InputGroup.js';
import '../input/defaultProps.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';
import '../input/style/css.js';
import '../input/type.js';
import '../button/Button.js';
import '../_util/useRipple.js';
import '../_util/setStyle.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import '../loading/style/css.js';
import '../loading/type.js';
import '../button/defaultProps.js';
import '../button/style/css.js';
import '../button/type.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var CheckboxGroup = Checkbox.Group;
var RadioGroup = Radio.Group;
function TableFilterController(props) {
  var tFilterValue = props.tFilterValue,
      innerFilterValue = props.innerFilterValue,
      tableFilterClasses = props.tableFilterClasses,
      isFocusClass = props.isFocusClass,
      column = props.column;

  var _useGlobalIcon = useGlobalIcon({
    FilterIcon: FilterIcon
  }),
      FilterIcon$1 = _useGlobalIcon.FilterIcon;

  var triggerElementRef = useRef(null);

  var _useLocaleReceiver = useLocaleReceiver("table"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      filterPopupVisible = _useState2[0],
      setFilterPopupVisible = _useState2[1];

  var onFilterPopupVisibleChange = function onFilterPopupVisibleChange(visible) {
    var _props$onVisibleChang;

    setFilterPopupVisible(visible);
    (_props$onVisibleChang = props.onVisibleChange) === null || _props$onVisibleChang === void 0 ? void 0 : _props$onVisibleChang.call(props, visible);
  };

  var getFilterContent = function getFilterContent(column2) {
    var _column2$filter, _column2$filter2, _column2$filter3, _column2$filter4, _column2$filter5;

    var types = ["single", "multiple", "input"];

    if (column2.type && !types.includes(column2.filter.type)) {
      log.error("Table", "TDesign Table Error: column.filter.type must be the following: ".concat(JSON.stringify(types)));
      return;
    }

    var Component = {
      single: RadioGroup,
      multiple: CheckboxGroup,
      input: Input
    }[column2.filter.type];
    if (!Component && !(column2 !== null && column2 !== void 0 && (_column2$filter = column2.filter) !== null && _column2$filter !== void 0 && _column2$filter.component)) return;

    var filterComponentProps = _objectSpread(_objectSpread({
      options: ["single", "multiple"].includes(column2.filter.type) ? (_column2$filter2 = column2.filter) === null || _column2$filter2 === void 0 ? void 0 : _column2$filter2.list : void 0
    }, ((_column2$filter3 = column2.filter) === null || _column2$filter3 === void 0 ? void 0 : _column2$filter3.props) || {}), {}, {
      value: innerFilterValue === null || innerFilterValue === void 0 ? void 0 : innerFilterValue[column2.colKey],
      onChange: function onChange(val) {
        var _props$onInnerFilterC;

        (_props$onInnerFilterC = props.onInnerFilterChange) === null || _props$onInnerFilterC === void 0 ? void 0 : _props$onInnerFilterC.call(props, val, column2);
      }
    });

    if ((_column2$filter4 = column2.filter) !== null && _column2$filter4 !== void 0 && _column2$filter4.confirmEvents) {
      column2.filter.confirmEvents.forEach(function (event) {
        filterComponentProps[event] = function () {
          var _props$onConfirm;

          setFilterPopupVisible(false);
          (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 ? void 0 : _props$onConfirm.call(props, column2);
        };
      });
    }

    var FilterComponent = (column2 === null || column2 === void 0 ? void 0 : (_column2$filter5 = column2.filter) === null || _column2$filter5 === void 0 ? void 0 : _column2$filter5.component) || Component;
    return /* @__PURE__ */React.createElement("div", {
      className: tableFilterClasses.contentInner
    }, /* @__PURE__ */React.createElement(FilterComponent, _objectSpread({
      value: innerFilterValue === null || innerFilterValue === void 0 ? void 0 : innerFilterValue[column2.colKey]
    }, filterComponentProps)));
  };

  var getBottomButtons = function getBottomButtons(column2) {
    if (!column2.filter.showConfirmAndReset) return;
    return /* @__PURE__ */React.createElement("div", {
      className: tableFilterClasses.bottomButtons
    }, /* @__PURE__ */React.createElement(Button, {
      theme: "default",
      size: "small",
      onClick: function onClick() {
        var _props$onReset;

        setFilterPopupVisible(false);
        (_props$onReset = props.onReset) === null || _props$onReset === void 0 ? void 0 : _props$onReset.call(props, column2);
      }
    }, locale.resetText), /* @__PURE__ */React.createElement(Button, {
      theme: "primary",
      size: "small",
      onClick: function onClick() {
        var _props$onConfirm2;

        setFilterPopupVisible(false);
        (_props$onConfirm2 = props.onConfirm) === null || _props$onConfirm2 === void 0 ? void 0 : _props$onConfirm2.call(props, column2);
      }
    }, locale.confirmText));
  };

  if (!column.filter || column.filter && !Object.keys(column.filter).length) return null;
  var defaultFilterIcon = t(locale.filterIcon) || /* @__PURE__ */React.createElement(FilterIcon$1, null);
  return /* @__PURE__ */React.createElement("div", {
    className: classNames([tableFilterClasses.icon, _defineProperty({}, isFocusClass, !isEmpty_1(tFilterValue === null || tFilterValue === void 0 ? void 0 : tFilterValue[column.colKey]))])
  }, /* @__PURE__ */React.createElement(Popup, {
    visible: filterPopupVisible,
    destroyOnClose: true,
    trigger: "click",
    placement: "bottom-right",
    showArrow: true,
    overlayClassName: tableFilterClasses.popup,
    onVisibleChange: function onVisibleChange(val) {
      return onFilterPopupVisibleChange(val);
    },
    content: /* @__PURE__ */React.createElement("div", {
      className: tableFilterClasses.popupContent
    }, getFilterContent(column), getBottomButtons(column))
  }, /* @__PURE__ */React.createElement("div", {
    ref: triggerElementRef
  }, props.filterIcon || defaultFilterIcon)));
}

export { TableFilterController as default };
