// tdesign-react/esm/dropdown/DropdownItem.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import React, { forwardRef, useRef } from 'react';
import classNames from 'classnames';
import useConfig from '../hooks/useConfig.js';
import useRipple from '../_util/useRipple.js';
import { dropdownItemDefaultProps } from './defaultProps.js';
import { pxCompat } from '../_util/helper.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-e54bb0dc.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-e237f0f2.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-ebf3a272.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../_util/useAnimation.js';
import '../_util/setStyle.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-3c98169f.js';
import '../_util/easing.js';
import '../_chunks/dep-8368bb87.js';
import '../_chunks/dep-9615f03f.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-fc5ba3d5.js';
import '../_chunks/dep-c8bfa11e.js';
import '../_chunks/dep-199b1581.js';
import '../_chunks/dep-f54ff1e5.js';
import '../_chunks/dep-e1d4dd61.js';
import '../_chunks/dep-236ef238.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var DropdownItem = /*#__PURE__*/forwardRef(function (props, ref) {
  var _classNames;

  var children = props.children,
      className = props.className,
      active = props.active,
      disabled = props.disabled,
      value = props.value,
      theme = props.theme,
      prefixIcon = props.prefixIcon,
      maxColumnWidth = props.maxColumnWidth,
      minColumnWidth = props.minColumnWidth,
      onClick = props.onClick,
      style = props.style,
      isSubmenu = props.isSubmenu;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var dropdownItemRef = useRef(null);
  useRipple(isSubmenu ? null : ref || dropdownItemRef);

  var handleItemClick = function handleItemClick(e) {
    onClick === null || onClick === void 0 ? void 0 : onClick(value, e);
  };

  return /* @__PURE__ */React.createElement(React.Fragment, null, /* @__PURE__ */React.createElement("li", {
    className: classNames(className, "".concat(classPrefix, "-dropdown__item--theme-").concat(theme), (_classNames = {}, _defineProperty(_classNames, "".concat(classPrefix, "-dropdown__item--active"), active), _defineProperty(_classNames, "".concat(classPrefix, "-dropdown__item--disabled"), disabled), _classNames)),
    onClick: handleItemClick,
    style: _objectSpread({
      maxWidth: pxCompat(maxColumnWidth),
      minWidth: pxCompat(minColumnWidth)
    }, style),
    ref: ref || dropdownItemRef
  }, prefixIcon ? /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-dropdown__item-icon")
  }, prefixIcon) : null, children));
});
DropdownItem.displayName = "DropdownItem";
DropdownItem.defaultProps = dropdownItemDefaultProps;

export { DropdownItem as default };
