// tdesign-react/esm/dropdown/Dropdown.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import React, { useState } from 'react';
import classNames from 'classnames';
import { o as omit_1 } from '../_chunks/dep-27f96743.js';
import { Popup } from '../popup/index.js';
import DropdownMenu from './DropdownMenu.js';
import DropdownItem from './DropdownItem.js';
import { dropdownDefaultProps } from './defaultProps.js';
import useConfig from '../hooks/useConfig.js';
import useDropdownOptions from './hooks/useDropdownOptions.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-c8bfa11e.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-199b1581.js';
import '../_chunks/dep-e237f0f2.js';
import '../_chunks/dep-1a36642a.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-e54bb0dc.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-614f9a08.js';
import '../_chunks/dep-aadcae4d.js';
import '../_chunks/dep-93a10d8c.js';
import '../_chunks/dep-236ef238.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-3576b6cd.js';
import '../_chunks/dep-ebf3a272.js';
import '../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../hooks/useControlled.js';
import '../_chunks/dep-f54ff1e5.js';
import '../_chunks/dep-e1d4dd61.js';
import '../_util/noop.js';
import '../_util/useAnimation.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../common/Portal.js';
import 'react-dom';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-3c98169f.js';
import '../_util/easing.js';
import '../popup/hooks/useTrigger.js';
import 'react-is';
import '../popup/utils/ref.js';
import '../_util/composeRefs.js';
import '../popup/utils/transition.js';
import '../_util/useMutationObserver.js';
import '../_chunks/dep-236b0359.js';
import '../_chunks/dep-6d5c0a90.js';
import '../_util/useWindowSize.js';
import '../popup/defaultProps.js';
import '../popup/style/index.js';
import '../popup/type.js';
import 'tdesign-icons-react';
import '../divider/index.js';
import '../divider/Divider.js';
import '../_chunks/dep-8368bb87.js';
import '../divider/defaultProps.js';
import '../divider/style/index.js';
import '../divider/type.js';
import '../hooks/useGlobalIcon.js';
import '../_util/useRipple.js';
import '../_util/setStyle.js';
import '../_util/helper.js';
import '../_chunks/dep-fc5ba3d5.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var Dropdown = function Dropdown(props) {
  var _props$popupProps = props.popupProps,
      popupProps = _props$popupProps === void 0 ? {} : _props$popupProps,
      disabled = props.disabled,
      placement = props.placement,
      trigger = props.trigger,
      className = props.className,
      children = props.children,
      hideAfterItemClick = props.hideAfterItemClick,
      propsOptions = props.options,
      style = props.style;
  var arrayChildren = React.Children.toArray(children);

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isPopupVisible = _useState2[0],
      togglePopupVisible = _useState2[1];

  var dropdownClass = "".concat(classPrefix, "-dropdown");
  var options = useDropdownOptions(arrayChildren, propsOptions);

  var handleMenuClick = function handleMenuClick(data, context) {
    var _props$onClick;

    if (hideAfterItemClick) togglePopupVisible(false);
    props === null || props === void 0 ? void 0 : (_props$onClick = props.onClick) === null || _props$onClick === void 0 ? void 0 : _props$onClick.call(props, data, context);
  };

  var handleVisibleChange = function handleVisibleChange(visible, context) {
    var _popupProps$onVisible;

    togglePopupVisible(visible);
    popupProps === null || popupProps === void 0 ? void 0 : (_popupProps$onVisible = popupProps.onVisibleChange) === null || _popupProps$onVisible === void 0 ? void 0 : _popupProps$onVisible.call(popupProps, visible, context);
  };

  var renderContent = /* @__PURE__ */React.createElement(DropdownMenu, _objectSpread(_objectSpread({}, props), {}, {
    options: options,
    onClick: handleMenuClick
  }));

  var popupParams = _objectSpread(_objectSpread({
    disabled: disabled,
    placement: placement,
    trigger: trigger,
    showArrow: false,
    content: renderContent
  }, omit_1(popupProps, "onVisibleChange")), {}, {
    overlayInnerClassName: classNames(dropdownClass, className, popupProps === null || popupProps === void 0 ? void 0 : popupProps.overlayInnerClassName),
    overlayInnerStyle: style
  });

  return /* @__PURE__ */React.createElement(Popup, _objectSpread({
    expandAnimation: true,
    destroyOnClose: true,
    visible: isPopupVisible,
    onVisibleChange: handleVisibleChange
  }, popupParams), arrayChildren === null || arrayChildren === void 0 ? void 0 : arrayChildren[0]);
};

Dropdown.DropdownItem = DropdownItem;
Dropdown.DropdownMenu = DropdownMenu;
Dropdown.displayName = "Dropdown";
Dropdown.defaultProps = dropdownDefaultProps;

export { Dropdown as default };
