// tdesign-react/es/select-input/useSingle.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import React, { useRef, useMemo } from 'react';
import { i as isObject_1 } from '../_chunks/dep-e54bb0dc.js';
import { p as pick_1 } from '../_chunks/dep-c776db26.js';
import classNames from 'classnames';
import { Input } from '../input/index.js';
import { Loading } from '../loading/index.js';
import useConfig from '../hooks/useConfig.js';
import useControlled from '../hooks/useControlled.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-93a10d8c.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-199b1581.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-e237f0f2.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-c8bfa11e.js';
import '../_chunks/dep-afbd242f.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-3576b6cd.js';
import '../_chunks/dep-aadcae4d.js';
import '../_chunks/dep-ebf3a272.js';
import '../input/Input.js';
import '../_chunks/dep-8368bb87.js';
import 'tdesign-icons-react';
import '../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../hooks/useGlobalIcon.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-d1b4ab24.js';
import '../input/InputGroup.js';
import '../locale/LocalReceiver.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';
import '../input/defaultProps.js';
import '../_chunks/dep-f54ff1e5.js';
import '../_chunks/dep-e1d4dd61.js';
import '../_chunks/dep-236ef238.js';
import '../_util/noop.js';
import '../input/style/css.js';
import '../input/type.js';
import '../loading/Loading.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-3c98169f.js';
import '../_util/easing.js';
import '../common/Portal.js';
import 'react-dom';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import '../loading/style/css.js';
import '../loading/type.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var COMMON_PROPERTIES = ["status", "clearable", "disabled", "label", "placeholder", "readonly", "suffix", "suffixIcon", "onPaste", "onEnter", "onMouseenter", "onMouseleave"];
var DEFAULT_KEYS = {
  label: "label",
  value: "value"
};

function getInputValue(value, keys) {
  var iKeys = keys || DEFAULT_KEYS;
  return isObject_1(value) ? value[iKeys.label] : value;
}

function useSingle(props) {
  var value = props.value,
      keys = props.keys,
      loading = props.loading,
      disabled = props.disabled;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var inputRef = useRef();

  var _useControlled = useControlled(props, "inputValue", props.onInputChange),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      inputValue = _useControlled2[0],
      setInputValue = _useControlled2[1];

  var showLoading = useMemo(function () {
    return !disabled && loading;
  }, [loading, disabled]);

  var commonInputProps = _objectSpread(_objectSpread({}, pick_1(props, COMMON_PROPERTIES)), {}, {
    suffixIcon: showLoading ? /* @__PURE__ */React.createElement(Loading, {
      loading: true,
      size: "small"
    }) : props.suffixIcon
  });

  var onInnerClear = function onInnerClear(context) {
    var _context$e, _props$onClear;

    context === null || context === void 0 ? void 0 : (_context$e = context.e) === null || _context$e === void 0 ? void 0 : _context$e.stopPropagation();
    (_props$onClear = props.onClear) === null || _props$onClear === void 0 ? void 0 : _props$onClear.call(props, context);
    setInputValue("", {
      trigger: "clear"
    });
  };

  var onInnerInputChange = function onInnerInputChange(value2, context) {
    if (props.allowInput) {
      setInputValue(value2, _objectSpread(_objectSpread({}, context), {}, {
        trigger: "input"
      }));
    }
  };

  var renderSelectSingle = function renderSelectSingle(popupVisible) {
    var _props$inputProps;

    var singleValueDisplay = !props.multiple ? props.valueDisplay : null;
    var displayedValue = popupVisible && props.allowInput ? inputValue : getInputValue(value, keys);
    return /* @__PURE__ */React.createElement(Input, _objectSpread(_objectSpread(_objectSpread({
      ref: inputRef
    }, commonInputProps), {}, {
      autoWidth: props.autoWidth,
      placeholder: singleValueDisplay ? "" : props.placeholder,
      value: singleValueDisplay ? void 0 : displayedValue,
      label: /* @__PURE__ */React.createElement(React.Fragment, null, props.label, singleValueDisplay),
      onChange: onInnerInputChange,
      readonly: !props.allowInput,
      onClear: onInnerClear,
      onBlur: function onBlur(val, context) {
        var _props$onBlur;

        (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 ? void 0 : _props$onBlur.call(props, value, _objectSpread(_objectSpread({}, context), {}, {
          inputValue: val
        }));
      },
      onFocus: function onFocus(val, context) {
        var _props$onFocus;

        (_props$onFocus = props.onFocus) === null || _props$onFocus === void 0 ? void 0 : _props$onFocus.call(props, value, _objectSpread(_objectSpread({}, context), {}, {
          inputValue: val
        }));
        !popupVisible && setInputValue(getInputValue(value, keys), _objectSpread(_objectSpread({}, context), {}, {
          trigger: "input"
        }));
      },
      onEnter: function onEnter(val, context) {
        var _props$onEnter;

        (_props$onEnter = props.onEnter) === null || _props$onEnter === void 0 ? void 0 : _props$onEnter.call(props, value, _objectSpread(_objectSpread({}, context), {}, {
          inputValue: val
        }));
      }
    }, props.inputProps), {}, {
      inputClass: classNames((_props$inputProps = props.inputProps) === null || _props$inputProps === void 0 ? void 0 : _props$inputProps.className, _defineProperty({}, "".concat(classPrefix, "-input--focused"), popupVisible))
    }));
  };

  return {
    inputRef: inputRef,
    commonInputProps: commonInputProps,
    onInnerClear: onInnerClear,
    renderSelectSingle: renderSelectSingle
  };
}

export { useSingle as default };
