// tdesign-react/lib/menu/MenuItem.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import React, { useRef, useContext } from 'react';
import classNames from 'classnames';
import useConfig from '../hooks/useConfig.js';
import useRipple from '../_util/useRipple.js';
import { MenuContext } from './MenuContext.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../_util/useAnimation.js';
import '../_util/setStyle.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-8d5556d7.js';
import '../_util/easing.js';
import '../_util/noop.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var MenuItem = function MenuItem(props) {
  var _classNames;

  var content = props.content,
      _props$children = props.children,
      children = _props$children === void 0 ? content : _props$children,
      disabled = props.disabled,
      href = props.href,
      _props$target = props.target,
      target = _props$target === void 0 ? "_self" : _props$target,
      value = props.value,
      className = props.className,
      style = props.style,
      icon = props.icon,
      onClick = props.onClick;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var menuItemRef = useRef();
  useRipple(menuItemRef);

  var _useContext = useContext(MenuContext),
      onChange = _useContext.onChange,
      setState = _useContext.setState,
      active = _useContext.active;

  var handleClick = function handleClick(e) {
    e.stopPropagation();
    if (disabled) return;
    onClick && onClick({
      e: e
    });
    onChange(value);
    setState({
      active: value
    });
  };

  return /* @__PURE__ */React.createElement("li", {
    ref: menuItemRef,
    className: classNames("".concat(classPrefix, "-menu__item"), className, (_classNames = {}, _defineProperty(_classNames, "".concat(classPrefix, "-is-disabled"), disabled), _defineProperty(_classNames, "".concat(classPrefix, "-is-active"), value === active), _defineProperty(_classNames, "".concat(classPrefix, "-menu__item--plain"), !icon), _classNames)),
    style: _objectSpread({}, style),
    onClick: handleClick
  }, icon, href ? /* @__PURE__ */React.createElement("a", {
    href: href,
    target: target,
    className: classNames("".concat(classPrefix, "-menu__item-link"))
  }, /* @__PURE__ */React.createElement("span", {
    className: "".concat(classPrefix, "-menu__content")
  }, children)) : /* @__PURE__ */React.createElement("span", {
    className: "".concat(classPrefix, "-menu__content")
  }, children));
};

MenuItem.displayName = "MenuItem";

export { MenuItem as default };
