// tdesign-react/lib/menu/Menu.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import React from 'react';
import classNames from 'classnames';
import useConfig from '../hooks/useConfig.js';
import forwardRefWithStatics from '../_util/forwardRefWithStatics.js';
import { MenuContext } from './MenuContext.js';
import { DEFAULT_MENU_WIDTH } from './_util/constant.js';
import useMenuContext from './hooks/useMenuContext.js';
import { menuDefaultProps } from './defaultProps.js';
import HeadMenu from './HeadMenu.js';
import SubMenu from './SubMenu.js';
import MenuItem from './MenuItem.js';
import MenuGroup from './MenuGroup.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import 'hoist-non-react-statics';
import '../_util/noop.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-d1b4ab24.js';
import '../_chunks/dep-9615f03f.js';
import './_util/checkSubMenuChildExpanded.js';
import './_util/type.js';
import '../tabs/index.js';
import '../tabs/Tabs.js';
import '../tabs/TabNav.js';
import 'tdesign-icons-react';
import '../_chunks/dep-05d10538.js';
import '../_chunks/dep-652c1652.js';
import '../_chunks/dep-b989427d.js';
import '../tabs/useTabClass.js';
import '../tabs/TabNavItem.js';
import '../_util/useRipple.js';
import '../_util/useAnimation.js';
import '../_util/setStyle.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-8d5556d7.js';
import '../_util/easing.js';
import '../hooks/useGlobalIcon.js';
import '../tabs/TabBar.js';
import '../_common/js/tabs/base.js';
import '../tabs/TabPanel.js';
import '../tabs/defaultProps.js';
import './_util/checkSubMenuActive.js';
import './_util/checkMenuType.js';
import './_util/checkSubMenuChildrenActive.js';
import './_util/getSubMenuChildStyle.js';
import '../common/FakeArrow.js';
import './_util/cacularPaddingLeft.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Menu = forwardRefWithStatics(function (props, ref) {
  var _classNames;

  var className = props.className,
      _props$width = props.width,
      width = _props$width === void 0 ? DEFAULT_MENU_WIDTH : _props$width,
      children = props.children,
      theme = props.theme,
      logo = props.logo,
      operations = props.operations,
      style = props.style;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useMenuContext = useMenuContext(_objectSpread(_objectSpread({}, props), {}, {
    children: children,
    mode: "accordion"
  })),
      value = _useMenuContext.value;

  var menuWidthArr = Array.isArray(width) ? width : [width, DEFAULT_MENU_WIDTH[1]];
  var collapsed = value.collapsed;
  return /* @__PURE__ */React.createElement(MenuContext.Provider, {
    value: value
  }, /* @__PURE__ */React.createElement("div", {
    ref: ref,
    className: classNames("".concat(classPrefix, "-default-menu"), className, (_classNames = {}, _defineProperty(_classNames, "".concat(classPrefix, "-is-collapsed"), collapsed), _defineProperty(_classNames, "".concat(classPrefix, "-menu--dark"), theme === "dark"), _classNames)),
    style: _objectSpread({
      width: collapsed ? menuWidthArr[1] : menuWidthArr[0]
    }, style)
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-default-menu__inner")
  }, logo && /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-menu__logo")
  }, logo), /* @__PURE__ */React.createElement("ul", {
    className: classNames("".concat(classPrefix, "-menu"), _defineProperty({}, "".concat(classPrefix, "-menu--scroll"), !collapsed))
  }, children), operations && /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-menu__operations")
  }, operations))));
}, {
  HeadMenu: HeadMenu,
  SubMenu: SubMenu,
  MenuItem: MenuItem,
  MenuGroup: MenuGroup
});
Menu.displayName = "Menu";
Menu.defaultProps = menuDefaultProps;

export { Menu as default };
