// tdesign-react/esm/date-picker/base/Table.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../../_chunks/dep-9615f03f.js';
import React from 'react';
import classNames from 'classnames';
import { useLocaleReceiver } from '../../locale/LocalReceiver.js';
import useConfig from '../../hooks/useConfig.js';
import DatePickerCell from './Cell.js';
import { parseToDayjs } from '../../_common/js/date-picker/format.js';
import '../../_chunks/dep-7909fefa.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../_chunks/dep-1cce5a60.js';
import '../../_chunks/dep-13a70a46.js';
import '../../_chunks/dep-dc1ce7ed.js';
import '../../_chunks/dep-4e2181de.js';
import '../../_chunks/dep-c197ce4f.js';
import '../../_chunks/dep-e430e2ee.js';
import '../../_chunks/dep-4a64abb1.js';
import '../../_chunks/dep-4e0da095.js';
import '../../_chunks/dep-e54bb0dc.js';
import '../../_chunks/dep-b2626e28.js';
import '../../_chunks/dep-5a2778c1.js';
import '../../_chunks/dep-a2b97df8.js';
import '../../_chunks/dep-7aeeb170.js';
import '../../_chunks/dep-2afd805c.js';
import '../../_chunks/dep-e237f0f2.js';
import '../../_chunks/dep-4ed2dce5.js';
import '../../_chunks/dep-c8db35c0.js';
import '../../_chunks/dep-96aef4a9.js';
import '../../_chunks/dep-557a1b0f.js';
import '../../_chunks/dep-264b0e55.js';
import '../../_chunks/dep-f5693bf6.js';
import '../../_chunks/dep-fdd1d1a6.js';
import '../../_chunks/dep-7ff8b18e.js';
import '../../_chunks/dep-ebf3a272.js';
import '../../_chunks/dep-6a2482c6.js';
import '../../_chunks/dep-a980ac05.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import '../../_common/js/date-picker/utils.js';
import 'dayjs';
import '../../_chunks/dep-76ad5996.js';
import '../../_chunks/dep-6d5c0a90.js';
import '../../_chunks/dep-199b1581.js';
import '../../_chunks/dep-236ef238.js';
import '../../_common/js/log/index.js';
import '../../_common/js/log/log.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var DatePickerTable = function DatePickerTable(props) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var value = props.value,
      format = props.format,
      mode = props.mode,
      data = props.data,
      time = props.time,
      onCellClick = props.onCellClick,
      onCellMouseEnter = props.onCellMouseEnter,
      onCellMouseLeave = props.onCellMouseLeave,
      firstDayOfWeek = props.firstDayOfWeek;

  var _useLocaleReceiver = useLocaleReceiver("datePicker"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var weekdays = t(local.weekdays);
  var weekAbbreviation = t(local.weekAbbreviation);
  var weekArr = [];
  var wi = firstDayOfWeek - 1;
  var len = weekdays.length;

  while (weekArr.length < len) {
    weekArr.push(weekdays[wi]);
    wi = (wi + len + 1) % len;
  }

  if (mode === "week") weekArr.unshift(weekAbbreviation);
  var showThead = mode === "date" || mode === "week";

  var weekRowClass = function weekRowClass(value2, format2, targetValue) {
    if (mode !== "week") return {};

    if (Array.isArray(value2)) {
      var _startObj$year, _startObj$week, _endObj$year, _endObj$week, _ref;

      if (!value2.length) return {};

      var _value2$map = value2.map(function (v) {
        return v && parseToDayjs(v, format2);
      }),
          _value2$map2 = _slicedToArray(_value2$map, 2),
          startObj = _value2$map2[0],
          endObj = _value2$map2[1];

      var startYear = startObj === null || startObj === void 0 ? void 0 : (_startObj$year = startObj.year) === null || _startObj$year === void 0 ? void 0 : _startObj$year.call(startObj);
      var startWeek = startObj === null || startObj === void 0 ? void 0 : (_startObj$week = startObj.week) === null || _startObj$week === void 0 ? void 0 : _startObj$week.call(startObj);
      var endYear = endObj === null || endObj === void 0 ? void 0 : (_endObj$year = endObj.year) === null || _endObj$year === void 0 ? void 0 : _endObj$year.call(endObj);
      var endWeek = endObj === null || endObj === void 0 ? void 0 : (_endObj$week = endObj.week) === null || _endObj$week === void 0 ? void 0 : _endObj$week.call(endObj);
      var targetObj = parseToDayjs(targetValue, format2);
      var targetYear = targetObj.year();
      var targetWeek = targetObj.week();
      var isActive = targetYear === startYear && targetWeek === startWeek || targetYear === endYear && targetWeek === endWeek;
      var isRange = targetYear >= startYear && targetYear <= endYear && targetWeek > startWeek && targetWeek < endWeek;
      return _ref = {}, _defineProperty(_ref, "".concat(classPrefix, "-date-picker__table-").concat(mode, "-row--active"), isActive), _defineProperty(_ref, "".concat(classPrefix, "-date-picker__table-").concat(mode, "-row--range"), isRange), _ref;
    }

    return _defineProperty({}, "".concat(classPrefix, "-date-picker__table-").concat(mode, "-row--active"), parseToDayjs(value2, format2).week() === parseToDayjs(targetValue, format2).week());
  };

  return /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-date-picker__table"),
    onMouseLeave: function onMouseLeave(e) {
      return onCellMouseLeave === null || onCellMouseLeave === void 0 ? void 0 : onCellMouseLeave({
        e: e
      });
    }
  }, /* @__PURE__ */React.createElement("table", null, showThead && /* @__PURE__ */React.createElement("thead", null, /* @__PURE__ */React.createElement("tr", null, weekArr.map(function (value2, i) {
    return /* @__PURE__ */React.createElement("th", {
      key: i
    }, value2);
  }))), /* @__PURE__ */React.createElement("tbody", null, data.map(function (row, i) {
    return /* @__PURE__ */React.createElement("tr", {
      key: i,
      className: classNames("".concat(classPrefix, "-date-picker__table-").concat(mode, "-row"), _objectSpread({}, weekRowClass(value, format, row[0].value)))
    }, row.map(function (col, j) {
      return /* @__PURE__ */React.createElement(DatePickerCell, _objectSpread(_objectSpread({}, col), {}, {
        key: j,
        time: time,
        onClick: onCellClick,
        onMouseEnter: onCellMouseEnter
      }));
    }));
  }))));
};

DatePickerTable.displayName = "DatePickerTable";

export { DatePickerTable as default };
