// tdesign-react/cjs/rate/Rate.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../_chunks/dep-8052f095.js');
var defineProperty = require('../_chunks/dep-4eb93341.js');
var slicedToArray = require('../_chunks/dep-9c5d4f93.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var tooltip_index = require('../tooltip/index.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var rate_defaultProps = require('./defaultProps.js');
require('../_chunks/dep-2205decf.js');
require('../_chunks/dep-606cfe2a.js');
require('../tooltip/Tooltip.js');
require('../_chunks/dep-1a0462cc.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../_util/useAnimation.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../common/Portal.js');
require('react-dom');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-9a2dbbb1.js');
require('../_util/easing.js');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../popup/utils/ref.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../_util/useMutationObserver.js');
require('../_chunks/dep-db2b3323.js');
require('../_chunks/dep-52c5813c.js');
require('../_chunks/dep-09805e35.js');
require('../_util/useWindowSize.js');
require('../popup/defaultProps.js');
require('../_chunks/dep-c76c2a08.js');
require('../_chunks/dep-dc6c808e.js');
require('../_chunks/dep-76deafcc.js');
require('../_chunks/dep-afce8659.js');
require('../_util/noop.js');
require('../tooltip/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Rate = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var allowHalf = props.allowHalf,
      color = props.color,
      count = props.count,
      disabled = props.disabled,
      gap = props.gap,
      showText = props.showText,
      size = props.size,
      texts = props.texts,
      icon = props.icon,
      className = props.className,
      style = props.style,
      onChange = props.onChange;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
    StarFilledIcon: tdesignIconsReact.StarFilledIcon
  }),
      StarFilledIcon = _useGlobalIcon.StarFilledIcon;

  var _useControlled = hooks_useControlled["default"](props, "value", onChange),
      _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
      _useControlled2$ = _useControlled2[0],
      starValue = _useControlled2$ === void 0 ? 0 : _useControlled2$,
      setStarValue = _useControlled2[1];

  var _useState = React.useState(void 0),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      _useState2$ = _useState2[0],
      hoverValue = _useState2$ === void 0 ? void 0 : _useState2$,
      setHoverValue = _useState2[1];

  var displayValue = hoverValue || starValue;
  var rootRef = React__default["default"].useRef(null);
  var activeColor = Array.isArray(color) ? color[0] : color;
  var defaultColor = Array.isArray(color) ? color[1] : "var(--td-bg-color-component)";

  var RateIcon = function RateIcon(props2) {
    if ( /*#__PURE__*/React__default["default"].isValidElement(icon)) {
      return /*#__PURE__*/React__default["default"].cloneElement(icon, props2);
    }

    return /* @__PURE__ */React__default["default"].createElement(StarFilledIcon, _objectSpread({}, props2));
  };

  var getStarValue = function getStarValue(event, index) {
    if (allowHalf) {
      var rootNode = rootRef.current;

      var _rootNode$getBounding = rootNode.getBoundingClientRect(),
          left = _rootNode$getBounding.left;

      var firstStar = rootNode.firstChild;

      var _firstStar$getBoundin = firstStar.getBoundingClientRect(),
          width = _firstStar$getBoundin.width;

      var clientX = event.clientX;
      var starMiddle = width * (index - 0.5) + gap * (index - 1);
      if (clientX - left >= starMiddle) return index;
      if (clientX - left < starMiddle) return index - 0.5;
    }

    return index;
  };

  var mouseEnterHandler = function mouseEnterHandler(event, index) {
    if (disabled) return;
    setHoverValue(getStarValue(event, index));
  };

  var mouseLeaveHandler = function mouseLeaveHandler() {
    if (disabled) return;
    setHoverValue(void 0);
  };

  var clickHandler = function clickHandler(event, index) {
    if (disabled) return;
    setStarValue(getStarValue(event, index));
  };

  var getStarCls = function getStarCls(index) {
    if (allowHalf && index + 0.5 === displayValue) return "".concat(classPrefix, "-rate__item--half");
    if (index >= displayValue) return "";
    if (index < displayValue) return "".concat(classPrefix, "-rate__item--full");
  };

  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    style: style,
    className: classNames__default["default"]("".concat(classPrefix, "-rate"), className),
    onMouseLeave: mouseLeaveHandler
  }, /* @__PURE__ */React__default["default"].createElement("ul", {
    className: "".concat(classPrefix, "-rate__list"),
    style: {
      gap: gap
    },
    ref: rootRef
  }, toConsumableArray._toConsumableArray(Array(count)).map(function (_, index) {
    return /* @__PURE__ */React__default["default"].createElement("li", {
      key: index,
      className: classNames__default["default"]("".concat(classPrefix, "-rate__item"), getStarCls(index)),
      onClick: function onClick(event) {
        return clickHandler(event, index + 1);
      },
      onMouseMove: function onMouseMove(event) {
        return mouseEnterHandler(event, index + 1);
      }
    }, showText ? /* @__PURE__ */React__default["default"].createElement(tooltip_index.Tooltip, {
      key: index,
      content: texts[displayValue - 1]
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-rate__star-top")
    }, /* @__PURE__ */React__default["default"].createElement(RateIcon, {
      size: size,
      color: activeColor
    })), /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-rate__star-bottom")
    }, /* @__PURE__ */React__default["default"].createElement(RateIcon, {
      size: size,
      color: defaultColor
    }))) : /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-rate__star-top")
    }, /* @__PURE__ */React__default["default"].createElement(RateIcon, {
      size: size,
      color: activeColor
    })), /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-rate__star-bottom")
    }, /* @__PURE__ */React__default["default"].createElement(RateIcon, {
      size: size,
      color: defaultColor
    }))));
  })), showText && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-rate__text")
  }, texts[displayValue - 1]));
});
Rate.displayName = "Rate";
Rate.defaultProps = rate_defaultProps.rateDefaultProps;

exports["default"] = Rate;
