// tdesign-react/es/drawer/Drawer.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import React, { forwardRef, useRef, useState, useImperativeHandle, useEffect, useCallback } from 'react';
import classNames from 'classnames';
import { CloseIcon } from 'tdesign-icons-react';
import { addClass, removeClass } from '../_util/dom.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import getScrollbarWidth from '../_util/getScrollbarWidth.js';
import hasScrollbar from '../_util/hasScrollBar.js';
import DrawerWrapper from './DrawerWrapper.js';
import { Button } from '../button/index.js';
import useConfig from '../hooks/useConfig.js';
import useGlobalIcon from '../hooks/useGlobalIcon.js';
import { drawerDefaultProps } from './defaultProps.js';
import useDrag from './hooks/useDrag.js';
import '../_chunks/dep-7909fefa.js';
import 'raf';
import '../_chunks/dep-3c98169f.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-e237f0f2.js';
import '../_util/easing.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';
import '../_chunks/dep-13a70a46.js';
import '../_chunks/dep-dc1ce7ed.js';
import '../_chunks/dep-4e2181de.js';
import '../_chunks/dep-c197ce4f.js';
import '../_chunks/dep-e430e2ee.js';
import '../_chunks/dep-e54bb0dc.js';
import '../_chunks/dep-b2626e28.js';
import '../_chunks/dep-5a2778c1.js';
import '../_chunks/dep-a2b97df8.js';
import '../_chunks/dep-7aeeb170.js';
import '../_chunks/dep-2afd805c.js';
import '../_chunks/dep-4ed2dce5.js';
import '../_chunks/dep-c8db35c0.js';
import '../_chunks/dep-96aef4a9.js';
import '../_chunks/dep-557a1b0f.js';
import '../_chunks/dep-f5693bf6.js';
import '../_chunks/dep-fdd1d1a6.js';
import '../_chunks/dep-7ff8b18e.js';
import '../_chunks/dep-ebf3a272.js';
import '../_chunks/dep-6a2482c6.js';
import '../_chunks/dep-a980ac05.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../common/Portal.js';
import 'react-dom';
import '../button/Button.js';
import '../_chunks/dep-8368bb87.js';
import '../_util/useRipple.js';
import '../_util/useAnimation.js';
import '../_util/setStyle.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-d1b4ab24.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import '../loading/style/css.js';
import '../loading/type.js';
import '../button/defaultProps.js';
import '../button/style/css.js';
import '../button/type.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var CloseTriggerType = {
  CLICK_OVERLAY: "overlay",
  CLICK_CLOSE_BTN: "close-btn",
  CLICK_CANCEL_BTN: "cancel",
  KEYDOWN_ESC: "esc"
};
var Drawer = /*#__PURE__*/forwardRef(function (props, ref) {
  var _classnames;

  var className = props.className,
      style = props.style,
      visible = props.visible,
      _props$attach = props.attach,
      attach = _props$attach === void 0 ? "" : _props$attach,
      showOverlay = props.showOverlay,
      propsSize = props.size,
      placement = props.placement,
      onCancel = props.onCancel,
      onConfirm = props.onConfirm,
      onClose = props.onClose,
      onCloseBtnClick = props.onCloseBtnClick,
      onOverlayClick = props.onOverlayClick,
      onEscKeydown = props.onEscKeydown,
      showInAttachedElement = props.showInAttachedElement,
      closeOnOverlayClick = props.closeOnOverlayClick,
      closeOnEscKeydown = props.closeOnEscKeydown,
      children = props.children,
      header = props.header,
      body = props.body,
      footer = props.footer,
      closeBtn = props.closeBtn,
      cancelBtn = props.cancelBtn,
      confirmBtn = props.confirmBtn,
      zIndex = props.zIndex,
      destroyOnClose = props.destroyOnClose,
      mode = props.mode,
      _props$preventScrollT = props.preventScrollThrough,
      preventScrollThrough = _props$preventScrollT === void 0 ? true : _props$preventScrollT,
      sizeDraggable = props.sizeDraggable;

  var _useLocaleReceiver = useLocaleReceiver("drawer"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var _useGlobalIcon = useGlobalIcon({
    CloseIcon: CloseIcon
  }),
      CloseIcon$1 = _useGlobalIcon.CloseIcon;

  var size = propsSize !== null && propsSize !== void 0 ? propsSize : local.size;
  var confirmText = t(local.confirm);
  var cancelText = t(local.cancel);

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var containerRef = useRef();
  var contentWrapperRef = useRef();
  var drawerWrapperRef = useRef();
  var prefixCls = "".concat(classPrefix, "-drawer");
  var lockCls = "".concat(prefixCls, "--lock");
  var transform = visible ? "translate(0px)" : "";
  var closeIcon = /*#__PURE__*/React.isValidElement(closeBtn) ? closeBtn : /* @__PURE__ */React.createElement(CloseIcon$1, null);

  var _useDrag = useDrag(placement, sizeDraggable),
      dragSizeValue = _useDrag.dragSizeValue,
      enableDrag = _useDrag.enableDrag,
      draggableLineStyles = _useDrag.draggableLineStyles;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isDestroyOnClose = _useState2[0],
      setIsDestroyOnClose = _useState2[1];

  useImperativeHandle(ref, function () {
    return containerRef.current;
  });
  useEffect(function () {
    if (preventScrollThrough) {
      if (visible && !showInAttachedElement) {
        addClass(document.body, lockCls);
      } else {
        removeClass(document.body, lockCls);
      }
    }
  }, [visible, showInAttachedElement, lockCls, preventScrollThrough]);
  useEffect(function () {
    if (!destroyOnClose || !visible) return;
    setIsDestroyOnClose(false);
  }, [visible, destroyOnClose]);
  var getSizeValue = useCallback(function (size2) {
    if (dragSizeValue) return dragSizeValue;
    var defaultSize = isNaN(Number(size2)) ? size2 : "".concat(size2, "px");
    return {
      small: "300px",
      medium: "500px",
      large: "760px"
    }[size2] || defaultSize;
  }, [dragSizeValue]);
  useEffect(function () {
    var documentBodyCssText = "";

    if (visible) {
      if (attach !== "" && hasScrollbar()) {
        var scrollbarWidth = getScrollbarWidth();
        documentBodyCssText = "overflow: hidden; width: calc(100% - ".concat(scrollbarWidth, "px);");

        if (mode !== "push") {
          document.body.style.cssText = documentBodyCssText;
        }
      }
    }

    if (mode === "push") {
      drawerWrapperRef.current.style.cssText = "transition: margin 300ms cubic-bezier(0.7, 0.3, 0.1, 1) 0s";
      var marginStr = {
        left: "margin: 0 0 0 ".concat(getSizeValue(size)),
        right: "margin: 0 0 0 -".concat(getSizeValue(size)),
        top: "margin: ".concat(getSizeValue(size), " 0 0 0"),
        bottom: "margin: -".concat(getSizeValue(size), " 0 0 0")
      }[placement];

      if (visible) {
        drawerWrapperRef.current.style.cssText += marginStr;
      } else {
        drawerWrapperRef.current.style.cssText = drawerWrapperRef.current.style.cssText.replace(/margin:.+;/, "");
      }
    }

    if (contentWrapperRef.current) {
      containerRef.current.focus();
      contentWrapperRef.current.style.transform = transform;
    }
  }, [attach, mode, transform, visible, placement, size, getSizeValue]);

  function onMaskClick(e) {
    onOverlayClick === null || onOverlayClick === void 0 ? void 0 : onOverlayClick({
      e: e
    });
    (closeOnOverlayClick !== null && closeOnOverlayClick !== void 0 ? closeOnOverlayClick : local.closeOnOverlayClick) && (onClose === null || onClose === void 0 ? void 0 : onClose({
      e: e,
      trigger: CloseTriggerType.CLICK_OVERLAY
    }));
  }

  function onClickCloseBtn(e) {
    onCloseBtnClick === null || onCloseBtnClick === void 0 ? void 0 : onCloseBtnClick({
      e: e
    });
    onClose === null || onClose === void 0 ? void 0 : onClose({
      e: e,
      trigger: CloseTriggerType.CLICK_CLOSE_BTN
    });
  }

  function onKeyDownEsc(e) {
    if (e.key !== "Escape") return;
    onEscKeydown === null || onEscKeydown === void 0 ? void 0 : onEscKeydown({
      e: e
    });
    (closeOnEscKeydown !== null && closeOnEscKeydown !== void 0 ? closeOnEscKeydown : local.closeOnEscKeydown) && (onClose === null || onClose === void 0 ? void 0 : onClose({
      e: e,
      trigger: CloseTriggerType.KEYDOWN_ESC
    }));
  }

  function onCancelClick(e) {
    onCancel === null || onCancel === void 0 ? void 0 : onCancel({
      e: e
    });
    onClose === null || onClose === void 0 ? void 0 : onClose({
      e: e,
      trigger: CloseTriggerType.CLICK_CANCEL_BTN
    });
  }

  function onConfirmClick(e) {
    onConfirm === null || onConfirm === void 0 ? void 0 : onConfirm({
      e: e
    });
  }

  function onTransitionEnd() {
    if (!visible) {
      document.body.style.display = "none";
      document.body.style.overflow = "";
      document.body.style.width = "";
      document.body.style.margin = "";
      document.body.style.display = "block";
      destroyOnClose && setIsDestroyOnClose(true);
    }
  }

  var drawerClass = classNames(prefixCls, className, "".concat(prefixCls, "--").concat(placement), (_classnames = {}, _defineProperty(_classnames, "".concat(prefixCls, "--open"), visible), _defineProperty(_classnames, "".concat(prefixCls, "--attach"), showInAttachedElement), _defineProperty(_classnames, "".concat(prefixCls, "--without-mask"), !showOverlay), _classnames));
  var contentWrapperClass = classNames("".concat(prefixCls, "__content-wrapper"), "".concat(prefixCls, "__content-wrapper--").concat(placement));
  var getContentWrapperStyle = useCallback(function () {
    return {
      transform: visible ? "translateX(0)" : void 0,
      width: ["left", "right"].includes(placement) ? getSizeValue(size) : "",
      height: ["top", "bottom"].includes(placement) ? getSizeValue(size) : ""
    };
  }, [visible, placement, size, getSizeValue]);

  var _useState3 = useState(getContentWrapperStyle()),
      _useState4 = _slicedToArray(_useState3, 2),
      contentWrapperStyle = _useState4[0],
      setContentWrapperStyle = _useState4[1];

  useEffect(function () {
    setContentWrapperStyle(getContentWrapperStyle());
  }, [getContentWrapperStyle]);

  function getFooter() {
    if (footer !== true) return footer;
    var defaultCancelBtn = /* @__PURE__ */React.createElement(Button, {
      theme: "default",
      onClick: onCancelClick,
      className: "".concat(prefixCls, "__cancel")
    }, cancelBtn && typeof cancelBtn === "string" ? cancelBtn : cancelText);
    var defaultConfirmBtn = /* @__PURE__ */React.createElement(Button, {
      theme: "primary",
      onClick: onConfirmClick,
      className: "".concat(prefixCls, "__confirm")
    }, confirmBtn && typeof confirmBtn === "string" ? confirmBtn : confirmText);
    var renderCancelBtn = cancelBtn && /*#__PURE__*/React.isValidElement(cancelBtn) ? cancelBtn : defaultCancelBtn;
    var renderConfirmBtn = confirmBtn && /*#__PURE__*/React.isValidElement(confirmBtn) ? confirmBtn : defaultConfirmBtn;
    var footerStyle = {
      display: "flex",
      justifyContent: placement === "right" ? "flex-start" : "flex-end"
    };
    return /* @__PURE__ */React.createElement("div", {
      style: footerStyle
    }, placement === "right" && renderConfirmBtn, renderCancelBtn, placement !== "right" && renderConfirmBtn);
  }

  var renderOverlay = showOverlay && /* @__PURE__ */React.createElement("div", {
    className: "".concat(prefixCls, "__mask"),
    onClick: onMaskClick
  });
  var renderCloseBtn = closeBtn && /* @__PURE__ */React.createElement("div", {
    onClick: onClickCloseBtn,
    className: "".concat(prefixCls, "__close-btn")
  }, closeIcon);
  var renderHeader = header && /* @__PURE__ */React.createElement("div", {
    className: "".concat(prefixCls, "__header")
  }, header);
  var renderBody = /* @__PURE__ */React.createElement("div", {
    className: "".concat(prefixCls, "__body")
  }, body || children);
  var renderFooter = footer && /* @__PURE__ */React.createElement("div", {
    className: "".concat(prefixCls, "__footer")
  }, getFooter());
  if (isDestroyOnClose && !visible) return null;
  return /* @__PURE__ */React.createElement(DrawerWrapper, {
    attach: attach,
    ref: drawerWrapperRef
  }, /* @__PURE__ */React.createElement("div", {
    ref: containerRef,
    className: drawerClass,
    style: _objectSpread({
      zIndex: zIndex
    }, style),
    tabIndex: -1,
    onKeyDown: onKeyDownEsc,
    onTransitionEnd: onTransitionEnd
  }, renderOverlay, /* @__PURE__ */React.createElement("div", {
    ref: contentWrapperRef,
    className: contentWrapperClass,
    style: contentWrapperStyle
  }, renderCloseBtn, renderHeader, renderBody, renderFooter, sizeDraggable && /* @__PURE__ */React.createElement("div", {
    style: draggableLineStyles,
    onMouseDown: enableDrag
  }))));
});
Drawer.displayName = "Drawer";
Drawer.defaultProps = drawerDefaultProps;

export { CloseTriggerType, Drawer as default };
