// tdesign-react/lib/tree/useControllable.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import useControlled from '../hooks/useControlled.js';
import '../_chunks/dep-7909fefa.js';
import 'react';
import '../_chunks/dep-12260644.js';
import '../_chunks/dep-ec84085d.js';
import '../_chunks/dep-66273dbc.js';
import '../_chunks/dep-2ac048c0.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-b989427d.js';
import '../_chunks/dep-7f6754ac.js';
import '../_util/noop.js';

function useControllable(props) {
  var _useControlled = useControlled(props, "value", props.onChange),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      value = _useControlled2[0],
      onChange = _useControlled2[1];

  var _useControlled3 = useControlled(props, "expanded", props.onExpand),
      _useControlled4 = _slicedToArray(_useControlled3, 2),
      expanded = _useControlled4[0],
      onExpand = _useControlled4[1];

  var _useControlled5 = useControlled(props, "actived", props.onActive),
      _useControlled6 = _slicedToArray(_useControlled5, 2),
      actived = _useControlled6[0],
      onActive = _useControlled6[1];

  return {
    value: value,
    onChange: onChange,
    expanded: expanded,
    onExpand: onExpand,
    actived: actived,
    onActive: onActive
  };
}

export { useControllable as default };
