// tdesign-react/lib/tree/useTreeConfig.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import { useMemo } from 'react';
import useConfig from '../hooks/useConfig.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-1e32006c.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';

function useTreeConfig() {
  var _useConfig = useConfig(),
      prefix = _useConfig.classPrefix;

  var _useLocaleReceiver = useLocaleReceiver("tree"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      _locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  return useMemo(function () {
    var tree = "".concat(prefix, "-tree");
    var treeClassNames = {
      icon: "".concat(prefix, "-icon"),
      folderIcon: "".concat(prefix, "-folder-icon"),
      actived: "".concat(prefix, "-is-active"),
      disabled: "".concat(prefix, "-is-disabled"),
      treeIconRight: "".concat(prefix, "-icon-arrow-right"),
      treeIconDown: "".concat(prefix, "-icon-arrow-down"),
      tree: tree,
      treeFx: "".concat(tree, "--transition"),
      treeBlockNode: "".concat(tree, "--block-node"),
      treeEmpty: "".concat(tree, "__empty"),
      treeList: "".concat(tree, "__list"),
      treeNode: "".concat(tree, "__item"),
      treeNodeOpen: "".concat(tree, "__item--open"),
      treeHoverable: "".concat(tree, "--hoverable"),
      treeCheckable: "".concat(tree, "--checkable"),
      treeLabel: "".concat(tree, "__label"),
      treeLabelStrictly: "".concat(tree, "__label--strictly"),
      treeIcon: "".concat(tree, "__icon"),
      treeIconDefault: "".concat(tree, "__icon--default"),
      treeSpace: "".concat(tree, "__space"),
      treeOperations: "".concat(tree, "__operations"),
      line: "".concat(tree, "__line"),
      lineIsLeaf: "".concat(tree, "__line--leaf"),
      lineIsFirst: "".concat(tree, "__line--first"),
      treeNodeShow: "".concat(tree, "__item--show"),
      treeClickable: "".concat(tree, "__item--clickable"),
      loading: "".concat(prefix, "-icon-loading ").concat(prefix, "-icon-loading-blue"),
      toggleEnter: "".concat(tree, "__item--enter-active"),
      toggleLeave: "".concat(tree, "__item--leave-active")
    };
    var transitionNames = {
      treeNode: "".concat(prefix, "-tree-toggle")
    };
    var transitionClassNames = {
      enter: treeClassNames.toggleEnter,
      exit: treeClassNames.toggleLeave
    };
    return {
      treeClassNames: treeClassNames,
      transitionNames: transitionNames,
      transitionClassNames: transitionClassNames,
      transitionDuration: 300,
      locale: function locale(key) {
        return t(_locale[key]);
      }
    };
  }, [_locale, prefix, t]);
}

export { useTreeConfig };
