// tdesign-react/cjs/space/Space.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-4eb93341.js');
var React = require('react');
var classNames = require('classnames');
var reactIs = require('react-is');
var hooks_useConfig = require('../hooks/useConfig.js');
var space_defaultProps = require('./defaultProps.js');
require('../_chunks/dep-4344eec8.js');
require('../_chunks/dep-fe984d70.js');
require('../_chunks/dep-46cf36fd.js');
require('../_chunks/dep-f5e18a2f.js');
require('../_chunks/dep-53ba6729.js');
require('../_chunks/dep-61a8a7b0.js');
require('../_chunks/dep-2a90f794.js');
require('../_chunks/dep-aab723b3.js');
require('../_chunks/dep-384b291c.js');
require('../_chunks/dep-a0b5d8f6.js');
require('../_chunks/dep-12656997.js');
require('../_chunks/dep-1fff9729.js');
require('../_chunks/dep-bb60493d.js');
require('../_chunks/dep-72020528.js');
require('../_chunks/dep-ad854ba5.js');
require('../_chunks/dep-eebdbd74.js');
require('../_chunks/dep-8f18a7c4.js');
require('../_chunks/dep-9cd0fde8.js');
require('../_chunks/dep-d3ad6e52.js');
require('../_chunks/dep-aafeb50a.js');
require('../_chunks/dep-6fa7a9e9.js');
require('../_chunks/dep-994ec160.js');
require('../_chunks/dep-9429a38a.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var SizeMap = {
  small: "8px",
  medium: "16px",
  large: "24px"
};

var toArray = function toArray(children) {
  var ret = [];
  React__default["default"].Children.forEach(children, function (child) {
    if (child === void 0 || child === null) {
      return;
    }

    if (Array.isArray(child)) {
      ret = ret.concat(toArray(child));
    } else if (reactIs.isFragment(child) && child.props) {
      ret = ret.concat(toArray(child.props.children));
    } else {
      ret.push(child);
    }
  });
  return ret;
};

var Space = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _classNames;

  var className = props.className,
      style = props.style,
      align = props.align,
      direction = props.direction,
      size = props.size,
      breakLine = props.breakLine,
      separator = props.separator;

  var _useConfig = hooks_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var renderStyle = React.useMemo(function () {
    var renderGap = "";

    if (Array.isArray(size)) {
      renderGap = size.map(function (s) {
        if (typeof s === "number") return "".concat(s, "px");
        if (typeof s === "string") return SizeMap[s] || s;
        return s;
      }).join(" ");
    } else if (typeof size === "string") {
      renderGap = SizeMap[size] || size;
    } else if (typeof size === "number") {
      renderGap = "".concat(size, "px");
    }

    return _objectSpread(_objectSpread({
      gap: renderGap
    }, breakLine ? {
      flexWrap: "wrap"
    } : {}), style);
  }, [style, size, breakLine]);

  function renderChildren() {
    var children = toArray(props.children);
    var childCount = React__default["default"].Children.count(children);
    return React__default["default"].Children.map(children, function (child, index) {
      var showSeparator = index + 1 !== childCount && separator;
      return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("div", {
        className: "".concat(classPrefix, "-space-item")
      }, child), showSeparator && /* @__PURE__ */React__default["default"].createElement("div", {
        className: "".concat(classPrefix, "-space-item-separator")
      }, separator));
    });
  }

  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    style: renderStyle,
    className: classNames__default["default"]("".concat(classPrefix, "-space"), className, (_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-space-align-").concat(align), align), defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-space-").concat(direction), direction), _classNames))
  }, renderChildren());
});
Space.displayName = "Space";
Space.defaultProps = space_defaultProps.spaceDefaultProps;

exports.SizeMap = SizeMap;
exports["default"] = Space;
