// tdesign-react/lib/list/List.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import React from 'react';
import classNames from 'classnames';
import { i as isString_1 } from '../_chunks/dep-8d5556d7.js';
import useConfig from '../hooks/useConfig.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import forwardRefWithStatics from '../_util/forwardRefWithStatics.js';
import noop from '../_util/noop.js';
import { Loading } from '../loading/index.js';
import ListItem from './ListItem.js';
import ListItemMeta from './ListItemMeta.js';
import { listDefaultProps } from './defaultProps.js';
import '../_chunks/dep-7909fefa.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import 'hoist-non-react-statics';
import '../loading/Loading.js';
import '../_util/dom.js';
import 'raf';
import '../_util/easing.js';
import '../common/Portal.js';
import 'react-dom';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-d1b4ab24.js';
import '../_chunks/dep-8368bb87.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';

var List = forwardRefWithStatics(function (props, ref) {
  var _classNames, _classNames2;

  var header = props.header,
      footer = props.footer,
      asyncLoading = props.asyncLoading,
      size = props.size,
      split = props.split,
      stripe = props.stripe,
      layout = props.layout,
      children = props.children,
      className = props.className,
      _props$onLoadMore = props.onLoadMore,
      onLoadMore = _props$onLoadMore === void 0 ? noop : _props$onLoadMore,
      _props$onScroll = props.onScroll,
      onScroll = _props$onScroll === void 0 ? noop : _props$onScroll,
      style = props.style;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useLocaleReceiver = useLocaleReceiver("list"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var handleClickLoad = function handleClickLoad(e) {
    if (asyncLoading === "load-more") {
      onLoadMore({
        e: e
      });
    }
  };

  var handleScroll = function handleScroll(event) {
    var currentTarget = event.currentTarget;
    var scrollTop = currentTarget.scrollTop,
        offsetHeight = currentTarget.offsetHeight,
        scrollHeight = currentTarget.scrollHeight;
    var scrollBottom = scrollHeight - scrollTop - offsetHeight;
    onScroll({
      e: event,
      scrollTop: scrollTop,
      scrollBottom: scrollBottom
    });
  };

  var loadElement = isString_1(asyncLoading) ? /* @__PURE__ */React.createElement("div", {
    className: classNames("".concat(classPrefix, "-list__load"), (_classNames = {}, _defineProperty(_classNames, "".concat(classPrefix, "-list__load--loading"), asyncLoading === "loading"), _defineProperty(_classNames, "".concat(classPrefix, "-list__load--load-more"), asyncLoading === "load-more"), _classNames)),
    onClick: handleClickLoad
  }, asyncLoading === "loading" && /* @__PURE__ */React.createElement("div", null, /* @__PURE__ */React.createElement(Loading, {
    loading: true
  }), /* @__PURE__ */React.createElement("span", null, t(local.loadingText))), asyncLoading === "load-more" && /* @__PURE__ */React.createElement("span", null, t(local.loadingMoreText))) : asyncLoading;
  return /* @__PURE__ */React.createElement("div", {
    ref: ref,
    style: style,
    onScroll: handleScroll,
    className: classNames("".concat(classPrefix, "-list"), className, (_classNames2 = {}, _defineProperty(_classNames2, "".concat(classPrefix, "-list--split"), split), _defineProperty(_classNames2, "".concat(classPrefix, "-list--stripe"), stripe), _defineProperty(_classNames2, "".concat(classPrefix, "-list--vertical-action"), layout === "vertical"), _defineProperty(_classNames2, "".concat(classPrefix, "-size-s"), size === "small"), _defineProperty(_classNames2, "".concat(classPrefix, "-size-l"), size === "large"), _classNames2))
  }, header && /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-list__header")
  }, header), /* @__PURE__ */React.createElement("ul", {
    className: "".concat(classPrefix, "-list__inner")
  }, children), asyncLoading && loadElement, footer && /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-list__footer")
  }, footer));
}, {
  ListItem: ListItem,
  ListItemMeta: ListItemMeta
});
List.displayName = "List";
List.defaultProps = listDefaultProps;

export { List as default };
