// tdesign-react/lib/color-picker/components/panel/alpha.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../../_chunks/dep-1e32006c.js';
import { _ as _objectWithoutProperties } from '../../../_chunks/dep-8368bb87.js';
import React from 'react';
import classNames from 'classnames';
import ColorSlider from './slider.js';
import '../../const.js';
import '../../../_util/useDrag.js';

var _excluded = ["color", "baseClassName", "onChange"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var Alpha = function Alpha(props) {
  var color = props.color,
      baseClassName = props.baseClassName,
      onChange = props.onChange,
      rest = _objectWithoutProperties(props, _excluded);

  var handleChange = function handleChange(v, isDragEnd) {
    onChange(v / 100, isDragEnd);
  };

  var railStyle = {
    background: "linear-gradient(to right, rgba(0, 0, 0, 0), ".concat(props.color.rgb, ")")
  };
  return /* @__PURE__ */React.createElement(ColorSlider, _objectSpread({
    baseClassName: baseClassName,
    className: classNames(["".concat(baseClassName, "__alpha"), "".concat(baseClassName, "--bg-alpha")]),
    color: color,
    value: color.alpha * 100,
    onChange: handleChange,
    railStyle: railStyle,
    type: "alpha",
    maxValue: 100
  }, rest));
};

var AlphaSlider = /*#__PURE__*/React.memo(Alpha);

export { AlphaSlider as default };
