// tdesign-react/lib/color-picker/ColorPicker.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-1e32006c.js';
import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import { _ as _objectWithoutProperties } from '../_chunks/dep-8368bb87.js';
import React, { useState, useRef } from 'react';
import { Popup } from '../popup/index.js';
import useClassNames from './hooks/useClassNames.js';
import useControlled from '../hooks/useControlled.js';
import ColorTrigger from './components/trigger.js';
import ColorPanel from './components/panel/index.js';
import useClickOutside from '../_util/useClickOutside.js';
import { colorPickerDefaultProps } from './defaultProps.js';
import '../_chunks/dep-7909fefa.js';
import '../popup/Popup.js';
import 'react-transition-group';
import 'classnames';
import 'react-popper';
import '../_util/useAnimation.js';
import '../hooks/useConfig.js';
import '../_chunks/dep-918e41e2.js';
import '../_chunks/dep-734a21a0.js';
import '../_chunks/dep-68f7a306.js';
import '../_chunks/dep-a74f6fff.js';
import '../_chunks/dep-d23d346c.js';
import '../_chunks/dep-9ea787ca.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-0e2dfd87.js';
import '../_chunks/dep-f50ff346.js';
import '../_chunks/dep-51592cde.js';
import '../_chunks/dep-ae94c4e7.js';
import '../_chunks/dep-7f6754ac.js';
import '../_chunks/dep-2509667d.js';
import '../_chunks/dep-b817c90e.js';
import '../_chunks/dep-fe9825b6.js';
import '../_chunks/dep-ebc05da7.js';
import '../_chunks/dep-3160df29.js';
import '../_chunks/dep-c91b3852.js';
import '../_chunks/dep-9e44c34c.js';
import '../_chunks/dep-0c62d47a.js';
import '../_chunks/dep-e2fe68b7.js';
import '../_chunks/dep-44a03722.js';
import '../_chunks/dep-f5616354.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../common/Portal.js';
import 'react-dom';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-8d5556d7.js';
import '../_util/easing.js';
import '../popup/hooks/useTrigger.js';
import 'react-is';
import '../popup/utils/ref.js';
import '../_util/composeRefs.js';
import '../popup/utils/transition.js';
import '../_util/useMutationObserver.js';
import '../_chunks/dep-05d10538.js';
import '../_chunks/dep-652c1652.js';
import '../_chunks/dep-b989427d.js';
import '../_util/useWindowSize.js';
import '../popup/defaultProps.js';
import '../_chunks/dep-12260644.js';
import '../_chunks/dep-ec84085d.js';
import '../_chunks/dep-66273dbc.js';
import '../_chunks/dep-2ac048c0.js';
import '../_util/noop.js';
import '../input/index.js';
import '../input/Input.js';
import 'tdesign-icons-react';
import '../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../hooks/useGlobalIcon.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-59b9253f.js';
import '../_chunks/dep-d1b4ab24.js';
import '../input/InputGroup.js';
import '../locale/LocalReceiver.js';
import '../input/defaultProps.js';
import '../_common/js/color-picker/color.js';
import '../_chunks/dep-cc25e35c.js';
import 'tinycolor2';
import '../_common/js/color-picker/cmyk.js';
import '../_common/js/color-picker/gradient.js';
import '../_util/useCommonClassName.js';
import './components/panel/header.js';
import './const.js';
import '../radio/index.js';
import '../_chunks/dep-c0f2b2c6.js';
import '../common/Check.js';
import '../_util/helper.js';
import '../_chunks/dep-8fe8d164.js';
import '../radio/defaultProps.js';
import './components/panel/linear-gradient.js';
import '../_chunks/dep-91b4e41f.js';
import '../_chunks/dep-51754e76.js';
import '../_chunks/dep-6be1e491.js';
import '../_chunks/dep-c4881228.js';
import '../input-number/index.js';
import '../input-number/InputNumber.js';
import '../button/index.js';
import '../button/Button.js';
import '../_util/useRipple.js';
import '../_util/setStyle.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import '../button/defaultProps.js';
import '../input-number/useInputNumber.js';
import '../_common/js/input-number/number.js';
import '../_chunks/dep-31043569.js';
import '../_common/js/input-number/large-number.js';
import '../_common/js/log/log.js';
import '../input-number/defaultProps.js';
import './components/panel/saturation.js';
import '../_util/useDrag.js';
import './components/panel/hue.js';
import './components/panel/slider.js';
import './components/panel/alpha.js';
import './components/panel/format/index.js';
import 'lodash';
import '../select/index.js';
import '../select/base/Select.js';
import '../_chunks/dep-1f300d2c.js';
import '../_chunks/dep-e103618e.js';
import '../select/util/helper.js';
import '../select/base/OptionGroup.js';
import '../select/defaultProps.js';
import '../select/base/Option.js';
import '../common/FakeArrow.js';
import '../select-input/index.js';
import '../select-input/SelectInput.js';
import '../select-input/useSingle.js';
import '../_chunks/dep-5e5947bc.js';
import '../_chunks/dep-dcadd2e9.js';
import '../_chunks/dep-9f757569.js';
import '../select-input/useMultiple.js';
import '../tag-input/index.js';
import '../tag-input/TagInput.js';
import '../_util/useDragSorter.js';
import '../tag-input/useTagScroll.js';
import '../tag-input/useTagList.js';
import '../tag/index.js';
import '../tag/Tag.js';
import '../tag/CheckTag.js';
import '../tag/defaultProps.js';
import '../tag-input/useHover.js';
import '../tag-input/defaultProps.js';
import '../select-input/useOverlayInnerStyle.js';
import '../select-input/defaultProps.js';
import '../select/base/PopupContent.js';
import './components/panel/format/inputs.js';
import './components/panel/format/config.js';
import './components/panel/swatches.js';

var _excluded = ["popupProps", "disabled", "inputProps", "onChange", "colorModes"],
    _excluded2 = ["overlayClassName", "overlayInnerStyle"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var ColorPicker = function ColorPicker(props) {
  var baseClassName = useClassNames();

  var popupProps = props.popupProps,
      disabled = props.disabled,
      inputProps = props.inputProps,
      onChange = props.onChange,
      colorModes = props.colorModes,
      rest = _objectWithoutProperties(props, _excluded);

  var _ref = popupProps || {},
      overlayClassName = _ref.overlayClassName,
      _ref$overlayInnerStyl = _ref.overlayInnerStyle,
      overlayInnerStyle = _ref$overlayInnerStyl === void 0 ? {} : _ref$overlayInnerStyl,
      restPopupProps = _objectWithoutProperties(_ref, _excluded2);

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      visible = _useState2[0],
      setVisible = _useState2[1];

  var _useControlled = useControlled(props, "value", onChange),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      innerValue = _useControlled2[0],
      setInnerValue = _useControlled2[1];

  var triggerRef = useRef();
  var colorPanelRef = useRef();

  var popProps = _objectSpread(_objectSpread({
    placement: "bottom-left",
    expandAnimation: true,
    trigger: "click",
    visible: visible
  }, restPopupProps), {}, {
    overlayClassName: [baseClassName, overlayClassName],
    overlayInnerStyle: _objectSpread({
      padding: 0
    }, overlayInnerStyle)
  });

  useClickOutside([triggerRef, colorPanelRef], function () {
    setVisible(false);
  }, true);
  return /* @__PURE__ */React.createElement(Popup, _objectSpread(_objectSpread({}, popProps), {}, {
    onVisibleChange: function onVisibleChange(v) {
      return setVisible(v);
    },
    content: !disabled && /* @__PURE__ */React.createElement(ColorPanel, _objectSpread(_objectSpread({}, rest), {}, {
      disabled: disabled,
      value: innerValue,
      colorModes: colorModes,
      togglePopup: setVisible,
      onChange: function onChange(value, context) {
        return setInnerValue(value, context);
      },
      ref: colorPanelRef
    }))
  }), /* @__PURE__ */React.createElement("div", {
    className: "".concat(baseClassName, "__trigger"),
    ref: triggerRef
  }, /* @__PURE__ */React.createElement(ColorTrigger, {
    disabled: disabled,
    inputProps: inputProps,
    value: innerValue,
    onChange: setInnerValue
  })));
};

ColorPicker.displayName = "ColorPicker";
ColorPicker.defaultProps = colorPickerDefaultProps;
var _ColorPicker = /*#__PURE__*/React.memo(ColorPicker);

export { _ColorPicker as default };
