// tdesign-react/es/tree/useControllable.js
/**
 * tdesign v0.41.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-9615f03f.js';
import useControlled from '../hooks/useControlled.js';
import '../_chunks/dep-7909fefa.js';
import 'react';
import '../_chunks/dep-f54ff1e5.js';
import '../_chunks/dep-1cce5a60.js';
import '../_chunks/dep-e1d4dd61.js';
import '../_chunks/dep-236ef238.js';
import '../_chunks/dep-c8bfa11e.js';
import '../_chunks/dep-4a64abb1.js';
import '../_chunks/dep-4e0da095.js';
import '../_chunks/dep-264b0e55.js';
import '../_chunks/dep-199b1581.js';
import '../_chunks/dep-e237f0f2.js';
import '../_util/noop.js';

function useControllable(props) {
  var _useControlled = useControlled(props, "value", props.onChange),
      _useControlled2 = _slicedToArray(_useControlled, 2),
      value = _useControlled2[0],
      onChange = _useControlled2[1];

  var _useControlled3 = useControlled(props, "expanded", props.onExpand),
      _useControlled4 = _slicedToArray(_useControlled3, 2),
      expanded = _useControlled4[0],
      onExpand = _useControlled4[1];

  var _useControlled5 = useControlled(props, "actived", props.onActive),
      _useControlled6 = _slicedToArray(_useControlled5, 2),
      actived = _useControlled6[0],
      onActive = _useControlled6[1];

  return {
    value: value,
    onChange: onChange,
    expanded: expanded,
    onExpand: onExpand,
    actived: actived,
    onActive: onActive
  };
}

export { useControllable as default };
